// ra-ui-materialui/dist/KeyboardShortcut.js
import * as React from 'react';
import { styled, Typography, } from '@mui/material';
import clsx from 'clsx';
export const KeyboardShortcut = ({ className, keyboardShortcut, ...rest }) => {
    if (!keyboardShortcut) {
        return null;
    }
    return (React.createElement(Root, { className: clsx(KeyboardShortcutClasses.root, className), ...rest }, keyboardShortcut
        .split('>')
        .map((sequence, sequenceIndex, sequences) => (React.createElement(React.Fragment, { key: `${sequence}-${sequenceIndex}` },
        sequence.split('+').map((key, keyIndex) => (React.createElement(React.Fragment, { key: `${key}-${keyIndex}` },
            React.createElement(Typography, { component: "kbd", className: KeyboardShortcutClasses.kbd, key: key }, KeyMap[key]
                ? KeyMap[key]
                : key.toUpperCase())))),
        sequenceIndex < sequences.length - 1 ? (React.createElement(React.Fragment, null, "\u00A0")) : null)))));
};
const KeyMap = {
    meta: '⌘',
    mod: '⌘',
    ctrl: '⌃',
    shift: '⇧',
    alt: '⌥',
    enter: '⏎',
    esc: '⎋',
    escape: '⎋',
    backspace: '⌫',
    delete: '⌦',
    tab: '⇥',
    space: '␣',
    up: '↑',
    down: '↓',
    left: '←',
    right: '→',
    home: '↖',
    end: '↘',
    pageup: '⇞',
    pagedown: '⇟',
};
const PREFIX = 'RaKeyboardShortcut';
const KeyboardShortcutClasses = {
    root: `${PREFIX}-root`,
    kbd: `${PREFIX}-kbd`,
};
const Root = styled('div')(({ theme }) => ({
    opacity: 0.7,
    [`& .${KeyboardShortcutClasses.kbd}`]: {
        padding: '4px 5px',
        display: 'inline-block',
        whiteSpace: 'nowrap',
        margin: '0 1px',
        fontSize: '11px',
        lineHeight: '10px',
        color: theme.palette.text.primary,
        verticalAlign: 'middle',
        border: `1px solid ${theme.palette.divider}`,
        borderRadius: 6,
        boxShadow: `inset 0 -1px 0 ${theme.palette.divider}`,
    },
}));
