// ra-ui-materialui/dist/KeyboardShortcut.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyboardShortcut = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const KeyboardShortcut = ({ className, keyboardShortcut, ...rest }) => {
    if (!keyboardShortcut) {
        return null;
    }
    return (React.createElement(Root, { className: (0, clsx_1.default)(KeyboardShortcutClasses.root, className), ...rest }, keyboardShortcut
        .split('>')
        .map((sequence, sequenceIndex, sequences) => (React.createElement(React.Fragment, { key: `${sequence}-${sequenceIndex}` },
        sequence.split('+').map((key, keyIndex) => (React.createElement(React.Fragment, { key: `${key}-${keyIndex}` },
            React.createElement(material_1.Typography, { component: "kbd", className: KeyboardShortcutClasses.kbd, key: key }, KeyMap[key]
                ? KeyMap[key]
                : key.toUpperCase())))),
        sequenceIndex < sequences.length - 1 ? (React.createElement(React.Fragment, null, "\u00A0")) : null)))));
};
exports.KeyboardShortcut = KeyboardShortcut;
const KeyMap = {
    meta: '⌘',
    mod: '⌘',
    ctrl: '⌃',
    shift: '⇧',
    alt: '⌥',
    enter: '⏎',
    esc: '⎋',
    escape: '⎋',
    backspace: '⌫',
    delete: '⌦',
    tab: '⇥',
    space: '␣',
    up: '↑',
    down: '↓',
    left: '←',
    right: '→',
    home: '↖',
    end: '↘',
    pageup: '⇞',
    pagedown: '⇟',
};
const PREFIX = 'RaKeyboardShortcut';
const KeyboardShortcutClasses = {
    root: `${PREFIX}-root`,
    kbd: `${PREFIX}-kbd`,
};
const Root = (0, material_1.styled)('div')(({ theme }) => ({
    opacity: 0.7,
    [`& .${KeyboardShortcutClasses.kbd}`]: {
        padding: '4px 5px',
        display: 'inline-block',
        whiteSpace: 'nowrap',
        margin: '0 1px',
        fontSize: '11px',
        lineHeight: '10px',
        color: theme.palette.text.primary,
        verticalAlign: 'middle',
        border: `1px solid ${theme.palette.divider}`,
        borderRadius: 6,
        boxShadow: `inset 0 -1px 0 ${theme.palette.divider}`,
    },
}));
