// ra-ui-materialui/dist/Labeled.js
import * as React from 'react';
import { Stack, Typography, } from '@mui/material';
import { styled } from '@mui/material/styles';
import clsx from 'clsx';
import { FieldTitle } from 'ra-core';
import { useThemeProps } from '@mui/system';
/**
 * Wrap a field or an input with a label if necessary.
 *
 * The label is displayed if:
 * - the field or input has a label prop that is not false, or
 * - the field or input has a source prop
 *
 * @example
 * <Labeled>
 *     <FooComponent source="title" />
 * </Labeled>
 */
export const Labeled = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { children, className = '', color, component = 'span', fullWidth, isRequired, label, resource, source, TypographyProps, ...rest } = props;
    return (React.createElement(Root
    // @ts-ignore https://github.com/mui/material-ui/issues/29875
    , { 
        // @ts-ignore https://github.com/mui/material-ui/issues/29875
        component: component, className: clsx(className, {
            [LabeledClasses.fullWidth]: fullWidth,
        }), ...rest },
        label !== false &&
            children.props.label !== false &&
            typeof children.type !== 'string' &&
            // @ts-ignore
            children.type?.displayName !== 'Labeled' &&
            // @ts-ignore
            children.type?.displayName !== 'Labeled' ? (React.createElement(Typography, { sx: color
                ? { color }
                : {
                    color: theme => (theme.vars || theme).palette.text
                        .secondary,
                }, className: LabeledClasses.label, ...TypographyProps },
            React.createElement(FieldTitle, { label: label || children.props.label, source: source || children.props.source, resource: resource, isRequired: isRequired }))) : null,
        children));
};
Labeled.displayName = 'Labeled';
const PREFIX = 'RaLabeled';
export const LabeledClasses = {
    label: `${PREFIX}-label`,
    fullWidth: `${PREFIX}-fullWidth`,
};
const Root = styled(Stack, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'inline-flex',
    marginBottom: '0.2em',
    [`&.${LabeledClasses.fullWidth}`]: {
        width: '100%',
    },
    [`& .${LabeledClasses.label}`]: {
        fontSize: '0.75em',
        marginBottom: '0.2em',
    },
});
