// ra-ui-materialui/dist/Labeled.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LabeledClasses = exports.Labeled = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const system_1 = require("@mui/system");
/**
 * Wrap a field or an input with a label if necessary.
 *
 * The label is displayed if:
 * - the field or input has a label prop that is not false, or
 * - the field or input has a source prop
 *
 * @example
 * <Labeled>
 *     <FooComponent source="title" />
 * </Labeled>
 */
const Labeled = (inProps) => {
    const props = (0, system_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, className = '', color, component = 'span', fullWidth, isRequired, label, resource, source, TypographyProps, ...rest } = props;
    return (React.createElement(Root
    // @ts-ignore https://github.com/mui/material-ui/issues/29875
    , { 
        // @ts-ignore https://github.com/mui/material-ui/issues/29875
        component: component, className: (0, clsx_1.default)(className, {
            [exports.LabeledClasses.fullWidth]: fullWidth,
        }), ...rest },
        label !== false &&
            children.props.label !== false &&
            typeof children.type !== 'string' &&
            // @ts-ignore
            children.type?.displayName !== 'Labeled' &&
            // @ts-ignore
            children.type?.displayName !== 'Labeled' ? (React.createElement(material_1.Typography, { sx: color
                ? { color }
                : {
                    color: theme => (theme.vars || theme).palette.text
                        .secondary,
                }, className: exports.LabeledClasses.label, ...TypographyProps },
            React.createElement(ra_core_1.FieldTitle, { label: label || children.props.label, source: source || children.props.source, resource: resource, isRequired: isRequired }))) : null,
        children));
};
exports.Labeled = Labeled;
exports.Labeled.displayName = 'Labeled';
const PREFIX = 'RaLabeled';
exports.LabeledClasses = {
    label: `${PREFIX}-label`,
    fullWidth: `${PREFIX}-fullWidth`,
};
const Root = (0, styles_1.styled)(material_1.Stack, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'inline-flex',
    marginBottom: '0.2em',
    [`&.${exports.LabeledClasses.fullWidth}`]: {
        width: '100%',
    },
    [`& .${exports.LabeledClasses.label}`]: {
        fontSize: '0.75em',
        marginBottom: '0.2em',
    },
});
