// ra-ui-materialui/dist/Link.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkClasses = exports.Link = void 0;
const React = __importStar(require("react"));
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const Link = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { to, children, className, ...rest } = props;
    return (React.createElement(StyledMuiLink, { component: ra_core_1.LinkBase, to: to, className: (0, clsx_1.default)(exports.LinkClasses.link, className), ...rest }, children));
};
exports.Link = Link;
const PREFIX = 'RaLink';
exports.LinkClasses = {
    link: `${PREFIX}-link`,
};
const StyledMuiLink = (0, material_1.styled)(material_1.Link, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({}); // @see https://mui.com/material-ui/guides/typescript/#complications-with-the-component-prop
