// ra-ui-materialui/dist/Offline.js
import * as React from 'react';
import { Alert, styled, Typography, } from '@mui/material';
import { useGetResourceLabel, useResourceContext, useResourceTranslation, } from 'ra-core';
import clsx from 'clsx';
export const Offline = (props) => {
    const { icon, message: messageProp, variant = 'standard', ...rest } = props;
    const resource = useResourceContext(props);
    const getResourceLabel = useGetResourceLabel();
    const message = useResourceTranslation({
        baseI18nKey: 'ra.notification.offline',
        resourceI18nKey: resource
            ? `resources.${resource}.notification.offline`
            : undefined,
        userText: messageProp,
        options: {
            name: resource ? getResourceLabel(resource, 0) : undefined,
            _: 'No connectivity. Could not fetch data.',
        },
    });
    return (React.createElement(Root, { className: clsx(OfflineClasses.root, {
            [OfflineClasses.inline]: variant === 'inline',
        }), severity: "warning", variant: variant === 'inline' ? 'outlined' : variant, icon: variant === 'inline' ? false : icon, ...rest },
        React.createElement(Typography, { variant: "body2" }, message)));
};
const PREFIX = 'RaOffline';
export const OfflineClasses = {
    root: `${PREFIX}-root`,
    inline: `${PREFIX}-inline`,
};
const Root = styled(Alert, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    [`&.${OfflineClasses.inline}`]: {
        border: 'none',
        display: 'inline-flex',
        padding: 0,
        margin: 0,
    },
}));
