// ra-ui-materialui/dist/list/BulkActionsToolbar.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BulkActionsToolbarClasses = exports.BulkActionsToolbar = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const Toolbar_1 = __importDefault(require("@mui/material/Toolbar"));
const Typography_1 = __importDefault(require("@mui/material/Typography"));
const IconButton_1 = __importDefault(require("@mui/material/IconButton"));
const Close_1 = __importDefault(require("@mui/icons-material/Close"));
const ra_core_1 = require("ra-core");
const TopToolbar_1 = __importDefault(require("../layout/TopToolbar.cjs"));
const button_1 = require("../button/index.cjs");
const defaultSelectAllButton = React.createElement(button_1.SelectAllButton, null);
const BulkActionsToolbar = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { label = 'ra.action.bulk_actions', children, className, selectAllButton, ...rest } = props;
    const { selectedIds = [], onUnselectItems } = (0, ra_core_1.useListContext)();
    const translate = (0, ra_core_1.useTranslate)();
    const handleUnselectAllClick = (0, react_1.useCallback)(() => {
        onUnselectItems();
    }, [onUnselectItems]);
    return (React.createElement(Root, { className: className },
        React.createElement(Toolbar_1.default, { "data-test": "bulk-actions-toolbar", className: (0, clsx_1.default)(exports.BulkActionsToolbarClasses.toolbar, {
                [exports.BulkActionsToolbarClasses.collapsed]: selectedIds.length === 0,
            }), ...(0, ra_core_1.sanitizeListRestProps)(rest) },
            React.createElement("div", { className: exports.BulkActionsToolbarClasses.title },
                React.createElement(IconButton_1.default, { className: exports.BulkActionsToolbarClasses.icon, "aria-label": translate('ra.action.unselect'), title: translate('ra.action.unselect'), onClick: handleUnselectAllClick, color: "primary", size: "small" },
                    React.createElement(Close_1.default, { fontSize: "small" })),
                React.createElement(Typography_1.default, { sx: {
                        color: theme => (theme.vars || theme).palette.text.primary,
                    }, variant: "body1" }, translate(label, {
                    _: label,
                    smart_count: selectedIds.length,
                })),
                selectAllButton !== false
                    ? (0, react_1.isValidElement)(selectAllButton)
                        ? selectAllButton
                        : defaultSelectAllButton
                    : null),
            React.createElement(TopToolbar_1.default, { className: exports.BulkActionsToolbarClasses.topToolbar }, children))));
};
exports.BulkActionsToolbar = BulkActionsToolbar;
const PREFIX = 'RaBulkActionsToolbar';
exports.BulkActionsToolbarClasses = {
    toolbar: `${PREFIX}-toolbar`,
    topToolbar: `${PREFIX}-topToolbar`,
    buttons: `${PREFIX}-buttons`,
    collapsed: `${PREFIX}-collapsed`,
    title: `${PREFIX}-title`,
    icon: `${PREFIX}-icon`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    position: 'relative',
    [`& .${exports.BulkActionsToolbarClasses.toolbar}`]: {
        position: 'absolute',
        left: 0,
        right: 0,
        zIndex: 3,
        color: (theme.vars || theme).palette.primary.contrastText,
        justifyContent: 'space-between',
        backgroundColor: theme.vars
            ? theme.vars.palette.bulkActionsToolbarBackgroundColor
            : theme.palette.mode === 'light'
                ? (0, styles_1.lighten)(theme.palette.primary.light, 0.8)
                : (0, styles_1.darken)(theme.palette.primary.dark, 0.5),
        minHeight: theme.spacing(6),
        height: theme.spacing(6),
        paddingRight: theme.spacing(2),
        transform: `translateY(${theme.spacing(-6)})`,
        transition: `${theme.transitions.create('height')}, ${theme.transitions.create('min-height')}, ${theme.transitions.create('transform')}`,
        borderTopLeftRadius: theme.shape.borderRadius,
        borderTopRightRadius: theme.shape.borderRadius,
    },
    [`& .${exports.BulkActionsToolbarClasses.topToolbar}`]: {
        padding: theme.spacing(0.5),
        minHeight: 'auto',
        [theme.breakpoints.down('sm')]: {
            backgroundColor: 'transparent',
        },
    },
    [`& .${exports.BulkActionsToolbarClasses.buttons}`]: {},
    [`& .${exports.BulkActionsToolbarClasses.collapsed}`]: {
        minHeight: 0,
        height: 0,
        transform: `translateY(0)`,
        overflowY: 'hidden',
    },
    [`& .${exports.BulkActionsToolbarClasses.title}`]: {
        display: 'flex',
        flex: '0 0 auto',
        alignItems: 'center',
        gap: theme.spacing(1),
    },
    [`& .${exports.BulkActionsToolbarClasses.icon}`]: {
        marginLeft: '-0.5em',
    },
}));
