// ra-ui-materialui/dist/list/Empty.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmptyClasses = exports.Empty = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const Inbox_1 = __importDefault(require("@mui/icons-material/Inbox"));
const ra_core_1 = require("ra-core");
const button_1 = require("../button/index.cjs");
const Empty = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className } = props;
    const { hasCreate } = (0, ra_core_1.useResourceDefinition)(props);
    const resource = (0, ra_core_1.useResourceContext)(props);
    const translate = (0, ra_core_1.useTranslate)();
    const getResourceLabel = (0, ra_core_1.useGetResourceLabel)();
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 0,
        _: resource ? getResourceLabel(resource, 0) : undefined,
    });
    const emptyMessage = translate('ra.page.empty', { name: resourceName });
    const inviteMessage = translate('ra.page.invite');
    return (React.createElement(Root, { className: className },
        React.createElement("div", { className: exports.EmptyClasses.message },
            React.createElement(Inbox_1.default, { className: exports.EmptyClasses.icon }),
            React.createElement(material_1.Typography, { variant: "h4", paragraph: true }, translate(`resources.${resource}.empty`, {
                _: emptyMessage,
            })),
            hasCreate && (React.createElement(material_1.Typography, { variant: "body1" }, translate(`resources.${resource}.invite`, {
                _: inviteMessage,
            })))),
        hasCreate && (React.createElement("div", { className: exports.EmptyClasses.toolbar },
            React.createElement(button_1.CreateButton, { variant: "contained" })))));
};
exports.Empty = Empty;
const PREFIX = 'RaEmpty';
exports.EmptyClasses = {
    message: `${PREFIX}-message`,
    icon: `${PREFIX}-icon`,
    toolbar: `${PREFIX}-toolbar`,
};
const Root = (0, styles_1.styled)('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flex: 1,
    [`& .${exports.EmptyClasses.message}`]: {
        textAlign: 'center',
        margin: '0 1em',
        color: (theme.vars || theme).palette.text.disabled,
    },
    [`& .${exports.EmptyClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
    [`& .${exports.EmptyClasses.toolbar}`]: {
        textAlign: 'center',
        marginTop: '2em',
    },
}));
