// ra-ui-materialui/dist/list/Empty.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Typography } from '@mui/material';
import Inbox from '@mui/icons-material/Inbox';
import { useTranslate, useResourceDefinition, useResourceContext, useGetResourceLabel, } from 'ra-core';
import { CreateButton } from "../button/index.js";
export const Empty = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className } = props;
    const { hasCreate } = useResourceDefinition(props);
    const resource = useResourceContext(props);
    const translate = useTranslate();
    const getResourceLabel = useGetResourceLabel();
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 0,
        _: resource ? getResourceLabel(resource, 0) : undefined,
    });
    const emptyMessage = translate('ra.page.empty', { name: resourceName });
    const inviteMessage = translate('ra.page.invite');
    return (React.createElement(Root, { className: className },
        React.createElement("div", { className: EmptyClasses.message },
            React.createElement(Inbox, { className: EmptyClasses.icon }),
            React.createElement(Typography, { variant: "h4", paragraph: true }, translate(`resources.${resource}.empty`, {
                _: emptyMessage,
            })),
            hasCreate && (React.createElement(Typography, { variant: "body1" }, translate(`resources.${resource}.invite`, {
                _: inviteMessage,
            })))),
        hasCreate && (React.createElement("div", { className: EmptyClasses.toolbar },
            React.createElement(CreateButton, { variant: "contained" })))));
};
const PREFIX = 'RaEmpty';
export const EmptyClasses = {
    message: `${PREFIX}-message`,
    icon: `${PREFIX}-icon`,
    toolbar: `${PREFIX}-toolbar`,
};
const Root = styled('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flex: 1,
    [`& .${EmptyClasses.message}`]: {
        textAlign: 'center',
        margin: '0 1em',
        color: (theme.vars || theme).palette.text.disabled,
    },
    [`& .${EmptyClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
    [`& .${EmptyClasses.toolbar}`]: {
        textAlign: 'center',
        marginTop: '2em',
    },
}));
