// ra-ui-materialui/dist/list/listFieldTypes.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.listFieldTypes = void 0;
const React = __importStar(require("react"));
const datatable_1 = require("./datatable/index.cjs");
const SingleFieldList_1 = require("./SingleFieldList.cjs");
const field_1 = require("../field/index.cjs");
exports.listFieldTypes = {
    table: {
        component: props => {
            return React.createElement(datatable_1.DataTable, { ...props });
        },
        representation: (_props, children) => `        <DataTable>
${children.map(child => `            ${child.getRepresentation()}`).join('\n')}
        </DataTable>`,
    },
    array: {
        component: ({ children, ...props }) => {
            const childrenArray = React.Children.toArray(children);
            return (React.createElement(datatable_1.DataTable.Col, { ...props },
                React.createElement(field_1.ArrayField, { ...props },
                    React.createElement(SingleFieldList_1.SingleFieldList, null,
                        React.createElement(field_1.ChipField, { source: childrenArray.length > 0 &&
                                React.isValidElement(childrenArray[0]) &&
                                childrenArray[0].props.source })))));
        },
        representation: (props, children) => `<DataTable.Col source="${props.source}">
                <ArrayField source="${props.source}">
                    <SingleFieldList>
                        <ChipField source="${children.length > 0 && children[0].getProps().source}" />
                    </SingleFieldList>
                </ArrayField>
            </DataTable.Col>`,
    },
    scalar_array: {
        component: (props) => {
            return (React.createElement(datatable_1.DataTable.Col, { ...props },
                React.createElement(field_1.TextArrayField, { ...props })));
        },
        representation: props => `<DataTable.Col source="${props.source}">
                <TextArrayField source="${props.source}" />
            </DataTable.Col>`,
    },
    boolean: {
        component: props => (React.createElement(datatable_1.DataTable.Col, { ...props },
            React.createElement(field_1.BooleanField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <BooleanField source="${props.source}" />
            </DataTable.Col>`,
    },
    date: {
        component: props => (React.createElement(datatable_1.DataTable.Col, { ...props },
            React.createElement(field_1.DateField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <DateField source="${props.source}" />
            </DataTable.Col>`,
    },
    email: {
        component: props => (React.createElement(datatable_1.DataTable.Col, { ...props },
            React.createElement(field_1.EmailField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <EmailField source="${props.source}" />
            </DataTable.Col>`,
    },
    id: {
        component: props => React.createElement(datatable_1.DataTable.Col, { ...props }),
        representation: props => `<DataTable.Col source="${props.source}" />`,
    },
    number: {
        component: datatable_1.DataTable.NumberCol,
        representation: props => `<DataTable.NumberCol source="${props.source}" />`,
    },
    reference: {
        component: props => (React.createElement(datatable_1.DataTable.Col, { ...props },
            React.createElement(field_1.ReferenceField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <ReferenceField source="${props.source}" reference="${props.reference}" />
            </DataTable.Col>`,
    },
    referenceChild: {
        component: () => React.createElement(field_1.TextField, { source: "id" }),
        representation: () => `<TextField source="id" />`,
    },
    referenceArray: {
        component: props => (React.createElement(datatable_1.DataTable.Col, { ...props },
            React.createElement(field_1.ReferenceArrayField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <ReferenceArrayField source="${props.source}" reference="${props.reference}" />
            </DataTable.Col>`,
    },
    referenceArrayChild: {
        component: () => (React.createElement(SingleFieldList_1.SingleFieldList, null,
            React.createElement(field_1.ChipField, { source: "id" }))),
        representation: () => `<SingleFieldList>
                <ChipField source="id" />
            </SingleFieldList>`,
    },
    richText: undefined, // never display a rich text field in a datagrid
    string: {
        component: datatable_1.DataTable.Col,
        representation: props => `<DataTable.Col source="${props.source}" />`,
    },
    url: {
        component: props => (React.createElement(datatable_1.DataTable.Col, { ...props },
            React.createElement(field_1.UrlField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <UrlField source="${props.source}" />
            </DataTable.Col>`,
    },
};
