// ra-ui-materialui/dist/list/ListToolbar.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListToolbar = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const filter_1 = require("./filter/index.cjs");
exports.ListToolbar = (0, react_1.memo)(inProps => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { filters, actions, className, ...rest } = props;
    return Array.isArray(filters) ? (React.createElement(ra_core_1.FilterContext.Provider, { value: filters },
        React.createElement(Root, { className: className },
            React.createElement(filter_1.FilterForm, null),
            React.createElement("span", null),
            actions))) : (React.createElement(Root, { className: className },
        filters &&
            React.cloneElement(filters, {
                ...rest,
                context: 'form',
            }),
        React.createElement("span", null),
        actions &&
            React.cloneElement(actions, {
                ...rest,
                filters,
                ...actions.props,
            })));
});
const PREFIX = 'RaListToolbar';
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    position: 'relative',
    justifyContent: 'space-between',
    alignItems: 'flex-end',
    width: '100%',
    [theme.breakpoints.down('md')]: {
        flexWrap: 'wrap',
    },
    [theme.breakpoints.down('sm')]: {
        backgroundColor: (theme.vars || theme).palette.background.paper,
        flexWrap: 'inherit',
        flexDirection: 'column-reverse',
    },
}));
