// ra-ui-materialui/dist/list/ListView.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListClasses = exports.ListView = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const Card_1 = __importDefault(require("@mui/material/Card"));
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const Title_1 = require("../layout/Title.cjs");
const ListToolbar_1 = require("./ListToolbar.cjs");
const pagination_1 = require("./pagination/index.cjs");
const ListActions_1 = require("./ListActions.cjs");
const Empty_1 = require("./Empty.cjs");
const Offline_1 = require("../Offline.cjs");
const defaultActions = React.createElement(ListActions_1.ListActions, null);
const defaultPagination = React.createElement(pagination_1.Pagination, null);
const defaultEmpty = React.createElement(Empty_1.Empty, null);
const DefaultComponent = Card_1.default;
const defaultOffline = React.createElement(Offline_1.Offline, null);
const ListView = (props) => {
    const { actions = defaultActions, aside, filters, emptyWhileLoading, pagination = defaultPagination, children, className, component: Content = DefaultComponent, title, empty = defaultEmpty, render, offline = defaultOffline, error, ...rest } = props;
    const listContext = (0, ra_core_1.useListContext)();
    const { defaultTitle, data, error: errorState, isPaused, isPending, isPlaceholderData, filterValues, resource, total, hasNextPage, hasPreviousPage, } = listContext;
    const showOffline = isPaused &&
        (isPending || isPlaceholderData) &&
        offline !== false &&
        offline !== undefined;
    const showError = errorState && error !== false && error !== undefined;
    if ((!children && !render) ||
        (!data && isPending && !isPaused && emptyWhileLoading)) {
        return null;
    }
    const renderList = () => {
        return (React.createElement("div", { className: (0, clsx_1.default)(exports.ListClasses.main, {
                [exports.ListClasses.noActions]: !(filters || actions) || showOffline || showError,
            }) },
            filters || actions ? (React.createElement(ListToolbar_1.ListToolbar, { className: exports.ListClasses.actions, filters: filters, actions: actions })) : null,
            React.createElement(Content, { className: exports.ListClasses.content }, showOffline
                ? offline
                : showError
                    ? error
                    : render
                        ? render(listContext)
                        : children),
            pagination !== false ? pagination : null));
    };
    const renderEmpty = () => empty !== false && React.createElement("div", { className: exports.ListClasses.noResults }, empty);
    const shouldRenderEmptyPage = !errorState &&
        // the list is not loading data for the first time
        !isPending &&
        // the API returned no data (using either normal or partial pagination)
        (total === 0 ||
            (total == null &&
                hasPreviousPage === false &&
                hasNextPage === false &&
                // @ts-ignore FIXME total may be undefined when using partial pagination but the ListControllerResult type is wrong about it
                data.length === 0)) &&
        // the user didn't set any filters
        !Object.keys(filterValues).length &&
        // there is an empty page component
        empty !== false;
    return (React.createElement(Root, { className: (0, clsx_1.default)('list-page', className), ...rest },
        title !== false && (React.createElement(Title_1.Title, { title: title, defaultTitle: defaultTitle, preferenceKey: `${resource}.list.title` })),
        shouldRenderEmptyPage ? renderEmpty() : renderList(),
        aside));
};
exports.ListView = ListView;
const PREFIX = 'RaList';
exports.ListClasses = {
    main: `${PREFIX}-main`,
    content: `${PREFIX}-content`,
    actions: `${PREFIX}-actions`,
    noActions: `${PREFIX}-noActions`,
    noResults: `${PREFIX}-noResults`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    [`& .${exports.ListClasses.main}`]: {
        flex: '1 1 auto',
        display: 'flex',
        flexDirection: 'column',
    },
    [`& .${exports.ListClasses.content}`]: {
        position: 'relative',
        [theme.breakpoints.down('sm')]: {
            boxShadow: 'none',
        },
        overflow: 'inherit',
    },
    [`& .${exports.ListClasses.noActions}`]: {
        marginTop: '1em',
    },
    [`& .${exports.ListClasses.actions}`]: {},
    [`& .${exports.ListClasses.noResults}`]: {
        flex: 1,
    },
}));
