// ra-ui-materialui/dist/list/ListToolbar.js
import * as React from 'react';
import { memo } from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { FilterContext } from 'ra-core';
import { FilterForm } from "./filter/index.js";
export const ListToolbar = memo(inProps => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { filters, actions, className, ...rest } = props;
    return Array.isArray(filters) ? (React.createElement(FilterContext.Provider, { value: filters },
        React.createElement(Root, { className: className },
            React.createElement(FilterForm, null),
            React.createElement("span", null),
            actions))) : (React.createElement(Root, { className: className },
        filters &&
            React.cloneElement(filters, {
                ...rest,
                context: 'form',
            }),
        React.createElement("span", null),
        actions &&
            React.cloneElement(actions, {
                ...rest,
                filters,
                ...actions.props,
            })));
});
const PREFIX = 'RaListToolbar';
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    position: 'relative',
    justifyContent: 'space-between',
    alignItems: 'flex-end',
    width: '100%',
    [theme.breakpoints.down('md')]: {
        flexWrap: 'wrap',
    },
    [theme.breakpoints.down('sm')]: {
        backgroundColor: (theme.vars || theme).palette.background.paper,
        flexWrap: 'inherit',
        flexDirection: 'column-reverse',
    },
}));
