// ra-ui-materialui/dist/list/listFieldTypes.js
import * as React from 'react';
import { DataTable } from "./datatable/index.js";
import { SingleFieldList } from "./SingleFieldList.js";
import { ArrayField, BooleanField, ChipField, DateField, EmailField, ReferenceField, ReferenceArrayField, UrlField, TextField, TextArrayField, } from "../field/index.js";
export const listFieldTypes = {
    table: {
        component: props => {
            return React.createElement(DataTable, { ...props });
        },
        representation: (_props, children) => `        <DataTable>
${children.map(child => `            ${child.getRepresentation()}`).join('\n')}
        </DataTable>`,
    },
    array: {
        component: ({ children, ...props }) => {
            const childrenArray = React.Children.toArray(children);
            return (React.createElement(DataTable.Col, { ...props },
                React.createElement(ArrayField, { ...props },
                    React.createElement(SingleFieldList, null,
                        React.createElement(ChipField, { source: childrenArray.length > 0 &&
                                React.isValidElement(childrenArray[0]) &&
                                childrenArray[0].props.source })))));
        },
        representation: (props, children) => `<DataTable.Col source="${props.source}">
                <ArrayField source="${props.source}">
                    <SingleFieldList>
                        <ChipField source="${children.length > 0 && children[0].getProps().source}" />
                    </SingleFieldList>
                </ArrayField>
            </DataTable.Col>`,
    },
    scalar_array: {
        component: (props) => {
            return (React.createElement(DataTable.Col, { ...props },
                React.createElement(TextArrayField, { ...props })));
        },
        representation: props => `<DataTable.Col source="${props.source}">
                <TextArrayField source="${props.source}" />
            </DataTable.Col>`,
    },
    boolean: {
        component: props => (React.createElement(DataTable.Col, { ...props },
            React.createElement(BooleanField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <BooleanField source="${props.source}" />
            </DataTable.Col>`,
    },
    date: {
        component: props => (React.createElement(DataTable.Col, { ...props },
            React.createElement(DateField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <DateField source="${props.source}" />
            </DataTable.Col>`,
    },
    email: {
        component: props => (React.createElement(DataTable.Col, { ...props },
            React.createElement(EmailField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <EmailField source="${props.source}" />
            </DataTable.Col>`,
    },
    id: {
        component: props => React.createElement(DataTable.Col, { ...props }),
        representation: props => `<DataTable.Col source="${props.source}" />`,
    },
    number: {
        component: DataTable.NumberCol,
        representation: props => `<DataTable.NumberCol source="${props.source}" />`,
    },
    reference: {
        component: props => (React.createElement(DataTable.Col, { ...props },
            React.createElement(ReferenceField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <ReferenceField source="${props.source}" reference="${props.reference}" />
            </DataTable.Col>`,
    },
    referenceChild: {
        component: () => React.createElement(TextField, { source: "id" }),
        representation: () => `<TextField source="id" />`,
    },
    referenceArray: {
        component: props => (React.createElement(DataTable.Col, { ...props },
            React.createElement(ReferenceArrayField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <ReferenceArrayField source="${props.source}" reference="${props.reference}" />
            </DataTable.Col>`,
    },
    referenceArrayChild: {
        component: () => (React.createElement(SingleFieldList, null,
            React.createElement(ChipField, { source: "id" }))),
        representation: () => `<SingleFieldList>
                <ChipField source="id" />
            </SingleFieldList>`,
    },
    richText: undefined, // never display a rich text field in a datagrid
    string: {
        component: DataTable.Col,
        representation: props => `<DataTable.Col source="${props.source}" />`,
    },
    url: {
        component: props => (React.createElement(DataTable.Col, { ...props },
            React.createElement(UrlField, { ...props }))),
        representation: props => `<DataTable.Col source="${props.source}">
                <UrlField source="${props.source}" />
            </DataTable.Col>`,
    },
};
