// ra-ui-materialui/dist/list/ListGuesser.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListGuesserView = exports.ListGuesser = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const Inbox_1 = __importDefault(require("@mui/icons-material/Inbox"));
const ra_core_1 = require("ra-core");
const ListView_1 = require("./ListView.cjs");
const listFieldTypes_1 = require("./listFieldTypes.cjs");
const inflection_1 = require("inflection");
/**
 * List component rendering a <DataTable> based on the result of the
 * dataProvider.getList() call.
 *
 * The result (choice and type of columns) isn't configurable, but the
 * <ListGuesser> outputs the <DataTable> it has guessed to the console so that
 * developers can start from there.
 *
 * To be used as the list prop of a <Resource>.
 *
 * @example
 *
 * import { Admin, Resource, ListGuesser } from 'react-admin';
 *
 * const App = () => (
 *     <Admin dataProvider={myDataProvider}>
 *         <Resource name="posts" list={ListGuesser} />
 *     </Admin>
 * );
 */
const ListGuesser = (props) => {
    const { debounce, disableAuthentication, disableSyncWithLocation, empty, exporter, filter, filterDefaultValues, perPage, queryOptions, resource, sort, ...rest } = props;
    // force a rerender of this component when any list parameter changes
    // otherwise the ListBase won't be rerendered when the sort changes
    // and the following check won't be performed
    (0, ra_core_1.useLocation)();
    // keep previous data, unless the resource changes
    const resourceFromContext = (0, ra_core_1.useResourceContext)(props);
    const previousResource = (0, ra_core_1.usePrevious)(resourceFromContext);
    const keepPreviousData = previousResource === resourceFromContext;
    return (React.createElement(ra_core_1.ListBase, { debounce: debounce, disableAuthentication: disableAuthentication, disableSyncWithLocation: disableSyncWithLocation, empty: empty === undefined ? React.createElement(ListGuesserEmpty, null) : empty, exporter: exporter, filter: filter, filterDefaultValues: filterDefaultValues, perPage: perPage, queryOptions: {
            placeholderData: previousData => keepPreviousData ? previousData : undefined,
        }, resource: resource, sort: sort },
        React.createElement(exports.ListGuesserView, { ...rest })));
};
exports.ListGuesser = ListGuesser;
const ListGuesserView = (props) => {
    const { data } = (0, ra_core_1.useListContext)();
    const resource = (0, ra_core_1.useResourceContext)();
    const [child, setChild] = (0, react_1.useState)(null);
    const { enableLog = process.env.NODE_ENV === 'development', ...rest } = props;
    (0, react_1.useEffect)(() => {
        setChild(null);
    }, [resource]);
    (0, react_1.useEffect)(() => {
        if (data && data.length > 0 && !child) {
            const inferredElements = (0, ra_core_1.getElementsFromRecords)(data, listFieldTypes_1.listFieldTypes);
            const inferredChild = new ra_core_1.InferredElement(listFieldTypes_1.listFieldTypes.table, null, inferredElements);
            const inferredChildElement = inferredChild.getElement();
            const representation = inferredChild.getRepresentation();
            if (!resource) {
                throw new Error('Cannot use <ListGuesser> outside of a ResourceContext');
            }
            if (!inferredChildElement || !representation) {
                return;
            }
            setChild(inferredChildElement);
            const components = ['List']
                .concat(Array.from(new Set(Array.from(representation.matchAll(/<([^/\s>]+)/g))
                .map(match => match[1])
                .filter(component => component !== 'span'))))
                .filter(component => !component.match(/[A-Za-z]+\.[A-Za-z]+/i))
                .sort();
            if (enableLog) {
                console.log(`Guessed List:

import { ${components.join(', ')} } from 'react-admin';

export const ${(0, inflection_1.capitalize)((0, inflection_1.singularize)(resource))}List = () => (
    <List>
${inferredChild.getRepresentation()}
    </List>
);`);
            }
        }
    }, [data, child, resource, enableLog]);
    return React.createElement(ListView_1.ListView, { ...rest }, child);
};
exports.ListGuesserView = ListGuesserView;
const ListGuesserEmpty = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: LIST_GUESSER_EMPTY_PREFIX,
    });
    const { className } = props;
    const translate = (0, ra_core_1.useTranslate)();
    return (React.createElement(ListGuesserEmptyRoot, { className: className },
        React.createElement(material_1.Box, { className: ListGuesserEmptyClasses.message },
            React.createElement(Inbox_1.default, { className: ListGuesserEmptyClasses.icon }),
            React.createElement(material_1.Typography, { variant: "h4", paragraph: true }, translate('ra.guesser.empty.title', {
                _: 'No data to display',
            })),
            React.createElement(material_1.Typography, { variant: "body1" }, translate('ra.guesser.empty.message', {
                _: 'Please check your data provider',
            })))));
};
const LIST_GUESSER_EMPTY_PREFIX = 'RaListGuesserEmpty';
const ListGuesserEmptyClasses = {
    message: `${LIST_GUESSER_EMPTY_PREFIX}-message`,
    icon: `${LIST_GUESSER_EMPTY_PREFIX}-icon`,
};
const ListGuesserEmptyRoot = (0, styles_1.styled)('span', {
    name: LIST_GUESSER_EMPTY_PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flex: 1,
    [`& .${ListGuesserEmptyClasses.message}`]: {
        textAlign: 'center',
        margin: '0 1em',
        color: (theme.vars || theme).palette.text.disabled,
    },
    [`& .${ListGuesserEmptyClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
}));
