// ra-ui-materialui/dist/list/ListView.js
import * as React from 'react';
import { styled, } from '@mui/material/styles';
import Card from '@mui/material/Card';
import clsx from 'clsx';
import { useListContext } from 'ra-core';
import { Title } from "../layout/Title.js";
import { ListToolbar } from "./ListToolbar.js";
import { Pagination as DefaultPagination } from "./pagination/index.js";
import { ListActions as DefaultActions } from "./ListActions.js";
import { Empty } from "./Empty.js";
import { Offline } from "../Offline.js";
const defaultActions = React.createElement(DefaultActions, null);
const defaultPagination = React.createElement(DefaultPagination, null);
const defaultEmpty = React.createElement(Empty, null);
const DefaultComponent = Card;
const defaultOffline = React.createElement(Offline, null);
export const ListView = (props) => {
    const { actions = defaultActions, aside, filters, emptyWhileLoading, pagination = defaultPagination, children, className, component: Content = DefaultComponent, title, empty = defaultEmpty, render, offline = defaultOffline, error, ...rest } = props;
    const listContext = useListContext();
    const { defaultTitle, data, error: errorState, isPaused, isPending, isPlaceholderData, filterValues, resource, total, hasNextPage, hasPreviousPage, } = listContext;
    const showOffline = isPaused &&
        (isPending || isPlaceholderData) &&
        offline !== false &&
        offline !== undefined;
    const showError = errorState && error !== false && error !== undefined;
    if ((!children && !render) ||
        (!data && isPending && !isPaused && emptyWhileLoading)) {
        return null;
    }
    const renderList = () => {
        return (React.createElement("div", { className: clsx(ListClasses.main, {
                [ListClasses.noActions]: !(filters || actions) || showOffline || showError,
            }) },
            filters || actions ? (React.createElement(ListToolbar, { className: ListClasses.actions, filters: filters, actions: actions })) : null,
            React.createElement(Content, { className: ListClasses.content }, showOffline
                ? offline
                : showError
                    ? error
                    : render
                        ? render(listContext)
                        : children),
            pagination !== false ? pagination : null));
    };
    const renderEmpty = () => empty !== false && React.createElement("div", { className: ListClasses.noResults }, empty);
    const shouldRenderEmptyPage = !errorState &&
        // the list is not loading data for the first time
        !isPending &&
        // the API returned no data (using either normal or partial pagination)
        (total === 0 ||
            (total == null &&
                hasPreviousPage === false &&
                hasNextPage === false &&
                // @ts-ignore FIXME total may be undefined when using partial pagination but the ListControllerResult type is wrong about it
                data.length === 0)) &&
        // the user didn't set any filters
        !Object.keys(filterValues).length &&
        // there is an empty page component
        empty !== false;
    return (React.createElement(Root, { className: clsx('list-page', className), ...rest },
        title !== false && (React.createElement(Title, { title: title, defaultTitle: defaultTitle, preferenceKey: `${resource}.list.title` })),
        shouldRenderEmptyPage ? renderEmpty() : renderList(),
        aside));
};
const PREFIX = 'RaList';
export const ListClasses = {
    main: `${PREFIX}-main`,
    content: `${PREFIX}-content`,
    actions: `${PREFIX}-actions`,
    noActions: `${PREFIX}-noActions`,
    noResults: `${PREFIX}-noResults`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    [`& .${ListClasses.main}`]: {
        flex: '1 1 auto',
        display: 'flex',
        flexDirection: 'column',
    },
    [`& .${ListClasses.content}`]: {
        position: 'relative',
        [theme.breakpoints.down('sm')]: {
            boxShadow: 'none',
        },
        overflow: 'inherit',
    },
    [`& .${ListClasses.noActions}`]: {
        marginTop: '1em',
    },
    [`& .${ListClasses.actions}`]: {},
    [`& .${ListClasses.noResults}`]: {
        flex: 1,
    },
}));
