// ra-ui-materialui/dist/list/SingleFieldList.js
import * as React from 'react';
import { Chip, Stack, styled, useThemeProps, } from '@mui/material';
import { sanitizeListRestProps, useListContextWithProps, useResourceContext, RecordRepresentation, useCreatePath, RecordContextProvider, } from 'ra-core';
import { LinearProgress } from "../layout/LinearProgress.js";
import { Link } from "../Link.js";
/**
 * Iterator component to be used to display a list of entities, using a single field
 *
 * @example Display all the orders by the current customer as a list of chips
 * <ReferenceManyField reference="orders" target="customer_id">
 *     <SingleFieldList />
 * </ReferenceManyField>

 * @example Choose the field to be used as text label
 * <ReferenceManyField reference="orders" target="customer_id">
 *     <SingleFieldList>
 *         <ChipField source="reference" />
 *     </SingleFieldList>
 * </ReferenceManyField>
 *
 * @example Customize the link type
 * // By default, it includes a link to the <Edit> page of the related record
 * // (`/orders/:id` in the previous example).
 * // Set the linkType prop to "show" to link to the <Show> page instead.
 * <ReferenceManyField reference="books" target="author_id">
 *     <SingleFieldList linkType="show">
 *         <ChipField source="title" />
 *     </SingleFieldList>
 * </ReferenceManyField>
 *
 * @example Disable the link
 * // You can also prevent `<SingleFieldList>` from adding link to children by
 * // setting `linkType` to false.
 * <ReferenceManyField reference="books" target="author_id">
 *     <SingleFieldList linkType={false}>
 *         <ChipField source="title" />
 *     </SingleFieldList>
 * </ReferenceManyField>
 */
export const SingleFieldList = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, children, empty, linkType = 'edit', gap = 1, direction = 'row', ...rest } = props;
    const { data, total, isPending } = useListContextWithProps(props);
    const resource = useResourceContext(props);
    const createPath = useCreatePath();
    if (isPending === true) {
        return React.createElement(LinearProgress, null);
    }
    if (data == null || data.length === 0 || total === 0) {
        return empty ?? null;
    }
    return (React.createElement(Root, { gap: gap, direction: direction, className: className, ...sanitizeListRestProps(rest) }, data.map((record, rowIndex) => {
        const resourceLinkPath = !linkType
            ? false
            : createPath({
                resource,
                type: linkType,
                id: record.id,
            });
        if (resourceLinkPath) {
            return (React.createElement(RecordContextProvider, { value: record, key: record.id ?? `row${rowIndex}` },
                React.createElement(Link, { className: SingleFieldListClasses.link, to: resourceLinkPath, onClick: stopPropagation }, children || (React.createElement(DefaultChildComponent, { clickable: true })))));
        }
        return (React.createElement(RecordContextProvider, { value: record, key: record.id ?? `row${rowIndex}` }, children || React.createElement(DefaultChildComponent, null)));
    })));
};
const PREFIX = 'RaSingleFieldList';
export const SingleFieldListClasses = {
    link: `${PREFIX}-link`,
};
const Root = styled(Stack, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flexWrap: 'wrap',
    [`& .${SingleFieldListClasses.link}`]: {
        textDecoration: 'none',
        '& > *': {
            color: (theme.vars || theme).palette.primary.main,
        },
    },
}));
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
const DefaultChildComponent = ({ clickable }) => (React.createElement(Chip, { sx: { cursor: 'inherit' }, size: "small", label: React.createElement(RecordRepresentation, null), clickable: clickable }));
