// ra-ui-materialui/dist/list/SingleFieldList.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingleFieldListClasses = exports.SingleFieldList = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const LinearProgress_1 = require("../layout/LinearProgress.cjs");
const Link_1 = require("../Link.cjs");
/**
 * Iterator component to be used to display a list of entities, using a single field
 *
 * @example Display all the orders by the current customer as a list of chips
 * <ReferenceManyField reference="orders" target="customer_id">
 *     <SingleFieldList />
 * </ReferenceManyField>

 * @example Choose the field to be used as text label
 * <ReferenceManyField reference="orders" target="customer_id">
 *     <SingleFieldList>
 *         <ChipField source="reference" />
 *     </SingleFieldList>
 * </ReferenceManyField>
 *
 * @example Customize the link type
 * // By default, it includes a link to the <Edit> page of the related record
 * // (`/orders/:id` in the previous example).
 * // Set the linkType prop to "show" to link to the <Show> page instead.
 * <ReferenceManyField reference="books" target="author_id">
 *     <SingleFieldList linkType="show">
 *         <ChipField source="title" />
 *     </SingleFieldList>
 * </ReferenceManyField>
 *
 * @example Disable the link
 * // You can also prevent `<SingleFieldList>` from adding link to children by
 * // setting `linkType` to false.
 * <ReferenceManyField reference="books" target="author_id">
 *     <SingleFieldList linkType={false}>
 *         <ChipField source="title" />
 *     </SingleFieldList>
 * </ReferenceManyField>
 */
const SingleFieldList = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, children, empty, linkType = 'edit', gap = 1, direction = 'row', ...rest } = props;
    const { data, total, isPending } = (0, ra_core_1.useListContextWithProps)(props);
    const resource = (0, ra_core_1.useResourceContext)(props);
    const createPath = (0, ra_core_1.useCreatePath)();
    if (isPending === true) {
        return React.createElement(LinearProgress_1.LinearProgress, null);
    }
    if (data == null || data.length === 0 || total === 0) {
        return empty ?? null;
    }
    return (React.createElement(Root, { gap: gap, direction: direction, className: className, ...(0, ra_core_1.sanitizeListRestProps)(rest) }, data.map((record, rowIndex) => {
        const resourceLinkPath = !linkType
            ? false
            : createPath({
                resource,
                type: linkType,
                id: record.id,
            });
        if (resourceLinkPath) {
            return (React.createElement(ra_core_1.RecordContextProvider, { value: record, key: record.id ?? `row${rowIndex}` },
                React.createElement(Link_1.Link, { className: exports.SingleFieldListClasses.link, to: resourceLinkPath, onClick: stopPropagation }, children || (React.createElement(DefaultChildComponent, { clickable: true })))));
        }
        return (React.createElement(ra_core_1.RecordContextProvider, { value: record, key: record.id ?? `row${rowIndex}` }, children || React.createElement(DefaultChildComponent, null)));
    })));
};
exports.SingleFieldList = SingleFieldList;
const PREFIX = 'RaSingleFieldList';
exports.SingleFieldListClasses = {
    link: `${PREFIX}-link`,
};
const Root = (0, material_1.styled)(material_1.Stack, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flexWrap: 'wrap',
    [`& .${exports.SingleFieldListClasses.link}`]: {
        textDecoration: 'none',
        '& > *': {
            color: (theme.vars || theme).palette.primary.main,
        },
    },
}));
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
const DefaultChildComponent = ({ clickable }) => (React.createElement(material_1.Chip, { sx: { cursor: 'inherit' }, size: "small", label: React.createElement(ra_core_1.RecordRepresentation, null), clickable: clickable }));
