// ra-ui-materialui/dist/list/SimpleList/SimpleList.js
import { Avatar, Box, List, ListItemAvatar, ListItemIcon, ListItemText, } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import { RecordsIterator, sanitizeListRestProps, useGetRecordRepresentation, useRecordContext, useResourceContext, useTranslate, WithListContext, } from 'ra-core';
import * as React from 'react';
import { isValidElement } from 'react';
import { ListNoResults } from "../ListNoResults.js";
import { SimpleListLoading } from "./SimpleListLoading.js";
import { SimpleListItem, } from "./SimpleListItem.js";
/**
 * The <SimpleList> component renders a list of records as a Material UI <List>.
 * It is usually used as a child of react-admin's <List> and <ReferenceManyField> components.
 *
 * Also widely used on Mobile.
 *
 * Props:
 * - primaryText: function returning a React element (or some text) based on the record
 * - secondaryText: same
 * - tertiaryText: same
 * - leftAvatar: function returning a React element based on the record
 * - leftIcon: same
 * - rightAvatar: same
 * - rightIcon: same
 * - linkType: deprecated - 'edit' or 'show', or a function returning 'edit' or 'show' based on the record
 * - rowClick: The action to trigger when the user clicks on a row.
 * - rowStyle: function returning a style object based on (record, index)
 * - rowSx: function returning a sx object based on (record, index)
 *
 * @example // Display all posts as a List
 * const postRowSx = (record, index) => ({
 *     backgroundColor: record.views >= 500 ? '#efe' : 'white',
 * });
 * export const PostList = () => (
 *     <List>
 *         <SimpleList
 *             primaryText={record => record.title}
 *             secondaryText={record => `${record.views} views`}
 *             tertiaryText={record =>
 *                 new Date(record.published_at).toLocaleDateString()
 *             }
 *             rowSx={postRowSx}
 *          />
 *     </List>
 * );
 */
export const SimpleList = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, empty = DefaultEmpty, hasBulkActions, leftAvatar, leftIcon, linkType, rowClick, primaryText, rightAvatar, rightIcon, secondaryText, tertiaryText, ref, rowSx, rowStyle, resource, ...rest } = props;
    return (React.createElement(WithListContext, { ...props, loading: React.createElement(SimpleListLoading, { className: className, hasLeftAvatarOrIcon: !!leftIcon || !!leftAvatar, hasRightAvatarOrIcon: !!rightIcon || !!rightAvatar, hasSecondaryText: !!secondaryText, hasTertiaryText: !!tertiaryText }), empty: empty ?? null, 
        // We need to keep passing data explicitly as it may have been passed down through props.
        render: ({ data }) => (React.createElement(Root, { className: className, ...sanitizeListRestProps(rest) },
            React.createElement(RecordsIterator, { data: data, render: (record, rowIndex) => (React.createElement(SimpleListItem, { key: record.id, rowIndex: rowIndex, linkType: linkType, rowClick: rowClick, rowSx: rowSx, rowStyle: rowStyle, resource: resource },
                    React.createElement(SimpleListItemContent, { leftAvatar: leftAvatar, leftIcon: leftIcon, primaryText: primaryText, rightAvatar: rightAvatar, rightIcon: rightIcon, secondaryText: secondaryText, tertiaryText: tertiaryText, rowIndex: rowIndex }))) }))) }));
};
const SimpleListItemContent = (props) => {
    const { leftAvatar, leftIcon, primaryText, rightAvatar, rightIcon, secondaryText, tertiaryText, } = props;
    const resource = useResourceContext(props);
    const record = useRecordContext(props);
    const getRecordRepresentation = useGetRecordRepresentation(resource);
    const translate = useTranslate();
    const renderAvatar = (record, avatarCallback) => {
        const avatarValue = avatarCallback(record, record.id);
        if (typeof avatarValue === 'string' &&
            (avatarValue.startsWith('http') || avatarValue.startsWith('data:'))) {
            return React.createElement(Avatar, { src: avatarValue });
        }
        else {
            return React.createElement(Avatar, null, avatarValue);
        }
    };
    if (!record)
        return null;
    return (React.createElement(React.Fragment, null,
        leftIcon && (React.createElement(ListItemIcon, null, leftIcon(record, record.id))),
        leftAvatar && (React.createElement(ListItemAvatar, null, renderAvatar(record, leftAvatar))),
        React.createElement(ListItemText, { primary: React.createElement("div", null,
                primaryText
                    ? typeof primaryText === 'string'
                        ? translate(primaryText, {
                            ...record,
                            _: primaryText,
                        })
                        : isValidElement(primaryText)
                            ? primaryText
                            : // @ts-ignore
                                primaryText(record, record.id)
                    : getRecordRepresentation(record),
                !!tertiaryText &&
                    (isValidElement(tertiaryText) ? (tertiaryText) : (React.createElement("span", { className: SimpleListClasses.tertiary }, typeof tertiaryText === 'string'
                        ? translate(tertiaryText, {
                            ...record,
                            _: tertiaryText,
                        })
                        : isValidElement(tertiaryText)
                            ? tertiaryText
                            : // @ts-ignore
                                tertiaryText(record, record.id))))), secondary: !!secondaryText &&
                (typeof secondaryText === 'string'
                    ? translate(secondaryText, {
                        ...record,
                        _: secondaryText,
                    })
                    : isValidElement(secondaryText)
                        ? secondaryText
                        : // @ts-ignore
                            secondaryText(record, record.id)) }),
        (rightAvatar || rightIcon) && (React.createElement(Box, null,
            rightAvatar && (React.createElement(Avatar, null, renderAvatar(record, rightAvatar))),
            rightIcon && (React.createElement(ListItemIcon, null, rightIcon(record, record.id)))))));
};
const PREFIX = 'RaSimpleList';
export const SimpleListClasses = {
    tertiary: `${PREFIX}-tertiary`,
};
const Root = styled(List, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${SimpleListClasses.tertiary}`]: { float: 'right', opacity: 0.541176 },
});
const DefaultEmpty = React.createElement(ListNoResults, null);
