// ra-ui-materialui/dist/list/SimpleList/SimpleList.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleListClasses = exports.SimpleList = void 0;
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const React = __importStar(require("react"));
const react_1 = require("react");
const ListNoResults_1 = require("../ListNoResults.cjs");
const SimpleListLoading_1 = require("./SimpleListLoading.cjs");
const SimpleListItem_1 = require("./SimpleListItem.cjs");
/**
 * The <SimpleList> component renders a list of records as a Material UI <List>.
 * It is usually used as a child of react-admin's <List> and <ReferenceManyField> components.
 *
 * Also widely used on Mobile.
 *
 * Props:
 * - primaryText: function returning a React element (or some text) based on the record
 * - secondaryText: same
 * - tertiaryText: same
 * - leftAvatar: function returning a React element based on the record
 * - leftIcon: same
 * - rightAvatar: same
 * - rightIcon: same
 * - linkType: deprecated - 'edit' or 'show', or a function returning 'edit' or 'show' based on the record
 * - rowClick: The action to trigger when the user clicks on a row.
 * - rowStyle: function returning a style object based on (record, index)
 * - rowSx: function returning a sx object based on (record, index)
 *
 * @example // Display all posts as a List
 * const postRowSx = (record, index) => ({
 *     backgroundColor: record.views >= 500 ? '#efe' : 'white',
 * });
 * export const PostList = () => (
 *     <List>
 *         <SimpleList
 *             primaryText={record => record.title}
 *             secondaryText={record => `${record.views} views`}
 *             tertiaryText={record =>
 *                 new Date(record.published_at).toLocaleDateString()
 *             }
 *             rowSx={postRowSx}
 *          />
 *     </List>
 * );
 */
const SimpleList = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, empty = DefaultEmpty, hasBulkActions, leftAvatar, leftIcon, linkType, rowClick, primaryText, rightAvatar, rightIcon, secondaryText, tertiaryText, ref, rowSx, rowStyle, resource, ...rest } = props;
    return (React.createElement(ra_core_1.WithListContext, { ...props, loading: React.createElement(SimpleListLoading_1.SimpleListLoading, { className: className, hasLeftAvatarOrIcon: !!leftIcon || !!leftAvatar, hasRightAvatarOrIcon: !!rightIcon || !!rightAvatar, hasSecondaryText: !!secondaryText, hasTertiaryText: !!tertiaryText }), empty: empty ?? null, 
        // We need to keep passing data explicitly as it may have been passed down through props.
        render: ({ data }) => (React.createElement(Root, { className: className, ...(0, ra_core_1.sanitizeListRestProps)(rest) },
            React.createElement(ra_core_1.RecordsIterator, { data: data, render: (record, rowIndex) => (React.createElement(SimpleListItem_1.SimpleListItem, { key: record.id, rowIndex: rowIndex, linkType: linkType, rowClick: rowClick, rowSx: rowSx, rowStyle: rowStyle, resource: resource },
                    React.createElement(SimpleListItemContent, { leftAvatar: leftAvatar, leftIcon: leftIcon, primaryText: primaryText, rightAvatar: rightAvatar, rightIcon: rightIcon, secondaryText: secondaryText, tertiaryText: tertiaryText, rowIndex: rowIndex }))) }))) }));
};
exports.SimpleList = SimpleList;
const SimpleListItemContent = (props) => {
    const { leftAvatar, leftIcon, primaryText, rightAvatar, rightIcon, secondaryText, tertiaryText, } = props;
    const resource = (0, ra_core_1.useResourceContext)(props);
    const record = (0, ra_core_1.useRecordContext)(props);
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    const translate = (0, ra_core_1.useTranslate)();
    const renderAvatar = (record, avatarCallback) => {
        const avatarValue = avatarCallback(record, record.id);
        if (typeof avatarValue === 'string' &&
            (avatarValue.startsWith('http') || avatarValue.startsWith('data:'))) {
            return React.createElement(material_1.Avatar, { src: avatarValue });
        }
        else {
            return React.createElement(material_1.Avatar, null, avatarValue);
        }
    };
    if (!record)
        return null;
    return (React.createElement(React.Fragment, null,
        leftIcon && (React.createElement(material_1.ListItemIcon, null, leftIcon(record, record.id))),
        leftAvatar && (React.createElement(material_1.ListItemAvatar, null, renderAvatar(record, leftAvatar))),
        React.createElement(material_1.ListItemText, { primary: React.createElement("div", null,
                primaryText
                    ? typeof primaryText === 'string'
                        ? translate(primaryText, {
                            ...record,
                            _: primaryText,
                        })
                        : (0, react_1.isValidElement)(primaryText)
                            ? primaryText
                            : // @ts-ignore
                                primaryText(record, record.id)
                    : getRecordRepresentation(record),
                !!tertiaryText &&
                    ((0, react_1.isValidElement)(tertiaryText) ? (tertiaryText) : (React.createElement("span", { className: exports.SimpleListClasses.tertiary }, typeof tertiaryText === 'string'
                        ? translate(tertiaryText, {
                            ...record,
                            _: tertiaryText,
                        })
                        : (0, react_1.isValidElement)(tertiaryText)
                            ? tertiaryText
                            : // @ts-ignore
                                tertiaryText(record, record.id))))), secondary: !!secondaryText &&
                (typeof secondaryText === 'string'
                    ? translate(secondaryText, {
                        ...record,
                        _: secondaryText,
                    })
                    : (0, react_1.isValidElement)(secondaryText)
                        ? secondaryText
                        : // @ts-ignore
                            secondaryText(record, record.id)) }),
        (rightAvatar || rightIcon) && (React.createElement(material_1.Box, null,
            rightAvatar && (React.createElement(material_1.Avatar, null, renderAvatar(record, rightAvatar))),
            rightIcon && (React.createElement(material_1.ListItemIcon, null, rightIcon(record, record.id)))))));
};
const PREFIX = 'RaSimpleList';
exports.SimpleListClasses = {
    tertiary: `${PREFIX}-tertiary`,
};
const Root = (0, styles_1.styled)(material_1.List, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${exports.SimpleListClasses.tertiary}`]: { float: 'right', opacity: 0.541176 },
});
const DefaultEmpty = React.createElement(ListNoResults_1.ListNoResults, null);
