// ra-ui-materialui/dist/list/SimpleList/SimpleListLoading.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import Avatar from '@mui/material/Avatar';
import { List, ListItem, ListItemAvatar, ListItemText, } from '@mui/material';
import { useTimeout } from 'ra-core';
import { Placeholder } from "../Placeholder.js";
export const SimpleListLoading = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, hasLeftAvatarOrIcon, hasRightAvatarOrIcon, hasSecondaryText, hasTertiaryText, nbFakeLines = 5, ...rest } = props;
    const oneSecondHasPassed = useTimeout(1000);
    return oneSecondHasPassed ? (React.createElement(StyledList, { className: className, ...rest }, times(nbFakeLines, key => (React.createElement(ListItem, { key: key, secondaryAction: hasRightAvatarOrIcon ? React.createElement(Avatar, null, "\u00A0") : null },
        hasLeftAvatarOrIcon && (React.createElement(ListItemAvatar, null,
            React.createElement(Avatar, null, "\u00A0"))),
        React.createElement(ListItemText, { primary: React.createElement("div", null,
                React.createElement(Placeholder, { className: SimpleListLoadingClasses.primary }),
                hasTertiaryText && (React.createElement("span", { className: SimpleListLoadingClasses.tertiary },
                    React.createElement(Placeholder, null)))), secondary: hasSecondaryText ? React.createElement(Placeholder, null) : undefined })))))) : null;
};
const PREFIX = 'RaSimpleListLoading';
export const SimpleListLoadingClasses = {
    primary: `${PREFIX}-primary`,
    tertiary: `${PREFIX}-tertiary`,
};
const StyledList = styled(List, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${SimpleListLoadingClasses.primary}`]: {
        width: '30vw',
        display: 'inline-block',
        marginBottom: theme.spacing(),
    },
    [`& .${SimpleListLoadingClasses.tertiary}`]: {
        float: 'right',
        opacity: 0.541176,
        minWidth: '10vw',
    },
}));
const times = (nbChildren, fn) => Array.from({ length: nbChildren }, (_, key) => fn(key));
