// ra-ui-materialui/dist/list/SimpleList/SimpleListLoading.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleListLoadingClasses = exports.SimpleListLoading = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const Avatar_1 = __importDefault(require("@mui/material/Avatar"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const Placeholder_1 = require("../Placeholder.cjs");
const SimpleListLoading = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, hasLeftAvatarOrIcon, hasRightAvatarOrIcon, hasSecondaryText, hasTertiaryText, nbFakeLines = 5, ...rest } = props;
    const oneSecondHasPassed = (0, ra_core_1.useTimeout)(1000);
    return oneSecondHasPassed ? (React.createElement(StyledList, { className: className, ...rest }, times(nbFakeLines, key => (React.createElement(material_1.ListItem, { key: key, secondaryAction: hasRightAvatarOrIcon ? React.createElement(Avatar_1.default, null, "\u00A0") : null },
        hasLeftAvatarOrIcon && (React.createElement(material_1.ListItemAvatar, null,
            React.createElement(Avatar_1.default, null, "\u00A0"))),
        React.createElement(material_1.ListItemText, { primary: React.createElement("div", null,
                React.createElement(Placeholder_1.Placeholder, { className: exports.SimpleListLoadingClasses.primary }),
                hasTertiaryText && (React.createElement("span", { className: exports.SimpleListLoadingClasses.tertiary },
                    React.createElement(Placeholder_1.Placeholder, null)))), secondary: hasSecondaryText ? React.createElement(Placeholder_1.Placeholder, null) : undefined })))))) : null;
};
exports.SimpleListLoading = SimpleListLoading;
const PREFIX = 'RaSimpleListLoading';
exports.SimpleListLoadingClasses = {
    primary: `${PREFIX}-primary`,
    tertiary: `${PREFIX}-tertiary`,
};
const StyledList = (0, styles_1.styled)(material_1.List, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.SimpleListLoadingClasses.primary}`]: {
        width: '30vw',
        display: 'inline-block',
        marginBottom: theme.spacing(),
    },
    [`& .${exports.SimpleListLoadingClasses.tertiary}`]: {
        float: 'right',
        opacity: 0.541176,
        minWidth: '10vw',
    },
}));
const times = (nbChildren, fn) => Array.from({ length: nbChildren }, (_, key) => fn(key));
