// ra-core/dist/form/groups/FormGroupsProvider.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormGroupsProvider = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const FormGroupsContext_1 = require("./FormGroupsContext.cjs");
/**
 * This component provides functions through context to manage form groups,
 * allowing to link or unlink an input to a group.
 * @see FormGroupContextProvider
 * @see useFormGroup
 * @see useFormGroups
 */
const FormGroupsProvider = ({ children }) => {
    const formGroups = (0, react_1.useRef)({});
    const subscribers = (0, react_1.useRef)({});
    const formContextValue = (0, react_1.useMemo)(() => ({
        /**
         * Register a subscriber function for the specified group. The subscriber
         * will be called whenever the group content changes (fields added or removed).
         */
        subscribe: (group, subscriber) => {
            if (!subscribers.current[group]) {
                subscribers.current[group] = [];
            }
            subscribers.current[group].push(subscriber);
            return () => {
                subscribers.current[group] = subscribers.current[group].filter(s => s !== subscriber);
            };
        },
        getGroupFields: name => formGroups.current[name] || [],
        registerGroup: name => {
            formGroups.current[name] = formGroups.current[name] || [];
        },
        unregisterGroup: name => {
            delete formGroups[name];
        },
        registerField: (source, group) => {
            if (group != null) {
                if (!(formGroups.current[group] || []).includes(source)) {
                    formGroups.current[group] = [
                        ...(formGroups.current[group] || []),
                        source,
                    ];
                    // Notify subscribers that the group fields have changed
                    if (subscribers.current[group]) {
                        subscribers.current[group].forEach(subscriber => subscriber());
                    }
                }
            }
        },
        unregisterField: (source, group) => {
            if (group != null) {
                if (!formGroups.current[group]) {
                    console.warn(`Invalid form group ${group}`);
                }
                else {
                    const fields = new Set(formGroups.current[group]);
                    fields.delete(source);
                    formGroups.current[group] = Array.from(fields);
                    // Notify subscribers that the group fields have changed
                    if (subscribers.current[group]) {
                        subscribers.current[group].forEach(subscriber => subscriber());
                    }
                }
            }
        },
    }), []);
    return (React.createElement(FormGroupsContext_1.FormGroupsContext.Provider, { value: formContextValue }, children));
};
exports.FormGroupsProvider = FormGroupsProvider;
