// ra-core/dist/test-ui/Confirm.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Confirm = void 0;
const React = __importStar(require("react"));
const Translate_1 = require("../i18n/Translate.cjs");
const Confirm = ({ isOpen, content, onClose, onConfirm, title, translateOptions = {}, titleTranslateOptions = translateOptions, contentTranslateOptions = translateOptions, }) => {
    return isOpen ? (React.createElement("div", { style: {
            position: 'fixed',
            top: 0,
            left: 0,
            right: 0,
            bottom: 0,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            backgroundColor: 'rgba(0, 0, 0, 0.3)',
        } },
        React.createElement("div", { style: {
                backgroundColor: 'rgba(0, 0, 0, 0.8)',
                color: 'white',
                padding: '1em',
            } },
            React.createElement("p", null, typeof title === 'string' ? (React.createElement(Translate_1.Translate, { i18nKey: title, options: {
                    _: title,
                    ...titleTranslateOptions,
                } })) : (title)),
            React.createElement("p", null, typeof content === 'string' ? (React.createElement(Translate_1.Translate, { i18nKey: content, options: {
                    _: content,
                    ...contentTranslateOptions,
                } })) : (content)),
            React.createElement("div", { style: { display: 'flex', gap: '1em' } },
                React.createElement("button", { onClick: onConfirm, type: "button" },
                    React.createElement(Translate_1.Translate, { i18nKey: "ra.action.confirm" }, "Confirm")),
                React.createElement("button", { onClick: onClose, type: "button" },
                    React.createElement(Translate_1.Translate, { i18nKey: "ra.action.cancel" }, "Cancel")))))) : null;
};
exports.Confirm = Confirm;
