// ra-core/dist/test-ui/Layout.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Layout = void 0;
const React = __importStar(require("react"));
const routing_1 = require("../routing/index.cjs");
const useResourceDefinitions_1 = require("../core/useResourceDefinitions.cjs");
const useTranslate_1 = require("../i18n/useTranslate.cjs");
const useRefresh_1 = require("../dataProvider/useRefresh.cjs");
const Notification_1 = require("./Notification.cjs");
const Layout = ({ children }) => {
    const resources = (0, useResourceDefinitions_1.useResourceDefinitions)();
    const translate = (0, useTranslate_1.useTranslate)();
    const refresh = (0, useRefresh_1.useRefresh)();
    return (React.createElement("div", { style: { padding: '1em', border: '1px solid #ccc' } },
        React.createElement("header", null,
            React.createElement("nav", { style: {
                    display: 'flex',
                    gap: '1em',
                    listStyle: 'none',
                    marginBlockStart: 0,
                    marginBlockEnd: 0,
                    paddingInlineStart: 0,
                    justifyContent: 'space-between',
                } },
                React.createElement("ul", { style: {
                        display: 'flex',
                        gap: '1em',
                        listStyle: 'none',
                        marginBlockStart: 0,
                        marginBlockEnd: 0,
                        paddingInlineStart: 0,
                    } }, Object.values(resources).map(resource => resource.hasList ? (React.createElement("li", { key: resource.name },
                    React.createElement(routing_1.LinkBase, { to: `/${resource.name}` }, `${resource.name
                        .substring(0, 1)
                        .toUpperCase()}${resource.name.substring(1)}`))) : null)),
                React.createElement("div", { style: {
                        display: 'flex',
                        gap: '1em',
                        listStyle: 'none',
                        marginBlockStart: 0,
                        marginBlockEnd: 0,
                        paddingInlineStart: 0,
                    } },
                    React.createElement("button", { onClick: () => refresh() }, translate('ra.action.refresh'))))),
        React.createElement("hr", null),
        React.createElement("main", null, children),
        React.createElement(Notification_1.Notification, null)));
};
exports.Layout = Layout;
