// ra-core/dist/test-ui/Pagination.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pagination = void 0;
const React = __importStar(require("react"));
const Translate_1 = require("../i18n/Translate.cjs");
const useTranslate_1 = require("../i18n/useTranslate.cjs");
const useListContext_1 = require("../controller/list/useListContext.cjs");
const Pagination = () => {
    const { page, perPage, total, setPage } = (0, useListContext_1.useListContext)();
    const translate = (0, useTranslate_1.useTranslate)();
    if (total === undefined) {
        return null;
    }
    const nbPages = Math.ceil(total / perPage) || 1;
    return (React.createElement("div", { style: { display: 'flex', justifyContent: 'space-between' } },
        React.createElement("div", null,
            React.createElement(Translate_1.Translate, { i18nKey: "ra.navigation.page_range_info", options: {
                    offsetBegin: (page - 1) * perPage + 1,
                    offsetEnd: Math.min(page * perPage, total),
                    total,
                } }, `${(page - 1) * perPage + 1}-${Math.min(page * perPage, total)} of ${total}`)),
        React.createElement("div", { style: {
                display: 'flex',
                gap: '0',
                alignItems: 'center',
                justifyContent: 'end',
            } },
            page > 1 && (React.createElement("button", { onClick: () => setPage(page - 1), type: "button" },
                React.createElement(Translate_1.Translate, { i18nKey: "ra.navigation.previous" }, "Previous"))),
            Array.from({ length: nbPages }, (_, i) => i + 1).map(p => (React.createElement("button", { key: p, onClick: () => {
                    setPage(p);
                }, style: {
                    fontWeight: p === page ? 'bold' : 'normal',
                }, "aria-label": translate('ra.navigation.page', {
                    page: p,
                }), type: "button" }, p))),
            page < nbPages && (React.createElement("button", { onClick: () => setPage(page + 1), type: "button" },
                React.createElement(Translate_1.Translate, { i18nKey: "ra.navigation.next" }, "Next"))))));
};
exports.Pagination = Pagination;
