// ra-core/dist/test-ui/SimpleList.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleList = void 0;
const React = __importStar(require("react"));
const record_1 = require("../controller/record/index.cjs");
const WithListContext_1 = require("../controller/list/WithListContext.cjs");
const SimpleList = ({ children, render, inline = false, }) => (React.createElement(WithListContext_1.WithListContext, { render: ({ data, isPending }) => isPending ? null : children ? (children) : (React.createElement("ul", { style: {
            display: 'flex',
            flexDirection: inline ? 'row' : 'column',
            gap: '1em',
            listStyle: 'none',
            marginBlockStart: 0,
            marginBlockEnd: 0,
            paddingInlineStart: 0,
        } }, data?.map(record => (React.createElement(record_1.RecordContextProvider, { key: record.id, value: record },
        React.createElement("li", null, render ? render(record) : children)))))) }));
exports.SimpleList = SimpleList;
