// ra-core/dist/test-ui/SimpleFormIterator.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleFormIterator = exports.SimpleFormIteratorItem = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const useTranslate_1 = require("../i18n/useTranslate.cjs");
const Translate_1 = require("../i18n/Translate.cjs");
const useFieldValue_1 = require("../util/useFieldValue.cjs");
const useWrappedSource_1 = require("../core/useWrappedSource.cjs");
const useArrayInput_1 = require("../controller/input/useArrayInput.cjs");
const useSimpleFormIterator_1 = require("../controller/input/useSimpleFormIterator.cjs");
const useSimpleFormIteratorItem_1 = require("../controller/input/useSimpleFormIteratorItem.cjs");
const SimpleFormIteratorBase_1 = require("../controller/input/SimpleFormIteratorBase.cjs");
const SimpleFormIteratorItemBase_1 = require("../controller/input/SimpleFormIteratorItemBase.cjs");
const Confirm_1 = require("./Confirm.cjs");
const controller_1 = require("../controller/index.cjs");
const util_1 = require("../util/index.cjs");
const DefaultAddItemButton = (props) => {
    const { add, source } = (0, useSimpleFormIterator_1.useSimpleFormIterator)();
    const { className, ...rest } = props;
    return (React.createElement("button", { type: "button", onClick: () => add(), className: [`button-add button-add-${source}`, className].join(' '), ...rest },
        React.createElement(Translate_1.Translate, { i18nKey: "ra.action.add" }, "Add")));
};
const DefaultRemoveItemButton = (props) => {
    const { remove, index } = (0, useSimpleFormIteratorItem_1.useSimpleFormIteratorItem)();
    const { source } = (0, useSimpleFormIterator_1.useSimpleFormIterator)();
    const { className, ...rest } = props;
    return (React.createElement("button", { type: "button", onClick: () => remove(), className: [
            `button-remove button-remove-${source}-${index}`,
            className,
        ].join(' '), ...rest },
        React.createElement(Translate_1.Translate, { i18nKey: "ra.action.remove" }, "Remove")));
};
const DefaultReOrderButtons = ({ className }) => {
    const { index, total, reOrder } = (0, useSimpleFormIteratorItem_1.useSimpleFormIteratorItem)();
    const { source } = (0, useSimpleFormIterator_1.useSimpleFormIterator)();
    return (React.createElement("span", { className: [
            `button-reorder button-reorder-${source}-${index}`,
            className,
        ].join(' ') },
        React.createElement("button", { type: "button", onClick: () => reOrder(index - 1), disabled: index <= 0 },
            React.createElement(Translate_1.Translate, { i18nKey: "ra.action.move_up" }, "Move Up")),
        React.createElement("button", { type: "button", onClick: () => reOrder(index + 1), disabled: total == null || index >= total - 1 },
            React.createElement(Translate_1.Translate, { i18nKey: "ra.action.move_down" }, "Move Down"))));
};
exports.SimpleFormIteratorItem = React.forwardRef(function SimpleFormIteratorItem(props, ref) {
    const { children, disabled, disableReordering, disableRemove, getItemLabel, index, inline, record, removeButton = React.createElement(DefaultRemoveItemButton, null), reOrderButtons = React.createElement(DefaultReOrderButtons, null), } = props;
    // Returns a boolean to indicate whether to disable the remove button for certain fields.
    // If disableRemove is a function, then call the function with the current record to
    // determining if the button should be disabled. Otherwise, use a boolean property that
    // enables or disables the button for all of the fields.
    const disableRemoveField = (record) => {
        if (typeof disableRemove === 'boolean') {
            return disableRemove;
        }
        return disableRemove && disableRemove(record);
    };
    const label = typeof getItemLabel === 'function' ? getItemLabel(index) : getItemLabel;
    return (React.createElement(SimpleFormIteratorItemBase_1.SimpleFormIteratorItemBase, { ...props },
        React.createElement("li", { ref: ref },
            React.createElement("div", { style: {
                    display: 'flex',
                    gap: '1rem',
                    alignItems: 'center',
                } },
                label != null && label !== false && React.createElement("span", null, label),
                React.createElement("div", { style: {
                        display: 'flex',
                        flexDirection: inline ? 'row' : 'column',
                        gap: '1rem',
                    } }, children),
                !disabled && (React.createElement("span", null,
                    !disableReordering && reOrderButtons,
                    !disableRemoveField(record) && removeButton))),
            React.createElement("hr", null))));
});
const SimpleFormIterator = (props) => {
    const { addButton = React.createElement(DefaultAddItemButton, null), removeButton, reOrderButtons, children, className, resource, disabled, disableAdd = false, disableClear, disableRemove = false, disableReordering, inline, getItemLabel = false, fullWidth, } = props;
    const finalSource = (0, useWrappedSource_1.useWrappedSource)('');
    if (!finalSource) {
        throw new Error('SimpleFormIterator can only be called within an iterator input like ArrayInput');
    }
    const [confirmIsOpen, setConfirmIsOpen] = (0, react_1.useState)(false);
    const { fields, remove } = (0, useArrayInput_1.useArrayInput)(props);
    const translate = (0, useTranslate_1.useTranslate)();
    const handleArrayClear = (0, react_1.useCallback)(() => {
        remove();
        setConfirmIsOpen(false);
    }, [remove]);
    const records = (0, useFieldValue_1.useFieldValue)({ source: finalSource });
    const getArrayInputNewItemDefaults = (0, controller_1.useGetArrayInputNewItemDefaults)(fields);
    const getItemDefaults = (0, util_1.useEvent)((item = undefined) => {
        if (item != null)
            return item;
        return getArrayInputNewItemDefaults(children);
    });
    return fields ? (React.createElement(SimpleFormIteratorBase_1.SimpleFormIteratorBase, { getItemDefaults: getItemDefaults, ...props },
        React.createElement("div", { className: [
                className,
                fullWidth ? 'fullwidth' : '',
                disabled ? 'disabled' : '',
            ].join(' ') },
            React.createElement("ul", { style: {
                    listStyle: 'none',
                    margin: 0,
                    padding: 0,
                    display: 'flex',
                    flexDirection: 'column',
                    gap: '1rem',
                    marginBottom: '1rem',
                } }, fields.map((member, index) => (React.createElement(React.Fragment, { key: member.id },
                React.createElement(exports.SimpleFormIteratorItem, { disabled: disabled, disableRemove: disableRemove, disableReordering: disableReordering, fields: fields, getItemLabel: getItemLabel, index: index, record: (records && records[index]) || {}, removeButton: removeButton, reOrderButtons: reOrderButtons, resource: resource, inline: inline }, children))))),
            !disabled &&
                !(disableAdd && (disableClear || disableRemove)) && (React.createElement("div", { style: {
                    display: 'flex',
                    gap: '1rem',
                    marginBottom: '1rem',
                } },
                !disableAdd && React.createElement("div", null, addButton),
                fields.length > 0 &&
                    !disableClear &&
                    !disableRemove && (React.createElement("div", null,
                    React.createElement(Confirm_1.Confirm, { isOpen: confirmIsOpen, title: translate('ra.action.clear_array_input'), content: translate('ra.message.clear_array_input'), onConfirm: handleArrayClear, onClose: () => setConfirmIsOpen(false) }),
                    React.createElement("button", { type: "button", onClick: () => setConfirmIsOpen(true) },
                        React.createElement(Translate_1.Translate, { i18nKey: "ra.action.clear_array_input" }, "Clear"))))))))) : null;
};
exports.SimpleFormIterator = SimpleFormIterator;
