// ra-core/dist/routing/useGetPathForRecord.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useGetPathForRecord = void 0;
const react_1 = require("react");
const useResourceContext_1 = require("../core/useResourceContext.cjs");
const useRecordContext_1 = require("../controller/record/useRecordContext.cjs");
const auth_1 = require("../auth/index.cjs");
const core_1 = require("../core/index.cjs");
const useCreatePath_1 = require("./useCreatePath.cjs");
/**
 * Get a path for a record, based on the current resource and the link type.
 *
 * Accepted link types are 'edit', 'show', a route string, false, or a function returning one of these types.
 *
 * @example
 * // basic usage (leverages RecordContext, ResourceContext and ResourceDefinitionContext)
 * const EditLink = () => {
 *   const path = useGetPathForRecord();
 *   return path ? <Link to={path}>Edit</Link> : null;
 * };
 *
 * // controlled mode
 * const EditLink = ({ record, resource }) => {
 *    const path = useGetPathForRecord({ record, resource, link: 'edit' });
 *    return path ? <Link to={path}>Edit</Link> : null;
 * };
 *
 * // the link option can be a function
 * const EditLink = ({ record, resource }) => {
 *   const path = useGetPathForRecord({ record, resource, link: (record, resource) => record.canEdit ? 'edit' : false });
 *   return path ? <Link to={path}>Edit</Link> : null;
 * };
 *
 * // the link option can be a function returning a promise
 * const EditLink = ({ record, resource }) => {
 *   const path = useGetPathForRecord({ record, resource, link: async (record, resource) => {
 *     const canEdit = await canEditRecord(record, resource);
 *     return canEdit ? 'edit' : false;
 *   }});
 *   return path ? <Link to={path}>Edit</Link> : null;
 * };
 */
const useGetPathForRecord = (options = {}) => {
    const { link } = options || {};
    const record = (0, useRecordContext_1.useRecordContext)(options);
    const resource = (0, useResourceContext_1.useResourceContext)(options);
    if (!resource) {
        throw new Error('Cannot generate a link for a record without a resource. You must use useGetPathForRecord within a ResourceContextProvider, or pass a resource prop.');
    }
    const resourceDefinition = (0, core_1.useResourceDefinition)(options);
    const createPath = (0, useCreatePath_1.useCreatePath)();
    const [path, setPath] = (0, react_1.useState)(link && typeof link !== 'function' && record != null
        ? createPath({
            resource,
            id: record.id,
            type: link,
        })
        : false);
    // in preparation for the default value, does the user have access to the show and edit pages?
    // (we can't run hooks conditionally, so we need to run them even though the link is specified)
    const { canAccess: canAccessShow } = (0, auth_1.useCanAccess)({
        action: 'show',
        resource,
        record,
        enabled: link == null && resourceDefinition.hasShow,
    });
    const { canAccess: canAccessEdit } = (0, auth_1.useCanAccess)({
        action: 'edit',
        resource,
        record,
        enabled: link == null && resourceDefinition.hasEdit,
    });
    (0, react_1.useEffect)(() => {
        if (!record)
            return;
        if (link === false) {
            setPath(false);
            return;
        }
        // Handle the inferred link type case
        if (link == null) {
            // We must check whether the resource has an edit view because if there is no
            // authProvider, canAccessShow will always be true
            if (resourceDefinition.hasShow && canAccessShow) {
                setPath(createPath({
                    resource,
                    id: record.id,
                    type: 'show',
                }));
                return;
            }
            // We must check whether the resource has an edit view because if there is no
            // authProvider, canAccessEdit will always be true
            if (resourceDefinition.hasEdit && canAccessEdit) {
                setPath(createPath({
                    resource,
                    id: record.id,
                    type: 'edit',
                }));
                return;
            }
        }
        // Handle the link function case
        if (typeof link === 'function') {
            const linkResult = link(record, resource);
            if (linkResult instanceof Promise) {
                linkResult.then(resolvedPath => setPath(resolvedPath));
                return;
            }
            setPath(linkResult
                ? createPath({
                    resource,
                    id: record.id,
                    type: linkResult,
                })
                : false);
            return;
        }
        // handle string case
        if (link) {
            setPath(createPath({
                resource,
                id: record.id,
                type: link,
            }));
        }
    }, [
        createPath,
        canAccessShow,
        canAccessEdit,
        link,
        record,
        resource,
        resourceDefinition.hasEdit,
        resourceDefinition.hasShow,
    ]);
    return path;
};
exports.useGetPathForRecord = useGetPathForRecord;
