// ra-ui-materialui/dist/input/CheckboxGroupInputItem.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxGroupInputItemClasses = exports.CheckboxGroupInputItem = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const CheckboxGroupInputItem = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { id, choice, className, fullWidth, onChange, optionText, optionValue, options, translateChoice, value, labelPlacement, inputRef, disableValue = 'disabled', ...rest } = props;
    const { getChoiceText, getChoiceValue, getDisableValue } = (0, ra_core_1.useChoices)({
        optionText,
        optionValue,
        translateChoice,
        disableValue,
    });
    const choiceName = getChoiceText(choice);
    const disabled = getDisableValue(choice);
    return (React.createElement(StyledFormControlLabel, { htmlFor: `${id}_${getChoiceValue(choice)}`, key: getChoiceValue(choice), onChange: onChange, className: className, inputRef: inputRef, control: React.createElement(material_1.Checkbox, { id: `${id}_${getChoiceValue(choice)}`, color: "primary", className: exports.CheckboxGroupInputItemClasses.checkbox, checked: value
                ? value.find(v => v == getChoiceValue(choice)) !== // eslint-disable-line eqeqeq
                    undefined
                : false, value: String(getChoiceValue(choice)), disabled: disabled, ...options }), label: choiceName, labelPlacement: labelPlacement, ...rest }));
};
exports.CheckboxGroupInputItem = CheckboxGroupInputItem;
const PREFIX = 'RaCheckboxGroupInputItem';
exports.CheckboxGroupInputItemClasses = {
    checkbox: `${PREFIX}-checkbox`,
};
const StyledFormControlLabel = (0, styles_1.styled)(material_1.FormControlLabel, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${exports.CheckboxGroupInputItemClasses.checkbox}`]: {
        height: 32,
    },
});
