// ra-ui-materialui/dist/input/CheckboxGroupInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxGroupInputClasses = exports.CheckboxGroupInput = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const react_1 = require("react");
const get_js_1 = __importDefault(require("lodash/get.js"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const sanitizeInputRestProps_1 = require("./sanitizeInputRestProps.cjs");
const CheckboxGroupInputItem_1 = require("./CheckboxGroupInputItem.cjs");
const InputHelperText_1 = require("./InputHelperText.cjs");
const Labeled_1 = require("../Labeled.cjs");
const layout_1 = require("../layout/index.cjs");
/**
 * An Input component for a checkbox group, using an array of objects for the options
 *
 * Pass possible options as an array of objects in the 'choices' attribute.
 *
 * The expected input must be an array of identifiers (e.g. [12, 31]) which correspond to
 * the 'optionValue' of 'choices' attribute objects.
 *
 * By default, the options are built from:
 *  - the 'id' property as the option value,
 *  - the 'name' property as the option text
 * @example
 * const choices = [
 *     { id: 12, name: 'Ray Hakt' },
 *     { id: 31, name: 'Ann Gullar' },
 *     { id: 42, name: 'Sean Phonee' },
 * ];
 * <CheckboxGroupInput source="recipients" choices={choices} />
 *
 * You can also customize the properties to use for the option name and value,
 * thanks to the 'optionText' and 'optionValue' attributes.
 * @example
 * const choices = [
 *    { _id: 123, full_name: 'Leo Tolstoi' },
 *    { _id: 456, full_name: 'Jane Austen' },
 * ];
 * <CheckboxGroupInput source="recipients" choices={choices} optionText="full_name" optionValue="_id" />
 *
 * `optionText` also accepts a function, so you can shape the option text at will:
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
 * <CheckboxGroupInput source="recipients" choices={choices} optionText={optionRenderer} />
 *
 * `optionText` also accepts a React Element, that can access
 * the related choice through the `useRecordContext` hook. You can use Field components there.
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const FullNameField = () => {
 *     const record = useRecordContext();
 *     return <span>{record.first_name} {record.last_name}</span>;
 * };
 *
 * <CheckboxGroupInput source="recipients" choices={choices} optionText={<FullNameField />}/>
 *
 * The choices are translated by default, so you can use translation identifiers as choices:
 * @example
 * const choices = [
 *    { id: 'programming', name: 'myroot.category.programming' },
 *    { id: 'lifestyle', name: 'myroot.category.lifestyle' },
 *    { id: 'photography', name: 'myroot.category.photography' },
 * ];
 *
 * However, in some cases (e.g. inside a `<ReferenceArrayInput>`), you may not want
 * the choice to be translated. In that case, set the `translateChoice` prop to false.
 * @example
 * <CheckboxGroupInput source="tags" choices={choices} translateChoice={false}/>
 *
 * The object passed as `options` props is passed to the Material UI <Checkbox> components
 *
 * You can disable some choices by providing a `disableValue` field which name is `disabled` by default
 * @example
 * const choices = [
 *    { id: 'programming', name: 'myroot.category.programming' },
 *    { id: 'lifestyle', name: 'myroot.category.lifestyle' },
 *    { id: 'photography', name: 'myroot.category.photography', disabled: true },
 * ];
 *
 *  @example
 * const choices = [
 *    { id: 'programming', name: 'myroot.category.programming' },
 *    { id: 'lifestyle', name: 'myroot.category.lifestyle' },
 *    { id: 'photography', name: 'myroot.category.photography', not_available: true },
 * ];
 * <CheckboxGroupInput source="tags" choices={choices} disableValue="not_available" />
 *
 */
const CheckboxGroupInput = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { choices: choicesProp, className, classes: classesOverride, format, helperText, label, labelPlacement, isLoading: isLoadingProp, isPending: isPendingProp, isFetching: isFetchingProp, margin = 'dense', onBlur, onChange, options, optionText, optionValue = 'id', parse, resource: resourceProp, row = true, source: sourceProp, translateChoice, validate, disableValue = 'disabled', disabled, readOnly, ...rest } = props;
    const { allChoices, isPending, error: fetchError, resource, source, isFromReference, } = (0, ra_core_1.useChoicesContext)({
        choices: choicesProp,
        isFetching: isFetchingProp,
        isLoading: isLoadingProp,
        isPending: isPendingProp,
        resource: resourceProp,
        source: sourceProp,
    });
    if (source === undefined) {
        throw new Error(`If you're not wrapping the CheckboxGroupInput inside a ReferenceArrayInput, you must provide the source prop`);
    }
    if (!isPending && !fetchError && allChoices === undefined) {
        throw new Error(`If you're not wrapping the CheckboxGroupInput inside a ReferenceArrayInput, you must provide the choices prop`);
    }
    const { field: { onChange: formOnChange, onBlur: formOnBlur, value, ref }, fieldState: { error, invalid }, id, isRequired, } = (0, ra_core_1.useInput)({
        format,
        parse,
        resource,
        source,
        validate,
        onChange,
        onBlur,
        disabled,
        readOnly,
        ...rest,
    });
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    const handleCheck = (0, react_1.useCallback)((event, isChecked) => {
        let newValue;
        if (allChoices &&
            allChoices.every(item => typeof (0, get_js_1.default)(item, optionValue) === 'number')) {
            try {
                // try to convert string value to number, e.g. '123'
                newValue = JSON.parse(event.target.value);
            }
            catch (e) {
                // impossible to convert value, e.g. 'abc'
                newValue = event.target.value;
            }
        }
        else {
            newValue = event.target.value;
        }
        if (isChecked) {
            formOnChange([...(value || []), ...[newValue]]);
        }
        else {
            formOnChange(value.filter(v => v != newValue)); // eslint-disable-line eqeqeq
        }
        formOnBlur(); // Ensure field is flagged as touched
    }, [allChoices, formOnChange, formOnBlur, optionValue, value]);
    if (isPending) {
        return (React.createElement(Labeled_1.Labeled, { id: id, label: label, source: source, resource: resource, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), isRequired: isRequired, ...rest },
            React.createElement(layout_1.LinearProgress, null)));
    }
    const renderHelperText = !!fetchError || helperText !== false || invalid;
    return (React.createElement(StyledFormControl, { component: "fieldset", margin: margin, error: fetchError || invalid, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), ...sanitizeRestProps(rest) },
        React.createElement(material_1.FormLabel, { component: "legend", className: exports.CheckboxGroupInputClasses.label },
            React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource, isRequired: isRequired })),
        React.createElement(material_1.FormGroup, { row: row }, allChoices?.map((choice, index) => (React.createElement(CheckboxGroupInputItem_1.CheckboxGroupInputItem, { key: (0, get_js_1.default)(choice, optionValue), choice: choice, id: id, onChange: handleCheck, options: options, optionText: optionText ??
                (isFromReference ? getRecordRepresentation : 'name'), optionValue: optionValue, translateChoice: translateChoice ?? !isFromReference, value: value, labelPlacement: labelPlacement, inputRef: index === 0 ? ref : undefined, disableValue: disableValue, disabled: disabled || readOnly, readOnly: readOnly, ...sanitizeRestProps(rest) })))),
        renderHelperText ? (React.createElement(material_1.FormHelperText, { error: !!fetchError || !!error, className: exports.CheckboxGroupInputClasses.helperText },
            React.createElement(InputHelperText_1.InputHelperText, { error: error?.message || fetchError?.message, helperText: helperText }))) : null));
};
exports.CheckboxGroupInput = CheckboxGroupInput;
const sanitizeRestProps = ({ refetch, setFilter, setPagination, setSort, loaded, touched, ...rest }) => (0, sanitizeInputRestProps_1.sanitizeInputRestProps)(rest);
const PREFIX = 'RaCheckboxGroupInput';
exports.CheckboxGroupInputClasses = {
    label: `${PREFIX}-label`,
    helperText: `${PREFIX}-helperText`,
};
const StyledFormControl = (0, styles_1.styled)(material_1.FormControl, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.CheckboxGroupInputClasses.label}`]: {
        transform: 'translate(0, 4px) scale(0.75)',
        transformOrigin: `top ${theme.direction === 'ltr' ? 'left' : 'right'}`,
    },
    [`& .${exports.CheckboxGroupInputClasses.helperText}`]: {
        marginLeft: 0,
        marginRight: 0,
    },
}));
