// ra-ui-materialui/dist/input/CheckboxGroupInputItem.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Checkbox, FormControlLabel, } from '@mui/material';
import { useChoices } from 'ra-core';
export const CheckboxGroupInputItem = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { id, choice, className, fullWidth, onChange, optionText, optionValue, options, translateChoice, value, labelPlacement, inputRef, disableValue = 'disabled', ...rest } = props;
    const { getChoiceText, getChoiceValue, getDisableValue } = useChoices({
        optionText,
        optionValue,
        translateChoice,
        disableValue,
    });
    const choiceName = getChoiceText(choice);
    const disabled = getDisableValue(choice);
    return (React.createElement(StyledFormControlLabel, { htmlFor: `${id}_${getChoiceValue(choice)}`, key: getChoiceValue(choice), onChange: onChange, className: className, inputRef: inputRef, control: React.createElement(Checkbox, { id: `${id}_${getChoiceValue(choice)}`, color: "primary", className: CheckboxGroupInputItemClasses.checkbox, checked: value
                ? value.find(v => v == getChoiceValue(choice)) !== // eslint-disable-line eqeqeq
                    undefined
                : false, value: String(getChoiceValue(choice)), disabled: disabled, ...options }), label: choiceName, labelPlacement: labelPlacement, ...rest }));
};
const PREFIX = 'RaCheckboxGroupInputItem';
export const CheckboxGroupInputItemClasses = {
    checkbox: `${PREFIX}-checkbox`,
};
const StyledFormControlLabel = styled(FormControlLabel, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${CheckboxGroupInputItemClasses.checkbox}`]: {
        height: 32,
    },
});
