// ra-ui-materialui/dist/input/DatagridInput.js
import * as React from 'react';
import { cloneElement, useCallback } from 'react';
import clsx from 'clsx';
import { FilterContext, ListContextProvider, useChoicesContext, useInput, } from 'ra-core';
import { styled, useThemeProps, } from '@mui/material/styles';
import { InputHelperText } from "./InputHelperText.js";
import { Datagrid, FilterButton, FilterForm } from "../list/index.js";
import { Pagination as DefaultPagination } from "../list/pagination/index.js";
import { sanitizeInputRestProps } from "./sanitizeInputRestProps.js";
const defaultPagination = React.createElement(DefaultPagination, null);
/**
 * WIP: This component is not yet ready to be used.
 *
 * An input for selecting items displayed in a datagrid
 *
 * @example
 * const membersFilters = [
 *     <TextInput label="Search" source="q" alwaysOn />,
 * ];
 * const TeamEdit = () => (
 *    <Edit>
 *        <SimpleForm>
 *            <TextInput source="name" />
 *            <ReferenceArrayInput
 *                source="members"
 *                reference="users"
 *                filter={{ is_retired: false }}
 *                perPage={50}
 *                sort={{ field: 'lastName', order: 'ASC' }}
 *            >
 *                <DatagridInput
 *                    filters={membersFilters}
 *                >
 *                    <TextField source="firstName" />
 *                    <TextField source="lastName" />
 *                </DatagridInput>
 *            </ReferenceArrayInput>
 *        </SimpleForm>
 *    </Edit>
 * );
 */
export const DatagridInput = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { choices, className, pagination = defaultPagination, filters, source: sourceProp, resource: resourceProp, ...rest } = props;
    const { allChoices, availableChoices, selectedChoices, error: fetchError, source, ...choicesContext } = useChoicesContext({
        choices,
        resource: resourceProp,
        source: sourceProp,
    });
    const { field, fieldState } = useInput({
        ...props,
        ...choicesContext,
        source,
    });
    const onSelect = useCallback((idsToAdd) => {
        field.onChange(idsToAdd);
    }, [field]);
    const onToggleItem = useCallback((id) => {
        if (field.value.includes(id)) {
            field.onChange(field.value.filter(item => item !== id));
        }
        else {
            field.onChange([...field.value, id]);
        }
    }, [field]);
    const onUnselectItems = useCallback(() => {
        field.onChange([]);
    }, [field]);
    const listContext = React.useMemo(() => ({
        ...choicesContext,
        data: availableChoices,
        total: availableChoices?.length,
        error: null,
        onSelect,
        onToggleItem,
        onUnselectItems,
        selectedIds: field.value,
    }), [
        availableChoices,
        choicesContext,
        field,
        onSelect,
        onToggleItem,
        onUnselectItems,
    ]);
    return (React.createElement(Root, { className: clsx('ra-input', `ra-input-${source}`, className) },
        React.createElement(ListContextProvider, { value: listContext },
            filters ? (Array.isArray(filters) ? (React.createElement(FilterContext.Provider, { value: filters },
                React.createElement(React.Fragment, null,
                    React.createElement(FilterForm, null),
                    React.createElement(FilterButton, null)))) : (React.createElement(React.Fragment, null,
                cloneElement(filters, {
                    context: 'form',
                }),
                cloneElement(filters, {
                    context: 'button',
                })))) : null,
            !fieldState.error && !fetchError && (React.createElement(React.Fragment, null,
                React.createElement(Datagrid, { ...sanitizeInputRestProps(rest) }),
                pagination !== false && pagination)),
            React.createElement(InputHelperText, { error: fieldState.error?.message || fetchError?.message }))));
};
const PREFIX = 'RaDatagridInput';
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
