// ra-ui-materialui/dist/input/DatagridInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatagridInput = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const styles_1 = require("@mui/material/styles");
const InputHelperText_1 = require("./InputHelperText.cjs");
const list_1 = require("../list/index.cjs");
const pagination_1 = require("../list/pagination/index.cjs");
const sanitizeInputRestProps_1 = require("./sanitizeInputRestProps.cjs");
const defaultPagination = React.createElement(pagination_1.Pagination, null);
/**
 * WIP: This component is not yet ready to be used.
 *
 * An input for selecting items displayed in a datagrid
 *
 * @example
 * const membersFilters = [
 *     <TextInput label="Search" source="q" alwaysOn />,
 * ];
 * const TeamEdit = () => (
 *    <Edit>
 *        <SimpleForm>
 *            <TextInput source="name" />
 *            <ReferenceArrayInput
 *                source="members"
 *                reference="users"
 *                filter={{ is_retired: false }}
 *                perPage={50}
 *                sort={{ field: 'lastName', order: 'ASC' }}
 *            >
 *                <DatagridInput
 *                    filters={membersFilters}
 *                >
 *                    <TextField source="firstName" />
 *                    <TextField source="lastName" />
 *                </DatagridInput>
 *            </ReferenceArrayInput>
 *        </SimpleForm>
 *    </Edit>
 * );
 */
const DatagridInput = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { choices, className, pagination = defaultPagination, filters, source: sourceProp, resource: resourceProp, ...rest } = props;
    const { allChoices, availableChoices, selectedChoices, error: fetchError, source, ...choicesContext } = (0, ra_core_1.useChoicesContext)({
        choices,
        resource: resourceProp,
        source: sourceProp,
    });
    const { field, fieldState } = (0, ra_core_1.useInput)({
        ...props,
        ...choicesContext,
        source,
    });
    const onSelect = (0, react_1.useCallback)((idsToAdd) => {
        field.onChange(idsToAdd);
    }, [field]);
    const onToggleItem = (0, react_1.useCallback)((id) => {
        if (field.value.includes(id)) {
            field.onChange(field.value.filter(item => item !== id));
        }
        else {
            field.onChange([...field.value, id]);
        }
    }, [field]);
    const onUnselectItems = (0, react_1.useCallback)(() => {
        field.onChange([]);
    }, [field]);
    const listContext = React.useMemo(() => ({
        ...choicesContext,
        data: availableChoices,
        total: availableChoices?.length,
        error: null,
        onSelect,
        onToggleItem,
        onUnselectItems,
        selectedIds: field.value,
    }), [
        availableChoices,
        choicesContext,
        field,
        onSelect,
        onToggleItem,
        onUnselectItems,
    ]);
    return (React.createElement(Root, { className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className) },
        React.createElement(ra_core_1.ListContextProvider, { value: listContext },
            filters ? (Array.isArray(filters) ? (React.createElement(ra_core_1.FilterContext.Provider, { value: filters },
                React.createElement(React.Fragment, null,
                    React.createElement(list_1.FilterForm, null),
                    React.createElement(list_1.FilterButton, null)))) : (React.createElement(React.Fragment, null,
                (0, react_1.cloneElement)(filters, {
                    context: 'form',
                }),
                (0, react_1.cloneElement)(filters, {
                    context: 'button',
                })))) : null,
            !fieldState.error && !fetchError && (React.createElement(React.Fragment, null,
                React.createElement(list_1.Datagrid, { ...(0, sanitizeInputRestProps_1.sanitizeInputRestProps)(rest) }),
                pagination !== false && pagination)),
            React.createElement(InputHelperText_1.InputHelperText, { error: fieldState.error?.message || fetchError?.message }))));
};
exports.DatagridInput = DatagridInput;
const PREFIX = 'RaDatagridInput';
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
