// ra-ui-materialui/dist/input/FileInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileInputClasses = exports.FileInput = void 0;
const react_1 = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const react_dropzone_1 = require("react-dropzone");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const Labeled_1 = require("../Labeled.cjs");
const FileInputPreview_1 = require("./FileInputPreview.cjs");
const sanitizeInputRestProps_1 = require("./sanitizeInputRestProps.cjs");
const InputHelperText_1 = require("./InputHelperText.cjs");
const FileInput = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { accept, children, className, format, helperText, inputProps: inputPropsOptions, maxSize, minSize, multiple = false, label, labelMultiple = 'ra.input.file.upload_several', labelSingle = 'ra.input.file.upload_single', options = {}, onChange: onChangeProp, onRemove: onRemoveProp, parse, placeholder, removeIcon, resource, source, validate, validateFileRemoval, disabled, readOnly, ...rest } = props;
    const { onDrop: onDropProp } = options;
    const translate = (0, ra_core_1.useTranslate)();
    // turn a browser dropped file structure into expected structure
    const transformFile = file => {
        if (!(file instanceof File)) {
            return file;
        }
        const preview = URL.createObjectURL(file);
        const transformedFile = {
            rawFile: file,
            src: preview,
            title: file.name,
        };
        return transformedFile;
    };
    const transformFiles = (files) => {
        if (!files) {
            return multiple ? [] : null;
        }
        if (Array.isArray(files)) {
            return files.map(transformFile);
        }
        return transformFile(files);
    };
    const { id, field: { onChange, onBlur, value }, fieldState, isRequired, } = (0, ra_core_1.useInput)({
        format: format || transformFiles,
        parse: parse || transformFiles,
        source,
        validate,
        disabled,
        readOnly,
        onChange: onChangeProp,
        ...rest,
    });
    const { error, invalid } = fieldState;
    const files = value ? (Array.isArray(value) ? value : [value]) : [];
    const onDrop = (newFiles, rejectedFiles, event) => {
        const updatedFiles = multiple ? [...files, ...newFiles] : [...newFiles];
        if (multiple) {
            onChange(updatedFiles);
            onBlur();
        }
        else {
            onChange(updatedFiles[0]);
            onBlur();
        }
        if (onDropProp) {
            onDropProp(newFiles, rejectedFiles, event);
        }
    };
    const onRemove = file => async () => {
        if (validateFileRemoval) {
            try {
                await validateFileRemoval(file);
            }
            catch (e) {
                return;
            }
        }
        if (multiple) {
            const filteredFiles = files.filter(stateFile => !(0, ra_core_1.shallowEqual)(stateFile, file));
            onChange(filteredFiles);
            onBlur();
        }
        else {
            onChange(null);
            onBlur();
        }
        if (onRemoveProp) {
            onRemoveProp(file);
        }
    };
    const childrenElement = children && (0, react_1.isValidElement)(react_1.Children.only(children))
        ? react_1.Children.only(children)
        : undefined;
    const { getRootProps, getInputProps } = (0, react_dropzone_1.useDropzone)({
        accept,
        maxSize,
        minSize,
        multiple,
        disabled: disabled || readOnly,
        ...options,
        onDrop,
    });
    const renderHelperText = helperText !== false || invalid;
    const theme = (0, styles_1.useTheme)();
    return (react_1.default.createElement(StyledLabeled, { htmlFor: id, label: label, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), source: source, resource: resource, isRequired: isRequired, color: invalid ? 'error' : undefined, sx: {
            cursor: disabled || readOnly ? 'default' : 'pointer',
            ...rest.sx,
        }, ...(0, sanitizeInputRestProps_1.sanitizeInputRestProps)(rest) },
        react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement("div", { ...getRootProps({
                    className: exports.FileInputClasses.dropZone,
                    'data-testid': 'dropzone',
                    style: {
                        color: disabled || readOnly
                            ? (theme.vars || theme).palette.text
                                .disabled
                            : inputPropsOptions?.color ||
                                (theme.vars || theme).palette.text
                                    .primary,
                        backgroundColor: disabled || readOnly
                            ? (theme.vars || theme).palette.action
                                .disabledBackground
                            : inputPropsOptions?.backgroundColor,
                    },
                }) },
                react_1.default.createElement("input", { id: id, name: id, ...getInputProps({
                        ...inputPropsOptions,
                    }) }),
                placeholder ? (placeholder) : multiple ? (react_1.default.createElement("p", null, translate(labelMultiple))) : (react_1.default.createElement("p", null, translate(labelSingle)))),
            renderHelperText ? (react_1.default.createElement(material_1.FormHelperText, { error: invalid },
                react_1.default.createElement(InputHelperText_1.InputHelperText, { error: error?.message, helperText: helperText }))) : null,
            children && (react_1.default.createElement("div", { className: "previews" }, files.map((file, index) => (react_1.default.createElement(FileInputPreview_1.FileInputPreview, { key: index, file: file, onRemove: onRemove(file), className: exports.FileInputClasses.removeButton, removeIcon: removeIcon },
                react_1.default.createElement(ra_core_1.RecordContextProvider, { value: file }, childrenElement)))))))));
};
exports.FileInput = FileInput;
const PREFIX = 'RaFileInput';
exports.FileInputClasses = {
    dropZone: `${PREFIX}-dropZone`,
    removeButton: `${PREFIX}-removeButton`,
};
const StyledLabeled = (0, styles_1.styled)(Labeled_1.Labeled, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    width: '100%',
    [`& .${exports.FileInputClasses.dropZone}`]: {
        background: (theme.vars || theme).palette.background.default,
        borderRadius: theme.shape.borderRadius,
        fontFamily: theme.typography.fontFamily,
        padding: theme.spacing(1),
        textAlign: 'center',
        color: (theme.vars || theme).palette.primary.contrastText,
    },
    [`& .${exports.FileInputClasses.removeButton}`]: {},
}));
