// ra-ui-materialui/dist/input/FileInput.js
import React, { Children, isValidElement, } from 'react';
import { styled, useTheme, useThemeProps, } from '@mui/material/styles';
import clsx from 'clsx';
import { useDropzone } from 'react-dropzone';
import { FormHelperText } from '@mui/material';
import { useInput, useTranslate, shallowEqual, RecordContextProvider, } from 'ra-core';
import { Labeled } from "../Labeled.js";
import { FileInputPreview } from "./FileInputPreview.js";
import { sanitizeInputRestProps } from "./sanitizeInputRestProps.js";
import { InputHelperText } from "./InputHelperText.js";
export const FileInput = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { accept, children, className, format, helperText, inputProps: inputPropsOptions, maxSize, minSize, multiple = false, label, labelMultiple = 'ra.input.file.upload_several', labelSingle = 'ra.input.file.upload_single', options = {}, onChange: onChangeProp, onRemove: onRemoveProp, parse, placeholder, removeIcon, resource, source, validate, validateFileRemoval, disabled, readOnly, ...rest } = props;
    const { onDrop: onDropProp } = options;
    const translate = useTranslate();
    // turn a browser dropped file structure into expected structure
    const transformFile = file => {
        if (!(file instanceof File)) {
            return file;
        }
        const preview = URL.createObjectURL(file);
        const transformedFile = {
            rawFile: file,
            src: preview,
            title: file.name,
        };
        return transformedFile;
    };
    const transformFiles = (files) => {
        if (!files) {
            return multiple ? [] : null;
        }
        if (Array.isArray(files)) {
            return files.map(transformFile);
        }
        return transformFile(files);
    };
    const { id, field: { onChange, onBlur, value }, fieldState, isRequired, } = useInput({
        format: format || transformFiles,
        parse: parse || transformFiles,
        source,
        validate,
        disabled,
        readOnly,
        onChange: onChangeProp,
        ...rest,
    });
    const { error, invalid } = fieldState;
    const files = value ? (Array.isArray(value) ? value : [value]) : [];
    const onDrop = (newFiles, rejectedFiles, event) => {
        const updatedFiles = multiple ? [...files, ...newFiles] : [...newFiles];
        if (multiple) {
            onChange(updatedFiles);
            onBlur();
        }
        else {
            onChange(updatedFiles[0]);
            onBlur();
        }
        if (onDropProp) {
            onDropProp(newFiles, rejectedFiles, event);
        }
    };
    const onRemove = file => async () => {
        if (validateFileRemoval) {
            try {
                await validateFileRemoval(file);
            }
            catch (e) {
                return;
            }
        }
        if (multiple) {
            const filteredFiles = files.filter(stateFile => !shallowEqual(stateFile, file));
            onChange(filteredFiles);
            onBlur();
        }
        else {
            onChange(null);
            onBlur();
        }
        if (onRemoveProp) {
            onRemoveProp(file);
        }
    };
    const childrenElement = children && isValidElement(Children.only(children))
        ? Children.only(children)
        : undefined;
    const { getRootProps, getInputProps } = useDropzone({
        accept,
        maxSize,
        minSize,
        multiple,
        disabled: disabled || readOnly,
        ...options,
        onDrop,
    });
    const renderHelperText = helperText !== false || invalid;
    const theme = useTheme();
    return (React.createElement(StyledLabeled, { htmlFor: id, label: label, className: clsx('ra-input', `ra-input-${source}`, className), source: source, resource: resource, isRequired: isRequired, color: invalid ? 'error' : undefined, sx: {
            cursor: disabled || readOnly ? 'default' : 'pointer',
            ...rest.sx,
        }, ...sanitizeInputRestProps(rest) },
        React.createElement(React.Fragment, null,
            React.createElement("div", { ...getRootProps({
                    className: FileInputClasses.dropZone,
                    'data-testid': 'dropzone',
                    style: {
                        color: disabled || readOnly
                            ? (theme.vars || theme).palette.text
                                .disabled
                            : inputPropsOptions?.color ||
                                (theme.vars || theme).palette.text
                                    .primary,
                        backgroundColor: disabled || readOnly
                            ? (theme.vars || theme).palette.action
                                .disabledBackground
                            : inputPropsOptions?.backgroundColor,
                    },
                }) },
                React.createElement("input", { id: id, name: id, ...getInputProps({
                        ...inputPropsOptions,
                    }) }),
                placeholder ? (placeholder) : multiple ? (React.createElement("p", null, translate(labelMultiple))) : (React.createElement("p", null, translate(labelSingle)))),
            renderHelperText ? (React.createElement(FormHelperText, { error: invalid },
                React.createElement(InputHelperText, { error: error?.message, helperText: helperText }))) : null,
            children && (React.createElement("div", { className: "previews" }, files.map((file, index) => (React.createElement(FileInputPreview, { key: index, file: file, onRemove: onRemove(file), className: FileInputClasses.removeButton, removeIcon: removeIcon },
                React.createElement(RecordContextProvider, { value: file }, childrenElement)))))))));
};
const PREFIX = 'RaFileInput';
export const FileInputClasses = {
    dropZone: `${PREFIX}-dropZone`,
    removeButton: `${PREFIX}-removeButton`,
};
const StyledLabeled = styled(Labeled, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    width: '100%',
    [`& .${FileInputClasses.dropZone}`]: {
        background: (theme.vars || theme).palette.background.default,
        borderRadius: theme.shape.borderRadius,
        fontFamily: theme.typography.fontFamily,
        padding: theme.spacing(1),
        textAlign: 'center',
        color: (theme.vars || theme).palette.primary.contrastText,
    },
    [`& .${FileInputClasses.removeButton}`]: {},
}));
