// ra-ui-materialui/dist/input/FileInputPreview.js
import * as React from 'react';
import { useEffect } from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import RemoveCircle from '@mui/icons-material/RemoveCircle';
import IconButton from '@mui/material/IconButton';
import { useTranslate } from 'ra-core';
export const FileInputPreview = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, onRemove, file, removeIcon: RemoveIcon = RemoveCircle, ...rest } = props;
    const translate = useTranslate();
    useEffect(() => {
        return () => {
            const preview = file.rawFile ? file.rawFile.preview : file.preview;
            if (preview) {
                window.URL.revokeObjectURL(preview);
            }
        };
    }, [file]);
    return (React.createElement(Root, { className: className, ...rest },
        React.createElement(IconButton, { className: FileInputPreviewClasses.removeButton, onClick: onRemove, "aria-label": translate('ra.action.delete'), title: translate('ra.action.delete'), size: "small" },
            React.createElement(RemoveIcon, { className: FileInputPreviewClasses.removeIcon })),
        children));
};
const PREFIX = 'RaFileInputPreview';
const FileInputPreviewClasses = {
    removeButton: `${PREFIX}-removeButton`,
    removeIcon: `${PREFIX}-removeIcon`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${FileInputPreviewClasses.removeButton}`]: {},
    [`& .${FileInputPreviewClasses.removeIcon}`]: {
        color: (theme.vars || theme).palette.error.main,
    },
}));
