// ra-ui-materialui/dist/input/ImageInput.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { FileInput, FileInputClasses } from "./FileInput.js";
export const ImageInput = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    return (React.createElement(StyledFileInput, { labelMultiple: "ra.input.image.upload_several", labelSingle: "ra.input.image.upload_single", ...props }));
};
const PREFIX = 'RaImageInput';
const StyledFileInput = styled(FileInput, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    width: '100%',
    [`& .${FileInputClasses.dropZone}`]: {
        background: (theme.vars || theme).palette.background.default,
        borderRadius: theme.shape.borderRadius,
        fontFamily: theme.typography.fontFamily,
        padding: theme.spacing(1),
        textAlign: 'center',
        color: (theme.vars || theme).palette.primary.contrastText,
    },
    [`& .${FileInputClasses.removeButton}`]: {
        display: 'inline-block',
        position: 'relative',
        '& button': {
            position: 'absolute',
            top: theme.spacing(1),
            right: theme.spacing(1),
            minWidth: theme.spacing(2),
            opacity: 0,
        },
        '&:hover button': {
            opacity: 1,
        },
    },
}));
