// ra-ui-materialui/dist/input/LoadingInput.js
import * as React from 'react';
import { CircularProgress, InputAdornment } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import { useTimeout } from 'ra-core';
import { ResettableTextField } from "./ResettableTextField.js";
/**
 * An input placeholder with a loading indicator
 *
 * Avoids visual jumps when replaced by a form input
 */
export const LoadingInput = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { fullWidth, label, helperText, margin, size, sx, timeout = 1000, variant, } = props;
    const oneSecondHasPassed = useTimeout(timeout);
    return (React.createElement(StyledResettableTextField, { sx: sx, label: label, helperText: helperText, fullWidth: fullWidth, variant: variant, margin: margin, size: size, disabled: true, onChange: () => { }, InputProps: {
            endAdornment: (React.createElement(InputAdornment, { position: "end" }, oneSecondHasPassed ? (React.createElement(CircularProgress, { color: "inherit", size: 20 })) : (
            // use an adornment of the same size to avoid visual jumps
            React.createElement("span", { style: { width: 20 } }, "\u00A0")))),
        } }));
};
const PREFIX = 'RaLoadingInput';
// make it look just like a regular input, even though it's disabled
// because the loading indicator is enough
const StyledResettableTextField = styled(ResettableTextField, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    '& .MuiInputLabel-root.Mui-disabled': {
        color: (theme.vars || theme).palette.text.secondary,
    },
    '& .MuiFilledInput-root.Mui-disabled': {
        background: (theme.vars || theme).palette.action.disabledBackground,
    },
    '& .MuiFilledInput-root.Mui-disabled:before': {
        borderBottomStyle: 'solid',
    },
}));
