// ra-ui-materialui/dist/input/NullableBooleanInput.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { MenuItem, TextField } from '@mui/material';
import clsx from 'clsx';
import { useInput, useTranslate, FieldTitle } from 'ra-core';
import { sanitizeInputRestProps } from "./sanitizeInputRestProps.js";
import { InputHelperText } from "./InputHelperText.js";
export const NullableBooleanInput = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, format = getStringFromBoolean, helperText, label, margin, onBlur, onChange, parse = getBooleanFromString, resource, disabled, readOnly, source, validate, variant, nullLabel = 'ra.boolean.null', falseLabel = 'ra.boolean.false', trueLabel = 'ra.boolean.true', ...rest } = props;
    const translate = useTranslate();
    const { field, fieldState: { error, invalid }, id, isRequired, } = useInput({
        format,
        parse,
        onBlur,
        onChange,
        resource,
        source,
        validate,
        disabled,
        readOnly,
        ...rest,
    });
    const renderHelperText = helperText !== false || invalid;
    return (React.createElement(StyledTextField, { id: id, size: "small", ...field, className: clsx('ra-input', `ra-input-${source}`, NullableBooleanInputClasses.input, className), select: true, disabled: disabled || readOnly, readOnly: readOnly, margin: margin, label: label !== '' && label !== false ? (React.createElement(FieldTitle, { label: label, source: source, resource: resource, isRequired: isRequired })) : null, error: invalid, helperText: renderHelperText ? (React.createElement(InputHelperText, { error: error?.message, helperText: helperText })) : null, variant: variant, ...sanitizeInputRestProps(rest) },
        React.createElement(MenuItem, { value: "" }, translate(nullLabel)),
        React.createElement(MenuItem, { value: "false" }, translate(falseLabel)),
        React.createElement(MenuItem, { value: "true" }, translate(trueLabel))));
};
const PREFIX = 'RaNullableBooleanInput';
export const NullableBooleanInputClasses = {
    input: `${PREFIX}-input`,
};
const StyledTextField = styled(TextField, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    minWidth: theme.spacing(20),
    [theme.breakpoints.down('sm')]: {
        [`&.${NullableBooleanInputClasses.input}`]: {
            width: '100%',
        },
    },
}));
const getBooleanFromString = (value) => {
    if (value === 'true')
        return true;
    if (value === 'false')
        return false;
    return null;
};
const getStringFromBoolean = (value) => {
    if (value === true)
        return 'true';
    if (value === false)
        return 'false';
    return '';
};
