// ra-ui-materialui/dist/input/NullableBooleanInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullableBooleanInputClasses = exports.NullableBooleanInput = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const sanitizeInputRestProps_1 = require("./sanitizeInputRestProps.cjs");
const InputHelperText_1 = require("./InputHelperText.cjs");
const NullableBooleanInput = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, format = getStringFromBoolean, helperText, label, margin, onBlur, onChange, parse = getBooleanFromString, resource, disabled, readOnly, source, validate, variant, nullLabel = 'ra.boolean.null', falseLabel = 'ra.boolean.false', trueLabel = 'ra.boolean.true', ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const { field, fieldState: { error, invalid }, id, isRequired, } = (0, ra_core_1.useInput)({
        format,
        parse,
        onBlur,
        onChange,
        resource,
        source,
        validate,
        disabled,
        readOnly,
        ...rest,
    });
    const renderHelperText = helperText !== false || invalid;
    return (React.createElement(StyledTextField, { id: id, size: "small", ...field, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, exports.NullableBooleanInputClasses.input, className), select: true, disabled: disabled || readOnly, readOnly: readOnly, margin: margin, label: label !== '' && label !== false ? (React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource, isRequired: isRequired })) : null, error: invalid, helperText: renderHelperText ? (React.createElement(InputHelperText_1.InputHelperText, { error: error?.message, helperText: helperText })) : null, variant: variant, ...(0, sanitizeInputRestProps_1.sanitizeInputRestProps)(rest) },
        React.createElement(material_1.MenuItem, { value: "" }, translate(nullLabel)),
        React.createElement(material_1.MenuItem, { value: "false" }, translate(falseLabel)),
        React.createElement(material_1.MenuItem, { value: "true" }, translate(trueLabel))));
};
exports.NullableBooleanInput = NullableBooleanInput;
const PREFIX = 'RaNullableBooleanInput';
exports.NullableBooleanInputClasses = {
    input: `${PREFIX}-input`,
};
const StyledTextField = (0, styles_1.styled)(material_1.TextField, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    minWidth: theme.spacing(20),
    [theme.breakpoints.down('sm')]: {
        [`&.${exports.NullableBooleanInputClasses.input}`]: {
            width: '100%',
        },
    },
}));
const getBooleanFromString = (value) => {
    if (value === 'true')
        return true;
    if (value === 'false')
        return false;
    return null;
};
const getStringFromBoolean = (value) => {
    if (value === true)
        return 'true';
    if (value === false)
        return 'false';
    return '';
};
