// ra-ui-materialui/dist/input/RadioButtonGroupInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RadioButtonGroupInputClasses = exports.RadioButtonGroupInput = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const get_js_1 = __importDefault(require("lodash/get.js"));
const ra_core_1 = require("ra-core");
const sanitizeInputRestProps_1 = require("./sanitizeInputRestProps.cjs");
const InputHelperText_1 = require("./InputHelperText.cjs");
const RadioButtonGroupInputItem_1 = require("./RadioButtonGroupInputItem.cjs");
const Labeled_1 = require("../Labeled.cjs");
const layout_1 = require("../layout/index.cjs");
/**
 * An Input component for a radio button group, using an array of objects for the options
 *
 * Pass possible options as an array of objects in the 'choices' attribute.
 *
 * By default, the options are built from:
 *  - the 'id' property as the option value,
 *  - the 'name' property as the option text
 * @example
 * const choices = [
 *    { id: 'M', name: 'Male' },
 *    { id: 'F', name: 'Female' },
 * ];
 * <RadioButtonGroupInput source="gender" choices={choices} />
 *
 * You can also customize the properties to use for the option name and value,
 * thanks to the 'optionText' and 'optionValue' attributes.
 * @example
 * const choices = [
 *    { _id: 123, full_name: 'Leo Tolstoi', sex: 'M' },
 *    { _id: 456, full_name: 'Jane Austen', sex: 'F' },
 * ];
 * <RadioButtonGroupInput source="author_id" choices={choices} optionText="full_name" optionValue="_id" />
 *
 * `optionText` also accepts a function, so you can shape the option text at will:
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
 * <CheckboxGroupInput source="recipients" choices={choices} optionText={optionRenderer} />
 *
 * `optionText` also accepts a React Element, that can access
 * the related choice through the `useRecordContext` hook. You can use Field components there.
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const FullNameField = () => {
 *     const record = useRecordContext();
 *     return (<span>{record.first_name} {record.last_name}</span>)
 * };
 * <RadioButtonGroupInput source="recipients" choices={choices} optionText={<FullNameField />}/>
 *
 * The choices are translated by default, so you can use translation identifiers as choices:
 * @example
 * const choices = [
 *    { id: 'M', name: 'myroot.gender.male' },
 *    { id: 'F', name: 'myroot.gender.female' },
 * ];
 *
 * However, in some cases (e.g. inside a `<ReferenceInput>`), you may not want
 * the choice to be translated. In that case, set the `translateChoice` prop to false.
 * @example
 * <RadioButtonGroupInput source="gender" choices={choices} translateChoice={false}/>
 *
 * The object passed as `options` props is passed to the Material UI <RadioButtonGroup> component
 *
 * You can disable some choices by providing a `disableValue` field which name is `disabled` by default
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 *    { id: 976, first_name: 'William', last_name: 'Rinkerd', disabled: true },
 * ];
 *
 *  @example
 * const choices = [
 *   { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *   { id: 456, first_name: 'Jane', last_name: 'Austen' },
 *   { id: 976, first_name: 'William', last_name: 'Rinkerd', not_available: true },
 * ];
 * <RadioButtonGroupInput source="gender" choices={choices} disableValue="not_available" />
 *
 */
const RadioButtonGroupInput = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { choices: choicesProp, className, format, helperText, isFetching: isFetchingProp, isLoading: isLoadingProp, isPending: isPendingProp, label, margin = 'dense', onBlur, onChange, options = defaultOptions, optionText, optionValue = 'id', parse, resource: resourceProp, row = true, source: sourceProp, translateChoice, validate, disableValue = 'disabled', disabled, readOnly, ...rest } = props;
    const { allChoices, isPending, error: fetchError, resource, source, isFromReference, } = (0, ra_core_1.useChoicesContext)({
        choices: choicesProp,
        isFetching: isFetchingProp,
        isLoading: isLoadingProp,
        isPending: isPendingProp,
        resource: resourceProp,
        source: sourceProp,
    });
    if (source === undefined) {
        throw new Error(`If you're not wrapping the RadioButtonGroupInput inside a ReferenceArrayInput, you must provide the source prop`);
    }
    if (!isPending && !fetchError && allChoices === undefined) {
        throw new Error(`If you're not wrapping the RadioButtonGroupInput inside a ReferenceArrayInput, you must provide the choices prop`);
    }
    const { id, isRequired, fieldState, field } = (0, ra_core_1.useInput)({
        format,
        onBlur,
        onChange,
        parse,
        resource,
        source,
        validate,
        disabled,
        readOnly,
        ...rest,
    });
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    const { error, invalid } = fieldState;
    if (isPending) {
        return (React.createElement(Labeled_1.Labeled, { htmlFor: id, label: label, source: source, resource: resource, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), isRequired: isRequired },
            React.createElement(layout_1.LinearProgress, null)));
    }
    const renderHelperText = !!fetchError || helperText !== false || invalid;
    return (React.createElement(StyledFormControl, { component: "fieldset", className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), margin: margin, error: fetchError || invalid, disabled: disabled || readOnly, readOnly: readOnly, ...sanitizeRestProps(rest) },
        React.createElement(material_1.FormLabel, { component: "legend", className: exports.RadioButtonGroupInputClasses.label },
            React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource, isRequired: isRequired })),
        React.createElement(material_1.RadioGroup, { id: id, row: row, ...field, ...options, ...sanitizeRestProps(rest) }, allChoices?.map(choice => (React.createElement(RadioButtonGroupInputItem_1.RadioButtonGroupInputItem, { key: (0, get_js_1.default)(choice, optionValue), choice: choice, optionText: optionText ??
                (isFromReference ? getRecordRepresentation : 'name'), optionValue: optionValue, source: id, translateChoice: translateChoice ?? !isFromReference, disableValue: disableValue })))),
        renderHelperText ? (React.createElement(material_1.FormHelperText, null,
            React.createElement(InputHelperText_1.InputHelperText, { error: error?.message || fetchError?.message, helperText: helperText }))) : null));
};
exports.RadioButtonGroupInput = RadioButtonGroupInput;
const sanitizeRestProps = ({ afterSubmit, allowNull, beforeSubmit, choices, className, crudGetMatching, crudGetOne, data, filter, filterToQuery, formatOnBlur, isEqual, limitChoicesToValue, multiple, name, pagination, perPage, ref, reference, refetch, render, setFilter, setPagination, setSort, sort, subscription, type, validateFields, validation, value, ...rest }) => (0, sanitizeInputRestProps_1.sanitizeInputRestProps)(rest);
const PREFIX = 'RaRadioButtonGroupInput';
exports.RadioButtonGroupInputClasses = {
    label: `${PREFIX}-label`,
};
const StyledFormControl = (0, styles_1.styled)(material_1.FormControl, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.RadioButtonGroupInputClasses.label}`]: {
        transform: 'translate(0, 5px) scale(0.75)',
        transformOrigin: `top ${theme.direction === 'ltr' ? 'left' : 'right'}`,
    },
}));
const defaultOptions = {};
