// ra-ui-materialui/dist/input/ResettableTextField.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResettableTextFieldStyles = exports.ResettableTextFieldClasses = exports.ResettableTextField = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const Clear_1 = __importDefault(require("@mui/icons-material/Clear"));
const ra_core_1 = require("ra-core");
/**
 * An override of the default Material UI TextField which is resettable
 */
exports.ResettableTextField = (0, react_1.forwardRef)((inProps, ref) => {
    // This ensures we inherit the theme props from the TextField component (variant for instance)
    const textFieldProps = (0, styles_1.useThemeProps)({
        props: inProps,
        name: 'MuiTextField',
    });
    const props = (0, styles_1.useThemeProps)({
        props: textFieldProps,
        name: PREFIX,
    });
    const { clearAlwaysVisible, InputProps, value, resettable, disabled, readOnly, variant, margin, className, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const { onChange } = props;
    const handleClickClearButton = (0, react_1.useCallback)(event => {
        event.preventDefault();
        onChange && onChange('');
    }, [onChange]);
    const { clearButton, clearIcon, inputAdornedEnd, selectAdornment, visibleClearIcon, } = exports.ResettableTextFieldClasses;
    const { endAdornment, ...InputPropsWithoutEndAdornment } = InputProps || {};
    if (clearAlwaysVisible && endAdornment) {
        throw new Error('ResettableTextField cannot display both an endAdornment and a clear button always visible');
    }
    const getEndAdornment = () => {
        if (!resettable) {
            return endAdornment;
        }
        else if (!value) {
            if (clearAlwaysVisible) {
                // show clear button, inactive
                return (React.createElement(material_1.InputAdornment, { position: "end", className: props.select ? selectAdornment : undefined },
                    React.createElement(material_1.IconButton, { className: clearButton, "aria-label": translate('ra.action.clear_input_value'), title: translate('ra.action.clear_input_value'), disabled: true, size: "large" },
                        React.createElement(Clear_1.default, { className: (0, clsx_1.default)(clearIcon, visibleClearIcon) }))));
            }
            else {
                if (endAdornment) {
                    return endAdornment;
                }
                else {
                    // show spacer
                    return (React.createElement(material_1.InputAdornment, { position: "end", className: props.select ? selectAdornment : undefined },
                        React.createElement("span", { className: clearButton }, "\u00A0")));
                }
            }
        }
        else {
            // show clear
            return (React.createElement(material_1.InputAdornment, { position: "end", className: props.select ? selectAdornment : undefined },
                React.createElement(material_1.IconButton, { className: clearButton, "aria-label": translate('ra.action.clear_input_value'), title: translate('ra.action.clear_input_value'), onClick: handleClickClearButton, onMouseDown: handleMouseDownClearButton, disabled: disabled || readOnly, size: "large" },
                    React.createElement(Clear_1.default, { className: (0, clsx_1.default)(clearIcon, {
                            [visibleClearIcon]: clearAlwaysVisible || value,
                        }) }))));
        }
    };
    return (React.createElement(StyledTextField, { value: value, InputProps: {
            readOnly: readOnly,
            classes: props.select && variant === 'filled'
                ? { adornedEnd: inputAdornedEnd }
                : {},
            endAdornment: getEndAdornment(),
            ...InputPropsWithoutEndAdornment,
        }, disabled: disabled || readOnly, variant: variant, margin: margin, className: className, ...rest, inputRef: ref }));
});
exports.ResettableTextField.displayName = 'ResettableTextField';
const handleMouseDownClearButton = event => {
    event.preventDefault();
};
const PREFIX = 'RaResettableTextField';
exports.ResettableTextFieldClasses = {
    clearIcon: `${PREFIX}-clearIcon`,
    visibleClearIcon: `${PREFIX}-visibleClearIcon`,
    clearButton: `${PREFIX}-clearButton`,
    selectAdornment: `${PREFIX}-selectAdornment`,
    inputAdornedEnd: `${PREFIX}-inputAdornedEnd`,
};
exports.ResettableTextFieldStyles = {
    [`& .${exports.ResettableTextFieldClasses.clearIcon}`]: {
        height: 16,
        width: 0,
    },
    [`& .${exports.ResettableTextFieldClasses.visibleClearIcon}`]: {
        width: 16,
    },
    [`& .${exports.ResettableTextFieldClasses.clearButton}`]: {
        height: 24,
        width: 24,
        padding: 0,
    },
    [`& .${exports.ResettableTextFieldClasses.selectAdornment}`]: {
        position: 'absolute',
        right: 24,
    },
    [`& .${exports.ResettableTextFieldClasses.inputAdornedEnd}`]: {
        paddingRight: 0,
    },
};
const StyledTextField = (0, styles_1.styled)(material_1.TextField, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(exports.ResettableTextFieldStyles);
