// ra-ui-materialui/dist/input/SelectArrayInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectArrayInputClasses = exports.SelectArrayInput = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const react_1 = require("react");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const InputHelperText_1 = require("./InputHelperText.cjs");
const layout_1 = require("../layout/index.cjs");
const Labeled_1 = require("../Labeled.cjs");
/**
 * An Input component for a select box allowing multiple selections, using an array of objects for the options
 *
 * Pass possible options as an array of objects in the 'choices' attribute.
 *
 * By default, the options are built from:
 *  - the 'id' property as the option value,
 *  - the 'name' property as the option text
 * @example
 * const choices = [
 *    { id: 'programming', name: 'Programming' },
 *    { id: 'lifestyle', name: 'Lifestyle' },
 *    { id: 'photography', name: 'Photography' },
 * ];
 * <SelectArrayInput source="tags" choices={choices} />
 *
 * You can also customize the properties to use for the option name and value,
 * thanks to the 'optionText' and 'optionValue' attributes.
 * @example
 * const choices = [
 *    { _id: 123, full_name: 'Leo Tolstoi', sex: 'M' },
 *    { _id: 456, full_name: 'Jane Austen', sex: 'F' },
 * ];
 * <SelectArrayInput source="authors" choices={choices} optionText="full_name" optionValue="_id" />
 *
 * `optionText` also accepts a function, so you can shape the option text at will:
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
 * <SelectArrayInput source="authors" choices={choices} optionText={optionRenderer} />
 *
 * `optionText` also accepts a React Element, that can access
 * the related choice through the `useRecordContext` hook. You can use Field components there.
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const FullNameField = () => {
 *     const record = useRecordContext();
 *     return (<span>{record.first_name} {record.last_name}</span>)
 * };
 * <SelectArrayInput source="authors" choices={choices} optionText={<FullNameField />}/>
 *
 * The choices are translated by default, so you can use translation identifiers as choices:
 * @example
 * const choices = [
 *    { id: 'programming', name: 'myroot.tags.programming' },
 *    { id: 'lifestyle', name: 'myroot.tags.lifestyle' },
 *    { id: 'photography', name: 'myroot.tags.photography' },
 * ];
 */
const SelectArrayInput = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { choices: choicesProp, className, create, createLabel, createValue, disableValue = 'disabled', format, helperText, label, InputLabelProps, isFetching: isFetchingProp, isLoading: isLoadingProp, isPending: isPendingProp, margin, onBlur, onChange, onCreate, options = defaultOptions, optionText, optionValue = 'id', parse, resource: resourceProp, size = 'small', source: sourceProp, translateChoice, validate, variant, disabled, readOnly, ...rest } = props;
    const inputLabel = (0, react_1.useRef)(null);
    const { allChoices, isPending, error: fetchError, source, resource, isFromReference, } = (0, ra_core_1.useChoicesContext)({
        choices: choicesProp,
        isLoading: isLoadingProp,
        isPending: isPendingProp,
        isFetching: isFetchingProp,
        resource: resourceProp,
        source: sourceProp,
    });
    const { field, isRequired, fieldState: { error, invalid }, id, } = (0, ra_core_1.useInput)({
        format,
        onBlur,
        onChange,
        parse,
        resource,
        source,
        validate,
        disabled,
        readOnly,
        ...rest,
    });
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    const { getChoiceText, getChoiceValue, getDisableValue } = (0, ra_core_1.useChoices)({
        optionText: optionText ??
            (isFromReference ? getRecordRepresentation : undefined),
        optionValue,
        disableValue,
        translateChoice: translateChoice ?? !isFromReference,
    });
    const handleChange = (0, react_1.useCallback)((eventOrChoice) => {
        // We might receive an event from the mui component
        // In this case, it will be the choice id
        if (eventOrChoice?.target) {
            // when used with different IDs types, unselection leads to double selection with both types
            // instead of the value being removed from the array
            // e.g. we receive eventOrChoice.target.value = [1, '2', 2] instead of [1] after removing 2
            // this snippet removes a value if it is present twice
            eventOrChoice.target.value = eventOrChoice.target.value.reduce((acc, value) => {
                // eslint-disable-next-line eqeqeq
                const index = acc.findIndex(v => v == value);
                return index < 0
                    ? [...acc, value]
                    : [...acc.slice(0, index), ...acc.slice(index + 1)];
            }, []);
            field.onChange(eventOrChoice);
        }
        else {
            // Or we might receive a choice directly, for instance a newly created one
            field.onChange([
                ...(field.value || []),
                getChoiceValue(eventOrChoice),
            ]);
        }
    }, [field, getChoiceValue]);
    const { getCreateItem, handleChange: handleChangeWithCreateSupport, createElement, } = (0, ra_core_1.useSupportCreateSuggestion)({
        create,
        createLabel,
        createValue,
        handleChange,
        onCreate,
        optionText,
    });
    const createItem = create || onCreate ? getCreateItem() : null;
    const finalChoices = create || onCreate
        ? [...(allChoices || []), createItem]
        : allChoices || [];
    const renderMenuItemOption = (0, react_1.useCallback)(choice => !!createItem &&
        choice?.id === createItem.id &&
        typeof optionText === 'function'
        ? createItem.name
        : getChoiceText(choice), [createItem, getChoiceText, optionText]);
    const renderMenuItem = (0, react_1.useCallback)(choice => {
        return choice ? (React.createElement(material_1.MenuItem, { key: getChoiceValue(choice), value: getChoiceValue(choice), disabled: getDisableValue(choice) }, renderMenuItemOption(!!createItem && choice?.id === createItem.id
            ? createItem
            : choice))) : null;
    }, [getChoiceValue, getDisableValue, renderMenuItemOption, createItem]);
    if (isPending) {
        return (React.createElement(Labeled_1.Labeled, { label: label, source: source, resource: resource, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), isRequired: isRequired },
            React.createElement(layout_1.LinearProgress, null)));
    }
    // Here wen ensure we always have an array and this array does not contain the default value (empty string)
    const finalValue = Array.isArray(field.value ?? [])
        ? field.value
        : field.value
            ? [field.value]
            : [];
    const outlinedInputProps = variant === 'outlined'
        ? {
            input: (React.createElement(material_1.OutlinedInput, { id: "select-multiple-chip", label: React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource, isRequired: isRequired }) })),
        }
        : {};
    const renderHelperText = !!fetchError || helperText !== false || invalid;
    return (React.createElement(React.Fragment, null,
        React.createElement(StyledFormControl, { margin: margin, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), error: fetchError || invalid, variant: variant, ...sanitizeRestProps(rest) },
            React.createElement(material_1.InputLabel, { ref: inputLabel, id: `${id}-outlined-label`, htmlFor: id, ...InputLabelProps },
                React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource, isRequired: isRequired })),
            React.createElement(material_1.Select, { id: id, labelId: `${id}-outlined-label`, label: React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource, isRequired: isRequired }), multiple: true, error: !!fetchError || invalid, renderValue: (selected) => (React.createElement("div", { className: exports.SelectArrayInputClasses.chips }, (Array.isArray(selected) ? selected : [])
                    .map(item => (allChoices || []).find(
                // eslint-disable-next-line eqeqeq
                choice => getChoiceValue(choice) == item))
                    .filter(item => !!item)
                    .map(item => (React.createElement(material_1.Chip, { key: getChoiceValue(item), label: renderMenuItemOption(item), className: exports.SelectArrayInputClasses.chip, size: "small" }))))), disabled: disabled || readOnly, readOnly: readOnly, "data-testid": "selectArray", size: size, ...field, ...options, onChange: handleChangeWithCreateSupport, value: finalValue, ...outlinedInputProps }, finalChoices.map(renderMenuItem)),
            renderHelperText ? (React.createElement(material_1.FormHelperText, { error: !!fetchError || !!error },
                React.createElement(InputHelperText_1.InputHelperText, { error: error?.message || fetchError?.message, helperText: helperText }))) : null),
        createElement));
};
exports.SelectArrayInput = SelectArrayInput;
const sanitizeRestProps = ({ alwaysOn, choices, defaultValue, disableValue, emptyText, enableGetChoices, filter, filterToQuery, initializeForm, initialValue, input, isRequired, label, limitChoicesToValue, loaded, locale, meta, onChange, options, optionValue, optionText, perPage, record, reference, resource, setFilter, setPagination, setSort, sort, source, textAlign, translate, translateChoice, validation, ...rest }) => rest;
const PREFIX = 'RaSelectArrayInput';
exports.SelectArrayInputClasses = {
    chips: `${PREFIX}-chips`,
    chip: `${PREFIX}-chip`,
};
const StyledFormControl = (0, styles_1.styled)(material_1.FormControl, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    minWidth: theme.spacing(20),
    [theme.breakpoints.down('sm')]: {
        width: '100%',
    },
    [`& .${exports.SelectArrayInputClasses.chips}`]: {
        display: 'flex',
        flexWrap: 'wrap',
    },
    [`& .${exports.SelectArrayInputClasses.chip}`]: {
        marginTop: theme.spacing(0.5),
        marginRight: theme.spacing(0.5),
    },
}));
const defaultOptions = {};
