// ra-ui-materialui/dist/input/SelectInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectInput = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const ResettableTextField_1 = require("./ResettableTextField.cjs");
const InputHelperText_1 = require("./InputHelperText.cjs");
const sanitizeInputRestProps_1 = require("./sanitizeInputRestProps.cjs");
const LoadingInput_1 = require("./LoadingInput.cjs");
/**
 * An Input component for a select box, using an array of objects for the options
 *
 * Pass possible options as an array of objects in the 'choices' attribute.
 *
 * By default, the options are built from:
 *  - the 'id' property as the option value,
 *  - the 'name' property as the option text
 * @example
 * const choices = [
 *    { id: 'M', name: 'Male' },
 *    { id: 'F', name: 'Female' },
 * ];
 * <SelectInput source="gender" choices={choices} />
 *
 * You can also customize the properties to use for the option name and value,
 * thanks to the 'optionText' and 'optionValue' attributes.
 * @example
 * const choices = [
 *    { _id: 123, full_name: 'Leo Tolstoi', sex: 'M' },
 *    { _id: 456, full_name: 'Jane Austen', sex: 'F' },
 * ];
 * <SelectInput source="author_id" choices={choices} optionText="full_name" optionValue="_id" />
 *
 * `optionText` also accepts a function, so you can shape the option text at will:
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
 * <SelectInput source="author_id" choices={choices} optionText={optionRenderer} />
 *
 * `optionText` also accepts a React Element, that can access
 * the related choice through the `useRecordContext` hook. You can use Field components there.
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 * ];
 * const FullNameField = () => {
 *     const record = useRecordContext();
 *     return <span>{record.first_name} {record.last_name}</span>;
 * }
 * <SelectInput source="author" choices={choices} optionText={<FullNameField />}/>
 *
 * The choices are translated by default, so you can use translation identifiers as choices:
 * @example
 * const choices = [
 *    { id: 'M', name: 'myroot.gender.male' },
 *    { id: 'F', name: 'myroot.gender.female' },
 * ];
 *
 * However, in some cases (e.g. inside a `<ReferenceInput>`), you may not want
 * the choice to be translated. In that case, set the `translateChoice` prop to false.
 * @example
 * <SelectInput source="gender" choices={choices} translateChoice={false}/>
 *
 * You can disable some choices by providing a `disableValue` field which name is `disabled` by default
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 *    { id: 976, first_name: 'William', last_name: 'Rinkerd', disabled: true },
 * ];
 *
 * @example
 * const choices = [
 *    { id: 123, first_name: 'Leo', last_name: 'Tolstoi' },
 *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
 *    { id: 976, first_name: 'William', last_name: 'Rinkerd', not_available: true },
 * ];
 * <SelectInput source="gender" choices={choices} disableValue="not_available" />
 *
 */
const SelectInput = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { choices: choicesProp, className, create, createLabel, createValue, createHintValue, defaultValue, disableValue = 'disabled', emptyText = '', emptyValue = '', format, filter, helperText, isFetching: isFetchingProp, isLoading: isLoadingProp, isPending: isPendingProp, label, margin = 'dense', onBlur, onChange, onCreate, optionText, optionValue, parse, resource: resourceProp, source: sourceProp, translateChoice, validate, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    (0, react_1.useEffect)(() => {
        if (emptyValue == null) {
            throw new Error(`emptyValue being set to null or undefined is not supported. Use parse to turn the empty string into null.`);
        }
    }, [emptyValue]);
    const { allChoices, isPending, error: fetchError, source, resource, isFromReference, } = (0, ra_core_1.useChoicesContext)({
        choices: choicesProp,
        isLoading: isLoadingProp,
        isFetching: isFetchingProp,
        isPending: isPendingProp,
        resource: resourceProp,
        source: sourceProp,
    });
    if (source === undefined) {
        throw new Error(`If you're not wrapping the SelectInput inside a ReferenceInput, you must provide the source prop`);
    }
    if (!isPending && !fetchError && allChoices === undefined) {
        throw new Error(`If you're not wrapping the SelectInput inside a ReferenceInput, you must provide the choices prop`);
    }
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    const { getChoiceText, getChoiceValue, getDisableValue } = (0, ra_core_1.useChoices)({
        optionText: optionText ??
            (isFromReference ? getRecordRepresentation : undefined),
        optionValue,
        disableValue,
        translateChoice: translateChoice ?? !isFromReference,
        createValue,
        createHintValue,
    });
    const { field, fieldState, id, isRequired } = (0, ra_core_1.useInput)({
        defaultValue,
        parse,
        format,
        onBlur,
        onChange,
        resource,
        source,
        validate,
        ...rest,
    });
    const { error, invalid } = fieldState;
    const renderEmptyItemOption = (0, react_1.useCallback)(() => {
        return typeof emptyText === 'string'
            ? emptyText === ''
                ? ' ' // em space, forces the display of an empty line of normal height
                : translate(emptyText, { _: emptyText })
            : emptyText;
    }, [emptyText, translate]);
    const renderMenuItemOption = (0, react_1.useCallback)(choice => getChoiceText(choice), [getChoiceText]);
    const handleChange = (0, react_1.useCallback)(async (eventOrChoice) => {
        if (typeof eventOrChoice === 'string') {
            if (eventOrChoice === '') {
                // called  by the reset button
                field.onChange(emptyValue);
            }
        }
        else if (eventOrChoice?.target) {
            // We might receive an event from the mui component
            // In this case, it will be the choice id
            field.onChange(eventOrChoice);
        }
        else {
            // Or we might receive a choice directly, for instance a newly created one
            field.onChange(getChoiceValue(eventOrChoice));
        }
    }, [field, getChoiceValue, emptyValue]);
    const { getCreateItem, handleChange: handleChangeWithCreateSupport, createElement, } = (0, ra_core_1.useSupportCreateSuggestion)({
        create,
        createLabel,
        createValue,
        createHintValue,
        handleChange,
        onCreate,
        optionText,
    });
    const createItem = create || onCreate ? getCreateItem() : null;
    const renderMenuItem = (0, react_1.useCallback)(choice => {
        return choice ? (React.createElement(material_1.MenuItem, { key: getChoiceValue(choice), value: getChoiceValue(choice), disabled: getDisableValue(choice) }, renderMenuItemOption(!!createItem && choice?.id === createItem.id
            ? createItem
            : choice))) : null;
    }, [getChoiceValue, getDisableValue, renderMenuItemOption, createItem]);
    if (isPending) {
        return (React.createElement(LoadingInput_1.LoadingInput, { label: label !== '' &&
                label !== false && (React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resourceProp, isRequired: isRequired })), sx: props.sx, helperText: React.createElement(InputHelperText_1.InputHelperText, { error: error?.message, helperText: helperText }), variant: props.variant, size: props.size, margin: props.margin, fullWidth: props.fullWidth }));
    }
    let finalChoices = fetchError ? [] : allChoices;
    if (create || onCreate) {
        finalChoices = [...finalChoices, createItem];
    }
    const renderHelperText = !!fetchError || helperText !== false || invalid;
    return (React.createElement(React.Fragment, null,
        React.createElement(StyledResettableTextField, { id: id, ...field, className: (0, clsx_1.default)('ra-input', `ra-input-${source}`, className), onChange: handleChangeWithCreateSupport, select: true, label: label !== '' && label !== false ? (React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resourceProp, isRequired: isRequired })) : null, clearAlwaysVisible: true, error: !!fetchError || invalid, helperText: renderHelperText ? (React.createElement(InputHelperText_1.InputHelperText, { error: error?.message || fetchError?.message, helperText: helperText })) : null, margin: margin, ...sanitizeRestProps(rest) },
            !isRequired && (React.createElement(material_1.MenuItem, { value: emptyValue, key: "null", "aria-label": translate('ra.action.clear_input_value'), title: translate('ra.action.clear_input_value') }, renderEmptyItemOption())),
            finalChoices.map(renderMenuItem)),
        createElement));
};
exports.SelectInput = SelectInput;
const sanitizeRestProps = ({ afterSubmit, allowNull, beforeSubmit, choices, className, crudGetMatching, crudGetOne, data, field, fieldState, formState, filter, filterToQuery, formatOnBlur, isEqual, limitChoicesToValue, multiple, name, pagination, perPage, ref, reference, refetch, render, setFilter, setPagination, setSort, shouldUnregister, sort, subscription, type, validateFields, validation, value, ...rest }) => (0, sanitizeInputRestProps_1.sanitizeInputRestProps)(rest);
const PREFIX = 'RaSelectInput';
const StyledResettableTextField = (0, styles_1.styled)(ResettableTextField_1.ResettableTextField, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    ...ResettableTextField_1.ResettableTextFieldStyles,
    minWidth: theme.spacing(20),
    '& .MuiFilledInput-root': { paddingRight: 0 },
}));
