// ra-ui-materialui/dist/input/TranslatableInputsTab.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranslatableInputsTabClasses = exports.TranslatableInputsTab = void 0;
const react_1 = __importDefault(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const inflection_1 = require("inflection");
const clsx_1 = __importDefault(require("clsx"));
/**
 * Single tab that selects a locale in a TranslatableInputs component.
 * @see TranslatableInputs
 */
const TranslatableInputsTab = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { groupKey = '', locale, ...rest } = props;
    const { isValid } = (0, ra_core_1.useFormGroup)(`${groupKey}${locale}`);
    const translate = (0, ra_core_1.useTranslate)();
    return (react_1.default.createElement(StyledTab, { id: `translatable-header-${groupKey}${locale}`, label: translate(`ra.locales.${locale}`, {
            _: (0, inflection_1.capitalize)(locale),
        }), className: (0, clsx_1.default)(exports.TranslatableInputsTabClasses.root, {
            [exports.TranslatableInputsTabClasses.error]: !isValid,
        }), ...rest }));
};
exports.TranslatableInputsTab = TranslatableInputsTab;
const PREFIX = 'RaTranslatableInputsTab';
exports.TranslatableInputsTabClasses = {
    root: `${PREFIX}-root`,
    error: `${PREFIX}-error`,
};
const StyledTab = (0, styles_1.styled)(material_1.Tab, { name: PREFIX })(({ theme }) => ({
    [`&.${exports.TranslatableInputsTabClasses.root}`]: {
        fontSize: '0.8em',
        minHeight: theme.spacing(3),
        minWidth: theme.spacing(6),
    },
    [`&.${exports.TranslatableInputsTabClasses.error}`]: {
        color: (theme.vars || theme).palette.error.main,
    },
}));
