// ra-ui-materialui/dist/input/TranslatableInputsTab.js
import React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Tab } from '@mui/material';
import { useFormGroup, useTranslate } from 'ra-core';
import { capitalize } from 'inflection';
import clsx from 'clsx';
/**
 * Single tab that selects a locale in a TranslatableInputs component.
 * @see TranslatableInputs
 */
export const TranslatableInputsTab = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { groupKey = '', locale, ...rest } = props;
    const { isValid } = useFormGroup(`${groupKey}${locale}`);
    const translate = useTranslate();
    return (React.createElement(StyledTab, { id: `translatable-header-${groupKey}${locale}`, label: translate(`ra.locales.${locale}`, {
            _: capitalize(locale),
        }), className: clsx(TranslatableInputsTabClasses.root, {
            [TranslatableInputsTabClasses.error]: !isValid,
        }), ...rest }));
};
const PREFIX = 'RaTranslatableInputsTab';
export const TranslatableInputsTabClasses = {
    root: `${PREFIX}-root`,
    error: `${PREFIX}-error`,
};
const StyledTab = styled(Tab, { name: PREFIX })(({ theme }) => ({
    [`&.${TranslatableInputsTabClasses.root}`]: {
        fontSize: '0.8em',
        minHeight: theme.spacing(3),
        minWidth: theme.spacing(6),
    },
    [`&.${TranslatableInputsTabClasses.error}`]: {
        color: (theme.vars || theme).palette.error.main,
    },
}));
