// ra-ui-materialui/dist/input/TranslatableInputsTabs.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranslatableInputsTabsClasses = exports.TranslatableInputsTabs = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const TranslatableInputsTab_1 = require("./TranslatableInputsTab.cjs");
/**
 * Default locale selector for the TranslatableInputs component. Generates a tab for each specified locale.
 * @see TranslatableInputs
 */
const TranslatableInputsTabs = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { groupKey, TabsProps: tabsProps } = props;
    const { locales, selectLocale, selectedLocale } = (0, ra_core_1.useTranslatableContext)();
    const handleChange = (event, newLocale) => {
        selectLocale(newLocale);
    };
    return (React.createElement(StyledAppBar, { color: "default", position: "static", className: exports.TranslatableInputsTabsClasses.root },
        React.createElement(material_1.Tabs, { value: selectedLocale, onChange: handleChange, indicatorColor: "primary", textColor: "primary", className: exports.TranslatableInputsTabsClasses.tabs, ...tabsProps }, locales.map(locale => (React.createElement(TranslatableInputsTab_1.TranslatableInputsTab, { key: locale, value: locale, locale: locale, groupKey: groupKey }))))));
};
exports.TranslatableInputsTabs = TranslatableInputsTabs;
const PREFIX = 'RaTranslatableInputsTabs';
exports.TranslatableInputsTabsClasses = {
    root: `${PREFIX}-root`,
    tabs: `${PREFIX}-tabs`,
};
const StyledAppBar = (0, styles_1.styled)(material_1.AppBar, { name: PREFIX })(({ theme }) => ({
    [`&.${exports.TranslatableInputsTabsClasses.root}`]: {
        boxShadow: 'none',
        borderRadius: 0,
        borderTopLeftRadius: theme.shape.borderRadius,
        borderTopRightRadius: theme.shape.borderRadius,
        border: `1px solid ${(theme.vars || theme).palette.divider}`,
    },
    [`& .${exports.TranslatableInputsTabsClasses.tabs}`]: {
        minHeight: theme.spacing(3),
    },
}));
