// ra-ui-materialui/dist/input/TranslatableInputsTabContent.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranslatableInputsTabContentClasses = exports.TranslatableInputsTabContent = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
/**
 * Default container for a group of translatable inputs inside a TranslatableInputs component.
 * @see TranslatableInputs
 */
const TranslatableInputsTabContent = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, groupKey = '', locale, ...other } = props;
    const { selectedLocale, getRecordForLocale } = (0, ra_core_1.useTranslatableContext)();
    const parentSourceContext = (0, ra_core_1.useSourceContext)();
    const record = (0, ra_core_1.useRecordContext)(props);
    // The SourceContext will be read by children of TranslatableInputs to compute their composed source and label
    //
    // <TranslatableInputs locales={['en', 'fr']} /> => SourceContext is "fr"
    //     <TextInput source="description" /> => final source for this input will be "description.fr"
    // </TranslatableInputs>
    const sourceContext = (0, react_1.useMemo)(() => ({
        getSource: (source) => {
            if (!source) {
                throw new Error('Children of TranslatableInputs must have a source');
            }
            return parentSourceContext.getSource(`${source}.${locale}`);
        },
        getLabel: (source) => {
            return parentSourceContext.getLabel(source);
        },
    }), [locale, parentSourceContext]);
    // As fields rely on the RecordContext to get their values and have no knowledge of the locale,
    // we need to create a new record with the values for the current locale only
    // Given the record { title: { en: 'title_en', fr: 'title_fr' } } and the locale 'fr',
    // the record for the locale 'fr' will be { title: 'title_fr' }
    const recordForLocale = (0, react_1.useMemo)(() => getRecordForLocale(record, locale), [getRecordForLocale, record, locale]);
    return (React.createElement(ra_core_1.FormGroupContextProvider, { name: `${groupKey}${locale}` },
        React.createElement(Root, { role: "tabpanel", id: `translatable-content-${groupKey}${locale}`, "aria-labelledby": `translatable-header-${groupKey}${locale}`, className: (0, clsx_1.default)(exports.TranslatableInputsTabContentClasses.root, {
                [exports.TranslatableInputsTabContentClasses.hidden]: selectedLocale !== locale,
            }), ...other },
            React.createElement(ra_core_1.SourceContextProvider, { value: sourceContext },
                React.createElement(ra_core_1.RecordContextProvider, { value: recordForLocale }, children)))));
};
exports.TranslatableInputsTabContent = TranslatableInputsTabContent;
const PREFIX = 'RaTranslatableInputsTabContent';
exports.TranslatableInputsTabContentClasses = {
    root: `${PREFIX}-root`,
    hidden: `${PREFIX}-hidden`,
};
const Root = (0, styles_1.styled)(material_1.Stack, { name: PREFIX })(({ theme }) => ({
    [`&.${exports.TranslatableInputsTabContentClasses.root}`]: {
        flexGrow: 1,
        paddingLeft: theme.spacing(2),
        paddingRight: theme.spacing(2),
        paddingTop: theme.spacing(1),
        paddingBottom: theme.spacing(1),
        borderRadius: 0,
        borderBottomLeftRadius: theme.shape.borderRadius,
        borderBottomRightRadius: theme.shape.borderRadius,
        border: `1px solid ${(theme.vars || theme).palette.divider}`,
        borderTop: 0,
    },
    [`&.${exports.TranslatableInputsTabContentClasses.hidden}`]: {
        display: 'none',
    },
}));
