// ra-core/dist/dataProvider/convertLegacyDataProvider.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dataFetchActions_1 = require("./dataFetchActions.cjs");
const defaultDataProvider_1 = require("./defaultDataProvider.cjs");
const fetchMap = {
    create: dataFetchActions_1.CREATE,
    delete: dataFetchActions_1.DELETE,
    deleteMany: dataFetchActions_1.DELETE_MANY,
    getList: dataFetchActions_1.GET_LIST,
    getMany: dataFetchActions_1.GET_MANY,
    getManyReference: dataFetchActions_1.GET_MANY_REFERENCE,
    getOne: dataFetchActions_1.GET_ONE,
    update: dataFetchActions_1.UPDATE,
    updateMany: dataFetchActions_1.UPDATE_MANY,
};
/**
 * Turn a function-based dataProvider to an object-based one
 *
 * Allows using legacy dataProviders transparently.
 *
 * @param {Function} legacyDataProvider A legacy dataProvider (type, resource, params) => Promise<any>
 *
 * @returns {Object} A dataProvider that react-admin can use
 */
const convertLegacyDataProvider = (legacyDataProvider) => {
    const proxy = new Proxy(defaultDataProvider_1.defaultDataProvider, {
        get(_, name) {
            return (resource, params) => {
                if (Object.keys(fetchMap).includes(name.toString())) {
                    const fetchType = fetchMap[name.toString()];
                    return legacyDataProvider(fetchType, resource, params);
                }
                return legacyDataProvider(name.toString(), resource, params);
            };
        },
        apply(_, __, args) {
            return legacyDataProvider.apply(legacyDataProvider, args);
        },
    });
    return proxy;
};
exports.default = convertLegacyDataProvider;
module.exports = exports.default;
