// ra-core/dist/dataProvider/useGetMany.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useGetMany = void 0;
const react_1 = require("react");
const react_query_1 = require("@tanstack/react-query");
const useDataProvider_1 = require("./useDataProvider.cjs");
const util_1 = require("../util/index.cjs");
/**
 * Call the dataProvider.getMany() method and return the resolved result
 * as well as the loading state.
 *
 * The return value updates according to the request state:
 *
 * - start: { isPending: true, refetch }
 * - success: { data: [data from store], isPending: false, refetch }
 * - error: { error: [error from response], isPending: false, refetch }
 *
 * This hook will return the cached result when called a second time
 * with the same parameters, until the response arrives.
 *
 * @param {string} resource The resource name, e.g. 'posts'
 * @param {Params} params The getMany parameters { ids, meta }
 * @param {Object} options Options object to pass to the queryClient.
 * May include side effects to be executed upon success or failure, e.g. { onSuccess: () => { refresh(); } }
 *
 * @typedef Params
 * @prop params.ids The ids to get, e.g. [123, 456, 789]
 * @prop params.meta Optional meta parameters
 *
 * @returns The current request state. Destructure as { data, error, isPending, refetch }.
 *
 * @example
 *
 * import { useGetMany } from 'react-admin';
 *
 * const PostTags = ({ post }) => {
 *     const { data, isPending, error } = useGetMany(
 *         'tags',
 *         { ids: post.tags },
 *     );
 *     if (isPending) { return <Loading />; }
 *     if (error) { return <p>ERROR</p>; }
 *     return <ul>{data.map(tag =>
 *         <li key={tag.id}>{tag.name}</li>
 *     )}</ul>;
 * };
 */
const useGetMany = (resource, params, options = {}) => {
    const { ids, meta } = params;
    const dataProvider = (0, useDataProvider_1.useDataProvider)();
    const queryClient = (0, react_query_1.useQueryClient)();
    const { onError = noop, onSuccess = noop, onSettled = noop, enabled, ...queryOptions } = options;
    const onSuccessEvent = (0, util_1.useEvent)(onSuccess);
    const onErrorEvent = (0, util_1.useEvent)(onError);
    const onSettledEvent = (0, util_1.useEvent)(onSettled);
    const result = (0, react_query_1.useQuery)({
        queryKey: [
            resource,
            'getMany',
            {
                ids: !ids || ids.length === 0 ? [] : ids.map(id => String(id)),
                meta,
            },
        ],
        queryFn: queryParams => {
            if (!ids || ids.length === 0) {
                // no need to call the dataProvider
                return Promise.resolve([]);
            }
            return dataProvider
                .getMany(resource, {
                ids,
                meta,
                signal: dataProvider.supportAbortSignal === true
                    ? queryParams.signal
                    : undefined,
            })
                .then(({ data }) => data);
        },
        placeholderData: () => {
            const records = !ids || ids.length === 0
                ? []
                : ids.map(id => queryClient.getQueryData([
                    resource,
                    'getOne',
                    { id: String(id), meta },
                ]));
            if (records.some(record => record === undefined)) {
                return undefined;
            }
            else {
                return records;
            }
        },
        retry: false,
        enabled: enabled ?? ids != null,
        ...queryOptions,
    });
    const metaValue = (0, react_1.useRef)(meta);
    const resourceValue = (0, react_1.useRef)(resource);
    (0, react_1.useEffect)(() => {
        metaValue.current = meta;
    }, [meta]);
    (0, react_1.useEffect)(() => {
        resourceValue.current = resource;
    }, [resource]);
    (0, react_1.useEffect)(() => {
        if (result.data === undefined ||
            result.error != null ||
            result.isFetching)
            return;
        // optimistically populate the getOne cache
        result.data.forEach(record => {
            queryClient.setQueryData([
                resourceValue.current,
                'getOne',
                { id: String(record.id), meta: metaValue.current },
            ], oldRecord => oldRecord ?? record);
        });
        onSuccessEvent(result.data);
    }, [
        queryClient,
        onSuccessEvent,
        result.data,
        result.error,
        result.isFetching,
    ]);
    (0, react_1.useEffect)(() => {
        if (result.error == null || result.isFetching)
            return;
        onErrorEvent(result.error);
    }, [onErrorEvent, result.error, result.isFetching]);
    (0, react_1.useEffect)(() => {
        if (result.status === 'pending' || result.isFetching)
            return;
        onSettledEvent(result.data, result.error);
    }, [
        onSettledEvent,
        result.data,
        result.error,
        result.status,
        result.isFetching,
    ]);
    return result;
};
exports.useGetMany = useGetMany;
const noop = () => undefined;
