// ra-ui-materialui/dist/list/pagination/Pagination.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pagination = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const PaginationActions_1 = require("./PaginationActions.cjs");
exports.Pagination = (0, react_1.memo)(props => {
    const { rowsPerPageOptions = DefaultRowsPerPageOptions, actions, limit = null, ...rest } = props;
    const { isPending, hasNextPage, page, perPage, total, setPage, setPerPage, } = (0, ra_core_1.useListPaginationContext)();
    const translate = (0, ra_core_1.useTranslate)();
    const isSmall = (0, material_1.useMediaQuery)((theme) => theme.breakpoints.down('md'));
    const totalPages = (0, react_1.useMemo)(() => {
        return total != null ? Math.ceil(total / perPage) : undefined;
    }, [perPage, total]);
    /**
     * Warning: Material UI's page is 0-based
     */
    const handlePageChange = (0, react_1.useCallback)((event, page) => {
        event && event.stopPropagation();
        if (page < 0 || (totalPages && page > totalPages - 1)) {
            throw new Error(translate('ra.navigation.page_out_of_boundaries', {
                page: page + 1,
            }));
        }
        setPage(page + 1);
    }, [totalPages, setPage, translate]);
    const handlePerPageChange = (0, react_1.useCallback)(event => {
        setPerPage(event.target.value);
    }, [setPerPage]);
    const labelDisplayedRows = (0, react_1.useCallback)(({ from, to, count }) => count === -1 && hasNextPage
        ? translate('ra.navigation.partial_page_range_info', {
            offsetBegin: from,
            offsetEnd: to,
            _: `%{from}-%{to} of more than %{to}`,
        })
        : translate('ra.navigation.page_range_info', {
            offsetBegin: from,
            offsetEnd: to,
            total: count === -1 ? to : count,
            _: `%{from}-%{to} of %{count === -1 ? to : count}`,
        }), [translate, hasNextPage]);
    const labelItem = (0, react_1.useCallback)(type => translate(`ra.navigation.${type}`, { _: `Go to ${type} page` }), [translate]);
    if (isPending) {
        return React.createElement(material_1.Toolbar, { variant: "dense" });
    }
    // Avoid rendering TablePagination if "page" value is invalid
    if (total === 0 || page < 1 || (total != null && page > totalPages)) {
        if (limit != null && process.env.NODE_ENV === 'development') {
            console.warn('The Pagination limit prop is deprecated. Empty state should be handled by the component displaying data (Datagrid, SimpleList).');
        }
        return null;
    }
    if (isSmall) {
        return (React.createElement(material_1.TablePagination, { count: total == null ? -1 : total, rowsPerPage: perPage, page: page - 1, onPageChange: handlePageChange, rowsPerPageOptions: emptyArray, component: "span", labelDisplayedRows: labelDisplayedRows, ...(0, ra_core_1.sanitizeListRestProps)(rest) }));
    }
    const ActionsComponent = actions
        ? actions // overridden by caller
        : !isPending && total != null
            ? PaginationActions_1.PaginationActions // regular navigation
            : undefined; // partial navigation (uses default TablePaginationActions)
    return (React.createElement(material_1.TablePagination, { count: total == null ? -1 : total, rowsPerPage: perPage, page: page - 1, onPageChange: handlePageChange, onRowsPerPageChange: handlePerPageChange, 
        // @ts-ignore
        ActionsComponent: ActionsComponent, nextIconButtonProps: {
            disabled: !hasNextPage,
        }, component: "span", labelRowsPerPage: translate('ra.navigation.page_rows_per_page'), labelDisplayedRows: labelDisplayedRows, getItemAriaLabel: labelItem, rowsPerPageOptions: rowsPerPageOptions, ...(0, ra_core_1.sanitizeListRestProps)(rest) }));
});
const DefaultRowsPerPageOptions = [5, 10, 25, 50];
const emptyArray = [];
