// ra-ui-materialui/dist/list/pagination/PaginationActions.js
import * as React from 'react';
import { memo } from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Pagination } from '@mui/material';
import { useTranslate } from 'ra-core';
export const PaginationActions = memo(inProps => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { page, rowsPerPage, count, onPageChange, size = 'small', className, ...rest } = props;
    const translate = useTranslate();
    const nbPages = Math.ceil(count / rowsPerPage) || 1;
    if (nbPages === 1) {
        return React.createElement(Root, { className: className });
    }
    const getItemAriaLabel = (type, page, selected) => {
        if (type === 'page') {
            return selected
                ? translate('ra.navigation.current_page', {
                    page,
                    _: `page ${page}`,
                })
                : translate('ra.navigation.page', {
                    page,
                    _: `Go to page ${page}`,
                });
        }
        return translate(`ra.navigation.${type}`, { _: `Go to ${type} page` });
    };
    return (React.createElement(Root, { className: className },
        React.createElement(Pagination, { size: size, count: nbPages, 
            // <TablePagination>, the parent, uses 0-based pagination
            // while <Pagination> uses 1-based pagination
            page: page + 1, onChange: (e, page) => onPageChange(e, page - 1), ...sanitizeRestProps(rest), getItemAriaLabel: getItemAriaLabel })));
});
const PREFIX = 'RaPaginationActions';
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    flexShrink: 0,
    ml: 4,
}));
const sanitizeRestProps = ({ nextIconButtonProps, backIconButtonProps, slotProps, ...rest }) => rest;
