// ra-ui-materialui/dist/list/pagination/PaginationActions.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaginationActions = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
exports.PaginationActions = (0, react_1.memo)(inProps => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { page, rowsPerPage, count, onPageChange, size = 'small', className, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const nbPages = Math.ceil(count / rowsPerPage) || 1;
    if (nbPages === 1) {
        return React.createElement(Root, { className: className });
    }
    const getItemAriaLabel = (type, page, selected) => {
        if (type === 'page') {
            return selected
                ? translate('ra.navigation.current_page', {
                    page,
                    _: `page ${page}`,
                })
                : translate('ra.navigation.page', {
                    page,
                    _: `Go to page ${page}`,
                });
        }
        return translate(`ra.navigation.${type}`, { _: `Go to ${type} page` });
    };
    return (React.createElement(Root, { className: className },
        React.createElement(material_1.Pagination, { size: size, count: nbPages, 
            // <TablePagination>, the parent, uses 0-based pagination
            // while <Pagination> uses 1-based pagination
            page: page + 1, onChange: (e, page) => onPageChange(e, page - 1), ...sanitizeRestProps(rest), getItemAriaLabel: getItemAriaLabel })));
});
const PREFIX = 'RaPaginationActions';
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    flexShrink: 0,
    ml: 4,
}));
const sanitizeRestProps = ({ nextIconButtonProps, backIconButtonProps, slotProps, ...rest }) => rest;
