// ra-ui-materialui/dist/list/datatable/ColumnsButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColumnsButton = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const RtlProvider_1 = require("@mui/system/RtlProvider");
const ViewWeek_1 = __importDefault(require("@mui/icons-material/ViewWeek"));
const styles_1 = require("@mui/material/styles");
/**
 * Renders a button that lets users show / hide columns in a DataTable
 *
 * @example
 * import { ColumnsButton, DataTable } from 'react-admin';
 *
 * const PostListActions = () => (
 *   <TopToolbar>
        <ColumnsButton />
        <FilterButton />
 *   </TopToolbar>
 * );
 *
 * const PostList = () => (
 *   <List actions={<PostListActions />}>
 *     <DataTable>
 *       <DataTable.Col source="title" />
 *       <DataTable.Col source="author" />
         ...
 *     </DataTable>
 *   </List>
 * );
 */
const ColumnsButton = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const resource = (0, ra_core_1.useResourceContext)(props);
    const storeKey = props.storeKey || `${resource}.datatable`;
    const [anchorEl, setAnchorEl] = React.useState(null);
    const isRtl = (0, RtlProvider_1.useRtl)();
    const translate = (0, ra_core_1.useTranslate)();
    const isXSmall = (0, material_1.useMediaQuery)((theme) => theme.breakpoints.down('sm'));
    const title = translate('ra.action.select_columns', { _: 'Columns' });
    const handleClick = (event) => {
        setAnchorEl(event.currentTarget);
    };
    const handleClose = () => {
        setAnchorEl(null);
    };
    return (React.createElement(Root, null,
        isXSmall ? (React.createElement(material_1.Tooltip, { title: title },
            React.createElement(material_1.IconButton, { "aria-label": title, color: "primary", onClick: handleClick, size: "large", ...sanitizeRestProps(props) },
                React.createElement(ViewWeek_1.default, null)))) : (React.createElement(material_1.Button, { size: "small", onClick: handleClick, startIcon: React.createElement(ViewWeek_1.default, null), ...sanitizeRestProps(props) }, title)),
        React.createElement(material_1.Popover, { open: Boolean(anchorEl), keepMounted: true, anchorEl: anchorEl, onClose: handleClose, anchorOrigin: {
                vertical: 'bottom',
                horizontal: isRtl ? 'right' : 'left',
            }, transformOrigin: isRtl ? RTL_ORIGIN : LTR_ORIGIN },
            React.createElement(material_1.Box, { id: `${storeKey}-columnsSelector`, sx: { px: 1, my: 0, minWidth: 200 } }))));
};
exports.ColumnsButton = ColumnsButton;
const RTL_ORIGIN = {
    vertical: 'top',
    horizontal: 'right',
};
const LTR_ORIGIN = {
    vertical: 'top',
    horizontal: 'left',
};
const PREFIX = 'RaColumnsButton';
const Root = (0, styles_1.styled)('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    '& .MuiButton-sizeSmall': {
        // fix for icon misalignment on small buttons, see https://github.com/mui/material-ui/pull/30240
        lineHeight: 1.5,
    },
});
const sanitizeRestProps = ({ resource, storeKey, ...rest }) => rest;
