// ra-ui-materialui/dist/list/datatable/ColumnsButton.js
import * as React from 'react';
import { useTranslate, useResourceContext } from 'ra-core';
import { Button, useMediaQuery, Tooltip, IconButton, Popover, Box, } from '@mui/material';
import { useRtl } from '@mui/system/RtlProvider';
import ViewWeekIcon from '@mui/icons-material/ViewWeek';
import { styled, useThemeProps, } from '@mui/material/styles';
/**
 * Renders a button that lets users show / hide columns in a DataTable
 *
 * @example
 * import { ColumnsButton, DataTable } from 'react-admin';
 *
 * const PostListActions = () => (
 *   <TopToolbar>
        <ColumnsButton />
        <FilterButton />
 *   </TopToolbar>
 * );
 *
 * const PostList = () => (
 *   <List actions={<PostListActions />}>
 *     <DataTable>
 *       <DataTable.Col source="title" />
 *       <DataTable.Col source="author" />
         ...
 *     </DataTable>
 *   </List>
 * );
 */
export const ColumnsButton = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const resource = useResourceContext(props);
    const storeKey = props.storeKey || `${resource}.datatable`;
    const [anchorEl, setAnchorEl] = React.useState(null);
    const isRtl = useRtl();
    const translate = useTranslate();
    const isXSmall = useMediaQuery((theme) => theme.breakpoints.down('sm'));
    const title = translate('ra.action.select_columns', { _: 'Columns' });
    const handleClick = (event) => {
        setAnchorEl(event.currentTarget);
    };
    const handleClose = () => {
        setAnchorEl(null);
    };
    return (React.createElement(Root, null,
        isXSmall ? (React.createElement(Tooltip, { title: title },
            React.createElement(IconButton, { "aria-label": title, color: "primary", onClick: handleClick, size: "large", ...sanitizeRestProps(props) },
                React.createElement(ViewWeekIcon, null)))) : (React.createElement(Button, { size: "small", onClick: handleClick, startIcon: React.createElement(ViewWeekIcon, null), ...sanitizeRestProps(props) }, title)),
        React.createElement(Popover, { open: Boolean(anchorEl), keepMounted: true, anchorEl: anchorEl, onClose: handleClose, anchorOrigin: {
                vertical: 'bottom',
                horizontal: isRtl ? 'right' : 'left',
            }, transformOrigin: isRtl ? RTL_ORIGIN : LTR_ORIGIN },
            React.createElement(Box, { id: `${storeKey}-columnsSelector`, sx: { px: 1, my: 0, minWidth: 200 } }))));
};
const RTL_ORIGIN = {
    vertical: 'top',
    horizontal: 'right',
};
const LTR_ORIGIN = {
    vertical: 'top',
    horizontal: 'left',
};
const PREFIX = 'RaColumnsButton';
const Root = styled('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    '& .MuiButton-sizeSmall': {
        // fix for icon misalignment on small buttons, see https://github.com/mui/material-ui/pull/30240
        lineHeight: 1.5,
    },
});
const sanitizeRestProps = ({ resource, storeKey, ...rest }) => rest;
