// ra-ui-materialui/dist/list/datatable/DataTableHeadCell.js
import * as React from 'react';
import { FieldTitle, useTranslate, useDataTableCallbacksContext, useDataTableSortContext, useDataTableStoreContext, useResourceContext, useStore, useTranslateLabel, } from 'ra-core';
import { TableCell, TableSortLabel, Tooltip, useThemeProps, } from '@mui/material';
import { styled } from '@mui/material/styles';
import clsx from 'clsx';
import { DataTableClasses } from "./DataTableRoot.js";
const oppositeOrder = {
    ASC: 'DESC',
    DESC: 'ASC',
};
const PREFIX = 'RaDataTableHeadCell';
export const DataTableHeadCell = React.memo(React.forwardRef((inProps, ref) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { cellSx, className, cellClassName, disableSort, headerClassName, field, render, source, label, sortByOrder, ...rest } = props;
    const { storeKey, defaultHiddenColumns } = useDataTableStoreContext();
    const sort = useDataTableSortContext();
    const { handleSort } = useDataTableCallbacksContext();
    const [hiddenColumns] = useStore(storeKey, defaultHiddenColumns);
    const resource = useResourceContext();
    const translate = useTranslate();
    const translateLabel = useTranslateLabel();
    const isColumnHidden = hiddenColumns.includes(source ?? label);
    if (isColumnHidden)
        return null;
    const nextSortOrder = sort && sort.field === source
        ? oppositeOrder[sort.order]
        : sortByOrder ?? 'ASC';
    const fieldLabel = translateLabel({
        label: typeof label === 'string' ? label : undefined,
        resource,
        source,
    });
    const sortLabel = translate('ra.sort.sort_by', {
        field: fieldLabel,
        field_lower_first: typeof fieldLabel === 'string'
            ? fieldLabel.charAt(0).toLowerCase() +
                fieldLabel.slice(1)
            : undefined,
        order: translate(`ra.sort.${nextSortOrder}`),
        _: translate('ra.action.sort'),
    });
    return (React.createElement(TableCellStyled, { ref: ref, className: clsx(DataTableClasses.headerCell, className, headerClassName, `column-${source}`), variant: "head", ...rest }, handleSort && sort && !disableSort && source ? (React.createElement(Tooltip, { title: sortLabel, placement: props.align === 'right'
            ? 'bottom-end'
            : 'bottom-start', enterDelay: 300 },
        React.createElement(TableSortLabel, { active: sort.field === source, direction: sort.order === 'ASC' ? 'asc' : 'desc', "data-field": source, "data-order": sortByOrder || 'ASC', onClick: handleSort },
            React.createElement(FieldTitle, { label: label, source: source, resource: resource })))) : (React.createElement(FieldTitle, { label: label, source: source, resource: resource }))));
}));
DataTableHeadCell.displayName = 'DataTableHeadCell';
const TableCellStyled = styled(TableCell, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({}));
