// ra-ui-materialui/dist/list/datatable/DataTableRoot.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTableRoot = exports.DataTableClasses = void 0;
const material_1 = require("@mui/material");
const PREFIX = 'RaDataTable';
exports.DataTableClasses = {
    root: `${PREFIX}-root`,
    table: `${PREFIX}-table`,
    tableWrapper: `${PREFIX}-tableWrapper`,
    thead: `${PREFIX}-thead`,
    tbody: `${PREFIX}-tbody`,
    headerRow: `${PREFIX}-headerRow`,
    headerCell: `${PREFIX}-headerCell`,
    checkbox: `${PREFIX}-checkbox`,
    row: `${PREFIX}-row`,
    clickableRow: `${PREFIX}-clickableRow`,
    rowEven: `${PREFIX}-rowEven`,
    rowOdd: `${PREFIX}-rowOdd`,
    rowCell: `${PREFIX}-rowCell`,
    selectable: `${PREFIX}-selectable`,
    expandHeader: `${PREFIX}-expandHeader`,
    expandIconCell: `${PREFIX}-expandIconCell`,
    expandIcon: `${PREFIX}-expandIcon`,
    expandable: `${PREFIX}-expandable`,
    expanded: `${PREFIX}-expanded`,
    expandRow: `${PREFIX}-expandRow`,
};
exports.DataTableRoot = (0, material_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.DataTableClasses.table}`]: {
        tableLayout: 'auto',
    },
    [`& .${exports.DataTableClasses.tableWrapper}`]: {},
    [`& .${exports.DataTableClasses.thead}`]: {},
    [`& .${exports.DataTableClasses.tbody}`]: {},
    [`& .${exports.DataTableClasses.headerRow}`]: {},
    [`& .${exports.DataTableClasses.headerCell}`]: {
        position: 'sticky',
        top: 0,
        zIndex: 2,
        backgroundColor: (theme.vars || theme).palette.background.paper,
        '&:first-of-type': {
            borderTopLeftRadius: theme.shape.borderRadius,
        },
        '&:last-child': {
            borderTopRightRadius: theme.shape.borderRadius,
        },
    },
    [`& .${exports.DataTableClasses.checkbox}`]: {},
    [`& .${exports.DataTableClasses.row}`]: {},
    [`& .${exports.DataTableClasses.clickableRow}`]: {
        cursor: 'pointer',
    },
    [`& .${exports.DataTableClasses.rowEven}`]: {},
    [`& .${exports.DataTableClasses.rowOdd}`]: {},
    [`& .${exports.DataTableClasses.rowCell}`]: {},
    [`& .${exports.DataTableClasses.expandable} > td`]: {
        borderBottom: 'unset',
    },
    [`& .${exports.DataTableClasses.expandHeader}`]: {
        paddingRight: 0,
        width: theme.spacing(4),
    },
    [`& .${exports.DataTableClasses.expandIconCell}`]: {
        paddingRight: 0,
        width: theme.spacing(4),
    },
    [`& .${exports.DataTableClasses.expandIcon}`]: {
        padding: theme.spacing(1),
        transform: 'rotate(-90deg)',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.short,
        }),
    },
    [`& .${exports.DataTableClasses.expandIcon}.${exports.DataTableClasses.expanded}`]: {
        transform: 'rotate(0deg)',
    },
    [`& .${exports.DataTableClasses.expandRow}`]: {},
}));
