// ra-ui-materialui/dist/list/datatable/DataTableHeadCell.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTableHeadCell = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const DataTableRoot_1 = require("./DataTableRoot.cjs");
const oppositeOrder = {
    ASC: 'DESC',
    DESC: 'ASC',
};
const PREFIX = 'RaDataTableHeadCell';
exports.DataTableHeadCell = React.memo(React.forwardRef((inProps, ref) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { cellSx, className, cellClassName, disableSort, headerClassName, field, render, source, label, sortByOrder, ...rest } = props;
    const { storeKey, defaultHiddenColumns } = (0, ra_core_1.useDataTableStoreContext)();
    const sort = (0, ra_core_1.useDataTableSortContext)();
    const { handleSort } = (0, ra_core_1.useDataTableCallbacksContext)();
    const [hiddenColumns] = (0, ra_core_1.useStore)(storeKey, defaultHiddenColumns);
    const resource = (0, ra_core_1.useResourceContext)();
    const translate = (0, ra_core_1.useTranslate)();
    const translateLabel = (0, ra_core_1.useTranslateLabel)();
    const isColumnHidden = hiddenColumns.includes(source ?? label);
    if (isColumnHidden)
        return null;
    const nextSortOrder = sort && sort.field === source
        ? oppositeOrder[sort.order]
        : sortByOrder ?? 'ASC';
    const fieldLabel = translateLabel({
        label: typeof label === 'string' ? label : undefined,
        resource,
        source,
    });
    const sortLabel = translate('ra.sort.sort_by', {
        field: fieldLabel,
        field_lower_first: typeof fieldLabel === 'string'
            ? fieldLabel.charAt(0).toLowerCase() +
                fieldLabel.slice(1)
            : undefined,
        order: translate(`ra.sort.${nextSortOrder}`),
        _: translate('ra.action.sort'),
    });
    return (React.createElement(TableCellStyled, { ref: ref, className: (0, clsx_1.default)(DataTableRoot_1.DataTableClasses.headerCell, className, headerClassName, `column-${source}`), variant: "head", ...rest }, handleSort && sort && !disableSort && source ? (React.createElement(material_1.Tooltip, { title: sortLabel, placement: props.align === 'right'
            ? 'bottom-end'
            : 'bottom-start', enterDelay: 300 },
        React.createElement(material_1.TableSortLabel, { active: sort.field === source, direction: sort.order === 'ASC' ? 'asc' : 'desc', "data-field": source, "data-order": sortByOrder || 'ASC', onClick: handleSort },
            React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource })))) : (React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource }))));
}));
exports.DataTableHeadCell.displayName = 'DataTableHeadCell';
const TableCellStyled = (0, styles_1.styled)(material_1.TableCell, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({}));
