// ra-ui-materialui/dist/list/datatable/DataTable.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTable = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const BulkActionsToolbar_1 = require("../BulkActionsToolbar.cjs");
const button_1 = require("../../button/index.cjs");
const ListNoResults_1 = require("../ListNoResults.cjs");
const DataTableRoot_1 = require("./DataTableRoot.cjs");
const DataTableLoading_1 = require("./DataTableLoading.cjs");
const DataTableBody_1 = require("./DataTableBody.cjs");
const DataTableHead_1 = require("./DataTableHead.cjs");
const DataTableColumn_1 = require("./DataTableColumn.cjs");
const DataTableNumberColumn_1 = require("./DataTableNumberColumn.cjs");
const ColumnsSelector_1 = require("./ColumnsSelector.cjs");
const DataTableRowSxContext_1 = require("./DataTableRowSxContext.cjs");
const DefaultFoot = (_props) => null;
const PREFIX = 'RaDataTable';
/**
 * The DataTable component renders a list of records as a table.
 * It is usually used as a child of the <List> and <ReferenceManyField> components.
 *
 * Props:
 *  - body
 *  - bulkActionToolbar
 *  - bulkActionButtons
 *  - children
 *  - empty
 *  - expand
 *  - footer
 *  - header
 *  - hover
 *  - isRowExpandable
 *  - isRowSelectable
 *  - rowClick
 *  - rowSx
 *  - size
 *  - sx
 *
 * @example // Display all posts as a data table
 * const postRowSx = (record, index) => ({
 *     backgroundColor: record.nb_views >= 500 ? '#efe' : 'white',
 * });
 * export const PostList = () => (
 *     <List>
 *         <DataTable rowSx={postRowSx}>
 *             <DataTable.Col source="id" />
 *             <DataTable.Col source="title" />
 *             <DataTable.Col source="body" />
 *             <DataTable.Col>
 *                 <EditButton />
 *             </DataTable.Col>
 *         </DataTable>
 *     </List>
 * );
 *
 * @example // Display all the comments of the current post as a data table
 * <ReferenceManyField reference="comments" target="post_id">
 *     <DataTable>
 *         <DataTable.Col source="id" />
 *         <DataTable.Col source="body" />
 *         <DataTable.Col source="created_at" />
 *         <DataTable.Col>
 *             <EditButton />
 *         </DataTable.Col>
 *     </DataTable>
 * </ReferenceManyField>
 *
 * @example // Usage outside of a <List> or a <ReferenceManyField>.
 *
 * const sort = { field: 'published_at', order: 'DESC' };
 *
 * export const MyCustomList = (props) => {
 *     const { data, total, isPending } = useGetList(
 *         'posts',
 *         { pagination: { page: 1, perPage: 10 }, sort: sort }
 *     );
 *
 *     return (
 *         <DataTable
 *             data={data}
 *             total={total}
 *             isPending={isPending}
 *             sort={sort}
 *             selectedIds={[]}
 *             setSort={() => {
 *                 console.log('set sort');
 *             }}
 *             onSelect={() => {
 *                 console.log('on select');
 *             }}
 *             onToggleItem={() => {
 *                 console.log('on toggle item');
 *             }}
 *         >
 *             <DataTable.Col source="id" />
 *             <DataTable.Col source="title" />
 *         </DataTable>
 *     );
 * }
 */
exports.DataTable = React.forwardRef(function DataTable(inProps, ref) {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const resourceFromContext = (0, ra_core_1.useResourceContext)(props);
    const { canAccess: canDelete } = (0, ra_core_1.useCanAccess)({
        resource: resourceFromContext,
        action: 'delete',
    });
    const { body: TableBody = DataTableBody_1.DataTableBody, head: TableHead = DataTableHead_1.DataTableHead, foot: TableFoot = DefaultFoot, children, className, empty, expand, bulkActionsToolbar, bulkActionButtons = canDelete ? defaultBulkActionButtons : false, rowSx, size = 'small', sx, ...rest } = props;
    const { onToggleItem } = (0, ra_core_1.useListContextWithProps)(props);
    const hasBulkActions = !!bulkActionButtons !== false;
    const storeKey = props.storeKey || `${resourceFromContext}.datatable`;
    const [columnRanks] = (0, ra_core_1.useStore)(`${storeKey}_columnRanks`);
    const columns = columnRanks
        ? reorderChildren(children, columnRanks)
        : children;
    const loading = (React.createElement(DataTableLoading_1.DataTableLoading, { className: className, expand: expand, hasBulkActions: hasBulkActions, nbChildren: React.Children.count(children), size: size }));
    return (React.createElement(ra_core_1.DataTableBase, { ...props, hasBulkActions: hasBulkActions, loading: loading, empty: empty === undefined ? (React.createElement(ListNoResults_1.ListNoResults, { resource: resourceFromContext })) : (empty) },
        React.createElement(DataTableRowSxContext_1.DataTableRowSxContext.Provider, { value: rowSx },
            React.createElement(DataTableRoot_1.DataTableRoot, { sx: sx, className: (0, clsx_1.default)(DataTableRoot_1.DataTableClasses.root, className) },
                onToggleItem &&
                    (bulkActionsToolbar ??
                        (bulkActionButtons !== false && (React.createElement(BulkActionsToolbar_1.BulkActionsToolbar, null, (0, react_1.isValidElement)(bulkActionButtons)
                            ? bulkActionButtons
                            : defaultBulkActionButtons)))),
                React.createElement("div", { className: DataTableRoot_1.DataTableClasses.tableWrapper },
                    React.createElement(material_1.Table, { ref: ref, className: DataTableRoot_1.DataTableClasses.table, size: size, ...sanitizeRestProps(rest) },
                        React.createElement(ra_core_1.DataTableRenderContext.Provider, { value: "header" },
                            React.createElement(TableHead, null, columns)),
                        React.createElement(TableBody, null, columns),
                        React.createElement(ra_core_1.DataTableRenderContext.Provider, { value: "footer" },
                            React.createElement(TableFoot, null, columns)))))),
        React.createElement(ra_core_1.DataTableRenderContext.Provider, { value: "columnsSelector" },
            React.createElement(ColumnsSelector_1.ColumnsSelector, null, children))));
});
exports.DataTable.Col = DataTableColumn_1.DataTableColumn;
exports.DataTable.NumberCol = DataTableNumberColumn_1.DataTableNumberColumn;
exports.DataTable.displayName = 'DataTable';
const defaultBulkActionButtons = React.createElement(button_1.BulkDeleteButton, null);
/**
 * Reorder children based on columnRanks
 *
 * Note that columnRanks may be shorter than the number of children
 */
const reorderChildren = (children, columnRanks) => React.Children.toArray(children).reduce((acc, child, index) => {
    const rank = columnRanks.indexOf(index);
    if (rank === -1) {
        // if the column is not in columnRanks, keep it at the same index
        acc[index] = child;
    }
    else {
        // if the column is in columnRanks, move it to the rank index
        acc[rank] = child;
    }
    return acc;
}, []);
const sanitizeRestProps = ({ sort, data, expandSingle, hiddenColumns, hover, isLoading, isPending, isRowExpandable, isRowSelectable, onSelect, onToggleItem, resource, rowClick, setSort, selectedIds, storeKey, total, ...rest }) => rest;
