// ra-ui-materialui/dist/list/datatable/DataTable.js
import * as React from 'react';
import { isValidElement, } from 'react';
import { DataTableBase, DataTableRenderContext, useCanAccess, useListContextWithProps, useResourceContext, useStore, } from 'ra-core';
import { Table, useThemeProps, } from '@mui/material';
import clsx from 'clsx';
import { BulkActionsToolbar } from "../BulkActionsToolbar.js";
import { BulkDeleteButton } from "../../button/index.js";
import { ListNoResults } from "../ListNoResults.js";
import { DataTableClasses, DataTableRoot } from "./DataTableRoot.js";
import { DataTableLoading } from "./DataTableLoading.js";
import { DataTableBody } from "./DataTableBody.js";
import { DataTableHead } from "./DataTableHead.js";
import { DataTableColumn } from "./DataTableColumn.js";
import { DataTableNumberColumn } from "./DataTableNumberColumn.js";
import { ColumnsSelector } from "./ColumnsSelector.js";
import { DataTableRowSxContext } from "./DataTableRowSxContext.js";
const DefaultFoot = (_props) => null;
const PREFIX = 'RaDataTable';
/**
 * The DataTable component renders a list of records as a table.
 * It is usually used as a child of the <List> and <ReferenceManyField> components.
 *
 * Props:
 *  - body
 *  - bulkActionToolbar
 *  - bulkActionButtons
 *  - children
 *  - empty
 *  - expand
 *  - footer
 *  - header
 *  - hover
 *  - isRowExpandable
 *  - isRowSelectable
 *  - rowClick
 *  - rowSx
 *  - size
 *  - sx
 *
 * @example // Display all posts as a data table
 * const postRowSx = (record, index) => ({
 *     backgroundColor: record.nb_views >= 500 ? '#efe' : 'white',
 * });
 * export const PostList = () => (
 *     <List>
 *         <DataTable rowSx={postRowSx}>
 *             <DataTable.Col source="id" />
 *             <DataTable.Col source="title" />
 *             <DataTable.Col source="body" />
 *             <DataTable.Col>
 *                 <EditButton />
 *             </DataTable.Col>
 *         </DataTable>
 *     </List>
 * );
 *
 * @example // Display all the comments of the current post as a data table
 * <ReferenceManyField reference="comments" target="post_id">
 *     <DataTable>
 *         <DataTable.Col source="id" />
 *         <DataTable.Col source="body" />
 *         <DataTable.Col source="created_at" />
 *         <DataTable.Col>
 *             <EditButton />
 *         </DataTable.Col>
 *     </DataTable>
 * </ReferenceManyField>
 *
 * @example // Usage outside of a <List> or a <ReferenceManyField>.
 *
 * const sort = { field: 'published_at', order: 'DESC' };
 *
 * export const MyCustomList = (props) => {
 *     const { data, total, isPending } = useGetList(
 *         'posts',
 *         { pagination: { page: 1, perPage: 10 }, sort: sort }
 *     );
 *
 *     return (
 *         <DataTable
 *             data={data}
 *             total={total}
 *             isPending={isPending}
 *             sort={sort}
 *             selectedIds={[]}
 *             setSort={() => {
 *                 console.log('set sort');
 *             }}
 *             onSelect={() => {
 *                 console.log('on select');
 *             }}
 *             onToggleItem={() => {
 *                 console.log('on toggle item');
 *             }}
 *         >
 *             <DataTable.Col source="id" />
 *             <DataTable.Col source="title" />
 *         </DataTable>
 *     );
 * }
 */
export const DataTable = React.forwardRef(function DataTable(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const resourceFromContext = useResourceContext(props);
    const { canAccess: canDelete } = useCanAccess({
        resource: resourceFromContext,
        action: 'delete',
    });
    const { body: TableBody = DataTableBody, head: TableHead = DataTableHead, foot: TableFoot = DefaultFoot, children, className, empty, expand, bulkActionsToolbar, bulkActionButtons = canDelete ? defaultBulkActionButtons : false, rowSx, size = 'small', sx, ...rest } = props;
    const { onToggleItem } = useListContextWithProps(props);
    const hasBulkActions = !!bulkActionButtons !== false;
    const storeKey = props.storeKey || `${resourceFromContext}.datatable`;
    const [columnRanks] = useStore(`${storeKey}_columnRanks`);
    const columns = columnRanks
        ? reorderChildren(children, columnRanks)
        : children;
    const loading = (React.createElement(DataTableLoading, { className: className, expand: expand, hasBulkActions: hasBulkActions, nbChildren: React.Children.count(children), size: size }));
    return (React.createElement(DataTableBase, { ...props, hasBulkActions: hasBulkActions, loading: loading, empty: empty === undefined ? (React.createElement(ListNoResults, { resource: resourceFromContext })) : (empty) },
        React.createElement(DataTableRowSxContext.Provider, { value: rowSx },
            React.createElement(DataTableRoot, { sx: sx, className: clsx(DataTableClasses.root, className) },
                onToggleItem &&
                    (bulkActionsToolbar ??
                        (bulkActionButtons !== false && (React.createElement(BulkActionsToolbar, null, isValidElement(bulkActionButtons)
                            ? bulkActionButtons
                            : defaultBulkActionButtons)))),
                React.createElement("div", { className: DataTableClasses.tableWrapper },
                    React.createElement(Table, { ref: ref, className: DataTableClasses.table, size: size, ...sanitizeRestProps(rest) },
                        React.createElement(DataTableRenderContext.Provider, { value: "header" },
                            React.createElement(TableHead, null, columns)),
                        React.createElement(TableBody, null, columns),
                        React.createElement(DataTableRenderContext.Provider, { value: "footer" },
                            React.createElement(TableFoot, null, columns)))))),
        React.createElement(DataTableRenderContext.Provider, { value: "columnsSelector" },
            React.createElement(ColumnsSelector, null, children))));
});
DataTable.Col = DataTableColumn;
DataTable.NumberCol = DataTableNumberColumn;
DataTable.displayName = 'DataTable';
const defaultBulkActionButtons = React.createElement(BulkDeleteButton, null);
/**
 * Reorder children based on columnRanks
 *
 * Note that columnRanks may be shorter than the number of children
 */
const reorderChildren = (children, columnRanks) => React.Children.toArray(children).reduce((acc, child, index) => {
    const rank = columnRanks.indexOf(index);
    if (rank === -1) {
        // if the column is not in columnRanks, keep it at the same index
        acc[index] = child;
    }
    else {
        // if the column is in columnRanks, move it to the rank index
        acc[rank] = child;
    }
    return acc;
}, []);
const sanitizeRestProps = ({ sort, data, expandSingle, hiddenColumns, hover, isLoading, isPending, isRowExpandable, isRowSelectable, onSelect, onToggleItem, resource, rowClick, setSort, selectedIds, storeKey, total, ...rest }) => rest;
