// ra-ui-materialui/dist/auth/AuthLayout.js
import * as React from 'react';
import { Card, styled, } from '@mui/material';
import { useThemeProps } from '@mui/system';
import clsx from 'clsx';
export const AuthLayout = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, ...rest } = props;
    return (React.createElement(Root, { className: clsx(AuthLayoutClasses.root, className), ...rest },
        React.createElement(Card, { className: AuthLayoutClasses.card }, children)));
};
const PREFIX = 'RaAuthLayout';
export const AuthLayoutClasses = {
    root: `${PREFIX}-root`,
    card: `${PREFIX}-card`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    display: 'flex',
    flexDirection: 'column',
    minHeight: '100vh',
    height: '1px',
    alignItems: 'center',
    justifyContent: 'flex-start',
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'cover',
    backgroundImage: 'radial-gradient(circle at 50% 14em, #313264 0%, #00023b 60%, #00023b 100%)',
    [`& .${AuthLayoutClasses.card}`]: {
        minWidth: 300,
        marginTop: '6em',
    },
}));
