// ra-ui-materialui/dist/auth/AuthError.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material';
import LockIcon from '@mui/icons-material/Lock';
import { useTranslate, LinkBase } from 'ra-core';
import { Button } from "../button/index.js";
export const AuthError = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, title = 'ra.page.error', message = 'ra.message.auth_error', ...rest } = props;
    const translate = useTranslate();
    return (React.createElement(Root, { className: className, ...rest },
        React.createElement("div", { className: AuthErrorClasses.message },
            React.createElement("h1", null, translate(title, { _: title })),
            React.createElement("div", null, translate(message, { _: message })),
            React.createElement(Button, { component: LinkBase, to: "/login", label: "ra.auth.sign_in" },
                React.createElement(LockIcon, null)))));
};
const PREFIX = 'RaAuthError';
export const AuthErrorClasses = {
    root: `${PREFIX}-root`,
    message: `${PREFIX}-message`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    [theme.breakpoints.up('md')]: {
        height: '100%',
    },
    [theme.breakpoints.down('xl')]: {
        height: '100vh',
        marginTop: '-3em',
    },
    [`& .${AuthErrorClasses.message}`]: {
        textAlign: 'center',
        fontFamily: 'Roboto, sans-serif',
        opacity: 0.5,
        margin: '0 1em',
    },
}));
