// ra-ui-materialui/dist/auth/AuthLayout.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthLayoutClasses = exports.AuthLayout = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const system_1 = require("@mui/system");
const clsx_1 = __importDefault(require("clsx"));
const AuthLayout = (inProps) => {
    const props = (0, system_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, ...rest } = props;
    return (React.createElement(Root, { className: (0, clsx_1.default)(exports.AuthLayoutClasses.root, className), ...rest },
        React.createElement(material_1.Card, { className: exports.AuthLayoutClasses.card }, children)));
};
exports.AuthLayout = AuthLayout;
const PREFIX = 'RaAuthLayout';
exports.AuthLayoutClasses = {
    root: `${PREFIX}-root`,
    card: `${PREFIX}-card`,
};
const Root = (0, material_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    display: 'flex',
    flexDirection: 'column',
    minHeight: '100vh',
    height: '1px',
    alignItems: 'center',
    justifyContent: 'flex-start',
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'cover',
    backgroundImage: 'radial-gradient(circle at 50% 14em, #313264 0%, #00023b 60%, #00023b 100%)',
    [`& .${exports.AuthLayoutClasses.card}`]: {
        minWidth: 300,
        marginTop: '6em',
    },
}));
