// ra-ui-materialui/dist/auth/LoginForm.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoginFormClasses = exports.LoginForm = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const input_1 = require("../input/index.cjs");
const LoginForm = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { redirectTo, className, sx, children } = props;
    const [loading, setLoading] = React.useState(false);
    const login = (0, ra_core_1.useLogin)();
    const translate = (0, ra_core_1.useTranslate)();
    const notify = (0, ra_core_1.useNotify)();
    const submit = (values) => {
        setLoading(true);
        login(values, redirectTo)
            .then(() => {
            setLoading(false);
        })
            .catch(error => {
            setLoading(false);
            notify(typeof error === 'string'
                ? error
                : typeof error === 'undefined' || !error.message
                    ? 'ra.auth.sign_in_error'
                    : error.message, {
                type: 'error',
                messageArgs: {
                    _: typeof error === 'string'
                        ? error
                        : error && error.message
                            ? error.message
                            : undefined,
                },
            });
        });
    };
    return (React.createElement(StyledForm, { onSubmit: submit, mode: "onChange", noValidate: true, className: className, sx: sx },
        React.createElement(material_1.CardContent, { className: exports.LoginFormClasses.content },
            children || (React.createElement(React.Fragment, null,
                React.createElement(input_1.TextInput, { autoFocus: true, source: "username", label: translate('ra.auth.username'), autoComplete: "username", validate: (0, ra_core_1.required)() }),
                React.createElement(input_1.PasswordInput, { source: "password", label: translate('ra.auth.password'), autoComplete: "current-password", validate: (0, ra_core_1.required)() }))),
            React.createElement(material_1.Button, { variant: "contained", type: "submit", color: "primary", disabled: loading, fullWidth: true, className: exports.LoginFormClasses.button }, loading ? (React.createElement(material_1.CircularProgress, { className: exports.LoginFormClasses.icon, size: 19, thickness: 3 })) : (translate('ra.auth.sign_in'))))));
};
exports.LoginForm = LoginForm;
const PREFIX = 'RaLoginForm';
exports.LoginFormClasses = {
    content: `${PREFIX}-content`,
    button: `${PREFIX}-button`,
    icon: `${PREFIX}-icon`,
};
const StyledForm = (0, styles_1.styled)(ra_core_1.Form, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.LoginFormClasses.content}`]: {
        width: 300,
        paddingBottom: `${theme.spacing(2)}!important`,
    },
    [`& .${exports.LoginFormClasses.button}`]: {
        marginTop: theme.spacing(2),
    },
    [`& .${exports.LoginFormClasses.icon}`]: {
        margin: theme.spacing(0.3),
    },
}));
