// ra-ui-materialui/dist/auth/LoginForm.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Button, CardContent, CircularProgress } from '@mui/material';
import { Form, required, useTranslate, useLogin, useNotify } from 'ra-core';
import { PasswordInput, TextInput } from "../input/index.js";
export const LoginForm = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { redirectTo, className, sx, children } = props;
    const [loading, setLoading] = React.useState(false);
    const login = useLogin();
    const translate = useTranslate();
    const notify = useNotify();
    const submit = (values) => {
        setLoading(true);
        login(values, redirectTo)
            .then(() => {
            setLoading(false);
        })
            .catch(error => {
            setLoading(false);
            notify(typeof error === 'string'
                ? error
                : typeof error === 'undefined' || !error.message
                    ? 'ra.auth.sign_in_error'
                    : error.message, {
                type: 'error',
                messageArgs: {
                    _: typeof error === 'string'
                        ? error
                        : error && error.message
                            ? error.message
                            : undefined,
                },
            });
        });
    };
    return (React.createElement(StyledForm, { onSubmit: submit, mode: "onChange", noValidate: true, className: className, sx: sx },
        React.createElement(CardContent, { className: LoginFormClasses.content },
            children || (React.createElement(React.Fragment, null,
                React.createElement(TextInput, { autoFocus: true, source: "username", label: translate('ra.auth.username'), autoComplete: "username", validate: required() }),
                React.createElement(PasswordInput, { source: "password", label: translate('ra.auth.password'), autoComplete: "current-password", validate: required() }))),
            React.createElement(Button, { variant: "contained", type: "submit", color: "primary", disabled: loading, fullWidth: true, className: LoginFormClasses.button }, loading ? (React.createElement(CircularProgress, { className: LoginFormClasses.icon, size: 19, thickness: 3 })) : (translate('ra.auth.sign_in'))))));
};
const PREFIX = 'RaLoginForm';
export const LoginFormClasses = {
    content: `${PREFIX}-content`,
    button: `${PREFIX}-button`,
    icon: `${PREFIX}-icon`,
};
const StyledForm = styled(Form, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${LoginFormClasses.content}`]: {
        width: 300,
        paddingBottom: `${theme.spacing(2)}!important`,
    },
    [`& .${LoginFormClasses.button}`]: {
        marginTop: theme.spacing(2),
    },
    [`& .${LoginFormClasses.icon}`]: {
        margin: theme.spacing(0.3),
    },
}));
