// ra-ui-materialui/dist/auth/Logout.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { useCallback } from 'react';
import { ListItemIcon, ListItemText, MenuItem, useMediaQuery, } from '@mui/material';
import ExitIcon from '@mui/icons-material/PowerSettingsNew';
import clsx from 'clsx';
import { useTranslate, useLogout, useAuthState } from 'ra-core';
/**
 * Logout button component, to be passed to the Admin component
 *
 * Used for the Logout Menu item in the sidebar
 */
export const Logout = React.forwardRef(function Logout(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, redirectTo, icon, ...rest } = props;
    const { authenticated } = useAuthState();
    const isXSmall = useMediaQuery((theme) => theme.breakpoints.down('sm'));
    const translate = useTranslate();
    const logout = useLogout();
    const handleClick = useCallback(() => logout(null, redirectTo, false), [redirectTo, logout]);
    if (!authenticated)
        return null;
    return (React.createElement(StyledMenuItem, { className: clsx('logout', className), onClick: handleClick, ref: ref, component: isXSmall ? 'span' : 'li', ...rest },
        React.createElement(ListItemIcon, { className: LogoutClasses.icon }, icon ? icon : React.createElement(ExitIcon, { fontSize: "small" })),
        React.createElement(ListItemText, null, translate('ra.auth.logout', { _: 'Logout' }))));
});
const PREFIX = 'RaLogout';
export const LogoutClasses = {
    icon: `${PREFIX}-icon`,
};
const StyledMenuItem = styled(MenuItem, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${LogoutClasses.icon}`]: {},
});
