// ra-ui-materialui/dist/auth/Logout.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogoutClasses = exports.Logout = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const react_1 = require("react");
const material_1 = require("@mui/material");
const PowerSettingsNew_1 = __importDefault(require("@mui/icons-material/PowerSettingsNew"));
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
/**
 * Logout button component, to be passed to the Admin component
 *
 * Used for the Logout Menu item in the sidebar
 */
exports.Logout = React.forwardRef(function Logout(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, redirectTo, icon, ...rest } = props;
    const { authenticated } = (0, ra_core_1.useAuthState)();
    const isXSmall = (0, material_1.useMediaQuery)((theme) => theme.breakpoints.down('sm'));
    const translate = (0, ra_core_1.useTranslate)();
    const logout = (0, ra_core_1.useLogout)();
    const handleClick = (0, react_1.useCallback)(() => logout(null, redirectTo, false), [redirectTo, logout]);
    if (!authenticated)
        return null;
    return (React.createElement(StyledMenuItem, { className: (0, clsx_1.default)('logout', className), onClick: handleClick, ref: ref, component: isXSmall ? 'span' : 'li', ...rest },
        React.createElement(material_1.ListItemIcon, { className: exports.LogoutClasses.icon }, icon ? icon : React.createElement(PowerSettingsNew_1.default, { fontSize: "small" })),
        React.createElement(material_1.ListItemText, null, translate('ra.auth.logout', { _: 'Logout' }))));
});
const PREFIX = 'RaLogout';
exports.LogoutClasses = {
    icon: `${PREFIX}-icon`,
};
const StyledMenuItem = (0, styles_1.styled)(material_1.MenuItem, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${exports.LogoutClasses.icon}`]: {},
});
