// ra-core/dist/controller/button/useDeleteWithConfirmController.js
import { useState } from 'react';
import { useDeleteController, useUnselect, } from "..//index.js";
import { useRedirect } from "../../routing/index.js";
import { useNotify } from "../../notification/index.js";
import { useResourceContext } from "../../core/index.js";
import { useTranslate } from "../../i18n/index.js";
import { useEvent } from "../../util/index.js";
/**
 * Prepare a set of callbacks for a delete button guarded by confirmation dialog
 * @deprecated prefer the useDeleteController hook instead
 *
 * @example
 *
 * const DeleteButton = ({
 *     resource,
 *     record,
 *     redirect,
 *     onClick,
 *     ...rest
 * }) => {
 *     const {
 *         open,
 *         isPending,
 *         handleDialogOpen,
 *         handleDialogClose,
 *         handleDelete,
 *     } = useDeleteWithConfirmController({
 *         resource,
 *         record,
 *         redirect,
 *         onClick,
 *     });
 *
 *     return (
 *         <Fragment>
 *             <Button
 *                 onClick={handleDialogOpen}
 *                 label="ra.action.delete"
 *                 {...rest}
 *             >
 *                 {icon}
 *             </Button>
 *             <Confirm
 *                 isOpen={open}
 *                 loading={isPending}
 *                 title="ra.message.delete_title"
 *                 content="ra.message.delete_content"
 *                 titleTranslateOptions={{
 *                     name: resource,
 *                     id: record.id,
 *                 }}
 *                 contentTranslateOptions={{
 *                     name: resource,
 *                     id: record.id,
 *                 }}
 *                 onConfirm={handleDelete}
 *                 onClose={handleDialogClose}
 *             />
 *         </Fragment>
 *     );
 * };
 */
const useDeleteWithConfirmController = (props) => {
    const { mutationMode, onClick, record, redirect: redirectTo = 'list', successMessage, mutationOptions = {}, ...rest } = props;
    const [open, setOpen] = useState(false);
    const resource = useResourceContext(props);
    const notify = useNotify();
    const unselect = useUnselect(resource);
    const redirect = useRedirect();
    const translate = useTranslate();
    const { isPending, handleDelete: controllerHandleDelete } = useDeleteController({
        mutationMode,
        mutationOptions: {
            onSuccess: () => {
                setOpen(false);
                notify(successMessage ??
                    `resources.${resource}.notifications.deleted`, {
                    type: 'info',
                    messageArgs: {
                        smart_count: 1,
                        _: translate('ra.notification.deleted', {
                            smart_count: 1,
                        }),
                    },
                    undoable: mutationMode === 'undoable',
                });
                record && unselect([record.id], true);
                redirect(redirectTo, resource);
            },
            onError: error => {
                setOpen(false);
                notify(typeof error === 'string'
                    ? error
                    : error?.message ||
                        'ra.notification.http_error', {
                    type: 'error',
                    messageArgs: {
                        _: typeof error === 'string'
                            ? error
                            : error?.message
                                ? error.message
                                : undefined,
                    },
                });
            },
            ...mutationOptions,
        },
        record,
        redirect: redirectTo,
        successMessage,
        ...rest,
    });
    const handleDialogOpen = useEvent((e) => {
        e.stopPropagation();
        setOpen(true);
    });
    const handleDialogClose = useEvent((e) => {
        e.stopPropagation();
        setOpen(false);
    });
    const handleDelete = useEvent((event) => {
        if (event && event.stopPropagation) {
            event.stopPropagation();
        }
        controllerHandleDelete();
        if (typeof onClick === 'function') {
            onClick(event);
        }
    });
    return {
        open,
        isPending,
        isLoading: isPending,
        handleDialogOpen,
        handleDialogClose,
        handleDelete,
    };
};
export default useDeleteWithConfirmController;
