// ra-core/dist/controller/button/useDeleteWithConfirmController.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("react");
const __1 = require("..//index.cjs");
const routing_1 = require("../../routing/index.cjs");
const notification_1 = require("../../notification/index.cjs");
const core_1 = require("../../core/index.cjs");
const i18n_1 = require("../../i18n/index.cjs");
const util_1 = require("../../util/index.cjs");
/**
 * Prepare a set of callbacks for a delete button guarded by confirmation dialog
 * @deprecated prefer the useDeleteController hook instead
 *
 * @example
 *
 * const DeleteButton = ({
 *     resource,
 *     record,
 *     redirect,
 *     onClick,
 *     ...rest
 * }) => {
 *     const {
 *         open,
 *         isPending,
 *         handleDialogOpen,
 *         handleDialogClose,
 *         handleDelete,
 *     } = useDeleteWithConfirmController({
 *         resource,
 *         record,
 *         redirect,
 *         onClick,
 *     });
 *
 *     return (
 *         <Fragment>
 *             <Button
 *                 onClick={handleDialogOpen}
 *                 label="ra.action.delete"
 *                 {...rest}
 *             >
 *                 {icon}
 *             </Button>
 *             <Confirm
 *                 isOpen={open}
 *                 loading={isPending}
 *                 title="ra.message.delete_title"
 *                 content="ra.message.delete_content"
 *                 titleTranslateOptions={{
 *                     name: resource,
 *                     id: record.id,
 *                 }}
 *                 contentTranslateOptions={{
 *                     name: resource,
 *                     id: record.id,
 *                 }}
 *                 onConfirm={handleDelete}
 *                 onClose={handleDialogClose}
 *             />
 *         </Fragment>
 *     );
 * };
 */
const useDeleteWithConfirmController = (props) => {
    const { mutationMode, onClick, record, redirect: redirectTo = 'list', successMessage, mutationOptions = {}, ...rest } = props;
    const [open, setOpen] = (0, react_1.useState)(false);
    const resource = (0, core_1.useResourceContext)(props);
    const notify = (0, notification_1.useNotify)();
    const unselect = (0, __1.useUnselect)(resource);
    const redirect = (0, routing_1.useRedirect)();
    const translate = (0, i18n_1.useTranslate)();
    const { isPending, handleDelete: controllerHandleDelete } = (0, __1.useDeleteController)({
        mutationMode,
        mutationOptions: {
            onSuccess: () => {
                setOpen(false);
                notify(successMessage ??
                    `resources.${resource}.notifications.deleted`, {
                    type: 'info',
                    messageArgs: {
                        smart_count: 1,
                        _: translate('ra.notification.deleted', {
                            smart_count: 1,
                        }),
                    },
                    undoable: mutationMode === 'undoable',
                });
                record && unselect([record.id], true);
                redirect(redirectTo, resource);
            },
            onError: error => {
                setOpen(false);
                notify(typeof error === 'string'
                    ? error
                    : error?.message ||
                        'ra.notification.http_error', {
                    type: 'error',
                    messageArgs: {
                        _: typeof error === 'string'
                            ? error
                            : error?.message
                                ? error.message
                                : undefined,
                    },
                });
            },
            ...mutationOptions,
        },
        record,
        redirect: redirectTo,
        successMessage,
        ...rest,
    });
    const handleDialogOpen = (0, util_1.useEvent)((e) => {
        e.stopPropagation();
        setOpen(true);
    });
    const handleDialogClose = (0, util_1.useEvent)((e) => {
        e.stopPropagation();
        setOpen(false);
    });
    const handleDelete = (0, util_1.useEvent)((event) => {
        if (event && event.stopPropagation) {
            event.stopPropagation();
        }
        controllerHandleDelete();
        if (typeof onClick === 'function') {
            onClick(event);
        }
    });
    return {
        open,
        isPending,
        isLoading: isPending,
        handleDialogOpen,
        handleDialogClose,
        handleDelete,
    };
};
exports.default = useDeleteWithConfirmController;
module.exports = exports.default;
