// ra-core/dist/controller/button/useUpdateController.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useUpdateController = void 0;
const react_1 = require("react");
const useUpdate_1 = require("../../dataProvider/useUpdate.cjs");
const useRedirect_1 = require("../../routing/useRedirect.cjs");
const useNotify_1 = require("../../notification/useNotify.cjs");
const useRecordContext_1 = require("../record/useRecordContext.cjs");
const useResourceContext_1 = require("../../core/useResourceContext.cjs");
const useTranslate_1 = require("../../i18n/useTranslate.cjs");
/**
 * Prepare a set of callbacks for an update button
 *
 * @example
 * const UpdateButton = ({
 *     resource,
 *     record,
 *     redirect,
 *     ...rest
 * }) => {
 *     const {
 *         isPending,
 *         handleUpdate,
 *     } = useUpdateController({
 *         mutationMode: 'pessimistic',
 *         resource,
 *         record,
 *         redirect,
 *     });
 *
 *     const [open, setOpen] = useState(false);
 *
 *     return (
 *         <Fragment>
 *             <Button
 *                 onClick={() => setOpen(true)}
 *                 label="ra.action.update"
 *                 {...rest}
 *             >
 *                 {icon}
 *             </Button>
 *             <Confirm
 *                 isOpen={open}
 *                 loading={isPending}
 *                 title="ra.message.update_title"
 *                 content="ra.message.update_content"
 *                 titleTranslateOptions={{
 *                     name: resource,
 *                     id: record.id,
 *                 }}
 *                 contentTranslateOptions={{
 *                     name: resource,
 *                     id: record.id,
 *                 }}
 *                 onConfirm={() => handleUpdate()}
 *                 onClose={() => setOpen(false)}
 *             />
 *         </Fragment>
 *     );
 * };
 */
const useUpdateController = (props) => {
    const { redirect: redirectTo = false, mutationMode = 'undoable', mutationOptions = {}, successMessage, } = props;
    const { meta: mutationMeta, ...otherMutationOptions } = mutationOptions;
    const record = (0, useRecordContext_1.useRecordContext)(props);
    const resource = (0, useResourceContext_1.useResourceContext)(props);
    const notify = (0, useNotify_1.useNotify)();
    const redirect = (0, useRedirect_1.useRedirect)();
    const translate = (0, useTranslate_1.useTranslate)();
    const [updateOne, { isPending }] = (0, useUpdate_1.useUpdate)(resource, undefined, {
        onSuccess: () => {
            notify(successMessage ??
                `resources.${resource}.notifications.updated`, {
                type: 'info',
                messageArgs: {
                    smart_count: 1,
                    _: translate('ra.notification.updated', {
                        smart_count: 1,
                    }),
                },
                undoable: mutationMode === 'undoable',
            });
            redirect(redirectTo, resource);
        },
        onError: (error) => {
            notify(typeof error === 'string'
                ? error
                : error?.message || 'ra.notification.http_error', {
                type: 'error',
                messageArgs: {
                    _: typeof error === 'string'
                        ? error
                        : error?.message,
                },
            });
        },
        mutationMode,
        ...otherMutationOptions,
    });
    const handleUpdate = (0, react_1.useCallback)((data) => {
        if (!record) {
            throw new Error('The record cannot be updated because no record has been passed');
        }
        updateOne(resource, {
            id: record.id,
            data,
            previousData: record,
            meta: mutationMeta,
        });
    }, [updateOne, mutationMeta, record, resource]);
    return (0, react_1.useMemo)(() => ({
        isPending,
        isLoading: isPending,
        handleUpdate,
    }), [isPending, handleUpdate]);
};
exports.useUpdateController = useUpdateController;
