// ra-ui-materialui/dist/detail/CreateView.js
import * as React from 'react';
import { Card, styled, } from '@mui/material';
import { useCreateContext } from 'ra-core';
import clsx from 'clsx';
import { Title } from "../layout/index.js";
export const CreateView = (props) => {
    const { actions, aside, children, render, className, component: Content = Card, title, ...rest } = props;
    const createContext = useCreateContext();
    const { resource, defaultTitle } = createContext;
    return (React.createElement(Root, { className: clsx('create-page', className), ...rest },
        title !== false && (React.createElement(Title, { title: title, defaultTitle: defaultTitle, preferenceKey: `${resource}.create.title` })),
        actions,
        React.createElement("div", { className: clsx(CreateClasses.main, {
                [CreateClasses.noActions]: !actions,
            }) },
            React.createElement(Content, { className: CreateClasses.card }, render ? render(createContext) : children),
            aside)));
};
const PREFIX = 'RaCreate';
export const CreateClasses = {
    main: `${PREFIX}-main`,
    noActions: `${PREFIX}-noActions`,
    card: `${PREFIX}-card`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${CreateClasses.main}`]: {
        display: 'flex',
    },
    [`& .${CreateClasses.noActions}`]: {
        [theme.breakpoints.up('sm')]: {
            marginTop: '1em',
        },
    },
    [`& .${CreateClasses.card}`]: {
        flex: '1 1 auto',
    },
}));
