// ra-ui-materialui/dist/detail/CreateView.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateClasses = exports.CreateView = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const clsx_1 = __importDefault(require("clsx"));
const layout_1 = require("../layout/index.cjs");
const CreateView = (props) => {
    const { actions, aside, children, render, className, component: Content = material_1.Card, title, ...rest } = props;
    const createContext = (0, ra_core_1.useCreateContext)();
    const { resource, defaultTitle } = createContext;
    return (React.createElement(Root, { className: (0, clsx_1.default)('create-page', className), ...rest },
        title !== false && (React.createElement(layout_1.Title, { title: title, defaultTitle: defaultTitle, preferenceKey: `${resource}.create.title` })),
        actions,
        React.createElement("div", { className: (0, clsx_1.default)(exports.CreateClasses.main, {
                [exports.CreateClasses.noActions]: !actions,
            }) },
            React.createElement(Content, { className: exports.CreateClasses.card }, render ? render(createContext) : children),
            aside)));
};
exports.CreateView = CreateView;
const PREFIX = 'RaCreate';
exports.CreateClasses = {
    main: `${PREFIX}-main`,
    noActions: `${PREFIX}-noActions`,
    card: `${PREFIX}-card`,
};
const Root = (0, material_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.CreateClasses.main}`]: {
        display: 'flex',
    },
    [`& .${exports.CreateClasses.noActions}`]: {
        [theme.breakpoints.up('sm')]: {
            marginTop: '1em',
        },
    },
    [`& .${exports.CreateClasses.card}`]: {
        flex: '1 1 auto',
    },
}));
