// ra-ui-materialui/dist/detail/editFieldTypes.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.editFieldTypes = void 0;
const React = __importStar(require("react"));
const form_1 = require("../form/index.cjs");
const input_1 = require("../input/index.cjs");
exports.editFieldTypes = {
    form: {
        component: form_1.SimpleForm,
        representation: (_, children) => `        <SimpleForm>
${children.map(child => `            ${child.getRepresentation()}`).join('\n')}
        </SimpleForm>`,
    },
    array: {
        component: ({ children, ...props }) => (React.createElement(input_1.ArrayInput, { ...props },
            React.createElement(input_1.SimpleFormIterator, null, children))),
        representation: (props, children) => `<ArrayInput source="${props.source}"><SimpleFormIterator>${children
            .map(child => child.getRepresentation())
            .join('\n')}</SimpleFormIterator></ArrayInput>`,
    },
    scalar_array: {
        component: input_1.TextArrayInput,
        representation: (props) => `<TextArrayInput source="${props.source}" />`,
    },
    boolean: {
        component: input_1.BooleanInput,
        representation: (props) => `<BooleanInput source="${props.source}" />`,
    },
    date: {
        component: input_1.DateInput,
        representation: (props) => `<DateInput source="${props.source}" />`,
    },
    email: {
        component: input_1.TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    id: {
        component: input_1.TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    number: {
        component: input_1.NumberInput,
        representation: (props) => `<NumberInput source="${props.source}" />`,
    },
    reference: {
        component: input_1.ReferenceInput,
        representation: (props) => `<ReferenceInput source="${props.source}" reference="${props.reference}" />`,
    },
    referenceArray: {
        component: input_1.ReferenceArrayInput,
        representation: (props) => `<ReferenceArrayInput source="${props.source}" reference="${props.reference}" />`,
    },
    referenceArrayChild: {
        component: (props) => (React.createElement(input_1.SelectInput, { optionText: "id", ...props })),
        representation: () => `<SelectInput optionText="id" />`,
    },
    richText: {
        component: input_1.TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    string: {
        component: input_1.TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    url: {
        component: input_1.TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
};
