// ra-ui-materialui/dist/detail/EditView.js
import * as React from 'react';
import { Card, CardContent, styled, } from '@mui/material';
import clsx from 'clsx';
import { useEditContext, useResourceDefinition, } from 'ra-core';
import { EditActions } from "./EditActions.js";
import { Title } from "../layout/index.js";
import { Offline } from "../Offline.js";
const defaultActions = React.createElement(EditActions, null);
const defaultOffline = React.createElement(Offline, null);
export const EditView = (props) => {
    const { actions, aside, children, render, className, component: Content = Card, emptyWhileLoading = false, offline = defaultOffline, error, title, ...rest } = props;
    const { hasShow } = useResourceDefinition();
    const editContext = useEditContext();
    const { resource, defaultTitle, record, isPaused, isPending, error: errorState, } = editContext;
    const finalActions = typeof actions === 'undefined' && hasShow ? defaultActions : actions;
    const showOffline = isPaused && isPending && offline !== undefined && offline !== false;
    const showError = errorState && error !== false && error !== undefined;
    if (!record &&
        isPending &&
        emptyWhileLoading &&
        !showOffline &&
        !showError) {
        return null;
    }
    return (React.createElement(Root, { className: clsx('edit-page', className), ...rest },
        title !== false && (React.createElement(Title, { title: title, defaultTitle: defaultTitle, preferenceKey: `${resource}.edit.title` })),
        finalActions,
        React.createElement("div", { className: clsx(EditClasses.main, {
                [EditClasses.noActions]: !finalActions,
            }) },
            React.createElement(Content, { className: EditClasses.card }, showOffline ? (offline) : showError ? (error) : render ? (render(editContext)) : record ? (children) : (React.createElement(CardContent, null, "\u00A0"))),
            aside)));
};
const PREFIX = 'RaEdit';
export const EditClasses = {
    main: `${PREFIX}-main`,
    noActions: `${PREFIX}-noActions`,
    card: `${PREFIX}-card`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${EditClasses.main}`]: {
        display: 'flex',
        alignItems: 'flex-start',
    },
    [`& .${EditClasses.noActions}`]: {
        marginTop: '1em',
    },
    [`& .${EditClasses.card}`]: {
        flex: '1 1 auto',
    },
});
