// ra-ui-materialui/dist/detail/editFieldTypes.js
import * as React from 'react';
import { SimpleForm } from "../form/index.js";
import { ArrayInput, BooleanInput, DateInput, NumberInput, ReferenceInput, ReferenceArrayInput, SelectInput, SimpleFormIterator, TextInput, TextArrayInput, } from "../input/index.js";
export const editFieldTypes = {
    form: {
        component: SimpleForm,
        representation: (_, children) => `        <SimpleForm>
${children.map(child => `            ${child.getRepresentation()}`).join('\n')}
        </SimpleForm>`,
    },
    array: {
        component: ({ children, ...props }) => (React.createElement(ArrayInput, { ...props },
            React.createElement(SimpleFormIterator, null, children))),
        representation: (props, children) => `<ArrayInput source="${props.source}"><SimpleFormIterator>${children
            .map(child => child.getRepresentation())
            .join('\n')}</SimpleFormIterator></ArrayInput>`,
    },
    scalar_array: {
        component: TextArrayInput,
        representation: (props) => `<TextArrayInput source="${props.source}" />`,
    },
    boolean: {
        component: BooleanInput,
        representation: (props) => `<BooleanInput source="${props.source}" />`,
    },
    date: {
        component: DateInput,
        representation: (props) => `<DateInput source="${props.source}" />`,
    },
    email: {
        component: TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    id: {
        component: TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    number: {
        component: NumberInput,
        representation: (props) => `<NumberInput source="${props.source}" />`,
    },
    reference: {
        component: ReferenceInput,
        representation: (props) => `<ReferenceInput source="${props.source}" reference="${props.reference}" />`,
    },
    referenceArray: {
        component: ReferenceArrayInput,
        representation: (props) => `<ReferenceArrayInput source="${props.source}" reference="${props.reference}" />`,
    },
    referenceArrayChild: {
        component: (props) => (React.createElement(SelectInput, { optionText: "id", ...props })),
        representation: () => `<SelectInput optionText="id" />`,
    },
    richText: {
        component: TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    string: {
        component: TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
    url: {
        component: TextInput,
        representation: (props) => `<TextInput source="${props.source}" />`,
    },
};
