// ra-ui-materialui/dist/detail/EditView.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditClasses = exports.EditView = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const EditActions_1 = require("./EditActions.cjs");
const layout_1 = require("../layout/index.cjs");
const Offline_1 = require("../Offline.cjs");
const defaultActions = React.createElement(EditActions_1.EditActions, null);
const defaultOffline = React.createElement(Offline_1.Offline, null);
const EditView = (props) => {
    const { actions, aside, children, render, className, component: Content = material_1.Card, emptyWhileLoading = false, offline = defaultOffline, error, title, ...rest } = props;
    const { hasShow } = (0, ra_core_1.useResourceDefinition)();
    const editContext = (0, ra_core_1.useEditContext)();
    const { resource, defaultTitle, record, isPaused, isPending, error: errorState, } = editContext;
    const finalActions = typeof actions === 'undefined' && hasShow ? defaultActions : actions;
    const showOffline = isPaused && isPending && offline !== undefined && offline !== false;
    const showError = errorState && error !== false && error !== undefined;
    if (!record &&
        isPending &&
        emptyWhileLoading &&
        !showOffline &&
        !showError) {
        return null;
    }
    return (React.createElement(Root, { className: (0, clsx_1.default)('edit-page', className), ...rest },
        title !== false && (React.createElement(layout_1.Title, { title: title, defaultTitle: defaultTitle, preferenceKey: `${resource}.edit.title` })),
        finalActions,
        React.createElement("div", { className: (0, clsx_1.default)(exports.EditClasses.main, {
                [exports.EditClasses.noActions]: !finalActions,
            }) },
            React.createElement(Content, { className: exports.EditClasses.card }, showOffline ? (offline) : showError ? (error) : render ? (render(editContext)) : record ? (children) : (React.createElement(material_1.CardContent, null, "\u00A0"))),
            aside)));
};
exports.EditView = EditView;
const PREFIX = 'RaEdit';
exports.EditClasses = {
    main: `${PREFIX}-main`,
    noActions: `${PREFIX}-noActions`,
    card: `${PREFIX}-card`,
};
const Root = (0, material_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${exports.EditClasses.main}`]: {
        display: 'flex',
        alignItems: 'flex-start',
    },
    [`& .${exports.EditClasses.noActions}`]: {
        marginTop: '1em',
    },
    [`& .${exports.EditClasses.card}`]: {
        flex: '1 1 auto',
    },
});
