// ra-ui-materialui/dist/detail/Edit.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Edit = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const styles_1 = require("@mui/material/styles");
const EditView_1 = require("./EditView.cjs");
const layout_1 = require("../layout/index.cjs");
/**
 * Page component for the Edit view
 *
 * The `<Edit>` component renders the page title and actions,
 * fetches the record from the data provider.
 * It is not responsible for rendering the actual form -
 * that's the job of its child component (usually `<SimpleForm>`),
 * to which it passes the `record` as prop.
 *
 * The <Edit> component accepts the following props:
 *
 * - children: Component rendering  the Form Layout
 * - render: Alternative to children. A function to render the Form Layout. Receives the edit context as its argument.
 * - actions
 * - aside
 * - component
 * - title
 * - mutationMode
 * - mutationOptions
 *
 * @example
 *
 * // in src/posts.js
 * import * as React from "react";
 * import { Edit, SimpleForm, TextInput } from 'react-admin';
 *
 * export const PostEdit = () => (
 *     <Edit>
 *         <SimpleForm>
 *             <TextInput source="title" />
 *         </SimpleForm>
 *     </Edit>
 * );
 *
 * // in src/App.js
 * import * as React from "react";
 * import { Admin, Resource } from 'react-admin';
 *
 * import { PostEdit } from './posts';
 *
 * const App = () => (
 *     <Admin dataProvider={...}>
 *         <Resource name="posts" edit={PostEdit} />
 *     </Admin>
 * );
 * export default App;
 */
const Edit = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { resource, id, mutationMode, mutationOptions, queryOptions, redirect, transform, disableAuthentication, authLoading = defaultAuthLoading, loading, error, redirectOnError, ...rest } = props;
    if (!props.render && !props.children) {
        throw new Error('<Edit> requires either a `render` prop or `children` prop');
    }
    return (React.createElement(ra_core_1.EditBase, { resource: resource, id: id, mutationMode: mutationMode, mutationOptions: mutationOptions, queryOptions: queryOptions, redirect: redirect, transform: transform, disableAuthentication: disableAuthentication, authLoading: authLoading, loading: loading, 
        // Disable redirect on error as it is handled by EditView to display the error in the EditView container
        redirectOnError: redirectOnError ?? (error ? false : undefined), 
        // Disable offline support from EditBase as it is handled by EditView to keep the EditView container
        offline: false },
        React.createElement(EditView_1.EditView, { error: error, ...rest })));
};
exports.Edit = Edit;
const defaultAuthLoading = React.createElement(layout_1.Loading, null);
const PREFIX = 'RaEdit'; // Types declared in EditView.
