// ra-ui-materialui/dist/detail/showFieldTypes.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.showFieldTypes = void 0;
const React = __importStar(require("react"));
const field_1 = require("../field/index.cjs");
const SimpleShowLayout_1 = require("./SimpleShowLayout.cjs");
const list_1 = require("../list/index.cjs");
exports.showFieldTypes = {
    show: {
        component: (props) => (React.createElement(SimpleShowLayout_1.SimpleShowLayout, { ...props })),
        representation: (_, children) => `        <SimpleShowLayout>
${children.map(child => `            ${child.getRepresentation()}`).join('\n')}
        </SimpleShowLayout>`,
    },
    array: {
        component: ({ children, ...props }) => (React.createElement(field_1.ArrayField, { ...props },
            React.createElement(list_1.DataTable, null, children && children.length > 0
                ? children.map((child, index) => (React.createElement(list_1.DataTable.Col, { key: index, ...child.props }, child)))
                : children))),
        representation: (props, children) => `<ArrayField source="${props.source}">
                <DataTable>
                    ${children
            .map(child => `<DataTable.Col source="${child.getProps().source}">
                        ${child.getRepresentation()}
                    </DataTable.Col>`)
            .join('\n                    ')}
                </DataTable>
            </ArrayField>`,
    },
    scalar_array: {
        component: field_1.TextArrayField,
        representation: (props) => `<TextArrayField source="${props.source}" />`,
    },
    boolean: {
        component: field_1.BooleanField,
        representation: (props) => `<BooleanField source="${props.source}" />`,
    },
    date: {
        component: field_1.DateField,
        representation: (props) => `<DateField source="${props.source}" />`,
    },
    email: {
        component: field_1.EmailField,
        representation: (props) => `<EmailField source="${props.source}" />`,
    },
    id: {
        component: field_1.TextField,
        representation: (props) => `<TextField source="${props.source}" />`,
    },
    number: {
        component: field_1.NumberField,
        representation: (props) => `<NumberField source="${props.source}" />`,
    },
    reference: {
        component: field_1.ReferenceField,
        representation: (props) => `<ReferenceField source="${props.source}" reference="${props.reference}" />`,
    },
    referenceChild: {
        component: (props) => React.createElement(field_1.TextField, { source: "id", ...props }),
        representation: () => `<TextField source="id" />`,
    },
    referenceArray: {
        component: field_1.ReferenceArrayField,
        representation: (props) => `<ReferenceArrayField source="${props.source}" reference="${props.reference}" />`,
    },
    referenceArrayChild: {
        component: () => (React.createElement(list_1.SingleFieldList, null,
            React.createElement(field_1.ChipField, { source: "id" }))),
        representation: () => `<SingleFieldList><ChipField source="id" /></SingleFieldList>`,
    },
    richText: {
        component: field_1.RichTextField,
        representation: (props) => `<RichTextField source="${props.source}" />`,
    },
    string: {
        component: field_1.TextField,
        representation: (props) => `<TextField source="${props.source}" />`,
    },
    url: {
        component: field_1.UrlField,
        representation: (props) => `<UrlField source="${props.source}" />`,
    },
};
