// ra-ui-materialui/dist/detail/ShowView.js
import * as React from 'react';
import { Card, styled, } from '@mui/material';
import clsx from 'clsx';
import { useShowContext, useResourceDefinition, } from 'ra-core';
import { ShowActions } from "./ShowActions.js";
import { Title } from "../layout/index.js";
import { Offline } from "../Offline.js";
const defaultActions = React.createElement(ShowActions, null);
const defaultOffline = React.createElement(Offline, null);
export const ShowView = (props) => {
    const { actions, aside, children, render, className, component: Content = Card, emptyWhileLoading = false, offline = defaultOffline, error, title, ...rest } = props;
    const showContext = useShowContext();
    const { resource, defaultTitle, isPaused, isPending, record, error: errorState, } = showContext;
    const { hasEdit } = useResourceDefinition();
    const finalActions = typeof actions === 'undefined' && hasEdit ? defaultActions : actions;
    const showOffline = isPaused && isPending && offline !== undefined && offline !== false;
    const showError = errorState && error !== false && error !== undefined;
    if (!record &&
        isPending &&
        emptyWhileLoading &&
        !showOffline &&
        !showError) {
        return null;
    }
    return (React.createElement(Root, { className: clsx('show-page', className), ...rest },
        title !== false && (React.createElement(Title, { title: title, defaultTitle: defaultTitle, preferenceKey: `${resource}.show.title` })),
        finalActions !== false && finalActions,
        React.createElement("div", { className: clsx(ShowClasses.main, {
                [ShowClasses.noActions]: !finalActions,
            }) },
            React.createElement(Content, { className: ShowClasses.card }, showOffline
                ? offline
                : showError
                    ? error
                    : render
                        ? render(showContext)
                        : children),
            aside)));
};
const PREFIX = 'RaShow';
export const ShowClasses = {
    main: `${PREFIX}-main`,
    noActions: `${PREFIX}-noActions`,
    card: `${PREFIX}-card`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${ShowClasses.main}`]: {
        display: 'flex',
    },
    [`& .${ShowClasses.noActions}`]: {
        marginTop: '1em',
    },
    [`& .${ShowClasses.card}`]: {
        flex: '1 1 auto',
    },
});
