// ra-ui-materialui/dist/detail/showFieldTypes.js
import * as React from 'react';
import { ArrayField, BooleanField, DateField, EmailField, NumberField, ReferenceField, ReferenceArrayField, RichTextField, TextField, UrlField, ChipField, TextArrayField, } from "../field/index.js";
import { SimpleShowLayout } from "./SimpleShowLayout.js";
import { DataTable, SingleFieldList } from "../list/index.js";
export const showFieldTypes = {
    show: {
        component: (props) => (React.createElement(SimpleShowLayout, { ...props })),
        representation: (_, children) => `        <SimpleShowLayout>
${children.map(child => `            ${child.getRepresentation()}`).join('\n')}
        </SimpleShowLayout>`,
    },
    array: {
        component: ({ children, ...props }) => (React.createElement(ArrayField, { ...props },
            React.createElement(DataTable, null, children && children.length > 0
                ? children.map((child, index) => (React.createElement(DataTable.Col, { key: index, ...child.props }, child)))
                : children))),
        representation: (props, children) => `<ArrayField source="${props.source}">
                <DataTable>
                    ${children
            .map(child => `<DataTable.Col source="${child.getProps().source}">
                        ${child.getRepresentation()}
                    </DataTable.Col>`)
            .join('\n                    ')}
                </DataTable>
            </ArrayField>`,
    },
    scalar_array: {
        component: TextArrayField,
        representation: (props) => `<TextArrayField source="${props.source}" />`,
    },
    boolean: {
        component: BooleanField,
        representation: (props) => `<BooleanField source="${props.source}" />`,
    },
    date: {
        component: DateField,
        representation: (props) => `<DateField source="${props.source}" />`,
    },
    email: {
        component: EmailField,
        representation: (props) => `<EmailField source="${props.source}" />`,
    },
    id: {
        component: TextField,
        representation: (props) => `<TextField source="${props.source}" />`,
    },
    number: {
        component: NumberField,
        representation: (props) => `<NumberField source="${props.source}" />`,
    },
    reference: {
        component: ReferenceField,
        representation: (props) => `<ReferenceField source="${props.source}" reference="${props.reference}" />`,
    },
    referenceChild: {
        component: (props) => React.createElement(TextField, { source: "id", ...props }),
        representation: () => `<TextField source="id" />`,
    },
    referenceArray: {
        component: ReferenceArrayField,
        representation: (props) => `<ReferenceArrayField source="${props.source}" reference="${props.reference}" />`,
    },
    referenceArrayChild: {
        component: () => (React.createElement(SingleFieldList, null,
            React.createElement(ChipField, { source: "id" }))),
        representation: () => `<SingleFieldList><ChipField source="id" /></SingleFieldList>`,
    },
    richText: {
        component: RichTextField,
        representation: (props) => `<RichTextField source="${props.source}" />`,
    },
    string: {
        component: TextField,
        representation: (props) => `<TextField source="${props.source}" />`,
    },
    url: {
        component: UrlField,
        representation: (props) => `<UrlField source="${props.source}" />`,
    },
};
