// ra-ui-materialui/dist/detail/Tab.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TabClasses = exports.Tab = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const material_1 = require("@mui/material");
const system_1 = require("@mui/system");
const ra_core_1 = require("ra-core");
const clsx_1 = __importDefault(require("clsx"));
const Labeled_1 = require("../Labeled.cjs");
/**
 * Tab element for the TabbedShowLayout.
 *
 * The `<Tab>` component accepts the following props:
 *
 * - label: The string displayed for each tab
 * - icon: The icon to show before the label (optional). Must be a component.
 * - path: The string used for custom urls
 *
 * It is also available as TabbedShowLayout.Tab.
 *
 * @example
 *     // in src/posts.js
 *     import * as React from "react";
 *     import FavoriteIcon from '@mui/icons-material/Favorite';
 *     import PersonPinIcon from '@mui/icons-material/PersonPin';
 *     import { Show, TabbedShowLayout, TextField } from 'react-admin';
 *
 *     export const PostShow = () => (
 *         <Show>
 *             <TabbedShowLayout>
 *                 <TabbedShowLayout.Tab label="Content" icon={<FavoriteIcon />}>
 *                     <TextField source="title" />
 *                     <TextField source="subtitle" />
 *                </TabbedShowLayout.Tab>
 *                 <TabbedShowLayout.Tab label="Metadata" icon={<PersonIcon />} path="metadata">
 *                     <TextField source="category" />
 *                </TabbedShowLayout.Tab>
 *             </TabbedShowLayout>
 *         </Show>
 *     );
 *
 *     // in src/App.js
 *     import * as React from "react";
 *     import { Admin, Resource } from 'react-admin';
 *
 *     import { PostShow } from './posts';
 *
 *     const App = () => (
 *         <Admin dataProvider={...}>
 *             <Resource name="posts" show={PostShow} />
 *         </Admin>
 *     );
 *     export default App;
 */
const Tab = (inProps) => {
    const props = (0, system_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, contentClassName, context, count, className, divider, icon, iconPosition, label, record, spacing = 1, syncWithLocation = true, value, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const location = (0, ra_core_1.useLocation)();
    const splatPathBase = (0, ra_core_1.useSplatPathBase)();
    const newPathName = value == null || value === ''
        ? splatPathBase
        : `${splatPathBase}/${value}`;
    const propsForLink = {
        component: ra_core_1.LinkBase,
        to: { ...location, pathname: newPathName },
    };
    const renderHeader = () => {
        let tabLabel = typeof label === 'string' ? translate(label, { _: label }) : label;
        if (count !== undefined) {
            tabLabel = (React.createElement("span", null,
                tabLabel,
                " (",
                count,
                ")"));
        }
        return (React.createElement(material_1.Tab, { key: `tab-header-${value}`, label: tabLabel, value: value, icon: icon, iconPosition: iconPosition, className: (0, clsx_1.default)('show-tab', className), ...(syncWithLocation ? propsForLink : {}), ...rest }));
    };
    const renderContent = () => (React.createElement(Root, { className: contentClassName, spacing: spacing, divider: divider }, React.Children.map(children, field => field && (0, react_1.isValidElement)(field) ? (React.createElement(Labeled_1.Labeled, { key: field.props.source, className: (0, clsx_1.default)('ra-field', field.props.source &&
            `ra-field-${field.props.source}`, exports.TabClasses.row, field.props.className) }, field)) : null)));
    return context === 'header' ? renderHeader() : renderContent();
};
exports.Tab = Tab;
const PREFIX = 'RaTab';
exports.TabClasses = {
    row: `${PREFIX}-row`,
};
const Root = (0, material_1.styled)(material_1.Stack, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    [`& .${exports.TabClasses.row}`]: {
        display: 'inline',
    },
}));
