// ra-ui-materialui/dist/detail/TabbedShowLayout.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TabbedShowLayoutClasses = exports.TabbedShowLayout = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const TabbedShowLayoutTabs_1 = require("./TabbedShowLayoutTabs.cjs");
const Tab_1 = require("./Tab.cjs");
/**
 * Layout for a Show view showing fields grouped in tabs and laid out in a single column.
 *
 * It pulls the record from the RecordContext. It renders a set of `<Tabs>`,
 * each of which contains a list of record fields in a single-column layout
 * (via Material UI's `<Stack>` component).
 * `<TabbedShowLayout>` delegates the actual rendering of fields to its children,
 * which should be `<TabbedShowLayout.Tab>` components.
 * `<TabbedShowLayout.Tab>` wraps each field inside a `<Labeled>` component to add a label.
 *
 * @example
 * // in src/posts.js
 * import * as React from "react";
 * import { Show, TabbedShowLayout, TextField } from 'react-admin';
 *
 * export const PostShow = () => (
 *     <Show>
 *         <TabbedShowLayout>
 *             <TabbedShowLayout.Tab label="Content">
 *                 <TextField source="title" />
 *                 <TextField source="subtitle" />
 *            </TabbedShowLayout.Tab>
 *             <TabbedShowLayout.Tab label="Metadata">
 *                 <TextField source="category" />
 *            </TabbedShowLayout.Tab>
 *         </TabbedShowLayout>
 *     </Show>
 * );
 *
 * // in src/App.js
 * import * as React from "react";
 * import { Admin, Resource } from 'react-admin';
 *
 * import { PostShow } from './posts';
 *
 * const App = () => (
 *     <Admin dataProvider={...}>
 *         <Resource name="posts" show={PostShow} />
 *     </Admin>
 * );
 *
 * @param {TabbedShowLayoutProps} props
 * @param {string} props.className A className to apply to the page content.
 * @param {ElementType} props.component The component to use as root component (div by default).
 * @param {ReactNode} props.divider An optional divider between each field, passed to `<Stack>`.
 * @param {number} props.spacing The spacing to use between each field, passed to `<Stack>`. Defaults to 1.
 * @param {Object} props.sx Custom style object.
 * @param {boolean} props.syncWithLocation Whether to update the URL when the tab changes. Defaults to true.
 * @param {ElementType} props.tabs A custom component for rendering tabs.
 */
const TabbedShowLayout = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, spacing, divider, syncWithLocation = true, tabs = DefaultTabs, value, ...rest } = props;
    const { Route, Routes, Outlet } = (0, ra_core_1.useRouterProvider)();
    const record = (0, ra_core_1.useRecordContext)(props);
    const nonNullChildren = react_1.Children.toArray(children).filter(child => child !== null);
    const [tabValue, setTabValue] = (0, react_1.useState)(0);
    const handleTabChange = (event, value) => {
        if (!syncWithLocation) {
            setTabValue(value);
        }
    };
    if (!record) {
        return null;
    }
    const renderTabHeaders = () => (0, react_1.cloneElement)(tabs, {
        onChange: handleTabChange,
        syncWithLocation,
        value: tabValue,
    }, nonNullChildren);
    return (React.createElement(ra_core_1.OptionalRecordContextProvider, { value: props.record },
        React.createElement(Root, { className: className, ...sanitizeRestProps(rest) }, syncWithLocation ? (React.createElement(Routes, null,
            React.createElement(Route, { path: "/*", element: React.createElement(React.Fragment, null,
                    renderTabHeaders(),
                    React.createElement(material_1.Divider, null),
                    React.createElement("div", { className: exports.TabbedShowLayoutClasses.content },
                        React.createElement(Outlet, null))) }, react_1.Children.map(nonNullChildren, (tab, index) => (0, react_1.isValidElement)(tab) ? (React.createElement(Route, { path: (0, TabbedShowLayoutTabs_1.getShowLayoutTabFullPath)(tab, index), element: (0, react_1.cloneElement)(tab, {
                    context: 'content',
                    spacing,
                    divider,
                }) })) : null)))) : (React.createElement(React.Fragment, null,
            renderTabHeaders(),
            React.createElement(material_1.Divider, null),
            React.createElement("div", { className: exports.TabbedShowLayoutClasses.content }, react_1.Children.map(nonNullChildren, (tab, index) => {
                if (!(0, react_1.isValidElement)(tab) ||
                    tabValue !== index) {
                    return null;
                }
                return (0, react_1.cloneElement)(tab, {
                    context: 'content',
                    spacing,
                    divider,
                });
            })))))));
};
exports.TabbedShowLayout = TabbedShowLayout;
exports.TabbedShowLayout.Tab = Tab_1.Tab;
const DefaultTabs = React.createElement(TabbedShowLayoutTabs_1.TabbedShowLayoutTabs, null);
const PREFIX = 'RaTabbedShowLayout';
exports.TabbedShowLayoutClasses = {
    content: `${PREFIX}-content`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flex: 1,
    [`& .${exports.TabbedShowLayoutClasses.content}`]: {
        padding: `${theme.spacing(1)} ${theme.spacing(2)}`,
    },
}));
const sanitizeRestProps = ({ record, resource, initialValues, staticContext, translate, tabs, ...rest }) => rest;
