// ra-core/dist/auth/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithPermissions = exports.useLogoutIfAccessDenied = exports.useAuthState = exports.usePermissions = exports.useGetPermissions = exports.useLogout = exports.useLogin = exports.convertLegacyAuthProvider = exports.useAuthProvider = void 0;
const useAuthProvider_1 = __importDefault(require("./useAuthProvider.cjs"));
exports.useAuthProvider = useAuthProvider_1.default;
const useAuthState_1 = __importDefault(require("./useAuthState.cjs"));
exports.useAuthState = useAuthState_1.default;
const usePermissions_1 = __importDefault(require("./usePermissions.cjs"));
exports.usePermissions = usePermissions_1.default;
const WithPermissions_1 = __importDefault(require("./WithPermissions.cjs"));
exports.WithPermissions = WithPermissions_1.default;
const useLogin_1 = __importDefault(require("./useLogin.cjs"));
exports.useLogin = useLogin_1.default;
const useLogout_1 = __importDefault(require("./useLogout.cjs"));
exports.useLogout = useLogout_1.default;
const useGetPermissions_1 = __importDefault(require("./useGetPermissions.cjs"));
exports.useGetPermissions = useGetPermissions_1.default;
const useLogoutIfAccessDenied_1 = __importDefault(require("./useLogoutIfAccessDenied.cjs"));
exports.useLogoutIfAccessDenied = useLogoutIfAccessDenied_1.default;
const convertLegacyAuthProvider_1 = __importDefault(require("./convertLegacyAuthProvider.cjs"));
exports.convertLegacyAuthProvider = convertLegacyAuthProvider_1.default;
__exportStar(require("./Authenticated.cjs"), exports);
__exportStar(require("./AuthContext.cjs"), exports);
__exportStar(require("./CanAccess.cjs"), exports);
__exportStar(require("./LogoutOnMount.cjs"), exports);
__exportStar(require("./types.cjs"), exports);
__exportStar(require("./useAuthenticated.cjs"), exports);
__exportStar(require("./useCanAccess.cjs"), exports);
__exportStar(require("./useCanAccessResources.cjs"), exports);
__exportStar(require("./useCanAccessCallback.cjs"), exports);
__exportStar(require("./useCheckAuth.cjs"), exports);
__exportStar(require("./useGetIdentity.cjs"), exports);
__exportStar(require("./useHandleAuthCallback.cjs"), exports);
__exportStar(require("./useIsAuthPending.cjs"), exports);
__exportStar(require("./useRequireAccess.cjs"), exports);
__exportStar(require("./addRefreshAuthToAuthProvider.cjs"), exports);
__exportStar(require("./addRefreshAuthToDataProvider.cjs"), exports);
