// ra-core/dist/auth/useAuthState.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("react");
const react_query_1 = require("@tanstack/react-query");
const useAuthProvider_1 = __importStar(require("./useAuthProvider.cjs"));
const useLogout_1 = __importDefault(require("./useLogout.cjs"));
const routing_1 = require("../routing/index.cjs");
const notification_1 = require("../notification/index.cjs");
const util_1 = require("../util/index.cjs");
const emptyParams = {};
/**
 * Hook for getting the authentication status
 *
 * Calls the authProvider.checkAuth() method asynchronously.
 *
 * The return value updates according to the authProvider request state:
 *
 * - isPending: true just after mount, while the authProvider is being called. false once the authProvider has answered.
 * - authenticated: true while loading. then true or false depending on the authProvider response.
 *
 * To avoid rendering a component and force waiting for the authProvider response, use the useAuthState() hook
 * instead of the useAuthenticated() hook.
 *
 * You can render different content depending on the authenticated status.
 *
 * @see useAuthenticated()
 *
 * @param {Object} params Any params you want to pass to the authProvider
 *
 * @param {Boolean} logoutOnFailure: Optional. Whether the user should be logged out if the authProvider fails to authenticate them. False by default.
 *
 * @returns The current auth check state. Destructure as { authenticated, error, isPending }.
 *
 * @example
 * import { useAuthState, Loading } from 'react-admin';
 *
 * const MyPage = () => {
 *     const { isPending, authenticated } = useAuthState();
 *     if (isPending) {
 *         return <Loading />;
 *     }
 *     if (authenticated) {
 *        return <AuthenticatedContent />;
 *     }
 *     return <AnonymousContent />;
 * };
 */
const useAuthState = (params = emptyParams, logoutOnFailure = false, queryOptions = emptyParams) => {
    const authProvider = (0, useAuthProvider_1.default)();
    const logout = (0, useLogout_1.default)();
    const basename = (0, routing_1.useBasename)();
    const notify = (0, notification_1.useNotify)();
    const { onSuccess, onError, onSettled, ...options } = queryOptions;
    const queryResult = (0, react_query_1.useQuery)({
        queryKey: ['auth', 'checkAuth', params],
        queryFn: async ({ signal }) => {
            // The authProvider is optional in react-admin
            if (!authProvider) {
                return true;
            }
            try {
                await authProvider.checkAuth({ ...params, signal });
                return true;
            }
            catch (error) {
                // This is necessary because react-query requires the error to be defined
                if (error != null) {
                    throw error;
                }
                throw new Error();
            }
        },
        retry: false,
        ...options,
    });
    const onSuccessEvent = (0, util_1.useEvent)(onSuccess ?? noop);
    const onSettledEvent = (0, util_1.useEvent)(onSettled ?? noop);
    const onErrorEvent = (0, util_1.useEvent)(onError ??
        ((error) => {
            if (!logoutOnFailure)
                return;
            const loginUrl = (0, routing_1.removeDoubleSlashes)(`${basename}/${useAuthProvider_1.defaultAuthParams.loginUrl}`);
            logout({}, error && error.redirectTo != null
                ? error.redirectTo
                : loginUrl);
            const shouldSkipNotify = error && error.message === false;
            !shouldSkipNotify &&
                notify(getErrorMessage(error, 'ra.auth.auth_check_error'), {
                    type: 'error',
                });
        }));
    (0, react_1.useEffect)(() => {
        if (queryResult.data === undefined || queryResult.isFetching)
            return;
        if (queryOptions.enabled === false)
            return;
        onSuccessEvent(queryResult.data);
    }, [
        onSuccessEvent,
        queryResult.data,
        queryResult.isFetching,
        queryOptions.enabled,
    ]);
    (0, react_1.useEffect)(() => {
        if (queryResult.error == null || queryResult.isFetching)
            return;
        if (queryOptions.enabled === false)
            return;
        onErrorEvent(queryResult.error);
    }, [
        onErrorEvent,
        queryResult.error,
        queryResult.isFetching,
        queryOptions.enabled,
    ]);
    (0, react_1.useEffect)(() => {
        if (queryResult.status === 'pending' || queryResult.isFetching)
            return;
        if (queryOptions.enabled === false)
            return;
        onSettledEvent(queryResult.data, queryResult.error);
    }, [
        onSettledEvent,
        queryResult.data,
        queryResult.error,
        queryResult.status,
        queryResult.isFetching,
        queryOptions.enabled,
    ]);
    const result = (0, react_1.useMemo)(() => {
        return {
            ...queryResult,
            authenticated: queryResult.error ? false : queryResult.data,
        };
    }, [queryResult]);
    return authProvider != null
        ? result
        : noAuthProviderQueryResult;
};
exports.default = useAuthState;
const getErrorMessage = (error, defaultMessage) => typeof error === 'string'
    ? error
    : typeof error === 'undefined' || !error.message
        ? defaultMessage
        : error.message;
const noop = () => { };
const noAuthProviderQueryResult = {
    authenticated: true,
    data: true,
    dataUpdatedAt: 0,
    error: null,
    errorUpdatedAt: 0,
    errorUpdateCount: 0,
    failureCount: 0,
    failureReason: null,
    fetchStatus: 'idle',
    isError: false,
    isInitialLoading: false,
    isLoading: false,
    isLoadingError: false,
    isFetched: true,
    isFetchedAfterMount: true,
    isFetching: false,
    isPaused: false,
    isPlaceholderData: false,
    isPending: false,
    isRefetchError: false,
    isRefetching: false,
    isStale: false,
    isSuccess: true,
    status: 'success',
    refetch: () => Promise.resolve(noAuthProviderQueryResult),
};
module.exports = exports.default;
