// ra-ui-materialui/dist/list/filter/FilterForm.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterFormClasses = exports.FilterFormBase = exports.FilterForm = void 0;
const styles_1 = require("@mui/material/styles");
const get_js_1 = __importDefault(require("lodash/get.js"));
const ra_core_1 = require("ra-core");
const React = __importStar(require("react"));
const react_1 = require("react");
const FilterFormInput_1 = require("./FilterFormInput.cjs");
const FilterForm = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { filters: filtersProps, ...rest } = props;
    const filters = (0, ra_core_1.useFilterContext)() || filtersProps;
    return (React.createElement(ra_core_1.FilterLiveForm, { formComponent: StyledForm, ...sanitizeRestProps(rest) },
        React.createElement(exports.FilterFormBase, { filters: filters })));
};
exports.FilterForm = FilterForm;
const FilterFormBase = (props) => {
    const { filters } = props;
    const resource = (0, ra_core_1.useResourceContext)(props);
    const { displayedFilters = {}, filterValues, hideFilter, } = (0, ra_core_1.useListContext)();
    (0, react_1.useEffect)(() => {
        if (!filters)
            return;
        filters.forEach((filter) => {
            if (filter.props.alwaysOn && filter.props.defaultValue) {
                throw new Error('Cannot use alwaysOn and defaultValue on a filter input. Please set the filterDefaultValues props on the <List> element instead.');
            }
        });
    }, [filters]);
    const getShownFilters = () => {
        if (!filters)
            return [];
        const values = filterValues;
        return filters.filter((filterElement) => {
            if (!React.isValidElement(filterElement)) {
                return false;
            }
            const filterValue = (0, get_js_1.default)(values, filterElement.props.source);
            return (filterElement.props.alwaysOn ||
                displayedFilters[filterElement.props.source] ||
                !isEmptyValue(filterValue));
        });
    };
    const handleHide = (0, react_1.useCallback)(event => hideFilter(event.currentTarget.dataset.key), [hideFilter]);
    return (React.createElement(React.Fragment, null,
        getShownFilters().map(filterElement => (React.createElement(FilterFormInput_1.FilterFormInput, { key: filterElement.key || filterElement.props.source, filterElement: filterElement, handleHide: handleHide, resource: resource, className: exports.FilterFormClasses.filterFormInput }))),
        React.createElement("div", { className: exports.FilterFormClasses.clearFix })));
};
exports.FilterFormBase = FilterFormBase;
const sanitizeRestProps = ({ hasCreate, resource, ...props }) => props;
const PREFIX = 'RaFilterForm';
exports.FilterFormClasses = {
    clearFix: `${PREFIX}-clearFix`,
    filterFormInput: `${PREFIX}-filterFormInput`,
};
const StyledForm = (0, styles_1.styled)('form', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flex: '0 1 auto',
    [theme.breakpoints.down('sm')]: {
        width: '100%',
    },
    [theme.breakpoints.up('sm')]: {
        minHeight: theme.spacing(8),
    },
    [theme.breakpoints.up('md')]: {
        flex: '0 1 100%',
    },
    flexWrap: 'wrap',
    alignItems: 'flex-end',
    pointerEvents: 'none',
    padding: `0 0 ${theme.spacing(0.5)} 0`,
    '& .MuiFormHelperText-root': { display: 'none' },
    [`& .${exports.FilterFormClasses.clearFix}`]: { clear: 'right' },
    [`& .${exports.FilterFormClasses.filterFormInput} .MuiFormControl-root`]: {
        marginTop: `${theme.spacing(1)}`,
    },
}));
const isEmptyValue = (filterValue) => {
    if (filterValue === '' || filterValue == null)
        return true;
    // If one of the value leaf is not empty
    // the value is considered not empty
    if (typeof filterValue === 'object') {
        return Object.keys(filterValue).every(key => isEmptyValue(filterValue[key]));
    }
    return false;
};
