// ra-ui-materialui/dist/list/filter/FilterButton.js
import * as React from 'react';
import { useState, useCallback, useRef, } from 'react';
import { Menu, MenuItem, ListItemIcon, ListItemText, styled, Divider, useThemeProps, } from '@mui/material';
import ClearIcon from '@mui/icons-material/Clear';
import BookmarkAddIcon from '@mui/icons-material/BookmarkAdd';
import BookmarkRemoveIcon from '@mui/icons-material/BookmarkRemove';
import BookmarkBorderIcon from '@mui/icons-material/BookmarkBorder';
import ContentFilter from '@mui/icons-material/FilterList';
import isEqual from 'lodash/isEqual.js';
import { useFilterContext, useListContext, useResourceContext, useTranslate, extractValidSavedQueries, useSavedQueries, useNavigate, } from 'ra-core';
import { stringify } from 'query-string';
import { FilterButtonMenuItem } from "./FilterButtonMenuItem.js";
import { Button } from "../../button/index.js";
import { AddSavedQueryDialog } from "./AddSavedQueryDialog.js";
import { RemoveSavedQueryDialog } from "./RemoveSavedQueryDialog.js";
export const FilterButton = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { filters: filtersProp, className, disableSaveQuery, size, variant, ...rest } = props;
    const filters = useFilterContext() || filtersProp;
    const resource = useResourceContext(props);
    const translate = useTranslate();
    if (!resource && !disableSaveQuery) {
        throw new Error('<FilterButton> must be called inside a ResourceContextProvider, or must provide a resource prop');
    }
    const [savedQueries] = useSavedQueries(resource || '');
    const navigate = useNavigate();
    const { displayedFilters = {}, filterValues, perPage, setFilters, showFilter, hideFilter, sort, } = useListContext();
    const hasFilterValues = !isEqual(filterValues, {});
    const validSavedQueries = extractValidSavedQueries(savedQueries);
    const hasSavedCurrentQuery = validSavedQueries.some(savedQuery => isEqual(savedQuery.value, {
        filter: filterValues,
        sort,
        perPage,
        displayedFilters,
    }));
    const [open, setOpen] = useState(false);
    const anchorEl = useRef();
    if (filters === undefined) {
        throw new Error('The <FilterButton> component requires the <List filters> prop to be set');
    }
    const allTogglableFilters = filters.filter((filterElement) => {
        return (React.isValidElement(filterElement) &&
            !filterElement.props.alwaysOn);
    });
    const handleClickButton = useCallback(event => {
        // This prevents ghost click.
        event.preventDefault();
        setOpen(true);
        anchorEl.current = event.currentTarget;
    }, [anchorEl, setOpen]);
    const handleRequestClose = useCallback(() => {
        setOpen(false);
    }, [setOpen]);
    const handleShow = useCallback(({ source, defaultValue }) => {
        showFilter(source, defaultValue === '' ? undefined : defaultValue);
        // We have to fallback to imperative code because the new FilterFormInput
        // has no way of knowing it has just been displayed (and thus that it should focus its input)
        setTimeout(() => {
            const inputElement = document.querySelector(`input[name='${source}']`);
            if (inputElement) {
                inputElement.focus();
            }
        }, 50);
        setOpen(false);
    }, [showFilter, setOpen]);
    const handleRemove = useCallback(({ source }) => {
        hideFilter(source);
        setOpen(false);
    }, [hideFilter, setOpen]);
    // add query dialog state
    const [addSavedQueryDialogOpen, setAddSavedQueryDialogOpen] = useState(false);
    const hideAddSavedQueryDialog = () => {
        setAddSavedQueryDialogOpen(false);
    };
    const showAddSavedQueryDialog = () => {
        setOpen(false);
        setAddSavedQueryDialogOpen(true);
    };
    // remove query dialog state
    const [removeSavedQueryDialogOpen, setRemoveSavedQueryDialogOpen] = useState(false);
    const hideRemoveSavedQueryDialog = () => {
        setRemoveSavedQueryDialogOpen(false);
    };
    const showRemoveSavedQueryDialog = () => {
        setOpen(false);
        setRemoveSavedQueryDialogOpen(true);
    };
    if (allTogglableFilters.length === 0 &&
        validSavedQueries.length === 0 &&
        !hasFilterValues) {
        return null;
    }
    return (React.createElement(Root, { className: className, ...sanitizeRestProps(rest) },
        React.createElement(Button, { className: "add-filter", label: "ra.action.add_filter", "aria-haspopup": "true", onClick: handleClickButton, variant: variant, size: size },
            React.createElement(ContentFilter, null)),
        React.createElement(Menu, { open: open, anchorEl: anchorEl.current, onClose: handleRequestClose },
            allTogglableFilters.map((filterElement, index) => (React.createElement(FilterButtonMenuItem, { key: filterElement.props.source, filter: filterElement, displayed: !!displayedFilters[filterElement.props.source], resource: resource, onShow: handleShow, onHide: handleRemove, autoFocus: index === 0 }))),
            (hasFilterValues || validSavedQueries.length > 0) && (React.createElement(Divider, null)),
            validSavedQueries.map((savedQuery, index) => isEqual(savedQuery.value, {
                filter: filterValues,
                sort,
                perPage,
                displayedFilters,
            }) ? (React.createElement(MenuItem, { onClick: showRemoveSavedQueryDialog, key: index },
                React.createElement(ListItemIcon, null,
                    React.createElement(BookmarkRemoveIcon, { fontSize: "small" })),
                React.createElement(ListItemText, null, translate('ra.saved_queries.remove_label_with_name', {
                    _: 'Remove query "%{name}"',
                    name: savedQuery.label,
                })))) : (React.createElement(MenuItem, { onClick: () => {
                    navigate({
                        search: stringify({
                            filter: JSON.stringify(savedQuery.value.filter),
                            sort: savedQuery.value.sort?.field,
                            order: savedQuery.value.sort?.order,
                            page: 1,
                            perPage: savedQuery.value.perPage,
                            displayedFilters: JSON.stringify(savedQuery.value.displayedFilters),
                        }),
                    });
                    setOpen(false);
                }, key: index },
                React.createElement(ListItemIcon, null,
                    React.createElement(BookmarkBorderIcon, { fontSize: "small" })),
                React.createElement(ListItemText, null, savedQuery.label)))),
            hasFilterValues &&
                !hasSavedCurrentQuery &&
                !disableSaveQuery && (React.createElement(MenuItem, { onClick: showAddSavedQueryDialog },
                React.createElement(ListItemIcon, null,
                    React.createElement(BookmarkAddIcon, { fontSize: "small" })),
                React.createElement(ListItemText, null, translate('ra.saved_queries.new_label', {
                    _: 'Save current query...',
                })))),
            hasFilterValues && (React.createElement(MenuItem, { onClick: () => {
                    setFilters({}, {});
                    setOpen(false);
                } },
                React.createElement(ListItemIcon, null,
                    React.createElement(ClearIcon, { fontSize: "small" })),
                React.createElement(ListItemText, null, translate('ra.action.remove_all_filters', {
                    _: 'Remove all filters',
                }))))),
        !disableSaveQuery && (React.createElement(React.Fragment, null,
            React.createElement(AddSavedQueryDialog, { open: addSavedQueryDialogOpen, onClose: hideAddSavedQueryDialog }),
            React.createElement(RemoveSavedQueryDialog, { open: removeSavedQueryDialogOpen, onClose: hideRemoveSavedQueryDialog })))));
};
/* eslint-disable @typescript-eslint/no-unused-vars */
const sanitizeRestProps = ({ displayedFilters = null, filterValues = null, showFilter = null, ...rest }) => rest;
const PREFIX = 'RaFilterButton';
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'inline-block',
});
