// ra-ui-materialui/dist/list/filter/FilterListItem.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterListItemClasses = exports.FilterListItem = void 0;
const React = __importStar(require("react"));
const react_is_1 = require("react-is");
const styles_1 = require("@mui/material/styles");
const react_1 = require("react");
const material_1 = require("@mui/material");
const CancelOutlined_1 = __importDefault(require("@mui/icons-material/CancelOutlined"));
const ra_core_1 = require("ra-core");
const matches_js_1 = __importDefault(require("lodash/matches.js"));
const pickBy_js_1 = __importDefault(require("lodash/pickBy.js"));
const arePropsEqual = (prevProps, nextProps) => prevProps.label === nextProps.label &&
    (0, ra_core_1.shallowEqual)(prevProps.value, nextProps.value);
/**
 * Button to enable/disable a list filter.
 *
 * Expects 2 props:
 *
 * - label: The text (or React element) to be displayed for this item.
 *   If it's a string, the component will translate it.
 * - value: An object to be merged into the filter value when enabling the filter
 *   (e.g. { is_published: true, published_at_gte: '2020-07-08' })
 *
 * @example
 *
 * import * as React from 'react';
 * import { Card, CardContent } from '@mui/material';
 * import MailIcon from '@mui/icons-material/MailOutline';
 * import { FilterList, FilterListItem } from 'react-admin';
 *
 * const FilterSidebar = () => (
 *     <Card>
 *         <CardContent>
 *             <FilterList
 *                 label="Subscribed to newsletter"
 *                 icon={<MailIcon />}
 *             >
 *                 <FilterListItem
 *                     label="Yes"
 *                     value={{ has_newsletter: true }}
 *                  />
 *                 <FilterListItem
 *                     label="No"
 *                     value={{ has_newsletter: false }}
 *                  />
 *             </FilterList>
 *         </CardContent>
 *     </Card>
 * );
 *
 * @example // The value prop can contain multiple keys
 *
 * import * as React from 'react';
 * import {
 *     endOfYesterday,
 *     startOfWeek,
 *     subWeeks,
 *     startOfMonth,
 *     subMonths,
 * } from 'date-fns';
 * import { Card, CardContent } from '@mui/material';
 * import AccessTimeIcon from '@mui/icons-material/AccessTime';
 * import { FilterList, FilterListItem } from 'react-admin';
 *
 * const FilterSidebar = () => (
 *     <Card>
 *         <CardContent>
 *             <FilterList
 *                 label="Last visited"
 *                 icon={<AccessTimeIcon />}
 *             >
 *                 <FilterListItem
 *                     label="Today"
 *                     value={{
 *                         last_seen_gte: endOfYesterday().toISOString(),
 *                         last_seen_lte: undefined,
 *                     }}
 *                 />
 *                 <FilterListItem
 *                     label="This week"
 *                     value={{
 *                         last_seen_gte: startOfWeek(
 *                             new Date()
 *                         ).toISOString(),
 *                         last_seen_lte: undefined,
 *                     }}
 *                 />
 *                 <FilterListItem
 *                     label="Last week"
 *                     value={{
 *                         last_seen_gte: subWeeks(
 *                             startOfWeek(new Date()),
 *                             1
 *                         ).toISOString(),
 *                         last_seen_lte: startOfWeek(
 *                             new Date()
 *                         ).toISOString(),
 *                     }}
 *                 />
 *                 <FilterListItem
 *                     label="This month"
 *                     value={{
 *                         last_seen_gte: startOfMonth(
 *                             new Date()
 *                         ).toISOString(),
 *                         last_seen_lte: undefined,
 *                     }}
 *                 />
 *                 <FilterListItem
 *                     label="Last month"
 *                     value={{
 *                         last_seen_gte: subMonths(
 *                             startOfMonth(new Date()),
 *                             1
 *                         ).toISOString(),
 *                         last_seen_lte: startOfMonth(
 *                             new Date()
 *                         ).toISOString(),
 *                     }}
 *                 />
 *                 <FilterListItem
 *                     label="Earlier"
 *                     value={{
 *                         last_seen_gte: undefined,
 *                         last_seen_lte: subMonths(
 *                             startOfMonth(new Date()),
 *                             1
 *                         ).toISOString(),
 *                     }}
 *                 />
 *             </FilterList>
 *         </CardContent>
 *     </Card>
 * );
 */
exports.FilterListItem = (0, react_1.memo)((inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { label, value, icon, isSelected: getIsSelected = DefaultIsSelected, toggleFilter: userToggleFilter = DefaultToggleFilter, ...rest } = props;
    const { filterValues, setFilters } = (0, ra_core_1.useListFilterContext)();
    const translate = (0, ra_core_1.useTranslate)();
    const toggleFilter = (0, ra_core_1.useEvent)(userToggleFilter);
    // We can't wrap this function with useEvent as it is called in the render phase
    const isSelected = getIsSelected(value, filterValues);
    const handleClick = () => setFilters(toggleFilter(value, filterValues));
    return (React.createElement(StyledListItem, { disablePadding: true, disableGutters: true, secondaryAction: isSelected ? (React.createElement(material_1.IconButton, { size: "small", onClick: event => {
                event.stopPropagation();
                handleClick();
            } },
            React.createElement(CancelOutlined_1.default, null))) : null, ...rest },
        React.createElement(material_1.ListItemButton, { onClick: handleClick, selected: isSelected, className: exports.FilterListItemClasses.listItemButton },
            icon && (React.createElement(material_1.ListItemIcon, { className: exports.FilterListItemClasses.listItemIcon }, icon)),
            React.createElement(material_1.ListItemText, { primary: typeof label === 'string' && !(0, react_is_1.isElement)(label)
                    ? translate(label, { _: label })
                    : label, className: exports.FilterListItemClasses.listItemText, "data-selected": isSelected ? 'true' : 'false' }))));
}, arePropsEqual);
const DefaultIsSelected = (value, filters) => (0, matches_js_1.default)((0, pickBy_js_1.default)(value, val => typeof val !== 'undefined'))(filters);
const DefaultToggleFilter = (value, filters) => {
    const isSelected = (0, matches_js_1.default)((0, pickBy_js_1.default)(value, val => typeof val !== 'undefined'))(filters);
    if (isSelected) {
        const keysToRemove = Object.keys(value);
        return Object.keys(filters).reduce((acc, key) => keysToRemove.includes(key)
            ? acc
            : { ...acc, [key]: filters[key] }, {});
    }
    return { ...filters, ...value };
};
const PREFIX = 'RaFilterListItem';
exports.FilterListItemClasses = {
    listItemButton: `${PREFIX}-listItemButton`,
    listItemText: `${PREFIX}-listItemText`,
    listItemIcon: `${PREFIX}-listItemIcon`,
};
const StyledListItem = (0, styles_1.styled)(material_1.ListItem, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${exports.FilterListItemClasses.listItemButton}`]: {
        paddingRight: '2em',
        paddingLeft: '2em',
    },
    [`& .${exports.FilterListItemClasses.listItemText}`]: {
        margin: 0,
    },
    [`& .${exports.FilterListItemClasses.listItemIcon}`]: {
        minWidth: 0,
        marginRight: '0.5em',
    },
});
