// ra-ui-materialui/dist/list/filter/FilterButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterButton = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const material_1 = require("@mui/material");
const Clear_1 = __importDefault(require("@mui/icons-material/Clear"));
const BookmarkAdd_1 = __importDefault(require("@mui/icons-material/BookmarkAdd"));
const BookmarkRemove_1 = __importDefault(require("@mui/icons-material/BookmarkRemove"));
const BookmarkBorder_1 = __importDefault(require("@mui/icons-material/BookmarkBorder"));
const FilterList_1 = __importDefault(require("@mui/icons-material/FilterList"));
const isEqual_js_1 = __importDefault(require("lodash/isEqual.js"));
const ra_core_1 = require("ra-core");
const query_string_1 = require("query-string");
const FilterButtonMenuItem_1 = require("./FilterButtonMenuItem.cjs");
const button_1 = require("../../button/index.cjs");
const AddSavedQueryDialog_1 = require("./AddSavedQueryDialog.cjs");
const RemoveSavedQueryDialog_1 = require("./RemoveSavedQueryDialog.cjs");
const FilterButton = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { filters: filtersProp, className, disableSaveQuery, size, variant, ...rest } = props;
    const filters = (0, ra_core_1.useFilterContext)() || filtersProp;
    const resource = (0, ra_core_1.useResourceContext)(props);
    const translate = (0, ra_core_1.useTranslate)();
    if (!resource && !disableSaveQuery) {
        throw new Error('<FilterButton> must be called inside a ResourceContextProvider, or must provide a resource prop');
    }
    const [savedQueries] = (0, ra_core_1.useSavedQueries)(resource || '');
    const navigate = (0, ra_core_1.useNavigate)();
    const { displayedFilters = {}, filterValues, perPage, setFilters, showFilter, hideFilter, sort, } = (0, ra_core_1.useListContext)();
    const hasFilterValues = !(0, isEqual_js_1.default)(filterValues, {});
    const validSavedQueries = (0, ra_core_1.extractValidSavedQueries)(savedQueries);
    const hasSavedCurrentQuery = validSavedQueries.some(savedQuery => (0, isEqual_js_1.default)(savedQuery.value, {
        filter: filterValues,
        sort,
        perPage,
        displayedFilters,
    }));
    const [open, setOpen] = (0, react_1.useState)(false);
    const anchorEl = (0, react_1.useRef)();
    if (filters === undefined) {
        throw new Error('The <FilterButton> component requires the <List filters> prop to be set');
    }
    const allTogglableFilters = filters.filter((filterElement) => {
        return (React.isValidElement(filterElement) &&
            !filterElement.props.alwaysOn);
    });
    const handleClickButton = (0, react_1.useCallback)(event => {
        // This prevents ghost click.
        event.preventDefault();
        setOpen(true);
        anchorEl.current = event.currentTarget;
    }, [anchorEl, setOpen]);
    const handleRequestClose = (0, react_1.useCallback)(() => {
        setOpen(false);
    }, [setOpen]);
    const handleShow = (0, react_1.useCallback)(({ source, defaultValue }) => {
        showFilter(source, defaultValue === '' ? undefined : defaultValue);
        // We have to fallback to imperative code because the new FilterFormInput
        // has no way of knowing it has just been displayed (and thus that it should focus its input)
        setTimeout(() => {
            const inputElement = document.querySelector(`input[name='${source}']`);
            if (inputElement) {
                inputElement.focus();
            }
        }, 50);
        setOpen(false);
    }, [showFilter, setOpen]);
    const handleRemove = (0, react_1.useCallback)(({ source }) => {
        hideFilter(source);
        setOpen(false);
    }, [hideFilter, setOpen]);
    // add query dialog state
    const [addSavedQueryDialogOpen, setAddSavedQueryDialogOpen] = (0, react_1.useState)(false);
    const hideAddSavedQueryDialog = () => {
        setAddSavedQueryDialogOpen(false);
    };
    const showAddSavedQueryDialog = () => {
        setOpen(false);
        setAddSavedQueryDialogOpen(true);
    };
    // remove query dialog state
    const [removeSavedQueryDialogOpen, setRemoveSavedQueryDialogOpen] = (0, react_1.useState)(false);
    const hideRemoveSavedQueryDialog = () => {
        setRemoveSavedQueryDialogOpen(false);
    };
    const showRemoveSavedQueryDialog = () => {
        setOpen(false);
        setRemoveSavedQueryDialogOpen(true);
    };
    if (allTogglableFilters.length === 0 &&
        validSavedQueries.length === 0 &&
        !hasFilterValues) {
        return null;
    }
    return (React.createElement(Root, { className: className, ...sanitizeRestProps(rest) },
        React.createElement(button_1.Button, { className: "add-filter", label: "ra.action.add_filter", "aria-haspopup": "true", onClick: handleClickButton, variant: variant, size: size },
            React.createElement(FilterList_1.default, null)),
        React.createElement(material_1.Menu, { open: open, anchorEl: anchorEl.current, onClose: handleRequestClose },
            allTogglableFilters.map((filterElement, index) => (React.createElement(FilterButtonMenuItem_1.FilterButtonMenuItem, { key: filterElement.props.source, filter: filterElement, displayed: !!displayedFilters[filterElement.props.source], resource: resource, onShow: handleShow, onHide: handleRemove, autoFocus: index === 0 }))),
            (hasFilterValues || validSavedQueries.length > 0) && (React.createElement(material_1.Divider, null)),
            validSavedQueries.map((savedQuery, index) => (0, isEqual_js_1.default)(savedQuery.value, {
                filter: filterValues,
                sort,
                perPage,
                displayedFilters,
            }) ? (React.createElement(material_1.MenuItem, { onClick: showRemoveSavedQueryDialog, key: index },
                React.createElement(material_1.ListItemIcon, null,
                    React.createElement(BookmarkRemove_1.default, { fontSize: "small" })),
                React.createElement(material_1.ListItemText, null, translate('ra.saved_queries.remove_label_with_name', {
                    _: 'Remove query "%{name}"',
                    name: savedQuery.label,
                })))) : (React.createElement(material_1.MenuItem, { onClick: () => {
                    navigate({
                        search: (0, query_string_1.stringify)({
                            filter: JSON.stringify(savedQuery.value.filter),
                            sort: savedQuery.value.sort?.field,
                            order: savedQuery.value.sort?.order,
                            page: 1,
                            perPage: savedQuery.value.perPage,
                            displayedFilters: JSON.stringify(savedQuery.value.displayedFilters),
                        }),
                    });
                    setOpen(false);
                }, key: index },
                React.createElement(material_1.ListItemIcon, null,
                    React.createElement(BookmarkBorder_1.default, { fontSize: "small" })),
                React.createElement(material_1.ListItemText, null, savedQuery.label)))),
            hasFilterValues &&
                !hasSavedCurrentQuery &&
                !disableSaveQuery && (React.createElement(material_1.MenuItem, { onClick: showAddSavedQueryDialog },
                React.createElement(material_1.ListItemIcon, null,
                    React.createElement(BookmarkAdd_1.default, { fontSize: "small" })),
                React.createElement(material_1.ListItemText, null, translate('ra.saved_queries.new_label', {
                    _: 'Save current query...',
                })))),
            hasFilterValues && (React.createElement(material_1.MenuItem, { onClick: () => {
                    setFilters({}, {});
                    setOpen(false);
                } },
                React.createElement(material_1.ListItemIcon, null,
                    React.createElement(Clear_1.default, { fontSize: "small" })),
                React.createElement(material_1.ListItemText, null, translate('ra.action.remove_all_filters', {
                    _: 'Remove all filters',
                }))))),
        !disableSaveQuery && (React.createElement(React.Fragment, null,
            React.createElement(AddSavedQueryDialog_1.AddSavedQueryDialog, { open: addSavedQueryDialogOpen, onClose: hideAddSavedQueryDialog }),
            React.createElement(RemoveSavedQueryDialog_1.RemoveSavedQueryDialog, { open: removeSavedQueryDialogOpen, onClose: hideRemoveSavedQueryDialog })))));
};
exports.FilterButton = FilterButton;
/* eslint-disable @typescript-eslint/no-unused-vars */
const sanitizeRestProps = ({ displayedFilters = null, filterValues = null, showFilter = null, ...rest }) => rest;
const PREFIX = 'RaFilterButton';
const Root = (0, material_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'inline-block',
});
