// ra-ui-materialui/dist/list/filter/SavedQueryFilterListItem.js
import * as React from 'react';
import { memo } from 'react';
import { IconButton, ListItem, ListItemButton, ListItemText, styled, useThemeProps, } from '@mui/material';
import CancelIcon from '@mui/icons-material/CancelOutlined';
import isEqual from 'lodash/isEqual.js';
import { useListContext } from 'ra-core';
const arePropsEqual = (prevProps, nextProps) => prevProps.label === nextProps.label &&
    isEqual(prevProps.value, nextProps.value);
export const SavedQueryFilterListItem = memo((inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, label, sx, value } = props;
    const { filterValues, sort, perPage, displayedFilters, setFilters, setPage, } = useListContext();
    const isSelected = isEqual(value, {
        filter: filterValues,
        sort,
        perPage,
        displayedFilters,
    });
    const addFilter = () => {
        setFilters(value.filter, value.displayedFilters);
        setPage(1);
    };
    const removeFilter = () => {
        setFilters({}, {});
        setPage(1);
    };
    const toggleFilter = () => isSelected ? removeFilter() : addFilter();
    return (React.createElement(StyledListItem, { className: className, sx: sx, disablePadding: true, disableGutters: true, secondaryAction: isSelected ? (React.createElement(IconButton, { size: "small", onClick: toggleFilter },
            React.createElement(CancelIcon, null))) : null },
        React.createElement(ListItemButton, { onClick: toggleFilter, selected: isSelected, className: SavedQueryFilterListItemClasses.listItemButton },
            React.createElement(ListItemText, { primary: label, className: SavedQueryFilterListItemClasses.listItemText }))));
}, arePropsEqual);
const PREFIX = 'RaSavedQueryFilterListItem';
export const SavedQueryFilterListItemClasses = {
    listItemButton: `${PREFIX}-listItemButton`,
    listItemText: `${PREFIX}-listItemText`,
};
const StyledListItem = styled(ListItem, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    [`& .${SavedQueryFilterListItemClasses.listItemButton}`]: {
        paddingRight: '2em',
        paddingLeft: '2em',
    },
    [`& .${SavedQueryFilterListItemClasses.listItemText}`]: {
        margin: 0,
    },
}));
