// ra-ui-materialui/dist/list/filter/SavedQueryFilterListItem.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SavedQueryFilterListItemClasses = exports.SavedQueryFilterListItem = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const material_1 = require("@mui/material");
const CancelOutlined_1 = __importDefault(require("@mui/icons-material/CancelOutlined"));
const isEqual_js_1 = __importDefault(require("lodash/isEqual.js"));
const ra_core_1 = require("ra-core");
const arePropsEqual = (prevProps, nextProps) => prevProps.label === nextProps.label &&
    (0, isEqual_js_1.default)(prevProps.value, nextProps.value);
exports.SavedQueryFilterListItem = (0, react_1.memo)((inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, label, sx, value } = props;
    const { filterValues, sort, perPage, displayedFilters, setFilters, setPage, } = (0, ra_core_1.useListContext)();
    const isSelected = (0, isEqual_js_1.default)(value, {
        filter: filterValues,
        sort,
        perPage,
        displayedFilters,
    });
    const addFilter = () => {
        setFilters(value.filter, value.displayedFilters);
        setPage(1);
    };
    const removeFilter = () => {
        setFilters({}, {});
        setPage(1);
    };
    const toggleFilter = () => isSelected ? removeFilter() : addFilter();
    return (React.createElement(StyledListItem, { className: className, sx: sx, disablePadding: true, disableGutters: true, secondaryAction: isSelected ? (React.createElement(material_1.IconButton, { size: "small", onClick: toggleFilter },
            React.createElement(CancelOutlined_1.default, null))) : null },
        React.createElement(material_1.ListItemButton, { onClick: toggleFilter, selected: isSelected, className: exports.SavedQueryFilterListItemClasses.listItemButton },
            React.createElement(material_1.ListItemText, { primary: label, className: exports.SavedQueryFilterListItemClasses.listItemText }))));
}, arePropsEqual);
const PREFIX = 'RaSavedQueryFilterListItem';
exports.SavedQueryFilterListItemClasses = {
    listItemButton: `${PREFIX}-listItemButton`,
    listItemText: `${PREFIX}-listItemText`,
};
const StyledListItem = (0, material_1.styled)(material_1.ListItem, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({
    [`& .${exports.SavedQueryFilterListItemClasses.listItemButton}`]: {
        paddingRight: '2em',
        paddingLeft: '2em',
    },
    [`& .${exports.SavedQueryFilterListItemClasses.listItemText}`]: {
        margin: 0,
    },
}));
