// ra-ui-materialui/dist/list/filter/SavedQueriesList.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SavedQueriesListClasses = exports.SavedQueriesList = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const BookmarkBorder_1 = __importDefault(require("@mui/icons-material/BookmarkBorder"));
const HelpOutline_1 = __importDefault(require("@mui/icons-material/HelpOutline"));
const ra_core_1 = require("ra-core");
const isEqual_js_1 = __importDefault(require("lodash/isEqual.js"));
const RemoveSavedQueryIconButton_1 = require("./RemoveSavedQueryIconButton.cjs");
const AddSavedQueryIconButton_1 = require("./AddSavedQueryIconButton.cjs");
const SavedQueryFilterListItem_1 = require("./SavedQueryFilterListItem.cjs");
const FilterList_1 = require("./FilterList.cjs");
/**
 * FilterList-like component allowing to save and restore a query (filters, sort, perPage).
 *
 * Use this component in a Filter sidebar to let users store custom queries
 * that they can reuse later. The saved queries will appear as FilterListItems,
 * and can be removed.
 *
 * This component uses useStore under the hood to store saved queries in
 * localStorage, one set of saved queries per resource.
 *
 * @example
 *
 * import { FilterList, FilterListItem, List, Datagrid, SavedQueriesList } from 'react-admin';
 * import { Card, CardContent } from '@mui/material';
 *
 * const PostFilterSidebar = () => (
 *     <Card>
 *         <CardContent>
 *             <SavedQueriesList />
 *             <FilterList label="Category" icon={<CategoryIcon />}>
 *                 ...
 *             </FilterList>
 *             <FilterList label="Published" icon={<DateRangeIcon />}>
 *                ...
 *             </FilterList>
 *             <FilterList label="Popularity" icon={<DateRangeIcon />}>
 *                ...
 *             </FilterList>
 *         </CardContent>
 *     </Card>
 * );
 *
 * const PostList = () => (
 *     <List aside={<PostFilterSidebar />}>
 *         <Datagrid>
 *             ...
 *         </Datagrid>
 *     </List>
 * );
 *
 */
const SavedQueriesList = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { icon = defaultIcon } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const { resource, filterValues, displayedFilters, sort, perPage } = (0, ra_core_1.useListContext)();
    const [savedQueries] = (0, ra_core_1.useSavedQueries)(resource);
    const validSavedQueries = (0, ra_core_1.extractValidSavedQueries)(savedQueries);
    const hasSavedCurrentFilterValue = validSavedQueries.some(savedQuery => (0, isEqual_js_1.default)(savedQuery.value, {
        filter: filterValues,
        sort,
        perPage,
        displayedFilters,
    }));
    const hasFilterValues = !(0, isEqual_js_1.default)(filterValues, {});
    return (React.createElement(Root, { label: "ra.saved_queries.label", icon: icon },
        hasSavedCurrentFilterValue ? (React.createElement(RemoveSavedQueryIconButton_1.RemoveSavedQueryIconButton, { className: exports.SavedQueriesListClasses.floatingIcon })) : hasFilterValues ? (React.createElement(AddSavedQueryIconButton_1.AddSavedQueryIconButton, { className: exports.SavedQueriesListClasses.floatingIcon })) : (React.createElement(material_1.Tooltip, { title: translate('ra.saved_queries.help'), className: exports.SavedQueriesListClasses.floatingTooltip },
            React.createElement(HelpOutline_1.default, null))),
        validSavedQueries.map((savedQuery, index) => (React.createElement(SavedQueryFilterListItem_1.SavedQueryFilterListItem, { label: savedQuery.label, value: savedQuery.value, key: index })))));
};
exports.SavedQueriesList = SavedQueriesList;
const PREFIX = 'RaSavedQueriesList';
exports.SavedQueriesListClasses = {
    floatingIcon: `${PREFIX}-floatingIcon`,
    floatingTooltip: `${PREFIX}-floatingTooltip`,
    titleContainer: `${PREFIX}-titleContainer`,
    titleIcon: `${PREFIX}-titleIcon`,
};
const Root = (0, material_1.styled)(FilterList_1.FilterList, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.SavedQueriesListClasses.floatingIcon}`]: {
        position: 'absolute',
        top: '-1.8em',
        right: 0,
    },
    [`& .${exports.SavedQueriesListClasses.floatingTooltip}`]: {
        position: 'absolute',
        top: '-1.2em',
        right: 3,
        color: (theme.vars || theme).palette.action.disabled,
    },
}));
const defaultIcon = React.createElement(BookmarkBorder_1.default, null);
