// ra-core/dist/controller/show/useShowController.js
import { useAuthenticated, useRequireAccess } from "../../auth/index.js";
import { useGetOne, } from "../../dataProvider/index.js";
import { useTranslate } from "../../i18n/index.js";
import { useRedirect, useParams } from "../../routing/index.js";
import { useNotify } from "../../notification/index.js";
import { useResourceContext, useGetResourceLabel, useGetRecordRepresentation, } from "../../core/index.js";
/**
 * Prepare data for the Show view.
 *
 * useShowController does a few things:
 * - it grabs the id from the URL and the resource name from the ResourceContext,
 * - it fetches the record via useGetOne,
 * - it prepares the page title.
 *
 * @param {Object} props The props passed to the Show component.
 *
 * @return {Object} controllerProps Fetched data and callbacks for the Show view
 *
 * @example
 *
 * import { useShowController } from 'react-admin';
 * import ShowView from './ShowView';
 *
 * const MyShow = () => {
 *     const controllerProps = useShowController();
 *     return <ShowView {...controllerProps} />;
 * };
 *
 * @example // useShowController can also take its parameters from props
 *
 * import { useShowController } from 'react-admin';
 * import ShowView from './ShowView';
 *
 * const MyShow = () => {
 *     const controllerProps = useShowController({ resource: 'posts', id: 1234 });
 *     return <ShowView {...controllerProps} />;
 * };
 */
export const useShowController = (props = {}) => {
    const { disableAuthentication = false, id: propsId, queryOptions = {}, redirectOnError = DefaultRedirectOnError, } = props;
    const resource = useResourceContext(props);
    if (!resource) {
        throw new Error(`useShowController requires a non-empty resource prop or context`);
    }
    const { isPending: isPendingAuthenticated } = useAuthenticated({
        enabled: !disableAuthentication,
    });
    const { isPending: isPendingCanAccess } = useRequireAccess({
        action: 'show',
        resource,
        enabled: !disableAuthentication && !isPendingAuthenticated,
    });
    const getRecordRepresentation = useGetRecordRepresentation(resource);
    const translate = useTranslate();
    const notify = useNotify();
    const redirect = useRedirect();
    const { id: routeId } = useParams();
    if (!routeId && !propsId) {
        throw new Error('useShowController requires an id prop or a route with an /:id? parameter.');
    }
    const id = propsId != null ? propsId : routeId;
    const { meta, ...otherQueryOptions } = queryOptions;
    const { data: record, error, isLoading, isFetching, isPaused, isPending, isPlaceholderData, refetch, } = useGetOne(resource, { id, meta }, {
        enabled: (!isPendingAuthenticated && !isPendingCanAccess) ||
            disableAuthentication,
        onError: () => {
            notify('ra.notification.item_doesnt_exist', {
                type: 'error',
            });
            redirect(redirectOnError, resource, id);
        },
        retry: false,
        ...otherQueryOptions,
    });
    // eslint-disable-next-line eqeqeq
    if (record && record.id && record.id != id) {
        throw new Error(`useShowController: Fetched record's id attribute (${record.id}) must match the requested 'id' (${id})`);
    }
    const getResourceLabel = useGetResourceLabel();
    const recordRepresentation = getRecordRepresentation(record);
    const defaultTitle = translate(`resources.${resource}.page.show`, {
        id,
        record,
        recordRepresentation: typeof recordRepresentation === 'string'
            ? recordRepresentation
            : '',
        _: translate('ra.page.show', {
            name: getResourceLabel(resource, 1),
            id,
            record,
            recordRepresentation: typeof recordRepresentation === 'string'
                ? recordRepresentation
                : '',
        }),
    });
    return {
        defaultTitle,
        error,
        isLoading,
        isFetching,
        isPaused,
        isPending,
        isPlaceholderData,
        record,
        redirectOnError,
        refetch,
        resource,
    };
};
const DefaultRedirectOnError = 'list';
