// ra-ui-materialui/dist/input/ArrayInput/ArrayInput.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayInputClasses = exports.ArrayInput = void 0;
const React = __importStar(require("react"));
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const react_hook_form_1 = require("react-hook-form");
const material_1 = require("@mui/material");
const get_js_1 = __importDefault(require("lodash/get.js"));
const LinearProgress_1 = require("../../layout/LinearProgress.cjs");
const InputHelperText_1 = require("../InputHelperText.cjs");
const sanitizeInputRestProps_1 = require("../sanitizeInputRestProps.cjs");
const Labeled_1 = require("../../Labeled.cjs");
/**
 * To edit arrays of data embedded inside a record, <ArrayInput> creates a list of sub-forms.
 *
 *  @example
 *
 *      import { ArrayInput, SimpleFormIterator, DateInput, TextInput } from 'react-admin';
 *
 *      <ArrayInput source="backlinks">
 *          <SimpleFormIterator>
 *              <DateInput source="date" />
 *              <TextInput source="url" />
 *          </SimpleFormIterator>
 *      </ArrayInput>
 *
 * <ArrayInput> allows the edition of embedded arrays, like the backlinks field
 * in the following post record:
 *
 * {
 *   id: 123
 *   backlinks: [
 *         {
 *             date: '2012-08-10T00:00:00.000Z',
 *             url: 'http://example.com/foo/bar.html',
 *         },
 *         {
 *             date: '2012-08-14T00:00:00.000Z',
 *             url: 'https://blog.johndoe.com/2012/08/12/foobar.html',
 *         }
 *    ]
 * }
 *
 * <ArrayInput> expects a single child, which must be a *form iterator* component.
 * A form iterator is a component accepting a fields object as passed by
 * react-hook-form-arrays's useFieldArray() hook, and defining a layout for
 * an array of fields. For instance, the <SimpleFormIterator> component
 * displays an array of fields in an unordered list (<ul>), one sub-form by
 * list item (<li>). It also provides controls for adding and removing
 * a sub-record (a backlink in this example).
 *
 * @see {@link https://react-hook-form.com/docs/usefieldarray}
 */
const ArrayInput = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, defaultValue = [], label, children, helperText, isPending, resource: resourceFromProps, source: arraySource, validate, variant, margin = 'dense', ...rest } = props;
    const parentSourceContext = (0, ra_core_1.useSourceContext)();
    const finalSource = parentSourceContext.getSource(arraySource);
    const { subscribe } = (0, react_hook_form_1.useFormContext)();
    const [error, setError] = React.useState();
    React.useEffect(() => {
        return subscribe({
            formState: { errors: true },
            callback: ({ errors }) => {
                const error = (0, get_js_1.default)(errors ?? {}, finalSource);
                setError(error);
            },
        });
    }, [finalSource, subscribe]);
    const renderHelperText = helperText !== false || !!error;
    if (isPending) {
        // We handle the loading state here instead of using the loading prop
        // of ArrayInputBase to avoid wrapping the content below inside Root
        return (React.createElement(Labeled_1.Labeled, { label: label, className: className },
            React.createElement(LinearProgress_1.LinearProgress, null)));
    }
    return (React.createElement(Root, { fullWidth: true, margin: margin, className: (0, clsx_1.default)('ra-input', `ra-input-${finalSource}`, exports.ArrayInputClasses.root, className), error: !!error, ...(0, sanitizeInputRestProps_1.sanitizeInputRestProps)(rest) },
        React.createElement(material_1.InputLabel, { component: "span", className: exports.ArrayInputClasses.label, shrink: true, error: !!error },
            React.createElement(ra_core_1.FieldTitle, { label: label, source: arraySource, resource: resourceFromProps, isRequired: (0, ra_core_1.isRequired)(validate) })),
        React.createElement(ra_core_1.ArrayInputBase, { ...props, defaultValue: defaultValue }, children),
        renderHelperText ? (React.createElement(material_1.FormHelperText, { error: !!error },
            React.createElement(InputHelperText_1.InputHelperText
            // root property is applicable to built-in validation only,
            // Resolvers are yet to support useFieldArray root level validation.
            // Reference: https://react-hook-form.com/docs/usefieldarray
            , { 
                // root property is applicable to built-in validation only,
                // Resolvers are yet to support useFieldArray root level validation.
                // Reference: https://react-hook-form.com/docs/usefieldarray
                error: error?.root?.message ?? error?.message, helperText: helperText }))) : null));
};
exports.ArrayInput = ArrayInput;
const PREFIX = 'RaArrayInput';
exports.ArrayInputClasses = {
    root: `${PREFIX}-root`,
    label: `${PREFIX}-label`,
};
const Root = (0, material_1.styled)(material_1.FormControl, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    marginTop: 0,
    [`& .${exports.ArrayInputClasses.label}`]: {
        position: 'relative',
        top: theme.spacing(0.5),
        left: theme.spacing(-1.5),
    },
    [`& .${exports.ArrayInputClasses.root}`]: {
        // nested ArrayInput
        paddingLeft: theme.spacing(2),
    },
}));
