// ra-ui-materialui/dist/input/ArrayInput/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleFormIteratorItemContext = exports.SimpleFormIteratorContext = exports.useSimpleFormIteratorItem = exports.useSimpleFormIterator = exports.useArrayInput = exports.ArrayInputContext = void 0;
__exportStar(require("./ArrayInput.cjs"), exports);
__exportStar(require("./SimpleFormIterator.cjs"), exports);
__exportStar(require("./SimpleFormIteratorItem.cjs"), exports);
__exportStar(require("./useSimpleFormIteratorStyles.cjs"), exports);
__exportStar(require("./AddItemButton.cjs"), exports);
__exportStar(require("./RemoveItemButton.cjs"), exports);
__exportStar(require("./ReOrderButtons.cjs"), exports);
/**
 * @deprecated Import from `ra-core` or `react-admin` instead
 * FIXME: remove the re-export in v6
 */
var ra_core_1 = require("ra-core");
Object.defineProperty(exports, "ArrayInputContext", { enumerable: true, get: function () { return ra_core_1.ArrayInputContext; } });
Object.defineProperty(exports, "useArrayInput", { enumerable: true, get: function () { return ra_core_1.useArrayInput; } });
Object.defineProperty(exports, "useSimpleFormIterator", { enumerable: true, get: function () { return ra_core_1.useSimpleFormIterator; } });
Object.defineProperty(exports, "useSimpleFormIteratorItem", { enumerable: true, get: function () { return ra_core_1.useSimpleFormIteratorItem; } });
Object.defineProperty(exports, "SimpleFormIteratorContext", { enumerable: true, get: function () { return ra_core_1.SimpleFormIteratorContext; } });
Object.defineProperty(exports, "SimpleFormIteratorItemContext", { enumerable: true, get: function () { return ra_core_1.SimpleFormIteratorItemContext; } });
