// ra-ui-materialui/dist/input/ArrayInput/SimpleFormIterator.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleFormIterator = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const get_js_1 = __importDefault(require("lodash/get.js"));
const useSimpleFormIteratorStyles_1 = require("./useSimpleFormIteratorStyles.cjs");
const SimpleFormIteratorItem_1 = require("./SimpleFormIteratorItem.cjs");
const AddItemButton_1 = require("./AddItemButton.cjs");
const SimpleFormIteratorClearButton_1 = require("./SimpleFormIteratorClearButton.cjs");
const SimpleFormIterator = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: useSimpleFormIteratorStyles_1.SimpleFormIteratorPrefix,
    });
    const { addButton = React.createElement(AddItemButton_1.AddItemButton, null), removeButton, reOrderButtons, children, className, disabled, disableAdd = false, disableClear, disableRemove = false, disableReordering, inline, getItemLabel = false, fullWidth, sx, } = props;
    const finalSource = (0, ra_core_1.useWrappedSource)('');
    if (!finalSource) {
        throw new Error('SimpleFormIterator can only be called within an iterator input like ArrayInput');
    }
    const { fields } = (0, ra_core_1.useArrayInput)(props);
    const record = (0, ra_core_1.useRecordContext)(props);
    const records = (0, get_js_1.default)(record, finalSource);
    const getArrayInputNewItemDefaults = (0, ra_core_1.useGetArrayInputNewItemDefaults)(fields);
    const getItemDefaults = (0, ra_core_1.useEvent)((item = undefined) => {
        if (item != null)
            return item;
        return getArrayInputNewItemDefaults(children);
    });
    return (React.createElement(ra_core_1.SimpleFormIteratorBase, { getItemDefaults: getItemDefaults, ...props },
        React.createElement(Root, { className: (0, clsx_1.default)(className, fullWidth && 'fullwidth', disabled && 'disabled'), sx: sx },
            React.createElement("ul", { className: useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.list }, fields.map((member, index) => (React.createElement(ra_core_1.RecordContextProvider, { key: member.id, value: (records && records[index]) || {} },
                React.createElement(SimpleFormIteratorItem_1.SimpleFormIteratorItem, { index: index, fields: fields, resource: props.resource, disabled: disabled, disableRemove: disableRemove, disableReordering: disableReordering, getItemLabel: getItemLabel, removeButton: removeButton, reOrderButtons: reOrderButtons, inline: inline }, children))))),
            !disabled &&
                !(disableAdd && (disableClear || disableRemove)) && (React.createElement("div", { className: useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.buttons },
                !disableAdd && (React.createElement("div", { className: useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.add }, addButton)),
                React.createElement("div", { className: useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.clear },
                    React.createElement(SimpleFormIteratorClearButton_1.SimpleFormIteratorClearButton, { disableClear: disableClear, disableRemove: disableRemove })))))));
};
exports.SimpleFormIterator = SimpleFormIterator;
const Root = (0, material_1.styled)('div', {
    name: useSimpleFormIteratorStyles_1.SimpleFormIteratorPrefix,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    '& > ul': {
        padding: 0,
        marginTop: 0,
        marginBottom: 0,
    },
    '& > ul > li:last-child': {
        // hide the last separator
        borderBottom: 'none',
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.line}`]: {
        display: 'flex',
        listStyleType: 'none',
        borderBottom: `solid 1px ${(theme.vars || theme).palette.divider}`,
        [theme.breakpoints.down('sm')]: { display: 'block' },
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.index}`]: {
        display: 'flex',
        alignItems: 'top',
        marginRight: theme.spacing(1),
        marginTop: theme.spacing(1),
        [theme.breakpoints.down('md')]: { display: 'none' },
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.form}`]: {
        minWidth: 0,
    },
    [`&.fullwidth > ul > li > .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.form}`]: {
        flex: 2,
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.inline}`]: {
        flexDirection: 'row',
        columnGap: '1em',
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.action}`]: {
        marginTop: theme.spacing(0.5),
        visibility: 'hidden',
        '@media(hover:none)': {
            visibility: 'visible',
        },
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.buttons}`]: {
        display: 'flex',
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.add}`]: {
        borderBottom: 'none',
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.clear}`]: {
        borderBottom: 'none',
    },
    [`& .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.line}:hover > .${useSimpleFormIteratorStyles_1.SimpleFormIteratorClasses.action}`]: {
        visibility: 'visible',
    },
}));
