// ra-ui-materialui/dist/input/ArrayInput/SimpleFormIterator.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material';
import clsx from 'clsx';
import { useWrappedSource, SimpleFormIteratorBase, RecordContextProvider, useArrayInput, useRecordContext, useEvent, useGetArrayInputNewItemDefaults, } from 'ra-core';
import get from 'lodash/get.js';
import { SimpleFormIteratorClasses, SimpleFormIteratorPrefix as PREFIX, } from "./useSimpleFormIteratorStyles.js";
import { SimpleFormIteratorItem, } from "./SimpleFormIteratorItem.js";
import { AddItemButton as DefaultAddItemButton } from "./AddItemButton.js";
import { SimpleFormIteratorClearButton } from "./SimpleFormIteratorClearButton.js";
export const SimpleFormIterator = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { addButton = React.createElement(DefaultAddItemButton, null), removeButton, reOrderButtons, children, className, disabled, disableAdd = false, disableClear, disableRemove = false, disableReordering, inline, getItemLabel = false, fullWidth, sx, } = props;
    const finalSource = useWrappedSource('');
    if (!finalSource) {
        throw new Error('SimpleFormIterator can only be called within an iterator input like ArrayInput');
    }
    const { fields } = useArrayInput(props);
    const record = useRecordContext(props);
    const records = get(record, finalSource);
    const getArrayInputNewItemDefaults = useGetArrayInputNewItemDefaults(fields);
    const getItemDefaults = useEvent((item = undefined) => {
        if (item != null)
            return item;
        return getArrayInputNewItemDefaults(children);
    });
    return (React.createElement(SimpleFormIteratorBase, { getItemDefaults: getItemDefaults, ...props },
        React.createElement(Root, { className: clsx(className, fullWidth && 'fullwidth', disabled && 'disabled'), sx: sx },
            React.createElement("ul", { className: SimpleFormIteratorClasses.list }, fields.map((member, index) => (React.createElement(RecordContextProvider, { key: member.id, value: (records && records[index]) || {} },
                React.createElement(SimpleFormIteratorItem, { index: index, fields: fields, resource: props.resource, disabled: disabled, disableRemove: disableRemove, disableReordering: disableReordering, getItemLabel: getItemLabel, removeButton: removeButton, reOrderButtons: reOrderButtons, inline: inline }, children))))),
            !disabled &&
                !(disableAdd && (disableClear || disableRemove)) && (React.createElement("div", { className: SimpleFormIteratorClasses.buttons },
                !disableAdd && (React.createElement("div", { className: SimpleFormIteratorClasses.add }, addButton)),
                React.createElement("div", { className: SimpleFormIteratorClasses.clear },
                    React.createElement(SimpleFormIteratorClearButton, { disableClear: disableClear, disableRemove: disableRemove })))))));
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    '& > ul': {
        padding: 0,
        marginTop: 0,
        marginBottom: 0,
    },
    '& > ul > li:last-child': {
        // hide the last separator
        borderBottom: 'none',
    },
    [`& .${SimpleFormIteratorClasses.line}`]: {
        display: 'flex',
        listStyleType: 'none',
        borderBottom: `solid 1px ${(theme.vars || theme).palette.divider}`,
        [theme.breakpoints.down('sm')]: { display: 'block' },
    },
    [`& .${SimpleFormIteratorClasses.index}`]: {
        display: 'flex',
        alignItems: 'top',
        marginRight: theme.spacing(1),
        marginTop: theme.spacing(1),
        [theme.breakpoints.down('md')]: { display: 'none' },
    },
    [`& .${SimpleFormIteratorClasses.form}`]: {
        minWidth: 0,
    },
    [`&.fullwidth > ul > li > .${SimpleFormIteratorClasses.form}`]: {
        flex: 2,
    },
    [`& .${SimpleFormIteratorClasses.inline}`]: {
        flexDirection: 'row',
        columnGap: '1em',
    },
    [`& .${SimpleFormIteratorClasses.action}`]: {
        marginTop: theme.spacing(0.5),
        visibility: 'hidden',
        '@media(hover:none)': {
            visibility: 'visible',
        },
    },
    [`& .${SimpleFormIteratorClasses.buttons}`]: {
        display: 'flex',
    },
    [`& .${SimpleFormIteratorClasses.add}`]: {
        borderBottom: 'none',
    },
    [`& .${SimpleFormIteratorClasses.clear}`]: {
        borderBottom: 'none',
    },
    [`& .${SimpleFormIteratorClasses.line}:hover > .${SimpleFormIteratorClasses.action}`]: {
        visibility: 'visible',
    },
}));
