// ra-core/dist/controller/list/useListParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNumberOrDefault = exports.getQuery = exports.hasCustomParams = exports.parseQueryFromLocation = exports.useListParams = void 0;
const react_1 = require("react");
const query_string_1 = require("query-string");
const debounce_js_1 = __importDefault(require("lodash/debounce.js"));
const store_1 = require("../../store/index.cjs");
const routing_1 = require("../../routing/index.cjs");
const queryReducer_1 = __importStar(require("./queryReducer.cjs"));
const removeEmpty_1 = __importDefault(require("../../util/removeEmpty.cjs"));
const hooks_1 = require("../../util/hooks.cjs");
/**
 * Get the list parameters (page, sort, filters) and modifiers.
 *
 * These parameters are merged from 3 sources:
 *   - the query string from the URL
 *   - the params stored in the state (from previous navigation)
 *   - the options passed to the hook (including the filter defaultValues)
 *
 * @returns {Array} A tuple [parameters, modifiers].
 * Destructure as [
 *    { page, perPage, sort, order, filter, filterValues, displayedFilters, requestSignature },
 *    { setFilters, hideFilter, showFilter, setPage, setPerPage, setSort }
 * ]
 *
 * @example
 *
 * const [listParams, listParamsActions] = useListParams({
 *      resource: 'posts',
 *      location: location // From react-router. Injected to your component by react-admin inside a List
 *      filterDefaultValues: {
 *          published: true
 *      },
 *      sort: {
 *          field: 'published_at',
 *          order: 'DESC'
 *      },
 *      perPage: 25
 * });
 *
 * const {
 *      page,
 *      perPage,
 *      sort,
 *      order,
 *      filter,
 *      filterValues,
 *      displayedFilters,
 *      requestSignature
 * } = listParams;
 *
 * const {
 *      setFilters,
 *      hideFilter,
 *      showFilter,
 *      setPage,
 *      setPerPage,
 *      setSort,
 * } = listParamsActions;
 */
const useListParams = ({ debounce = 500, disableSyncWithLocation = false, filterDefaultValues, perPage = 10, resource, sort = defaultSort, storeKey = disableSyncWithLocation ? false : `${resource}.listParams`, }) => {
    const location = (0, routing_1.useLocation)();
    const navigate = (0, routing_1.useNavigate)();
    const [localParams, setLocalParams] = (0, react_1.useState)(defaultParams);
    // As we can't conditionally call a hook, if the storeKey is false,
    // we'll ignore the params variable later on and won't call setParams either.
    const [params, setParams] = (0, store_1.useStore)(storeKey || `${resource}.listParams`, defaultParams);
    const tempParams = (0, react_1.useRef)();
    const isMounted = (0, hooks_1.useIsMounted)();
    const requestSignature = [
        location.search,
        resource,
        storeKey,
        JSON.stringify(!storeKey ? localParams : params),
        JSON.stringify(filterDefaultValues),
        JSON.stringify(sort),
        perPage,
        disableSyncWithLocation,
    ];
    const queryFromLocation = disableSyncWithLocation
        ? {}
        : (0, exports.parseQueryFromLocation)(location);
    const query = (0, react_1.useMemo)(() => (0, exports.getQuery)({
        queryFromLocation,
        params: !storeKey ? localParams : params,
        filterDefaultValues,
        sort,
        perPage,
    }), requestSignature // eslint-disable-line react-hooks/exhaustive-deps
    );
    // if the location includes params (for example from a link like
    // the categories products on the demo), we need to persist them in the
    // store as well so that we don't lose them after a redirection back
    // to the list
    (0, react_1.useEffect)(() => {
        if (Object.keys(queryFromLocation).length > 0) {
            setParams(query);
        }
    }, [location.search]); // eslint-disable-line
    const changeParams = (0, react_1.useCallback)(action => {
        // do not change params if the component is already unmounted
        // this is necessary because changeParams can be debounced, and therefore
        // executed after the component is unmounted
        if (!isMounted.current)
            return;
        if (!tempParams.current) {
            // no other changeParams action dispatched this tick
            tempParams.current = (0, queryReducer_1.default)(query, action);
            // schedule side effects for next tick
            setTimeout(() => {
                if (!tempParams.current) {
                    // the side effects were already processed by another changeParams
                    return;
                }
                if (disableSyncWithLocation && !storeKey) {
                    setLocalParams(tempParams.current);
                }
                else if (disableSyncWithLocation && !!storeKey) {
                    setParams(tempParams.current);
                }
                else {
                    // the useEffect above will apply the changes to the params in the store
                    navigate({
                        search: `?${(0, query_string_1.stringify)({
                            ...tempParams.current,
                            filter: JSON.stringify(tempParams.current.filter),
                            displayedFilters: JSON.stringify(tempParams.current.displayedFilters),
                        })}`,
                    }, {
                        state: {
                            _scrollToTop: action.type === queryReducer_1.SET_PAGE,
                        },
                    });
                }
                tempParams.current = undefined;
            }, 0);
        }
        else {
            // side effects already scheduled, just change the params
            tempParams.current = (0, queryReducer_1.default)(tempParams.current, action);
        }
    }, [...requestSignature, navigate] // eslint-disable-line react-hooks/exhaustive-deps
    );
    const setSort = (0, react_1.useCallback)((sort) => changeParams({
        type: queryReducer_1.SET_SORT,
        payload: sort,
    }), [changeParams]);
    const setPage = (0, react_1.useCallback)((newPage) => changeParams({ type: queryReducer_1.SET_PAGE, payload: newPage }), [changeParams]);
    const setPerPage = (0, react_1.useCallback)((newPerPage) => changeParams({ type: queryReducer_1.SET_PER_PAGE, payload: newPerPage }), [changeParams]);
    const filterValues = query.filter || emptyObject;
    const displayedFilterValues = query.displayedFilters || emptyObject;
    const debouncedSetFilters = (0, debounce_js_1.default)((filter, displayedFilters) => {
        changeParams({
            type: queryReducer_1.SET_FILTER,
            payload: {
                filter: (0, removeEmpty_1.default)(filter),
                displayedFilters,
            },
        });
    }, debounce);
    const setFilters = (0, react_1.useCallback)((filter, displayedFilters = undefined, debounce = false) => debounce
        ? debouncedSetFilters(filter, displayedFilters)
        : changeParams({
            type: queryReducer_1.SET_FILTER,
            payload: {
                filter: (0, removeEmpty_1.default)(filter),
                displayedFilters,
            },
        }), [changeParams] // eslint-disable-line react-hooks/exhaustive-deps
    );
    const hideFilter = (0, react_1.useCallback)((filterName) => {
        changeParams({
            type: queryReducer_1.HIDE_FILTER,
            payload: filterName,
        });
    }, [changeParams]);
    const showFilter = (0, react_1.useCallback)((filterName, defaultValue) => {
        changeParams({
            type: queryReducer_1.SHOW_FILTER,
            payload: {
                filterName,
                defaultValue,
            },
        });
    }, [changeParams]);
    return [
        {
            filterValues,
            requestSignature,
            ...query,
            displayedFilters: displayedFilterValues,
        },
        {
            changeParams,
            setPage,
            setPerPage,
            setSort,
            setFilters,
            hideFilter,
            showFilter,
        },
    ];
};
exports.useListParams = useListParams;
const parseObject = (query, field) => {
    if (query[field] && typeof query[field] === 'string') {
        try {
            query[field] = JSON.parse(query[field]);
        }
        catch (err) {
            delete query[field];
        }
    }
};
const parseQueryFromLocation = ({ search }) => {
    const query = (0, query_string_1.parse)(search);
    parseObject(query, 'filter');
    parseObject(query, 'displayedFilters');
    return query;
};
exports.parseQueryFromLocation = parseQueryFromLocation;
/**
 * Check if user has already set custom sort, page, or filters for this list
 *
 * User params come from the store as the params props. By default,
 * this object is:
 *
 * { filter: {}, order: null, page: 1, perPage: null, sort: null }
 *
 * To check if the user has custom params, we must compare the params
 * to these initial values.
 *
 * @param {Object} params
 */
const hasCustomParams = (params) => {
    return (params &&
        params.filter &&
        (Object.keys(params.filter).length > 0 ||
            params.order != null ||
            params.page !== 1 ||
            params.perPage != null ||
            params.sort != null));
};
exports.hasCustomParams = hasCustomParams;
/**
 * Merge list params from 3 different sources:
 *   - the query string
 *   - the params stored in the state (from previous navigation)
 *   - the props passed to the List component (including the filter defaultValues)
 */
const getQuery = ({ queryFromLocation, params, filterDefaultValues, sort, perPage, }) => {
    const query = Object.keys(queryFromLocation).length > 0
        ? queryFromLocation
        : (0, exports.hasCustomParams)(params)
            ? { ...params }
            : { filter: filterDefaultValues || {} };
    if (!query.sort) {
        query.sort = sort.field;
        query.order = sort.order;
    }
    if (query.perPage == null) {
        query.perPage = perPage;
    }
    if (query.page == null) {
        query.page = 1;
    }
    return {
        ...query,
        page: (0, exports.getNumberOrDefault)(query.page, 1),
        perPage: (0, exports.getNumberOrDefault)(query.perPage, 10),
    };
};
exports.getQuery = getQuery;
const getNumberOrDefault = (possibleNumber, defaultValue) => {
    if (typeof possibleNumber === 'undefined') {
        return defaultValue;
    }
    const parsedNumber = typeof possibleNumber === 'string'
        ? parseInt(possibleNumber, 10)
        : possibleNumber;
    return isNaN(parsedNumber) ? defaultValue : parsedNumber;
};
exports.getNumberOrDefault = getNumberOrDefault;
const emptyObject = {};
const defaultSort = {
    field: 'id',
    order: queryReducer_1.SORT_ASC,
};
const defaultParams = {};
