// ra-core/dist/controller/input/SimpleFormIteratorBase.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleFormIteratorBase = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const react_hook_form_1 = require("react-hook-form");
const useWrappedSource_1 = require("../../core/useWrappedSource.cjs");
const util_1 = require("../../util/index.cjs");
const useArrayInput_1 = require("./useArrayInput.cjs");
const SimpleFormIteratorContext_1 = require("./SimpleFormIteratorContext.cjs");
const DefaultGetItemDefaults = item => item;
const SimpleFormIteratorBase = (props) => {
    const { children, getItemDefaults: getItemDefaultsProp = DefaultGetItemDefaults, disableAutoFocus = false, } = props;
    const getItemDefaults = (0, util_1.useEvent)(getItemDefaultsProp);
    const finalSource = (0, useWrappedSource_1.useWrappedSource)('');
    if (!finalSource) {
        throw new Error('SimpleFormIterator can only be called within an iterator input like ArrayInput');
    }
    const { append, fields, move, remove } = (0, useArrayInput_1.useArrayInput)(props);
    const { trigger, getValues } = (0, react_hook_form_1.useFormContext)();
    const removeField = (0, util_1.useEvent)((index) => {
        remove(index);
        const isScalarArray = getValues(finalSource).every((value) => typeof value !== 'object');
        if (isScalarArray) {
            // Trigger validation on the Array to avoid ghost errors.
            // Otherwise, validation errors on removed fields might still be displayed
            trigger(finalSource);
        }
    });
    const addField = (0, util_1.useEvent)((item = undefined) => {
        append(getItemDefaults(item), { shouldFocus: !disableAutoFocus });
    });
    const handleReorder = (0, util_1.useEvent)((origin, destination) => {
        move(origin, destination);
    });
    const handleArrayClear = (0, util_1.useEvent)(() => {
        remove();
    });
    const context = (0, react_1.useMemo)(() => ({
        total: fields.length,
        add: addField,
        clear: handleArrayClear,
        remove: removeField,
        reOrder: handleReorder,
        source: finalSource,
    }), [
        addField,
        fields.length,
        handleArrayClear,
        handleReorder,
        removeField,
        finalSource,
    ]);
    if (!fields) {
        return null;
    }
    return (React.createElement(SimpleFormIteratorContext_1.SimpleFormIteratorContext.Provider, { value: context }, children));
};
exports.SimpleFormIteratorBase = SimpleFormIteratorBase;
