// ra-core/dist/form/FilterLiveForm.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFilterFormValues = exports.FilterLiveForm = void 0;
const React = __importStar(require("react"));
const isEqual_js_1 = __importDefault(require("lodash/isEqual.js"));
const cloneDeep_js_1 = __importDefault(require("lodash/cloneDeep.js"));
const get_js_1 = __importDefault(require("lodash/get.js"));
const mergeWith_js_1 = __importDefault(require("lodash/mergeWith.js"));
const set_js_1 = __importDefault(require("lodash/set.js"));
const react_1 = require("react");
const react_hook_form_1 = require("react-hook-form");
const core_1 = require("../core/index.cjs");
const FormGroupsProvider_1 = require("./groups/FormGroupsProvider.cjs");
const validation_1 = require("./validation/index.cjs");
const util_1 = require("../util/index.cjs");
const useListContext_1 = require("../controller/list/useListContext.cjs");
/**
 * This component offers a convenient way to create a form that automatically
 * updates the filters when the user changes its child input values.
 *
 * It fits nicely alongside a `<FilterList>` component, but you can also use it
 * at other places to create your own filter UI.
 *
 * @example
 * import MailIcon from '@mui/icons-material/MailOutline';
 * import TitleIcon from '@mui/icons-material/Title';
 * import { Card, CardContent } from '@mui/material';
 * import * as React from 'react';
 * import {
 *     FilterLiveForm,
 *     FilterList,
 *     FilterListItem,
 *     FilterListSection,
 *     TextInput,
 * } from 'react-admin';
 *
 * export const BookListAside = () => (
 *     <Card sx={{ order: -1, mr: 2, mt: 6, width: 250, height: 'fit-content' }}>
 *         <CardContent>
 *             <FilterList label="Subscribed to newsletter" icon={<MailIcon />}>
 *                 <FilterListItem label="Yes" value={{ has_newsletter: true }} />
 *                 <FilterListItem label="No" value={{ has_newsletter: false }} />
 *             </FilterList>
 *             <FilterListSection label="Title" icon={<TitleIcon />}>
 *                 <FilterLiveForm>
 *                     <TextInput source="title" resettable helperText={false} />
 *                 </FilterLiveForm>
 *             </FilterListSection>
 *         </CardContent>
 *     </Card>
 * );
 */
const FilterLiveForm = (props) => {
    const { filterValues, setFilters } = (0, useListContext_1.useListContext)();
    const resource = (0, core_1.useResourceContext)(props);
    const { debounce = 500, resolver, validate, children, formComponent: Component = HTMLForm, ...rest } = props;
    const finalResolver = resolver
        ? resolver
        : validate
            ? (0, validation_1.getSimpleValidationResolver)(validate)
            : undefined;
    const formContext = (0, react_hook_form_1.useForm)({
        mode: 'onChange',
        resolver: finalResolver,
        ...rest,
    });
    const { handleSubmit, getValues, reset, watch, formState } = formContext;
    const { isValid } = formState;
    const hasJustBeenModifiedByUser = React.useRef(false);
    // Reapply filterValues when they change externally
    (0, react_1.useEffect)(() => {
        // Unless users has just modified the form themselves in which case we want to avoid overriding it with
        // a previous value which was applied with a delay (debounce in List)
        if (hasJustBeenModifiedByUser.current) {
            hasJustBeenModifiedByUser.current = false;
            return;
        }
        const newValues = (0, exports.getFilterFormValues)(getValues(), filterValues);
        const previousValues = getValues();
        if (!(0, isEqual_js_1.default)(newValues, previousValues)) {
            reset(newValues);
        }
        // The reference to the filterValues object is not updated when it changes,
        // so we must stringify it to compare it by value.
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [JSON.stringify(filterValues), getValues, reset]);
    const onSubmit = (values) => {
        // Do not call setFilters if the form is invalid
        if (!isValid) {
            return;
        }
        setFilters(mergeObjNotArray(filterValues, values));
    };
    const debouncedOnSubmit = (0, util_1.useDebouncedEvent)(onSubmit, debounce || 0);
    // Submit the form on values change
    (0, react_1.useEffect)(() => {
        const { unsubscribe } = watch((values, { name }) => {
            // Check that the name is present to avoid setting filters when
            // watch was triggered by a reset
            if (name) {
                if ((0, get_js_1.default)(values, name) === '') {
                    const newValues = (0, cloneDeep_js_1.default)(values);
                    (0, set_js_1.default)(newValues, name, '');
                    hasJustBeenModifiedByUser.current = true;
                    debouncedOnSubmit(newValues);
                }
                else {
                    hasJustBeenModifiedByUser.current = true;
                    debouncedOnSubmit(values);
                }
            }
        });
        return () => unsubscribe();
    }, [watch, debouncedOnSubmit]);
    const sourceContext = React.useMemo(() => ({
        getSource: (source) => source,
        getLabel: (source) => `resources.${resource}.fields.${source}`,
    }), [resource]);
    return (React.createElement(react_hook_form_1.FormProvider, { ...formContext },
        React.createElement(FormGroupsProvider_1.FormGroupsProvider, null,
            React.createElement(core_1.SourceContextProvider, { value: sourceContext },
                React.createElement(Component, { onSubmit: handleSubmit(onSubmit) }, children)))));
};
exports.FilterLiveForm = FilterLiveForm;
const HTMLForm = (props) => (React.createElement("form", { ...props }));
// Lodash merge customizer to merge objects but not arrays
const mergeCustomizer = (objValue, srcValue) => {
    if (Array.isArray(srcValue)) {
        return srcValue;
    }
};
const mergeObjNotArray = (a, b) => (0, mergeWith_js_1.default)((0, cloneDeep_js_1.default)(a), b, mergeCustomizer);
/**
 * Because we are using controlled inputs with react-hook-form, we must provide a default value
 * for each input when resetting the form. (see https://react-hook-form.com/docs/useform/reset).
 * To ensure we don't provide undefined which will result to the current input value being reapplied
 * and due to the dynamic nature of the filter form, we rebuild the filter form values from its current
 * values and make sure to pass at least an empty string for each input.
 */
const getFilterFormValues = (formValues, filterValues) => {
    return Object.keys(formValues).reduce((acc, key) => {
        acc[key] = getInputValue(formValues, key, filterValues);
        return acc;
    }, {});
};
exports.getFilterFormValues = getFilterFormValues;
const getInputValue = (formValues, key, filterValues) => {
    if (formValues[key] === undefined || formValues[key] === null) {
        return (0, get_js_1.default)(filterValues, key, '');
    }
    if (Array.isArray(formValues[key])) {
        return (0, get_js_1.default)(filterValues, key, '');
    }
    if (formValues[key] instanceof Date) {
        return (0, get_js_1.default)(filterValues, key, '');
    }
    if (typeof formValues[key] === 'object') {
        const inputValues = Object.keys(formValues[key]).reduce((acc, innerKey) => {
            const nestedInputValue = getInputValue(formValues[key], innerKey, (filterValues || {})[key] ?? {});
            acc[innerKey] = nestedInputValue;
            return acc;
        }, {});
        if (!Object.keys(inputValues).length)
            return '';
        return inputValues;
    }
    return (0, get_js_1.default)(filterValues, key, '');
};
