// ra-ui-materialui/dist/form/SimpleFormConfigurable.js
import * as React from 'react';
import { useResourceContext, usePreference, useStore, useTranslate, } from 'ra-core';
import { Configurable } from "../preferences/index.js";
import { SimpleForm } from "./SimpleForm.js";
import { SimpleFormEditor } from "./SimpleFormEditor.js";
export const SimpleFormConfigurable = ({ preferenceKey, omit, ...props }) => {
    const translate = useTranslate();
    const resource = useResourceContext(props);
    const finalPreferenceKey = preferenceKey || `${resource}.simpleForm`;
    const [availableInputs, setAvailableInputs] = useStore(`preferences.${finalPreferenceKey}.availableInputs`, EMPTY_ARRAY);
    const [_, setOmit] = useStore(`preferences.${finalPreferenceKey}.omit`, omit);
    React.useEffect(() => {
        // first render, or the preference have been cleared
        const inputs = React.Children.map(props.children, (child, index) => React.isValidElement(child)
            ? {
                index: String(index),
                source: child.props.source,
                label: child.props.source || child.props.label
                    ? child.props.label
                    : translate('ra.configurable.SimpleForm.unlabeled', {
                        input: index,
                        _: `Unlabeled input #%{input}`,
                    }),
            }
            : null)?.filter(column => column != null) ?? EMPTY_ARRAY;
        if (inputs.length !== availableInputs.length) {
            setAvailableInputs(inputs);
            setOmit(omit || EMPTY_ARRAY);
        }
    }, [availableInputs]); // eslint-disable-line react-hooks/exhaustive-deps
    return (React.createElement(Configurable, { editor: React.createElement(SimpleFormEditor, null), preferenceKey: finalPreferenceKey, sx: {
            display: 'block',
            '&.RaConfigurable-editMode': {
                margin: '2px',
            },
        } },
        React.createElement(SimpleFormWithPreferences, { ...props })));
};
const EMPTY_ARRAY = [];
/**
 * This SimpleForm filters its children depending on preferences
 */
const SimpleFormWithPreferences = ({ children, ...props }) => {
    const [availableInputs] = usePreference('availableInputs', []);
    const [omit] = usePreference('omit', []);
    const [inputs] = usePreference('inputs', availableInputs
        .filter(input => !omit?.includes(input.source))
        .map(input => input.index));
    const childrenArray = React.Children.toArray(children);
    return (React.createElement(SimpleForm, { ...props }, inputs === undefined
        ? children
        : inputs.map(index => childrenArray[index])));
};
