// ra-ui-materialui/dist/form/SimpleFormConfigurable.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleFormConfigurable = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const preferences_1 = require("../preferences/index.cjs");
const SimpleForm_1 = require("./SimpleForm.cjs");
const SimpleFormEditor_1 = require("./SimpleFormEditor.cjs");
const SimpleFormConfigurable = ({ preferenceKey, omit, ...props }) => {
    const translate = (0, ra_core_1.useTranslate)();
    const resource = (0, ra_core_1.useResourceContext)(props);
    const finalPreferenceKey = preferenceKey || `${resource}.simpleForm`;
    const [availableInputs, setAvailableInputs] = (0, ra_core_1.useStore)(`preferences.${finalPreferenceKey}.availableInputs`, EMPTY_ARRAY);
    const [_, setOmit] = (0, ra_core_1.useStore)(`preferences.${finalPreferenceKey}.omit`, omit);
    React.useEffect(() => {
        // first render, or the preference have been cleared
        const inputs = React.Children.map(props.children, (child, index) => React.isValidElement(child)
            ? {
                index: String(index),
                source: child.props.source,
                label: child.props.source || child.props.label
                    ? child.props.label
                    : translate('ra.configurable.SimpleForm.unlabeled', {
                        input: index,
                        _: `Unlabeled input #%{input}`,
                    }),
            }
            : null)?.filter(column => column != null) ?? EMPTY_ARRAY;
        if (inputs.length !== availableInputs.length) {
            setAvailableInputs(inputs);
            setOmit(omit || EMPTY_ARRAY);
        }
    }, [availableInputs]); // eslint-disable-line react-hooks/exhaustive-deps
    return (React.createElement(preferences_1.Configurable, { editor: React.createElement(SimpleFormEditor_1.SimpleFormEditor, null), preferenceKey: finalPreferenceKey, sx: {
            display: 'block',
            '&.RaConfigurable-editMode': {
                margin: '2px',
            },
        } },
        React.createElement(SimpleFormWithPreferences, { ...props })));
};
exports.SimpleFormConfigurable = SimpleFormConfigurable;
const EMPTY_ARRAY = [];
/**
 * This SimpleForm filters its children depending on preferences
 */
const SimpleFormWithPreferences = ({ children, ...props }) => {
    const [availableInputs] = (0, ra_core_1.usePreference)('availableInputs', []);
    const [omit] = (0, ra_core_1.usePreference)('omit', []);
    const [inputs] = (0, ra_core_1.usePreference)('inputs', availableInputs
        .filter(input => !omit?.includes(input.source))
        .map(input => input.index));
    const childrenArray = React.Children.toArray(children);
    return (React.createElement(SimpleForm_1.SimpleForm, { ...props }, inputs === undefined
        ? children
        : inputs.map(index => childrenArray[index])));
};
