// ra-ui-materialui/dist/form/SimpleForm.js
import * as React from 'react';
import { Form } from 'ra-core';
import { Stack, CardContent, } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Toolbar } from "./Toolbar.js";
/**
 * Form with a one column layout, one input per line.
 *
 * Pass input components as children.
 *
 * @example
 *
 * import * as React from "react";
 * import { Create, Edit, SimpleForm, TextInput, DateInput, ReferenceManyField, Datagrid, TextField, DateField, EditButton } from 'react-admin';
 * import RichTextInput from 'ra-input-rich-text';
 *
 * export const PostCreate = () => (
 *     <Create>
 *         <SimpleForm>
 *             <TextInput source="title" />
 *             <TextInput source="teaser" options={{ multiline: true }} />
 *             <RichTextInput source="body" />
 *             <DateInput label="Publication date" source="published_at" defaultValue={new Date()} />
 *         </SimpleForm>
 *     </Create>
 * );
 *
 * @typedef {Object} Props the props you can use (other props are injected by Create or Edit)
 * @prop {ReactNode[]} children Input elements
 * @prop {Object} defaultValues
 * @prop {Function} validate
 * @prop {ReactNode} toolbar The element displayed at the bottom of the form, containing the SaveButton
 *
 * @param {Props} props
 */
export const SimpleForm = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, component: Component = DefaultComponent, sx, toolbar = DefaultToolbar, ...rest } = props;
    return (React.createElement(Form, { ...rest },
        React.createElement(Component, { className: className, sx: sx },
            React.createElement(Stack, { ...sanitizeRestProps(props), sx: {
                    alignItems: 'flex-start',
                } }, children)),
        toolbar));
};
const PREFIX = 'RaSimpleForm';
const DefaultComponent = styled(CardContent, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [theme.breakpoints.down('sm')]: {
        paddingBottom: '5em',
    },
}));
const DefaultToolbar = React.createElement(Toolbar, null);
const sanitizeRestProps = ({ children, className, component, criteriaMode, defaultValues, delayError, onSubmit, record, resource, reValidateMode, sx, toolbar, validate, resetOptions, resolver, sanitizeEmptyValues, shouldFocusError, shouldUnregister, shouldUseNativeValidation, warnWhenUnsavedChanges, ...props }) => props;
