// ra-ui-materialui/dist/form/TabbedFormView.js
import * as React from 'react';
import { Children, cloneElement, isValidElement, useState, } from 'react';
import clsx from 'clsx';
import { CardContent, Divider } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import { useResourceContext, useSplatPathBase, useRouterProvider, useLocation, useMatchPath, } from 'ra-core';
import { Toolbar } from "./Toolbar.js";
import { TabbedFormTabs, getTabbedFormTabFullPath } from "./TabbedFormTabs.js";
export const TabbedFormView = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, component: Component = DefaultComponent, formRootPathname, syncWithLocation = true, tabs = DefaultTabs, toolbar = DefaultToolbar, ...rest } = props;
    const { Route, Routes } = useRouterProvider();
    const location = useLocation();
    const matchPath = useMatchPath();
    const resource = useResourceContext(props);
    const [tabValue, setTabValue] = useState(0);
    const splatPathBase = useSplatPathBase();
    const handleTabChange = (event, value) => {
        if (!syncWithLocation) {
            setTabValue(value);
        }
        if (tabs.props.onChange) {
            tabs.props.onChange(event, value);
        }
    };
    const renderTabHeaders = () => cloneElement(tabs, {
        onChange: handleTabChange,
        syncWithLocation,
        url: formRootPathname,
        value: tabValue,
    }, children);
    return (React.createElement(Root, { className: clsx('tabbed-form', className), ...sanitizeRestProps(rest) },
        syncWithLocation ? (React.createElement(Routes, null,
            React.createElement(Route, { path: "/*", element: renderTabHeaders() }))) : (renderTabHeaders()),
        React.createElement(Divider, null),
        React.createElement(Component, null, Children.map(children, (tab, index) => {
            if (!tab) {
                return null;
            }
            const tabPath = getTabbedFormTabFullPath(tab, index);
            const hidden = syncWithLocation
                ? !matchPath(`${splatPathBase}/${tabPath}`, location.pathname)
                : tabValue !== index;
            return isValidElement(tab)
                ? React.cloneElement(tab, {
                    intent: 'content',
                    resource,
                    hidden,
                    value: syncWithLocation ? tabPath : index,
                })
                : null;
        })),
        toolbar !== false ? toolbar : null));
};
const DefaultTabs = React.createElement(TabbedFormTabs, null);
const DefaultComponent = ({ children }) => (React.createElement(CardContent, null, children));
const DefaultToolbar = React.createElement(Toolbar, null);
const PREFIX = 'RaTabbedForm';
export const TabbedFormClasses = {
    errorTabButton: `${PREFIX}-errorTabButton`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .MuiTab-root.${TabbedFormClasses.errorTabButton}`]: {
        color: (theme.vars || theme).palette.error.main,
    },
}));
const sanitizeRestProps = ({ record, resource, ...rest }) => rest;
