// ra-ui-materialui/dist/form/TabbedFormView.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TabbedFormClasses = exports.TabbedFormView = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const Toolbar_1 = require("./Toolbar.cjs");
const TabbedFormTabs_1 = require("./TabbedFormTabs.cjs");
const TabbedFormView = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, component: Component = DefaultComponent, formRootPathname, syncWithLocation = true, tabs = DefaultTabs, toolbar = DefaultToolbar, ...rest } = props;
    const { Route, Routes } = (0, ra_core_1.useRouterProvider)();
    const location = (0, ra_core_1.useLocation)();
    const matchPath = (0, ra_core_1.useMatchPath)();
    const resource = (0, ra_core_1.useResourceContext)(props);
    const [tabValue, setTabValue] = (0, react_1.useState)(0);
    const splatPathBase = (0, ra_core_1.useSplatPathBase)();
    const handleTabChange = (event, value) => {
        if (!syncWithLocation) {
            setTabValue(value);
        }
        if (tabs.props.onChange) {
            tabs.props.onChange(event, value);
        }
    };
    const renderTabHeaders = () => (0, react_1.cloneElement)(tabs, {
        onChange: handleTabChange,
        syncWithLocation,
        url: formRootPathname,
        value: tabValue,
    }, children);
    return (React.createElement(Root, { className: (0, clsx_1.default)('tabbed-form', className), ...sanitizeRestProps(rest) },
        syncWithLocation ? (React.createElement(Routes, null,
            React.createElement(Route, { path: "/*", element: renderTabHeaders() }))) : (renderTabHeaders()),
        React.createElement(material_1.Divider, null),
        React.createElement(Component, null, react_1.Children.map(children, (tab, index) => {
            if (!tab) {
                return null;
            }
            const tabPath = (0, TabbedFormTabs_1.getTabbedFormTabFullPath)(tab, index);
            const hidden = syncWithLocation
                ? !matchPath(`${splatPathBase}/${tabPath}`, location.pathname)
                : tabValue !== index;
            return (0, react_1.isValidElement)(tab)
                ? React.cloneElement(tab, {
                    intent: 'content',
                    resource,
                    hidden,
                    value: syncWithLocation ? tabPath : index,
                })
                : null;
        })),
        toolbar !== false ? toolbar : null));
};
exports.TabbedFormView = TabbedFormView;
const DefaultTabs = React.createElement(TabbedFormTabs_1.TabbedFormTabs, null);
const DefaultComponent = ({ children }) => (React.createElement(material_1.CardContent, null, children));
const DefaultToolbar = React.createElement(Toolbar_1.Toolbar, null);
const PREFIX = 'RaTabbedForm';
exports.TabbedFormClasses = {
    errorTabButton: `${PREFIX}-errorTabButton`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .MuiTab-root.${exports.TabbedFormClasses.errorTabButton}`]: {
        color: (theme.vars || theme).palette.error.main,
    },
}));
const sanitizeRestProps = ({ record, resource, ...rest }) => rest;
