// ra-ui-materialui/dist/field/BooleanField.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import DoneIcon from '@mui/icons-material/Done';
import ClearIcon from '@mui/icons-material/Clear';
import { Tooltip, Typography, } from '@mui/material';
import { useTranslate, useFieldValue, genericMemo } from 'ra-core';
import { sanitizeFieldRestProps } from "./sanitizeFieldRestProps.js";
const BooleanFieldImpl = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, emptyText, valueLabelTrue, valueLabelFalse, TrueIcon = DoneIcon, FalseIcon = ClearIcon, looseValue = false, ...rest } = props;
    const translate = useTranslate();
    const value = useFieldValue(props);
    const isTruthyValue = value === true || (looseValue && value);
    let ariaLabel = value ? valueLabelTrue : valueLabelFalse;
    if (!ariaLabel) {
        ariaLabel = isTruthyValue ? 'ra.boolean.true' : 'ra.boolean.false';
    }
    if (looseValue || value === false || value === true) {
        return (React.createElement(StyledTypography, { component: "span", variant: "body2", className: className, ...sanitizeFieldRestProps(rest) },
            React.createElement(Tooltip, { title: translate(ariaLabel, { _: ariaLabel }) }, isTruthyValue ? (TrueIcon ? (React.createElement(TrueIcon, { "data-testid": "true", fontSize: "small", className: classes.trueIcon })) : (React.createElement(React.Fragment, null))) : FalseIcon ? (React.createElement(FalseIcon, { "data-testid": "false", fontSize: "small", className: classes.falseIcon })) : (React.createElement(React.Fragment, null)))));
    }
    return (React.createElement(Typography, { component: "span", variant: "body2", className: className, ...sanitizeFieldRestProps(rest) }, emptyText && translate(emptyText, { _: emptyText })));
};
BooleanFieldImpl.displayName = 'BooleanFieldImpl';
export const BooleanField = genericMemo(BooleanFieldImpl);
const PREFIX = 'RaBooleanField';
const classes = {
    trueIcon: `${PREFIX}-trueIcon`,
    falseIcon: `${PREFIX}-falseIcon`,
};
const StyledTypography = styled(Typography, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'inline-flex',
    verticalAlign: 'middle',
    lineHeight: 0,
    [`& .${classes.trueIcon}`]: {},
    [`& .${classes.falseIcon}`]: {},
});
