// ra-ui-materialui/dist/field/BooleanField.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BooleanField = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const Done_1 = __importDefault(require("@mui/icons-material/Done"));
const Clear_1 = __importDefault(require("@mui/icons-material/Clear"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const sanitizeFieldRestProps_1 = require("./sanitizeFieldRestProps.cjs");
const BooleanFieldImpl = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, emptyText, valueLabelTrue, valueLabelFalse, TrueIcon = Done_1.default, FalseIcon = Clear_1.default, looseValue = false, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const value = (0, ra_core_1.useFieldValue)(props);
    const isTruthyValue = value === true || (looseValue && value);
    let ariaLabel = value ? valueLabelTrue : valueLabelFalse;
    if (!ariaLabel) {
        ariaLabel = isTruthyValue ? 'ra.boolean.true' : 'ra.boolean.false';
    }
    if (looseValue || value === false || value === true) {
        return (React.createElement(StyledTypography, { component: "span", variant: "body2", className: className, ...(0, sanitizeFieldRestProps_1.sanitizeFieldRestProps)(rest) },
            React.createElement(material_1.Tooltip, { title: translate(ariaLabel, { _: ariaLabel }) }, isTruthyValue ? (TrueIcon ? (React.createElement(TrueIcon, { "data-testid": "true", fontSize: "small", className: classes.trueIcon })) : (React.createElement(React.Fragment, null))) : FalseIcon ? (React.createElement(FalseIcon, { "data-testid": "false", fontSize: "small", className: classes.falseIcon })) : (React.createElement(React.Fragment, null)))));
    }
    return (React.createElement(material_1.Typography, { component: "span", variant: "body2", className: className, ...(0, sanitizeFieldRestProps_1.sanitizeFieldRestProps)(rest) }, emptyText && translate(emptyText, { _: emptyText })));
};
BooleanFieldImpl.displayName = 'BooleanFieldImpl';
exports.BooleanField = (0, ra_core_1.genericMemo)(BooleanFieldImpl);
const PREFIX = 'RaBooleanField';
const classes = {
    trueIcon: `${PREFIX}-trueIcon`,
    falseIcon: `${PREFIX}-falseIcon`,
};
const StyledTypography = (0, styles_1.styled)(material_1.Typography, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'inline-flex',
    verticalAlign: 'middle',
    lineHeight: 0,
    [`& .${classes.trueIcon}`]: {},
    [`& .${classes.falseIcon}`]: {},
});
