// ra-ui-materialui/dist/field/ImageField.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Box, Typography } from '@mui/material';
import get from 'lodash/get.js';
import { useFieldValue, useTranslate, } from 'ra-core';
import { sanitizeFieldRestProps } from "./sanitizeFieldRestProps.js";
export const ImageField = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, emptyText, src, title, ...rest } = props;
    const sourceValue = useFieldValue(props);
    const titleValue = useFieldValue({
        ...props,
        // @ts-ignore We ignore here because title might be a custom label or undefined instead of a field name
        source: title,
    })?.toString() ?? title;
    const translate = useTranslate();
    if (!sourceValue) {
        return emptyText ? (React.createElement(Typography, { component: "span", variant: "body2", className: className, ...sanitizeFieldRestProps(rest) }, emptyText && translate(emptyText, { _: emptyText }))) : (React.createElement(Typography, { component: "div", className: className, ...sanitizeFieldRestProps(rest) }));
    }
    if (Array.isArray(sourceValue)) {
        return (React.createElement(Root, { className: className, ...sanitizeFieldRestProps(rest) },
            React.createElement("ul", { className: ImageFieldClasses.list }, sourceValue.map((file, index) => {
                const fileTitleValue = title
                    ? get(file, title, title)
                    : title;
                const srcValue = src ? get(file, src, title) : title;
                return (React.createElement("li", { key: index },
                    React.createElement("img", { alt: fileTitleValue, title: fileTitleValue, src: srcValue, className: ImageFieldClasses.image })));
            }))));
    }
    return (React.createElement(Root, { className: className, ...sanitizeFieldRestProps(rest) },
        React.createElement("img", { title: titleValue, alt: titleValue, src: sourceValue?.toString(), className: ImageFieldClasses.image })));
};
// What? TypeScript loses the displayName if we don't set it explicitly
ImageField.displayName = 'ImageField';
const PREFIX = 'RaImageField';
export const ImageFieldClasses = {
    list: `${PREFIX}-list`,
    image: `${PREFIX}-image`,
};
const Root = styled(Box, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${ImageFieldClasses.list}`]: {
        display: 'flex',
        listStyleType: 'none',
    },
    [`& .${ImageFieldClasses.image}`]: {
        margin: '0.25rem',
        width: 200,
        height: 100,
        objectFit: 'contain',
    },
});
