// ra-ui-materialui/dist/field/ImageField.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageFieldClasses = exports.ImageField = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const get_js_1 = __importDefault(require("lodash/get.js"));
const ra_core_1 = require("ra-core");
const sanitizeFieldRestProps_1 = require("./sanitizeFieldRestProps.cjs");
const ImageField = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, emptyText, src, title, ...rest } = props;
    const sourceValue = (0, ra_core_1.useFieldValue)(props);
    const titleValue = (0, ra_core_1.useFieldValue)({
        ...props,
        // @ts-ignore We ignore here because title might be a custom label or undefined instead of a field name
        source: title,
    })?.toString() ?? title;
    const translate = (0, ra_core_1.useTranslate)();
    if (!sourceValue) {
        return emptyText ? (React.createElement(material_1.Typography, { component: "span", variant: "body2", className: className, ...(0, sanitizeFieldRestProps_1.sanitizeFieldRestProps)(rest) }, emptyText && translate(emptyText, { _: emptyText }))) : (React.createElement(material_1.Typography, { component: "div", className: className, ...(0, sanitizeFieldRestProps_1.sanitizeFieldRestProps)(rest) }));
    }
    if (Array.isArray(sourceValue)) {
        return (React.createElement(Root, { className: className, ...(0, sanitizeFieldRestProps_1.sanitizeFieldRestProps)(rest) },
            React.createElement("ul", { className: exports.ImageFieldClasses.list }, sourceValue.map((file, index) => {
                const fileTitleValue = title
                    ? (0, get_js_1.default)(file, title, title)
                    : title;
                const srcValue = src ? (0, get_js_1.default)(file, src, title) : title;
                return (React.createElement("li", { key: index },
                    React.createElement("img", { alt: fileTitleValue, title: fileTitleValue, src: srcValue, className: exports.ImageFieldClasses.image })));
            }))));
    }
    return (React.createElement(Root, { className: className, ...(0, sanitizeFieldRestProps_1.sanitizeFieldRestProps)(rest) },
        React.createElement("img", { title: titleValue, alt: titleValue, src: sourceValue?.toString(), className: exports.ImageFieldClasses.image })));
};
exports.ImageField = ImageField;
// What? TypeScript loses the displayName if we don't set it explicitly
exports.ImageField.displayName = 'ImageField';
const PREFIX = 'RaImageField';
exports.ImageFieldClasses = {
    list: `${PREFIX}-list`,
    image: `${PREFIX}-image`,
};
const Root = (0, styles_1.styled)(material_1.Box, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .${exports.ImageFieldClasses.list}`]: {
        display: 'flex',
        listStyleType: 'none',
    },
    [`& .${exports.ImageFieldClasses.image}`]: {
        margin: '0.25rem',
        width: 200,
        height: 100,
        objectFit: 'contain',
    },
});
