// ra-ui-materialui/dist/field/RecordField.js
import * as React from 'react';
import { styled } from '@mui/material/styles';
import { Stack, Typography, useThemeProps, } from '@mui/material';
import { FieldTitle, useRecordContext, useResourceContext, useTranslate, } from 'ra-core';
import clsx from 'clsx';
import { TextField } from "./TextField.js";
const PREFIX = 'RaRecordField';
export const RecordField = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, empty, field, label, render, source, sx, TypographyProps, variant, ...rest } = props;
    const resource = useResourceContext();
    const record = useRecordContext(props);
    const translate = useTranslate();
    if (!source && !label)
        return null;
    return (React.createElement(Root, { sx: sx, className: clsx(className, {
            [RecordFieldClasses.inline]: variant === 'inline',
        }), ...rest },
        label !== '' && label !== false ? (React.createElement(Typography, { className: RecordFieldClasses.label, ...TypographyProps },
            React.createElement(FieldTitle, { label: label, source: source, resource: resource, isRequired: false }))) : null,
        children ? (React.createElement("span", { className: RecordFieldClasses.value }, children)) : render ? (record && (React.createElement(Typography, { component: "span", variant: "body2", className: RecordFieldClasses.value }, render(record) ||
            (typeof empty === 'string'
                ? translate(empty, { _: empty })
                : empty)))) : field ? (React.createElement(field, {
            source,
            emptyText: empty,
            className: RecordFieldClasses.value,
        })) : source ? (React.createElement(TextField, { source: source, emptyText: empty, resource: resource, className: RecordFieldClasses.value })) : null));
};
export const RecordFieldClasses = {
    label: `${PREFIX}-label`,
    value: `${PREFIX}-value`,
    inline: `${PREFIX}-inline`,
};
const Root = styled(Stack, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`&.${RecordFieldClasses.inline}`]: {
        flexDirection: 'row',
    },
    [`& .${RecordFieldClasses.label}`]: {
        fontSize: '0.75rem',
        marginBottom: '0.2em',
        color: (theme.vars || theme).palette.text.secondary,
    },
    [`&.${RecordFieldClasses.inline} .${RecordFieldClasses.label}`]: {
        fontSize: '0.875rem',
        display: 'block',
        minWidth: 150,
    },
    [`& .${RecordFieldClasses.value}`]: {
        flex: 1,
    },
}));
