// ra-ui-materialui/dist/field/RecordField.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecordFieldClasses = exports.RecordField = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const clsx_1 = __importDefault(require("clsx"));
const TextField_1 = require("./TextField.cjs");
const PREFIX = 'RaRecordField';
const RecordField = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, empty, field, label, render, source, sx, TypographyProps, variant, ...rest } = props;
    const resource = (0, ra_core_1.useResourceContext)();
    const record = (0, ra_core_1.useRecordContext)(props);
    const translate = (0, ra_core_1.useTranslate)();
    if (!source && !label)
        return null;
    return (React.createElement(Root, { sx: sx, className: (0, clsx_1.default)(className, {
            [exports.RecordFieldClasses.inline]: variant === 'inline',
        }), ...rest },
        label !== '' && label !== false ? (React.createElement(material_1.Typography, { className: exports.RecordFieldClasses.label, ...TypographyProps },
            React.createElement(ra_core_1.FieldTitle, { label: label, source: source, resource: resource, isRequired: false }))) : null,
        children ? (React.createElement("span", { className: exports.RecordFieldClasses.value }, children)) : render ? (record && (React.createElement(material_1.Typography, { component: "span", variant: "body2", className: exports.RecordFieldClasses.value }, render(record) ||
            (typeof empty === 'string'
                ? translate(empty, { _: empty })
                : empty)))) : field ? (React.createElement(field, {
            source,
            emptyText: empty,
            className: exports.RecordFieldClasses.value,
        })) : source ? (React.createElement(TextField_1.TextField, { source: source, emptyText: empty, resource: resource, className: exports.RecordFieldClasses.value })) : null));
};
exports.RecordField = RecordField;
exports.RecordFieldClasses = {
    label: `${PREFIX}-label`,
    value: `${PREFIX}-value`,
    inline: `${PREFIX}-inline`,
};
const Root = (0, styles_1.styled)(material_1.Stack, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`&.${exports.RecordFieldClasses.inline}`]: {
        flexDirection: 'row',
    },
    [`& .${exports.RecordFieldClasses.label}`]: {
        fontSize: '0.75rem',
        marginBottom: '0.2em',
        color: (theme.vars || theme).palette.text.secondary,
    },
    [`&.${exports.RecordFieldClasses.inline} .${exports.RecordFieldClasses.label}`]: {
        fontSize: '0.875rem',
        display: 'block',
        minWidth: 150,
    },
    [`& .${exports.RecordFieldClasses.value}`]: {
        flex: 1,
    },
}));
