// ra-ui-materialui/dist/field/ReferenceManyCount.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReferenceManyCountClasses = exports.ReferenceManyCount = void 0;
const react_1 = __importDefault(require("react"));
const ra_core_1 = require("ra-core");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const Error_1 = __importDefault(require("@mui/icons-material/Error"));
const get_js_1 = __importDefault(require("lodash/get.js"));
const sanitizeFieldRestProps_1 = require("./sanitizeFieldRestProps.cjs");
const Link_1 = require("../Link.cjs");
const Offline_1 = require("../Offline.cjs");
/**
 * Fetch and render the number of records related to the current one
 *
 * Relies on dataProvider.getManyReference() returning a total property
 *
 * @example // Display the number of comments for the current post
 * <ReferenceManyCount reference="comments" target="post_id" />
 *
 * @example // Display the number of published comments for the current post
 * <ReferenceManyCount reference="comments" target="post_id" filter={{ is_published: true }} />
 *
 * @example // Display the number of comments for the current post, with a custom Typography variant
 * <ReferenceManyCount reference="comments" target="post_id" variant="h1" />
 */
const ReferenceManyCount = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, reference, target, filter, sort, link, resource, source = 'id', offline = defaultOffline, ...rest } = props;
    const record = (0, ra_core_1.useRecordContext)(props);
    const createPath = (0, ra_core_1.useCreatePath)();
    const body = (react_1.default.createElement(ra_core_1.ReferenceManyCountBase, { ...props, loading: react_1.default.createElement(material_1.CircularProgress, { size: 14 }), error: react_1.default.createElement(Error_1.default, { color: "error", fontSize: "small", titleAccess: "error" }), offline: offline }));
    return (react_1.default.createElement(StyledTypography, { className: (0, clsx_1.default)(className, exports.ReferenceManyCountClasses.root), component: "span", variant: "body2", ...(0, sanitizeFieldRestProps_1.sanitizeFieldRestProps)(rest) }, link && record ? (react_1.default.createElement(Link_1.Link, { className: exports.ReferenceManyCountClasses.link, to: {
            pathname: createPath({
                resource: reference,
                type: 'list',
            }),
            search: `filter=${JSON.stringify({
                ...(filter || {}),
                [target]: (0, get_js_1.default)(record, source),
            })}`,
        }, onClick: e => e.stopPropagation() }, body)) : (body)));
};
exports.ReferenceManyCount = ReferenceManyCount;
// This is a hack that replaces react support for defaultProps. We currently need this for the Datagrid.
exports.ReferenceManyCount.textAlign = 'right';
const defaultOffline = react_1.default.createElement(Offline_1.Offline, { variant: "inline" });
const PREFIX = 'RaReferenceManyCount';
exports.ReferenceManyCountClasses = {
    root: `${PREFIX}-root`,
    link: `${PREFIX}-link`,
};
const StyledTypography = (0, styles_1.styled)(material_1.Typography, {
    name: PREFIX,
    overridesResolver: (props, styles) => ({
        ['&']: styles.root,
        [`& .${exports.ReferenceManyCountClasses.link}`]: styles.link,
    }),
})({});
