// ra-ui-materialui/dist/field/ReferenceManyCount.js
import React from 'react';
import { useRecordContext, useCreatePath, ReferenceManyCountBase, } from 'ra-core';
import clsx from 'clsx';
import { Typography, CircularProgress } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import ErrorIcon from '@mui/icons-material/Error';
import get from 'lodash/get.js';
import { sanitizeFieldRestProps } from "./sanitizeFieldRestProps.js";
import { Link } from "../Link.js";
import { Offline } from "../Offline.js";
/**
 * Fetch and render the number of records related to the current one
 *
 * Relies on dataProvider.getManyReference() returning a total property
 *
 * @example // Display the number of comments for the current post
 * <ReferenceManyCount reference="comments" target="post_id" />
 *
 * @example // Display the number of published comments for the current post
 * <ReferenceManyCount reference="comments" target="post_id" filter={{ is_published: true }} />
 *
 * @example // Display the number of comments for the current post, with a custom Typography variant
 * <ReferenceManyCount reference="comments" target="post_id" variant="h1" />
 */
export const ReferenceManyCount = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, reference, target, filter, sort, link, resource, source = 'id', offline = defaultOffline, ...rest } = props;
    const record = useRecordContext(props);
    const createPath = useCreatePath();
    const body = (React.createElement(ReferenceManyCountBase, { ...props, loading: React.createElement(CircularProgress, { size: 14 }), error: React.createElement(ErrorIcon, { color: "error", fontSize: "small", titleAccess: "error" }), offline: offline }));
    return (React.createElement(StyledTypography, { className: clsx(className, ReferenceManyCountClasses.root), component: "span", variant: "body2", ...sanitizeFieldRestProps(rest) }, link && record ? (React.createElement(Link, { className: ReferenceManyCountClasses.link, to: {
            pathname: createPath({
                resource: reference,
                type: 'list',
            }),
            search: `filter=${JSON.stringify({
                ...(filter || {}),
                [target]: get(record, source),
            })}`,
        }, onClick: e => e.stopPropagation() }, body)) : (body)));
};
// This is a hack that replaces react support for defaultProps. We currently need this for the Datagrid.
ReferenceManyCount.textAlign = 'right';
const defaultOffline = React.createElement(Offline, { variant: "inline" });
const PREFIX = 'RaReferenceManyCount';
export const ReferenceManyCountClasses = {
    root: `${PREFIX}-root`,
    link: `${PREFIX}-link`,
};
const StyledTypography = styled(Typography, {
    name: PREFIX,
    overridesResolver: (props, styles) => ({
        ['&']: styles.root,
        [`& .${ReferenceManyCountClasses.link}`]: styles.link,
    }),
})({});
