// ra-ui-materialui/dist/field/ReferenceField.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReferenceFieldClasses = exports.ReferenceFieldView = exports.ReferenceField = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const Error_1 = __importDefault(require("@mui/icons-material/Error"));
const ra_core_1 = require("ra-core");
const clsx_1 = __importDefault(require("clsx"));
const layout_1 = require("../layout/index.cjs");
const Link_1 = require("../Link.cjs");
const utils_1 = require("@mui/utils");
const Offline_1 = require("../Offline.cjs");
/**
 * Fetch reference record, and render its representation, or delegate rendering to child component.
 *
 * The reference prop should be the name of one of the <Resource> components
 * added as <Admin> child.
 *
 * @example // using recordRepresentation
 * <ReferenceField label="User" source="userId" reference="users" />
 *
 * @example // using a Field component to represent the record
 * <ReferenceField label="User" source="userId" reference="users">
 *     <TextField source="name" />
 * </ReferenceField>
 *
 * @example // using a render prop to render the record
 * <ReferenceField label="User" source="userId" reference="users" render={
 *     (context) => (
 *         <p>{context.referenceRecord?.name}</p>
 *     )
 * }>
 * </ReferenceField>
 *
 * @example // By default, includes a link to the <Edit> page of the related record
 * // (`/users/:userId` in the previous example).
 * // Set the `link` prop to "show" to link to the <Show> page instead.
 * <ReferenceField label="User" source="userId" reference="users" link="show" />
 *
 * @example // You can also prevent `<ReferenceField>` from adding link to children
 * // by setting `link` to false.
 * <ReferenceField label="User" source="userId" reference="users" link={false} />
 *
 * @example // Alternatively, you can also pass a custom function to `link`.
 * // It must take reference and record as arguments and return a string
 * <ReferenceField label="User" source="userId" reference="users" link={(record, reference) => "/path/to/${reference}/${record}"} />
 *
 * @default
 * In previous versions of React-Admin, the prop `linkType` was used. It is now deprecated and replaced with `link`. However
 * backward-compatibility is still kept
 */
const ReferenceField = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, render, emptyText, empty, offline = defaultOffline, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    return (React.createElement(ra_core_1.ReferenceFieldBase, { ...rest, empty: emptyText ? (React.createElement(material_1.Typography, { component: "span", variant: "body2" }, emptyText && translate(emptyText, { _: emptyText }))) : typeof empty === 'string' ? (React.createElement(material_1.Typography, { component: "span", variant: "body2" }, empty && translate(empty, { _: empty }))) : (empty ?? null), offline: offline },
        React.createElement(PureReferenceFieldView, { ...rest, render: render }, children)));
};
exports.ReferenceField = ReferenceField;
const defaultOffline = React.createElement(Offline_1.Offline, { variant: "inline" });
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
const ReferenceFieldView = (props) => {
    const { children, render, className, emptyText, reference, sx, ...rest } = (0, styles_1.useThemeProps)({
        props: props,
        name: PREFIX,
    });
    const referenceFieldContext = (0, ra_core_1.useReferenceFieldContext)();
    const { error, link, isLoading, referenceRecord } = referenceFieldContext;
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(reference);
    if (error && referenceRecord == null) {
        return (React.createElement("div", null,
            React.createElement(Error_1.default, { role: "presentation", color: "error", fontSize: "small" }),
            React.createElement("span", { style: utils_1.visuallyHidden }, typeof error === 'string' ? error : error?.message)));
    }
    // We explicitly check isLoading here as the record may not have an id for the reference,
    // in which case, the query will not be enabled and isPending will be true
    // isLoading checks that we are actually loading the reference record
    if (isLoading) {
        return React.createElement(layout_1.LinearProgress, null);
    }
    const child = (render ? render(referenceFieldContext) : children) || (React.createElement(material_1.Typography, { component: "span", variant: "body2" }, getRecordRepresentation(referenceRecord)));
    if (link) {
        return (React.createElement(Root, { className: (0, clsx_1.default)(exports.ReferenceFieldClasses.root, className), sx: sx, ...rest },
            React.createElement(Link_1.Link, { to: link, className: exports.ReferenceFieldClasses.link, onClick: stopPropagation, state: { _scrollToTop: true } }, child)));
    }
    return (React.createElement(Root, { className: (0, clsx_1.default)(exports.ReferenceFieldClasses.root, className), sx: sx, ...rest }, child));
};
exports.ReferenceFieldView = ReferenceFieldView;
const PureReferenceFieldView = (0, ra_core_1.genericMemo)(exports.ReferenceFieldView);
const PREFIX = 'RaReferenceField';
exports.ReferenceFieldClasses = {
    root: `${PREFIX}-root`,
    link: `${PREFIX}-link`,
};
const Root = (0, styles_1.styled)('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => ({
        ['&']: styles.root,
        [`& .${exports.ReferenceFieldClasses.link}`]: styles.link,
    }),
})(({ theme }) => ({
    lineHeight: 'initial',
    [`& .${exports.ReferenceFieldClasses.link}`]: {
        '& > *': {
            color: (theme.vars || theme).palette.primary.main,
        },
    },
}));
