// ra-ui-materialui/dist/field/ReferenceArrayField.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReferenceArrayFieldClasses = exports.ReferenceArrayFieldView = exports.ReferenceArrayField = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const styles_1 = require("@mui/material/styles");
const layout_1 = require("../layout/index.cjs");
const SingleFieldList_1 = require("../list/SingleFieldList.cjs");
const Offline_1 = require("../Offline.cjs");
/**
 * A container component that fetches records from another resource specified
 * by an array of *ids* in current record.
 *
 * You must define the fields to be passed to the iterator component as children.
 *
 * @example Display all the products of the current order as datagrid
 * // order = {
 * //   id: 123,
 * //   product_ids: [456, 457, 458],
 * // }
 * <ReferenceArrayField label="Products" reference="products" source="product_ids">
 *     <Datagrid>
 *         <TextField source="id" />
 *         <TextField source="description" />
 *         <NumberField source="price" options={{ style: 'currency', currency: 'USD' }} />
 *         <EditButton />
 *     </Datagrid>
 * </ReferenceArrayField>
 *
 * @example Display all the categories of the current product as a list of chips
 * // product = {
 * //   id: 456,
 * //   category_ids: [11, 22, 33],
 * // }
 * <ReferenceArrayField label="Categories" reference="categories" source="category_ids">
 *     <SingleFieldList>
 *         <ChipField source="name" />
 *     </SingleFieldList>
 * </ReferenceArrayField>
 *
 * By default, restricts the displayed values to 1000. You can extend this limit
 * by setting the `perPage` prop.
 *
 * @example
 * <ReferenceArrayField perPage={10} reference="categories" source="category_ids">
 *    ...
 * </ReferenceArrayField>
 *
 * By default, the field displays the results in the order in which they are referenced
 * (i.e. in the order of the list of ids). You can change this order
 * by setting the `sort` prop (an object with `field` and `order` properties).
 *
 * @example
 * <ReferenceArrayField sort={{ field: 'name', order: 'ASC' }} reference="categories" source="category_ids">
 *    ...
 * </ReferenceArrayField>
 *
 * Also, you can filter the results to display only a subset of values. Use the
 * `filter` prop for that.
 *
 * @example
 * <ReferenceArrayField filter={{ is_published: true }} reference="categories" source="category_ids">
 *    ...
 * </ReferenceArrayField>
 */
const ReferenceArrayField = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { pagination, children, className, sx, render, offline = defaultOffline, ...controllerProps } = props;
    return (React.createElement(ra_core_1.ReferenceArrayFieldBase, { ...controllerProps },
        React.createElement(PureReferenceArrayFieldView, { className: className, offline: offline, pagination: pagination, render: render, sx: sx }, children)));
};
exports.ReferenceArrayField = ReferenceArrayField;
const defaultOffline = React.createElement(Offline_1.Offline, { variant: "inline" });
const ReferenceArrayFieldView = (props) => {
    const { children, render, offline, pagination, className, sx } = props;
    const listContext = (0, ra_core_1.useListContext)();
    const { isPaused, isPending, isPlaceholderData, total } = listContext;
    const shouldRenderPending = isPending && !isPaused;
    const shouldRenderOffline = isPaused &&
        (isPending || isPlaceholderData) &&
        offline !== undefined &&
        offline !== false;
    return (React.createElement(Root, { className: className, sx: sx }, shouldRenderPending ? (React.createElement(layout_1.LinearProgress, { className: exports.ReferenceArrayFieldClasses.progress })) : shouldRenderOffline ? (offline) : (React.createElement("span", null,
        (render ? render(listContext) : children) || (React.createElement(SingleFieldList_1.SingleFieldList, null)),
        pagination && total !== undefined ? pagination : null))));
};
exports.ReferenceArrayFieldView = ReferenceArrayFieldView;
const PREFIX = 'RaReferenceArrayField';
exports.ReferenceArrayFieldClasses = {
    progress: `${PREFIX}-progress`,
};
const Root = (0, styles_1.styled)('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'block',
    [`& .${exports.ReferenceArrayFieldClasses.progress}`]: {
        marginTop: theme.spacing(2),
    },
}));
const PureReferenceArrayFieldView = (0, ra_core_1.genericMemo)(exports.ReferenceArrayFieldView);
