// ra-ui-materialui/dist/field/ReferenceField.js
import * as React from 'react';
import { Typography } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import ErrorIcon from '@mui/icons-material/Error';
import { useGetRecordRepresentation, useTranslate, ReferenceFieldBase, useReferenceFieldContext, genericMemo, } from 'ra-core';
import clsx from 'clsx';
import { LinearProgress } from "../layout/index.js";
import { Link } from "../Link.js";
import { visuallyHidden } from '@mui/utils';
import { Offline } from "../Offline.js";
/**
 * Fetch reference record, and render its representation, or delegate rendering to child component.
 *
 * The reference prop should be the name of one of the <Resource> components
 * added as <Admin> child.
 *
 * @example // using recordRepresentation
 * <ReferenceField label="User" source="userId" reference="users" />
 *
 * @example // using a Field component to represent the record
 * <ReferenceField label="User" source="userId" reference="users">
 *     <TextField source="name" />
 * </ReferenceField>
 *
 * @example // using a render prop to render the record
 * <ReferenceField label="User" source="userId" reference="users" render={
 *     (context) => (
 *         <p>{context.referenceRecord?.name}</p>
 *     )
 * }>
 * </ReferenceField>
 *
 * @example // By default, includes a link to the <Edit> page of the related record
 * // (`/users/:userId` in the previous example).
 * // Set the `link` prop to "show" to link to the <Show> page instead.
 * <ReferenceField label="User" source="userId" reference="users" link="show" />
 *
 * @example // You can also prevent `<ReferenceField>` from adding link to children
 * // by setting `link` to false.
 * <ReferenceField label="User" source="userId" reference="users" link={false} />
 *
 * @example // Alternatively, you can also pass a custom function to `link`.
 * // It must take reference and record as arguments and return a string
 * <ReferenceField label="User" source="userId" reference="users" link={(record, reference) => "/path/to/${reference}/${record}"} />
 *
 * @default
 * In previous versions of React-Admin, the prop `linkType` was used. It is now deprecated and replaced with `link`. However
 * backward-compatibility is still kept
 */
export const ReferenceField = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { children, render, emptyText, empty, offline = defaultOffline, ...rest } = props;
    const translate = useTranslate();
    return (React.createElement(ReferenceFieldBase, { ...rest, empty: emptyText ? (React.createElement(Typography, { component: "span", variant: "body2" }, emptyText && translate(emptyText, { _: emptyText }))) : typeof empty === 'string' ? (React.createElement(Typography, { component: "span", variant: "body2" }, empty && translate(empty, { _: empty }))) : (empty ?? null), offline: offline },
        React.createElement(PureReferenceFieldView, { ...rest, render: render }, children)));
};
const defaultOffline = React.createElement(Offline, { variant: "inline" });
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
export const ReferenceFieldView = (props) => {
    const { children, render, className, emptyText, reference, sx, ...rest } = useThemeProps({
        props: props,
        name: PREFIX,
    });
    const referenceFieldContext = useReferenceFieldContext();
    const { error, link, isLoading, referenceRecord } = referenceFieldContext;
    const getRecordRepresentation = useGetRecordRepresentation(reference);
    if (error && referenceRecord == null) {
        return (React.createElement("div", null,
            React.createElement(ErrorIcon, { role: "presentation", color: "error", fontSize: "small" }),
            React.createElement("span", { style: visuallyHidden }, typeof error === 'string' ? error : error?.message)));
    }
    // We explicitly check isLoading here as the record may not have an id for the reference,
    // in which case, the query will not be enabled and isPending will be true
    // isLoading checks that we are actually loading the reference record
    if (isLoading) {
        return React.createElement(LinearProgress, null);
    }
    const child = (render ? render(referenceFieldContext) : children) || (React.createElement(Typography, { component: "span", variant: "body2" }, getRecordRepresentation(referenceRecord)));
    if (link) {
        return (React.createElement(Root, { className: clsx(ReferenceFieldClasses.root, className), sx: sx, ...rest },
            React.createElement(Link, { to: link, className: ReferenceFieldClasses.link, onClick: stopPropagation, state: { _scrollToTop: true } }, child)));
    }
    return (React.createElement(Root, { className: clsx(ReferenceFieldClasses.root, className), sx: sx, ...rest }, child));
};
const PureReferenceFieldView = genericMemo(ReferenceFieldView);
const PREFIX = 'RaReferenceField';
export const ReferenceFieldClasses = {
    root: `${PREFIX}-root`,
    link: `${PREFIX}-link`,
};
const Root = styled('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => ({
        ['&']: styles.root,
        [`& .${ReferenceFieldClasses.link}`]: styles.link,
    }),
})(({ theme }) => ({
    lineHeight: 'initial',
    [`& .${ReferenceFieldClasses.link}`]: {
        '& > *': {
            color: (theme.vars || theme).palette.primary.main,
        },
    },
}));
