// ra-ui-materialui/dist/field/TranslatableFieldsTabs.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { AppBar, Tabs } from '@mui/material';
import { useTranslatableContext } from 'ra-core';
import { TranslatableFieldsTab } from "./TranslatableFieldsTab.js";
/**
 * Default locale selector for the TranslatableFields component. Generates a tab for each specified locale.
 * @see TranslatableFields
 */
export const TranslatableFieldsTabs = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { groupKey, TabsProps: tabsProps, className } = props;
    const { locales, selectLocale, selectedLocale } = useTranslatableContext();
    const handleChange = (event, newLocale) => {
        selectLocale(newLocale);
    };
    return (React.createElement(StyledAppBar, { color: "default", position: "static", className: className },
        React.createElement(Tabs, { value: selectedLocale, onChange: handleChange, indicatorColor: "primary", textColor: "primary", ...tabsProps }, locales.map(locale => (React.createElement(TranslatableFieldsTab, { key: locale, value: locale, locale: locale, groupKey: groupKey }))))));
};
const PREFIX = 'RaTranslatableFieldsTabs';
const StyledAppBar = styled(AppBar, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    boxShadow: 'none',
    borderRadius: 0,
    borderTopLeftRadius: theme.shape.borderRadius,
    borderTopRightRadius: theme.shape.borderRadius,
    border: `1px solid ${(theme.vars || theme).palette.divider}`,
}));
