// ra-ui-materialui/dist/field/TranslatableFieldsTabContent.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranslatableFieldsTabContent = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const Labeled_1 = require("../Labeled.cjs");
/**
 * Default container for a group of translatable fields inside a TranslatableFields components.
 * @see TranslatableFields
 */
const TranslatableFieldsTabContent = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, groupKey = '', locale, record, resource: resourceProp, className, ...other } = props;
    const { selectedLocale, getRecordForLocale } = (0, ra_core_1.useTranslatableContext)();
    const addLabel = react_1.Children.count(children) > 1;
    const parentSourceContext = (0, ra_core_1.useOptionalSourceContext)();
    const resource = (0, ra_core_1.useResourceContext)(props);
    if (!resource) {
        throw new Error(`<TranslatableFieldsTabContent> was called outside of a ResourceContext and without a record prop. You must set the resource prop.`);
    }
    const sourceContext = (0, react_1.useMemo)(() => ({
        getSource: (source) => parentSourceContext
            ? parentSourceContext.getSource(`${source}.${locale}`)
            : `${source}.${locale}`,
        getLabel: (source) => parentSourceContext
            ? parentSourceContext.getLabel(source)
            : (0, ra_core_1.getResourceFieldLabelKey)(resource, source),
    }), [locale, parentSourceContext, resource]);
    // As fields rely on the RecordContext to get their values and have no knowledge of the locale,
    // we need to create a new record with the values for the current locale only
    // Given the record { title: { en: 'title_en', fr: 'title_fr' } } and the locale 'fr',
    // the record for the locale 'fr' will be { title: 'title_fr' }
    const recordForLocale = (0, react_1.useMemo)(() => getRecordForLocale(record, locale), [getRecordForLocale, record, locale]);
    return (React.createElement(Root, { role: "tabpanel", hidden: selectedLocale !== locale, id: `translatable-content-${groupKey}${locale}`, "aria-labelledby": `translatable-header-${groupKey}${locale}`, className: className, ...other },
        React.createElement(ra_core_1.RecordContextProvider, { value: recordForLocale },
            React.createElement(ra_core_1.SourceContextProvider, { value: sourceContext }, react_1.Children.map(children, field => field && (0, react_1.isValidElement)(field) ? (React.createElement("div", null, addLabel ? (React.createElement(Labeled_1.Labeled
            // Only pass the resource if it was overridden through props to avoid
            // the default inference to potentially override label set by SourceContext
            , { 
                // Only pass the resource if it was overridden through props to avoid
                // the default inference to potentially override label set by SourceContext
                resource: resourceProp, label: field.props.label, source: field.props.source }, field)) : (field))) : null)))));
};
exports.TranslatableFieldsTabContent = TranslatableFieldsTabContent;
const PREFIX = 'RaTranslatableFieldsTabContent';
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flexGrow: 1,
    padding: theme.spacing(2),
    borderRadius: 0,
    borderBottomLeftRadius: theme.shape.borderRadius,
    borderBottomRightRadius: theme.shape.borderRadius,
    border: `1px solid ${(theme.vars || theme).palette.divider}`,
    borderTop: 0,
}));
