// ra-ui-materialui/dist/field/TranslatableFields.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranslatableFields = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const TranslatableFieldsTabs_1 = require("./TranslatableFieldsTabs.cjs");
const TranslatableFieldsTabContent_1 = require("./TranslatableFieldsTabContent.cjs");
/**
 * Provides a way to show multiple languages for any field passed as children.
 * It expects the translatable values to have the following structure:
 * {
 *     name: {
 *         en: 'The english value',
 *         fr: 'The french value',
 *         tlh: 'The klingon value',
 *     },
 *     description: {
 *         en: 'The english value',
 *         fr: 'The french value',
 *         tlh: 'The klingon value',
 *     }
 * }
 *
 * @example <caption>Basic usage</caption>
 * <TranslatableFields locales={['en', 'fr']}>
 *     <TextField source={getSource('title')} />
 *     <TextField source={getSource('description')} />
 * </TranslatableFields>
 *
 * @example <caption>With a custom language selector</caption>
 * <TranslatableFields
 *     selector={<MyLanguageSelector />}
 *     locales={['en', 'fr']}
 * >
 *     <TextField source={getSource('title')} />
 * <TranslatableFields>
>
 *
 * const MyLanguageSelector = () => {
 *     const {
 *         locales,
 *         selectedLocale,
 *         selectLocale,
 *     } = useTranslatableContext();
 *
 *     return (
 *         <select onChange={selectLocale}>
 *             {locales.map((locale) => (
 *                 <option selected={locale.locale === selectedLocale}>
 *                     {locale.name}
 *                 </option>
 *             ))}
 *        </select>
 *     );
 * }
 *
 * @param props The component props
 * @param {string} props.defaultLocale The locale selected by default. Default to 'en'.
 * @param {string[]} props.locales An array of the possible locales in the form. For example [{ 'en', 'fr' }].
 * @param {ReactElement} props.selector The element responsible for selecting a locale. Defaults to Material UI tabs.
 */
const TranslatableFields = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { defaultLocale, locales, groupKey = '', selector = React.createElement(TranslatableFieldsTabs_1.TranslatableFieldsTabs, { groupKey: groupKey }), children, className, resource: resourceProp, } = props;
    const record = (0, ra_core_1.useRecordContext)(props);
    if (!record) {
        throw new Error(`<TranslatableFields> was called outside of a RecordContext and without a record prop. You must set the record prop.`);
    }
    const resource = (0, ra_core_1.useResourceContext)(props);
    if (!resource) {
        throw new Error(`<TranslatableFields> was called outside of a ResourceContext and without a record prop. You must set the resource prop.`);
    }
    const context = (0, ra_core_1.useTranslatable)({ defaultLocale, locales });
    return (React.createElement(Root, { className: className },
        React.createElement(ra_core_1.TranslatableContextProvider, { value: context },
            selector,
            locales.map(locale => (React.createElement(TranslatableFieldsTabContent_1.TranslatableFieldsTabContent, { key: locale, locale: locale, record: record, resource: resourceProp, groupKey: groupKey }, children))))));
};
exports.TranslatableFields = TranslatableFields;
const PREFIX = 'RaTranslatableFields';
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    flexGrow: 1,
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(0.5),
}));
