// ra-core/dist/util/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMutationMode = exports.useWhyDidYouUpdate = exports.warning = exports.removeKey = exports.removeEmpty = exports.Ready = exports.FieldTitle = exports.escapePath = void 0;
const escapePath_1 = __importDefault(require("./escapePath.cjs"));
exports.escapePath = escapePath_1.default;
const FieldTitle_1 = __importDefault(require("./FieldTitle.cjs"));
exports.FieldTitle = FieldTitle_1.default;
const removeEmpty_1 = __importDefault(require("./removeEmpty.cjs"));
exports.removeEmpty = removeEmpty_1.default;
const removeKey_1 = __importDefault(require("./removeKey.cjs"));
exports.removeKey = removeKey_1.default;
const Ready_1 = __importDefault(require("./Ready.cjs"));
exports.Ready = Ready_1.default;
const warning_1 = __importDefault(require("./warning.cjs"));
exports.warning = warning_1.default;
const useWhyDidYouUpdate_1 = __importDefault(require("./useWhyDidYouUpdate.cjs"));
exports.useWhyDidYouUpdate = useWhyDidYouUpdate_1.default;
const getMutationMode_1 = require("./getMutationMode.cjs");
Object.defineProperty(exports, "getMutationMode", { enumerable: true, get: function () { return getMutationMode_1.getMutationMode; } });
__exportStar(require("./getFieldLabelTranslationArgs.cjs"), exports);
__exportStar(require("./mergeRefs.cjs"), exports);
__exportStar(require("./useEvent.cjs"), exports);
__exportStar(require("./useDebouncedEvent.cjs"), exports);
__exportStar(require("./useFieldValue.cjs"), exports);
__exportStar(require("./asyncDebounce.cjs"), exports);
__exportStar(require("./hooks.cjs"), exports);
__exportStar(require("./shallowEqual.cjs"), exports);
__exportStar(require("./useCheckForApplicationUpdate.cjs"), exports);
__exportStar(require("./genericMemo.cjs"), exports);
