// ra-ui-materialui/dist/button/BulkExportButton.js
import * as React from 'react';
import { useCallback } from 'react';
import DownloadIcon from '@mui/icons-material/GetApp';
import { useBulkExport, useResourceContext, } from 'ra-core';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Button } from "./Button.js";
/**
 * Export the selected rows
 *
 * To be used inside the <Datagrid bulkActionButtons> prop.
 *
 * @example // basic usage
 * import { BulkDeleteButton, BulkExportButton, List, Datagrid } from 'react-admin';
 *
 * const PostBulkActionButtons = () => (
 *     <>
 *         <BulkExportButton />
 *         <BulkDeleteButton />
 *     </>
 * );
 *
 * export const PostList = () => (
 *     <List>
 *        <Datagrid bulkActionButtons={<PostBulkActionButtons />}>
 *          ...
 *       </Datagrid>
 *     </List>
 * );
 */
export const BulkExportButton = React.forwardRef(function BulkExportButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { onClick, label = 'ra.action.export', icon = defaultIcon, exporter: customExporter, meta, ...rest } = props;
    const resource = useResourceContext(props);
    const bulkExport = useBulkExport({
        exporter: customExporter,
        resource,
        meta,
    });
    const handleClick = useCallback(event => {
        bulkExport();
        if (typeof onClick === 'function') {
            onClick(event);
        }
    }, [bulkExport, onClick]);
    return (React.createElement(StyledButton, { ref: ref, onClick: handleClick, label: label, ...sanitizeRestProps(rest) }, icon));
});
const defaultIcon = React.createElement(DownloadIcon, null);
const sanitizeRestProps = ({ resource, ...rest }) => rest;
const PREFIX = 'RaBulkExportButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
