// ra-ui-materialui/dist/button/BulkDeleteWithConfirmButton.js
import * as React from 'react';
import { Fragment, useState } from 'react';
import ActionDelete from '@mui/icons-material/Delete';
import { styled, useThemeProps, } from '@mui/material/styles';
import { useResourceContext, useTranslate, useBulkDeleteController, useListContext, } from 'ra-core';
import { Confirm } from "../layout/index.js";
import { Button } from "./Button.js";
import { humanize, inflect } from 'inflection';
export const BulkDeleteWithConfirmButton = React.forwardRef(function BulkDeleteWithConfirmButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { confirmTitle = 'ra.message.bulk_delete_title', confirmContent = 'ra.message.bulk_delete_content', confirmColor = 'primary', icon = defaultIcon, label = 'ra.action.delete', mutationMode = 'pessimistic', mutationOptions, onClick, ...rest } = props;
    const { selectedIds } = useListContext();
    const { handleDelete, isPending } = useBulkDeleteController({
        mutationMode,
        ...rest,
        mutationOptions: {
            ...mutationOptions,
            onSettled(...args) {
                // In pessimistic mode, we wait for the mutation to be completed (either successfully or with an error) before closing
                if (mutationMode === 'pessimistic') {
                    setOpen(false);
                }
                mutationOptions?.onSettled?.(...args);
            },
        },
    });
    const [isOpen, setOpen] = useState(false);
    const resource = useResourceContext(props);
    const translate = useTranslate();
    const handleClick = (e) => {
        e.stopPropagation();
        setOpen(true);
    };
    const handleDialogClose = (e) => {
        e.stopPropagation();
        setOpen(false);
    };
    const handleConfirm = (e) => {
        e.stopPropagation();
        // We close the dialog immediately here for optimistic/undoable modes instead of in onSuccess/onError
        // to avoid reimplementing the default side effects
        if (mutationMode !== 'pessimistic') {
            setOpen(false);
        }
        handleDelete();
        if (typeof onClick === 'function') {
            onClick(e);
        }
    };
    return (React.createElement(Fragment, null,
        React.createElement(StyledButton, { ref: ref, onClick: handleClick, label: label, color: "error", ...sanitizeRestProps(rest) }, icon),
        React.createElement(Confirm, { isOpen: isOpen, loading: isPending, title: confirmTitle, content: confirmContent, confirmColor: confirmColor, titleTranslateOptions: {
                smart_count: selectedIds.length,
                name: translate(`resources.${resource}.forcedCaseName`, {
                    smart_count: selectedIds.length,
                    _: humanize(translate(`resources.${resource}.name`, {
                        smart_count: selectedIds.length,
                        _: resource
                            ? inflect(resource, selectedIds.length)
                            : undefined,
                    }), true),
                }),
            }, contentTranslateOptions: {
                smart_count: selectedIds.length,
                name: translate(`resources.${resource}.forcedCaseName`, {
                    smart_count: selectedIds.length,
                    _: humanize(translate(`resources.${resource}.name`, {
                        smart_count: selectedIds.length,
                        _: resource
                            ? inflect(resource, selectedIds.length)
                            : undefined,
                    }), true),
                }),
            }, onConfirm: handleConfirm, onClose: handleDialogClose })));
});
const sanitizeRestProps = ({ classes, label, resource, successMessage, mutationOptions, ...rest }) => rest;
const PREFIX = 'RaBulkDeleteWithConfirmButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(() => ({}));
const defaultIcon = React.createElement(ActionDelete, null);
