// ra-ui-materialui/dist/button/CloneButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloneButton = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const Queue_1 = __importDefault(require("@mui/icons-material/Queue"));
const query_string_1 = require("query-string");
const ra_core_1 = require("ra-core");
const styles_1 = require("@mui/material/styles");
const Button_1 = require("./Button.cjs");
exports.CloneButton = React.forwardRef(function CloneButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { label = 'ra.action.clone', scrollToTop = true, icon = defaultIcon, ...rest } = props;
    const resource = (0, ra_core_1.useResourceContext)(props);
    const record = (0, ra_core_1.useRecordContext)(props);
    const createPath = (0, ra_core_1.useCreatePath)();
    const pathname = createPath({ resource, type: 'create' });
    return (React.createElement(StyledButton, { component: ra_core_1.LinkBase, ref: ref, to: record
            ? {
                pathname,
                search: (0, query_string_1.stringify)({
                    source: JSON.stringify(omitId(record)),
                }),
            }
            : pathname, state: { _scrollToTop: scrollToTop }, label: label, onClick: stopPropagation, ...sanitizeRestProps(rest) }, icon));
});
const defaultIcon = React.createElement(Queue_1.default, null);
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
const omitId = ({ id, ...rest }) => rest;
const sanitizeRestProps = ({ resource, record, ...rest }) => rest;
exports.default = (0, react_1.memo)(exports.CloneButton);
const PREFIX = 'RaCloneButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
