// ra-ui-materialui/dist/button/CloneButton.js
import * as React from 'react';
import { memo } from 'react';
import Queue from '@mui/icons-material/Queue';
import { stringify } from 'query-string';
import { useResourceContext, useRecordContext, useCreatePath, LinkBase, } from 'ra-core';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Button } from "./Button.js";
export const CloneButton = React.forwardRef(function CloneButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { label = 'ra.action.clone', scrollToTop = true, icon = defaultIcon, ...rest } = props;
    const resource = useResourceContext(props);
    const record = useRecordContext(props);
    const createPath = useCreatePath();
    const pathname = createPath({ resource, type: 'create' });
    return (React.createElement(StyledButton, { component: LinkBase, ref: ref, to: record
            ? {
                pathname,
                search: stringify({
                    source: JSON.stringify(omitId(record)),
                }),
            }
            : pathname, state: { _scrollToTop: scrollToTop }, label: label, onClick: stopPropagation, ...sanitizeRestProps(rest) }, icon));
});
const defaultIcon = React.createElement(Queue, null);
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
const omitId = ({ id, ...rest }) => rest;
const sanitizeRestProps = ({ resource, record, ...rest }) => rest;
export default memo(CloneButton);
const PREFIX = 'RaCloneButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
