// ra-ui-materialui/dist/button/CreateButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateButtonClasses = void 0;
const React = __importStar(require("react"));
const Add_1 = __importDefault(require("@mui/icons-material/Add"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const isEqual_js_1 = __importDefault(require("lodash/isEqual.js"));
const merge_js_1 = __importDefault(require("lodash/merge.js"));
const ra_core_1 = require("ra-core");
const Button_1 = require("./Button.cjs");
/**
 * Opens the Create view of a given resource
 *
 * Renders as a regular button on desktop, and a Floating Action Button
 * on mobile.
 *
 * @example // basic usage
 * import { CreateButton } from 'react-admin';
 *
 * const CommentCreateButton = () => (
 *     <CreateButton label="Create comment" />
 * );
 */
const CreateButton = React.forwardRef(function CreateButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, icon = defaultIcon, label: labelProp, resource: resourceProp, scrollToTop = true, variant, to: locationDescriptor, state: initialState = {}, ...rest } = props;
    const resource = (0, ra_core_1.useResourceContext)(props);
    if (!resource) {
        throw new Error('<CreateButton> components should be used inside a <Resource> component or provided the resource prop.');
    }
    const { canAccess, isPending } = (0, ra_core_1.useCanAccess)({
        action: 'create',
        resource,
    });
    const createPath = (0, ra_core_1.useCreatePath)();
    const getResourceLabel = (0, ra_core_1.useGetResourceLabel)();
    const label = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.action.create`,
        baseI18nKey: 'ra.action.create',
        options: {
            name: getResourceLabel(resource, 1),
        },
        userText: labelProp,
    });
    const isSmall = (0, material_1.useMediaQuery)((theme) => theme.breakpoints.down('md'));
    const state = (0, merge_js_1.default)({}, scrollStates.get(String(scrollToTop)), initialState);
    // Duplicated behavior of Button component (legacy use) which will be removed in v5.
    const linkParams = getLinkParams(locationDescriptor);
    if (!canAccess || isPending) {
        return null;
    }
    return isSmall ? (React.createElement(StyledFab, { component: ra_core_1.LinkBase, ref: ref, to: createPath({ resource, type: 'create' }), state: state, 
        // @ts-ignore FabProps ships its own runtime palette `FabPropsColorOverrides` provoking an overlap error with `ButtonProps`
        color: "primary", className: (0, clsx_1.default)(exports.CreateButtonClasses.floating, className), "aria-label": typeof label === 'string' ? label : undefined, ...rest, ...linkParams }, icon)) : (React.createElement(StyledButton, { component: ra_core_1.LinkBase, ref: ref, to: createPath({ resource, type: 'create' }), state: state, className: (0, clsx_1.default)(exports.CreateButtonClasses.root, className), 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, variant: variant, ...rest, ...linkParams }, icon));
});
// avoids using useMemo to get a constant value for the link state
const scrollStates = new Map([
    ['true', { _scrollToTop: true }],
    ['false', {}],
]);
const defaultIcon = React.createElement(Add_1.default, null);
const PREFIX = 'RaCreateButton';
exports.CreateButtonClasses = {
    root: `${PREFIX}-root`,
    floating: `${PREFIX}-floating`,
};
const StyledFab = (0, styles_1.styled)(material_1.Fab, {
    name: PREFIX,
    overridesResolver: (_props, styles) => styles.root,
})(({ theme }) => ({
    [`&.${exports.CreateButtonClasses.floating}`]: {
        color: (theme.vars || theme).palette.primary.contrastText,
        margin: 0,
        top: 'auto',
        right: 20,
        bottom: 60,
        left: 'auto',
        position: 'fixed',
        zIndex: 1000,
    },
}));
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (_props, styles) => styles.root,
})({});
exports.default = React.memo(CreateButton, (prevProps, nextProps) => {
    return (prevProps.resource === nextProps.resource &&
        prevProps.label === nextProps.label &&
        prevProps.translate === nextProps.translate &&
        prevProps.disabled === nextProps.disabled &&
        (0, isEqual_js_1.default)(prevProps.to, nextProps.to) &&
        (0, isEqual_js_1.default)(prevProps.state, nextProps.state));
});
const getLinkParams = (locationDescriptor) => {
    // eslint-disable-next-line
    if (locationDescriptor == undefined) {
        return undefined;
    }
    if (typeof locationDescriptor === 'string') {
        return { to: locationDescriptor };
    }
    const { redirect, replace, state, ...to } = locationDescriptor;
    return {
        to,
        redirect,
        replace,
        state,
    };
};
