// ra-ui-materialui/dist/button/DeleteWithConfirmButton.js
import React, { Fragment, isValidElement } from 'react';
import ActionDelete from '@mui/icons-material/Delete';
import { styled, useThemeProps, } from '@mui/material/styles';
import clsx from 'clsx';
import { useRecordContext, useResourceContext, useTranslate, useGetRecordRepresentation, useResourceTranslation, useDeleteController, useNotify, useUnselect, useRedirect, } from 'ra-core';
import { humanize, singularize } from 'inflection';
import { Confirm } from "../layout/index.js";
import { Button } from "./Button.js";
export const DeleteWithConfirmButton = React.forwardRef(function DeleteWithConfirmButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, confirmTitle: confirmTitleProp, confirmContent: confirmContentProp, confirmColor = 'primary', icon = defaultIcon, label: labelProp, mutationMode = 'pessimistic', onClick, redirect: redirectTo = 'list', translateOptions = {}, titleTranslateOptions = translateOptions, contentTranslateOptions = translateOptions, mutationOptions, color = 'error', successMessage, ...rest } = props;
    const translate = useTranslate();
    const record = useRecordContext(props);
    const resource = useResourceContext(props);
    const notify = useNotify();
    const unselect = useUnselect(resource);
    const redirect = useRedirect();
    const [open, setOpen] = React.useState(false);
    if (!resource) {
        throw new Error('<DeleteWithConfirmButton> components should be used inside a <Resource> component or provided with a resource prop. (The <Resource> component set the resource prop for all its children).');
    }
    const { onSuccess, onError, ...otherMutationOptions } = mutationOptions || {};
    const { isPending, handleDelete } = useDeleteController({
        record,
        redirect: redirectTo,
        mutationMode,
        mutationOptions: {
            ...otherMutationOptions,
            onSuccess: (...args) => {
                setOpen(false);
                if (onSuccess) {
                    onSuccess(...args);
                }
                else {
                    notify(successMessage ??
                        `resources.${resource}.notifications.deleted`, {
                        type: 'info',
                        messageArgs: {
                            smart_count: 1,
                            _: translate('ra.notification.deleted', {
                                smart_count: 1,
                            }),
                        },
                        undoable: mutationMode === 'undoable',
                    });
                    record && unselect([record.id]);
                    redirect(redirectTo, resource);
                }
            },
            onError: (...args) => {
                setOpen(false);
                if (onError) {
                    onError(...args);
                }
                else {
                    const [error] = args;
                    notify(typeof error === 'string'
                        ? error
                        : error?.message ||
                            'ra.notification.http_error', {
                        type: 'error',
                        messageArgs: {
                            _: typeof error === 'string'
                                ? error
                                : error?.message
                                    ? error.message
                                    : undefined,
                        },
                    });
                }
            },
        },
        resource,
        successMessage,
    });
    const handleDialogOpen = event => {
        event.stopPropagation();
        setOpen(true);
    };
    const handleDialogClose = () => {
        setOpen(false);
    };
    const getRecordRepresentation = useGetRecordRepresentation(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: humanize(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? singularize(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if (isValidElement(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.action.delete`,
        baseI18nKey: 'ra.action.delete',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    const confirmTitle = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.message.delete_title`,
        baseI18nKey: 'ra.message.delete_title',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            ...titleTranslateOptions,
        },
        userText: confirmTitleProp,
    });
    const confirmContent = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.message.delete_content`,
        baseI18nKey: 'ra.message.delete_content',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            ...contentTranslateOptions,
        },
        userText: confirmContentProp,
    });
    return (React.createElement(Fragment, null,
        React.createElement(StyledButton, { ref: ref, onClick: handleDialogOpen, 
            // avoid double translation
            label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, className: clsx('ra-delete-button', className), key: "button", color: color, ...rest }, icon),
        React.createElement(Confirm, { isOpen: open, loading: isPending, title: React.createElement(React.Fragment, null, confirmTitle), content: React.createElement(React.Fragment, null, confirmContent), confirmColor: confirmColor, onConfirm: handleDelete, onClose: handleDialogClose })));
});
const defaultIcon = React.createElement(ActionDelete, null);
const PREFIX = 'RaDeleteWithConfirmButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
