// ra-ui-materialui/dist/button/DeleteWithConfirmButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteWithConfirmButton = void 0;
const react_1 = __importStar(require("react"));
const Delete_1 = __importDefault(require("@mui/icons-material/Delete"));
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const inflection_1 = require("inflection");
const layout_1 = require("../layout/index.cjs");
const Button_1 = require("./Button.cjs");
exports.DeleteWithConfirmButton = react_1.default.forwardRef(function DeleteWithConfirmButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, confirmTitle: confirmTitleProp, confirmContent: confirmContentProp, confirmColor = 'primary', icon = defaultIcon, label: labelProp, mutationMode = 'pessimistic', onClick, redirect: redirectTo = 'list', translateOptions = {}, titleTranslateOptions = translateOptions, contentTranslateOptions = translateOptions, mutationOptions, color = 'error', successMessage, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const record = (0, ra_core_1.useRecordContext)(props);
    const resource = (0, ra_core_1.useResourceContext)(props);
    const notify = (0, ra_core_1.useNotify)();
    const unselect = (0, ra_core_1.useUnselect)(resource);
    const redirect = (0, ra_core_1.useRedirect)();
    const [open, setOpen] = react_1.default.useState(false);
    if (!resource) {
        throw new Error('<DeleteWithConfirmButton> components should be used inside a <Resource> component or provided with a resource prop. (The <Resource> component set the resource prop for all its children).');
    }
    const { onSuccess, onError, ...otherMutationOptions } = mutationOptions || {};
    const { isPending, handleDelete } = (0, ra_core_1.useDeleteController)({
        record,
        redirect: redirectTo,
        mutationMode,
        mutationOptions: {
            ...otherMutationOptions,
            onSuccess: (...args) => {
                setOpen(false);
                if (onSuccess) {
                    onSuccess(...args);
                }
                else {
                    notify(successMessage ??
                        `resources.${resource}.notifications.deleted`, {
                        type: 'info',
                        messageArgs: {
                            smart_count: 1,
                            _: translate('ra.notification.deleted', {
                                smart_count: 1,
                            }),
                        },
                        undoable: mutationMode === 'undoable',
                    });
                    record && unselect([record.id]);
                    redirect(redirectTo, resource);
                }
            },
            onError: (...args) => {
                setOpen(false);
                if (onError) {
                    onError(...args);
                }
                else {
                    const [error] = args;
                    notify(typeof error === 'string'
                        ? error
                        : error?.message ||
                            'ra.notification.http_error', {
                        type: 'error',
                        messageArgs: {
                            _: typeof error === 'string'
                                ? error
                                : error?.message
                                    ? error.message
                                    : undefined,
                        },
                    });
                }
            },
        },
        resource,
        successMessage,
    });
    const handleDialogOpen = event => {
        event.stopPropagation();
        setOpen(true);
    };
    const handleDialogClose = () => {
        setOpen(false);
    };
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: (0, inflection_1.humanize)(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? (0, inflection_1.singularize)(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if ((0, react_1.isValidElement)(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.action.delete`,
        baseI18nKey: 'ra.action.delete',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    const confirmTitle = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.message.delete_title`,
        baseI18nKey: 'ra.message.delete_title',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            ...titleTranslateOptions,
        },
        userText: confirmTitleProp,
    });
    const confirmContent = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.message.delete_content`,
        baseI18nKey: 'ra.message.delete_content',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            ...contentTranslateOptions,
        },
        userText: confirmContentProp,
    });
    return (react_1.default.createElement(react_1.Fragment, null,
        react_1.default.createElement(StyledButton, { ref: ref, onClick: handleDialogOpen, 
            // avoid double translation
            label: react_1.default.createElement(react_1.default.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, className: (0, clsx_1.default)('ra-delete-button', className), key: "button", color: color, ...rest }, icon),
        react_1.default.createElement(layout_1.Confirm, { isOpen: open, loading: isPending, title: react_1.default.createElement(react_1.default.Fragment, null, confirmTitle), content: react_1.default.createElement(react_1.default.Fragment, null, confirmContent), confirmColor: confirmColor, onConfirm: handleDelete, onClose: handleDialogClose })));
});
const defaultIcon = react_1.default.createElement(Delete_1.default, null);
const PREFIX = 'RaDeleteWithConfirmButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
