// ra-ui-materialui/dist/button/DeleteWithUndoButton.js
import * as React from 'react';
import ActionDelete from '@mui/icons-material/Delete';
import clsx from 'clsx';
import { useDeleteController, useRecordContext, useResourceContext, useTranslate, useGetRecordRepresentation, useResourceTranslation, } from 'ra-core';
import { humanize, singularize } from 'inflection';
import { Button } from "./Button.js";
import { styled, useThemeProps, } from '@mui/material/styles';
export const DeleteWithUndoButton = React.forwardRef(function DeleteWithUndoButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { label: labelProp, className, icon = defaultIcon, onClick, redirect = 'list', mutationOptions, color = 'error', successMessage, ...rest } = props;
    const record = useRecordContext(props);
    const resource = useResourceContext(props);
    if (!resource) {
        throw new Error('<DeleteWithUndoButton> components should be used inside a <Resource> component or provided with a resource prop. (The <Resource> component set the resource prop for all its children).');
    }
    const { isPending, handleDelete } = useDeleteController({
        record,
        resource,
        redirect,
        mutationMode: 'undoable',
        mutationOptions,
        successMessage,
    });
    const handleClick = event => {
        event.stopPropagation();
        handleDelete();
        if (onClick) {
            onClick(event);
        }
    };
    const translate = useTranslate();
    const getRecordRepresentation = useGetRecordRepresentation(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: humanize(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? singularize(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if (React.isValidElement(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.action.delete`,
        baseI18nKey: 'ra.action.delete',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    return (React.createElement(StyledButton, { ref: ref, onClick: handleClick, disabled: isPending, 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, className: clsx('ra-delete-button', className), key: "button", color: color, ...rest }, icon));
});
const defaultIcon = React.createElement(ActionDelete, null);
const PREFIX = 'RaDeleteWithUndoButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
