// ra-ui-materialui/dist/button/DeleteWithUndoButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteWithUndoButton = void 0;
const React = __importStar(require("react"));
const Delete_1 = __importDefault(require("@mui/icons-material/Delete"));
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const inflection_1 = require("inflection");
const Button_1 = require("./Button.cjs");
const styles_1 = require("@mui/material/styles");
exports.DeleteWithUndoButton = React.forwardRef(function DeleteWithUndoButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { label: labelProp, className, icon = defaultIcon, onClick, redirect = 'list', mutationOptions, color = 'error', successMessage, ...rest } = props;
    const record = (0, ra_core_1.useRecordContext)(props);
    const resource = (0, ra_core_1.useResourceContext)(props);
    if (!resource) {
        throw new Error('<DeleteWithUndoButton> components should be used inside a <Resource> component or provided with a resource prop. (The <Resource> component set the resource prop for all its children).');
    }
    const { isPending, handleDelete } = (0, ra_core_1.useDeleteController)({
        record,
        resource,
        redirect,
        mutationMode: 'undoable',
        mutationOptions,
        successMessage,
    });
    const handleClick = event => {
        event.stopPropagation();
        handleDelete();
        if (onClick) {
            onClick(event);
        }
    };
    const translate = (0, ra_core_1.useTranslate)();
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: (0, inflection_1.humanize)(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? (0, inflection_1.singularize)(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if (React.isValidElement(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.action.delete`,
        baseI18nKey: 'ra.action.delete',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    return (React.createElement(StyledButton, { ref: ref, onClick: handleClick, disabled: isPending, 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, className: (0, clsx_1.default)('ra-delete-button', className), key: "button", color: color, ...rest }, icon));
});
const defaultIcon = React.createElement(Delete_1.default, null);
const PREFIX = 'RaDeleteWithUndoButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
