// ra-ui-materialui/dist/button/EditButton.js
import * as React from 'react';
import clsx from 'clsx';
import { styled, useThemeProps, } from '@mui/material/styles';
import ContentCreate from '@mui/icons-material/Create';
import { useResourceContext, useRecordContext, useCreatePath, useCanAccess, useGetResourceLabel, useGetRecordRepresentation, useResourceTranslation, LinkBase, } from 'ra-core';
import { Button } from "./Button.js";
/**
 * Opens the Edit view for the current record.
 *
 * Reads the record and resource from the context.
 *
 * @example // basic usage
 * import { EditButton } from 'react-admin';
 *
 * const CommentEditButton = () => (
 *     <EditButton label="Edit comment" />
 * );
 */
export const EditButton = React.forwardRef(function EditButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { icon = defaultIcon, label: labelProp, scrollToTop = true, className, ...rest } = props;
    const resource = useResourceContext(props);
    if (!resource) {
        throw new Error('<EditButton> components should be used inside a <Resource> component or provided with a resource prop. (The <Resource> component set the resource prop for all its children).');
    }
    const record = useRecordContext(props);
    const createPath = useCreatePath();
    const { canAccess, isPending } = useCanAccess({
        action: 'edit',
        resource,
        record,
    });
    const getResourceLabel = useGetResourceLabel();
    const getRecordRepresentation = useGetRecordRepresentation();
    const recordRepresentationValue = getRecordRepresentation(record);
    const recordRepresentation = typeof recordRepresentationValue === 'string'
        ? recordRepresentationValue
        : recordRepresentationValue?.toString();
    const label = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.action.edit`,
        baseI18nKey: 'ra.action.edit',
        options: {
            name: getResourceLabel(resource, 1),
            recordRepresentation,
        },
        userText: labelProp,
    });
    if (!record || !canAccess || isPending)
        return null;
    return (React.createElement(StyledButton, { component: LinkBase, ref: ref, to: createPath({ type: 'edit', resource, id: record.id }), state: scrollStates[String(scrollToTop)], 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, onClick: stopPropagation, className: clsx(EditButtonClasses.root, className), ...rest }, icon));
});
// avoids using useMemo to get a constant value for the link state
const scrollStates = {
    true: { _scrollToTop: true },
    false: {},
};
const defaultIcon = React.createElement(ContentCreate, null);
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
const PREFIX = 'RaEditButton';
export const EditButtonClasses = {
    root: `${PREFIX}-root`,
};
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (_props, styles) => styles.root,
})({});
