// ra-ui-materialui/dist/button/ListButton.js
import * as React from 'react';
import ActionList from '@mui/icons-material/List';
import { useResourceContext, useCreatePath, useCanAccess, useGetResourceLabel, useResourceTranslation, LinkBase, } from 'ra-core';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Button } from "./Button.js";
/**
 * Opens the List view of a given resource
 *
 * @example // basic usage
 * import { ListButton } from 'react-admin';
 *
 * const CommentListButton = () => (
 *     <ListButton label="Comments" />
 * );
 *
 * @example // linking back to the list from the Edit view
 * import { TopToolbar, ListButton, ShowButton, Edit } from 'react-admin';
 *
 * const PostEditActions = () => (
 *     <TopToolbar>
 *         <ListButton />
 *         <ShowButton />
 *     </TopToolbar>
 * );
 *
 * export const PostEdit = (props) => (
 *     <Edit actions={<PostEditActions />} {...props}>
 *         ...
 *     </Edit>
 * );
 */
export const ListButton = React.forwardRef(function ListButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { icon = defaultIcon, label: labelProp, resource: resourceProp, scrollToTop = true, ...rest } = props;
    const resource = useResourceContext(props);
    if (!resource) {
        throw new Error('<ListButton> components should be used inside a <Resource> component or provided the resource prop.');
    }
    const { canAccess, isPending } = useCanAccess({
        action: 'list',
        resource,
    });
    const createPath = useCreatePath();
    const getResourceLabel = useGetResourceLabel();
    const label = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.action.list`,
        baseI18nKey: 'ra.action.list',
        options: {
            name: getResourceLabel(resource, 1),
        },
        userText: labelProp,
    });
    if (!canAccess || isPending) {
        return null;
    }
    return (React.createElement(StyledButton, { component: LinkBase, ref: ref, to: createPath({ type: 'list', resource }), state: scrollStates[String(scrollToTop)], 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, ...rest }, icon));
});
// avoids using useMemo to get a constant value for the link state
const scrollStates = {
    true: { _scrollToTop: true },
    false: {},
};
const defaultIcon = React.createElement(ActionList, null);
const PREFIX = 'RaListButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
