// ra-ui-materialui/dist/button/ListButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListButton = void 0;
const React = __importStar(require("react"));
const List_1 = __importDefault(require("@mui/icons-material/List"));
const ra_core_1 = require("ra-core");
const styles_1 = require("@mui/material/styles");
const Button_1 = require("./Button.cjs");
/**
 * Opens the List view of a given resource
 *
 * @example // basic usage
 * import { ListButton } from 'react-admin';
 *
 * const CommentListButton = () => (
 *     <ListButton label="Comments" />
 * );
 *
 * @example // linking back to the list from the Edit view
 * import { TopToolbar, ListButton, ShowButton, Edit } from 'react-admin';
 *
 * const PostEditActions = () => (
 *     <TopToolbar>
 *         <ListButton />
 *         <ShowButton />
 *     </TopToolbar>
 * );
 *
 * export const PostEdit = (props) => (
 *     <Edit actions={<PostEditActions />} {...props}>
 *         ...
 *     </Edit>
 * );
 */
exports.ListButton = React.forwardRef(function ListButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { icon = defaultIcon, label: labelProp, resource: resourceProp, scrollToTop = true, ...rest } = props;
    const resource = (0, ra_core_1.useResourceContext)(props);
    if (!resource) {
        throw new Error('<ListButton> components should be used inside a <Resource> component or provided the resource prop.');
    }
    const { canAccess, isPending } = (0, ra_core_1.useCanAccess)({
        action: 'list',
        resource,
    });
    const createPath = (0, ra_core_1.useCreatePath)();
    const getResourceLabel = (0, ra_core_1.useGetResourceLabel)();
    const label = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.action.list`,
        baseI18nKey: 'ra.action.list',
        options: {
            name: getResourceLabel(resource, 1),
        },
        userText: labelProp,
    });
    if (!canAccess || isPending) {
        return null;
    }
    return (React.createElement(StyledButton, { component: ra_core_1.LinkBase, ref: ref, to: createPath({ type: 'list', resource }), state: scrollStates[String(scrollToTop)], 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, ...rest }, icon));
});
// avoids using useMemo to get a constant value for the link state
const scrollStates = {
    true: { _scrollToTop: true },
    false: {},
};
const defaultIcon = React.createElement(List_1.default, null);
const PREFIX = 'RaListButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
