// ra-ui-materialui/dist/button/PrevNextButtons.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrevNextButtonClasses = exports.PrevNextButtons = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const NavigateBefore_1 = __importDefault(require("@mui/icons-material/NavigateBefore"));
const NavigateNext_1 = __importDefault(require("@mui/icons-material/NavigateNext"));
const Error_1 = __importDefault(require("@mui/icons-material/Error"));
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const LinearProgress_1 = require("../layout/LinearProgress.cjs");
/**
 * A component used to render the previous and next buttons in a Show or Edit view.
 *
 * The `<PrevNextButtons>` component renders navigation buttons linking to
 * the next and previous records of the current resource, the current index
 * and the total number of records.
 *
 * It uses usePrevNextController to fetch the list of records.
 *
 * `<PrevNextButtons>` can be used anywhere a record context is provided
 * (often inside a `<Show>` or `<Edit>` component).
 *
 * @example <caption>navigate to edit view by default</caption>
 * <ShowButton />
 *
 * @example <caption>navigate to show view</caption>
 * <PrevNextButtons linkType="show" />
 *
 * @example <caption>custom storeKey</caption>
 * <PrevNextButtons storeKey="listStoreKey" />
 *
 * @example <caption>limit the number of records to fetch</caption>
 * <PrevNextButtons limit={500} />
 *
 * @example <caption>customize filters and sort order</caption>
 * <PrevNextButtons
 *     linkType="show"
 *     sort={{
 *         field: 'first_name',
 *         order: 'DESC',
 *     }}
 *     filter={{ q: 'East a' }}
 * />
 *
 * @example <caption>customize style</caption>
 * <PrevNextButtons
 *     sx={{
 *         color: 'blue',
 *         '& .RaPrevNextButton-list': {
 *             marginBottom: '20px',
 *             color: 'red',
 *         },
 *     }}
 * />
 *
 * @example <caption>in an edit view</caption>
 * import * as React from "react";
 * import { Edit, PrevNextButtons, ShowButton, SimpleForm, TopToolbar } from 'react-admin';
 *
 * const MyTopToolbar = ({ children }) => (
 *     <TopToolbar>
 *         {children}
 *     </TopToolbar>
 * );
 *
 * export const PostEdit = () => (
 *      <Edit
 *          actions={
 *              <MyTopToolbar>
 *                  <PrevNextButtons
 *                      sort={{
 *                          field: 'first_name',
 *                          order: 'DESC',
 *                      }}
 *                      filter={{ q: 'East a' }}
 *                  />
 *                  <ShowButton />
 *              </MyTopToolbar>
 *          }
 *      >
 *          <SimpleForm>...</SimpleForm>
 *      </Edit>
 * );
 */
const PrevNextButtons = (inProps) => {
    const props = (0, material_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { sx } = props;
    const { hasPrev, hasNext, prevPath, nextPath, index, total, error, isPending, } = (0, ra_core_1.usePrevNextController)(props);
    const translate = (0, ra_core_1.useTranslate)();
    if (isPending) {
        return (React.createElement(material_1.Box, { sx: {
                minHeight: 34,
                display: 'flex',
                alignItems: 'center',
            } },
            React.createElement(LinearProgress_1.LinearProgress, null)));
    }
    if (error) {
        return (React.createElement(Error_1.default, { color: "error", fontSize: "small", titleAccess: "error", "aria-errormessage": error.message }));
    }
    if (!hasPrev && !hasNext) {
        return (React.createElement(material_1.Box, { sx: {
                minHeight: 34,
            } }));
    }
    return (React.createElement(Root, { sx: sx, role: "navigation", direction: "row", className: (0, clsx_1.default)(exports.PrevNextButtonClasses.root) },
        hasPrev && prevPath ? (React.createElement(material_1.IconButton, { component: ra_core_1.LinkBase, to: prevPath, "aria-label": translate('ra.navigation.previous'), size: "small" },
            React.createElement(NavigateBefore_1.default, null))) : (React.createElement(material_1.IconButton, { "aria-label": translate('ra.navigation.previous'), disabled: true, size: "small" },
            React.createElement(NavigateBefore_1.default, null))),
        typeof index === 'number' && (React.createElement(material_1.Typography, { variant: "body2" },
            index + 1,
            " / ",
            total)),
        hasNext && nextPath ? (React.createElement(material_1.IconButton, { component: ra_core_1.LinkBase, to: nextPath, "aria-label": translate('ra.navigation.next'), size: "small" },
            React.createElement(NavigateNext_1.default, null))) : (React.createElement(material_1.IconButton, { "aria-label": translate('ra.navigation.next'), disabled: true, size: "small" },
            React.createElement(NavigateNext_1.default, null)))));
};
exports.PrevNextButtons = PrevNextButtons;
const PREFIX = 'RaPrevNextButton';
exports.PrevNextButtonClasses = {
    root: `${PREFIX}-root`,
};
const Root = (0, material_1.styled)(material_1.Stack, {
    name: PREFIX,
    overridesResolver: (_props, styles) => styles.root,
})({
    display: 'inline-flex',
    alignItems: 'center',
    gap: '0.5em',
});
