// ra-ui-materialui/dist/button/SaveButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SaveButton = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const Save_1 = __importDefault(require("@mui/icons-material/Save"));
const react_hook_form_1 = require("react-hook-form");
const ra_core_1 = require("ra-core");
/**
 * Submit button for resource forms (Edit and Create).
 *
 * @typedef {Object} Props the props you can use (other props are injected by the <Toolbar>)
 * @prop {string} className
 * @prop {string} label Button label. Defaults to 'ra.action.save', translated.
 * @prop {boolean} disabled Disable the button.
 * @prop {string} variant Material UI variant for the button. Defaults to 'contained'.
 * @prop {ReactNode} icon
 * @prop {function} mutationOptions Object of options passed to react-query.
 * @prop {function} transform Callback to execute before calling the dataProvider. Receives the data from the form, must return that transformed data. Can be asynchronous (and return a Promise)
 * @prop {boolean} alwaysEnable Force enabling the <SaveButton>. If it's not defined, the `<SaveButton>` will be enabled using `react-hook-form`'s `isValidating` state props and form context's `saving` prop (disabled if isValidating or saving, enabled otherwise).
 *
 * @param {Props} props
 *
 * @example // with custom success side effect
 *
 * const MySaveButton = props => {
 *     const notify = useNotify();
 *     const redirect = useRedirect();
 *     const onSuccess = (response) => {
 *         notify(`Post "${response.data.title}" saved!`);
 *         redirect('/posts');
 *     };
 *     return <SaveButton {...props} mutationOptions={{ onSuccess }} />;
 * }
 */
exports.SaveButton = React.forwardRef(function SaveButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { color = 'primary', icon = defaultIcon, invalid, label = 'ra.action.save', onClick, mutationOptions, disabled: disabledProp, type = 'submit', transform, variant = 'contained', alwaysEnable = false, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const form = (0, react_hook_form_1.useFormContext)();
    const saveContext = (0, ra_core_1.useSaveContext)();
    const { isValidating, isSubmitting } = (0, react_hook_form_1.useFormState)();
    // useFormState().isDirty might differ from useFormState().dirtyFields (https://github.com/react-hook-form/react-hook-form/issues/4740)
    const isDirty = (0, ra_core_1.useFormIsDirty)();
    // Use form isDirty, isValidating and form context saving to enable or disable the save button
    // if alwaysEnable is undefined and the form wasn't prefilled
    const recordFromLocation = (0, ra_core_1.useRecordFromLocation)();
    const disabled = valueOrDefault(alwaysEnable === false || alwaysEnable === undefined
        ? undefined
        : !alwaysEnable, disabledProp ||
        (!isDirty && recordFromLocation == null) ||
        isValidating ||
        isSubmitting);
    (0, ra_core_1.warning)(type === 'submit' &&
        ((mutationOptions &&
            (mutationOptions.onSuccess || mutationOptions.onError)) ||
            transform), 'Cannot use <SaveButton mutationOptions> props on a button of type "submit". To override the default mutation options on a particular save button, set the <SaveButton type="button"> prop, or set mutationOptions in the main view component (<Create> or <Edit>).');
    const handleSubmit = (0, react_1.useCallback)(async (values) => {
        let errors;
        if (saveContext?.save) {
            errors = await saveContext.save(values, {
                ...mutationOptions,
                transform,
            });
        }
        if (errors != null) {
            (0, ra_core_1.setSubmissionErrors)(errors, form.setError);
        }
    }, [form.setError, saveContext, mutationOptions, transform]);
    const handleClick = (0, react_1.useCallback)(async (event) => {
        if (onClick) {
            onClick(event);
        }
        if (event.defaultPrevented) {
            return;
        }
        if (type === 'button') {
            // this button doesn't submit the form, so it doesn't trigger useIsFormInvalid in <FormContent>
            // therefore we need to check for errors manually
            event.stopPropagation();
            await form.handleSubmit(handleSubmit)(event);
        }
    }, [onClick, type, form, handleSubmit]);
    const displayedLabel = label && translate(label, { _: label });
    return (React.createElement(StyledButton, { ref: ref, variant: variant, type: type, color: color, "aria-label": displayedLabel, disabled: disabled, onClick: handleClick, ...rest },
        isSubmitting ? (React.createElement(material_1.CircularProgress, { sx: circularProgressStyle, size: 14, thickness: 3, color: "inherit" })) : (icon),
        displayedLabel));
});
const circularProgressStyle = {
    '&.MuiCircularProgress-root': {
        marginRight: '10px',
        marginLeft: '2px',
    },
};
const defaultIcon = React.createElement(Save_1.default, null);
const PREFIX = 'RaSaveButton';
const StyledButton = (0, styles_1.styled)(material_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    position: 'relative',
    [`& .MuiSvgIcon-root, & .MuiIcon-root, & .MuiCircularProgress-root`]: {
        marginRight: theme.spacing(1),
    },
    [`& .MuiSvgIcon-root, & .MuiIcon-root`]: {
        fontSize: 18,
    },
}));
const valueOrDefault = (value, defaultValue) => typeof value === 'undefined' ? defaultValue : value;
