// ra-ui-materialui/dist/button/ShowButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const react_1 = require("react");
const RemoveRedEye_1 = __importDefault(require("@mui/icons-material/RemoveRedEye"));
const ra_core_1 = require("ra-core");
const styles_1 = require("@mui/material/styles");
const Button_1 = require("./Button.cjs");
/**
 * Opens the Show view of a given record
 *
 * @example // basic usage
 * import { ShowButton, useRecordContext } from 'react-admin';
 *
 * const CommentShowButton = () => {
 *     const record = useRecordContext();
 *     return (
 *         <ShowButton label="Show comment" record={record} />
 *     );
 * };
 */
const ShowButton = React.forwardRef(function ShowButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { icon = defaultIcon, label: labelProp, record: recordProp, resource: resourceProp, scrollToTop = true, ...rest } = props;
    const resource = (0, ra_core_1.useResourceContext)(props);
    if (!resource) {
        throw new Error('<ShowButton> components should be used inside a <Resource> component or provided the resource prop.');
    }
    const record = (0, ra_core_1.useRecordContext)(props);
    const createPath = (0, ra_core_1.useCreatePath)();
    const { canAccess, isPending } = (0, ra_core_1.useCanAccess)({
        action: 'show',
        resource,
        record,
    });
    const getResourceLabel = (0, ra_core_1.useGetResourceLabel)();
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)();
    const recordRepresentationValue = getRecordRepresentation(record);
    const recordRepresentation = typeof recordRepresentationValue === 'string'
        ? recordRepresentationValue
        : recordRepresentationValue?.toString();
    const label = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.action.show`,
        baseI18nKey: 'ra.action.show',
        options: {
            name: getResourceLabel(resource, 1),
            recordRepresentation,
        },
        userText: labelProp,
    });
    if (!record || !canAccess || isPending)
        return null;
    return (React.createElement(StyledButton, { component: ra_core_1.LinkBase, ref: ref, to: createPath({ type: 'show', resource, id: record.id }), state: scrollStates[String(scrollToTop)], 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, onClick: stopPropagation, ...rest }, icon));
});
// avoids using useMemo to get a constant value for the link state
const scrollStates = {
    true: { _scrollToTop: true },
    false: {},
};
const defaultIcon = React.createElement(RemoveRedEye_1.default, null);
// useful to prevent click bubbling in a datagrid with rowClick
const stopPropagation = e => e.stopPropagation();
const PureShowButton = (0, react_1.memo)(ShowButton, (prevProps, nextProps) => prevProps.resource === nextProps.resource &&
    (prevProps.record && nextProps.record
        ? prevProps.record.id === nextProps.record.id
        : prevProps.record == nextProps.record) && // eslint-disable-line eqeqeq
    prevProps.label === nextProps.label &&
    prevProps.disabled === nextProps.disabled);
exports.default = PureShowButton;
const PREFIX = 'RaShowButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({});
module.exports = exports.default;
