// ra-ui-materialui/dist/button/SkipNavigationButton.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkipNavigationButton = void 0;
const react_1 = __importDefault(require("react"));
const styles_1 = require("@mui/material/styles");
const Button_1 = require("./Button.cjs");
exports.SkipNavigationButton = react_1.default.forwardRef(function SkipNavigationButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    return (react_1.default.createElement(StyledButton, { ref: ref, onClick: skipToContent, className: 'skip-nav-button', label: "ra.navigation.skip_nav", variant: "contained", ...props }));
});
const PREFIX = 'RaSkipNavigationButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    position: 'fixed',
    padding: theme.spacing(1),
    backgroundColor: (theme.vars || theme).palette.background.default,
    color: (theme.vars || theme).palette.primary.contrastText,
    transition: theme.transitions.create(['top', 'opacity'], {
        easing: theme.transitions.easing.easeIn,
        duration: theme.transitions.duration.leavingScreen,
    }),
    left: theme.spacing(2),
    top: theme.spacing(-10),
    zIndex: 5000,
    '&:hover': {
        opacity: 0.8,
        backgroundColor: (theme.vars || theme).palette.background.default,
    },
    '&:focus': {
        top: theme.spacing(2),
        transition: theme.transitions.create(['top', 'opacity'], {
            easing: theme.transitions.easing.easeOut,
            duration: theme.transitions.duration.enteringScreen,
        }),
    },
}));
const skipToContent = () => {
    if (typeof document === 'undefined')
        return;
    const element = document.getElementById('main-content');
    if (!element) {
        if (process.env.NODE_ENV !== 'production') {
            console.warn('No element with id "main-content" was found. Ensure the element that contains your main content has an id of "main-content".');
        }
        return;
    }
    element.setAttribute('tabIndex', '-1');
    element.focus();
    element.blur();
    element.removeAttribute('tabIndex');
};
