// ra-ui-materialui/dist/button/SortButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const React = __importStar(require("react"));
const react_1 = require("react");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const Sort_1 = __importDefault(require("@mui/icons-material/Sort"));
const ArrowDropDown_1 = __importDefault(require("@mui/icons-material/ArrowDropDown"));
const ra_core_1 = require("ra-core");
/**
 * A button allowing to change the sort field and order.
 *
 * To be used inside a ListContext (e.g. inside a <List> or <ReferenceManyField>)
 *
 * Expects one 'fields' prop, containing an array of field strings that shall
 * be used and displayed for sorting.
 *
 * When users clicks on the <SortButton>, they see a dropdown list with the
 * proposed sorting fields. Once they click on one of these fields, the related
 * list refreshes, re-sorted.
 *
 * @example
 *
 * import * as React from 'react';
 * import { TopToolbar, SortButton, CreateButton, ExportButton } from 'react-admin';
 *
 * const ListActions = () => (
 *     <TopToolbar>
 *         <SortButton fields={['reference', 'sales', 'stock']} />
 *         <CreateButton />
 *         <ExportButton />
 *     </TopToolbar>
 * );
 */
const SortButton = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        name: PREFIX,
        props: inProps,
    });
    const { fields, label = 'ra.sort.sort_by', icon = defaultIcon, sx, className, resource: resourceProp, } = props;
    const { resource: resourceFromContext, sort, setSort, } = (0, ra_core_1.useListSortContext)();
    const resource = resourceProp || resourceFromContext;
    const translate = (0, ra_core_1.useTranslate)();
    const translateLabel = (0, ra_core_1.useTranslateLabel)();
    const isXSmall = (0, material_1.useMediaQuery)((theme) => theme.breakpoints.down('sm'));
    const [anchorEl, setAnchorEl] = React.useState(null);
    const handleClick = (event) => {
        setAnchorEl(event.currentTarget);
    };
    const handleClose = () => {
        setAnchorEl(null);
    };
    const handleChangeSort = (event) => {
        const field = event.currentTarget.dataset.sort;
        if (!field) {
            throw new Error('<SortButton> MenuItems should have a data-sort attribute');
        }
        setSort({
            field,
            order: field === sort.field ? inverseOrder(sort.order) : 'ASC',
        });
        setAnchorEl(null);
    };
    const fieldLabel = translateLabel({
        resource,
        source: sort.field,
    });
    const buttonLabel = translate(label, {
        field: fieldLabel,
        field_lower_first: typeof fieldLabel === 'string'
            ? fieldLabel.charAt(0).toLowerCase() + fieldLabel.slice(1)
            : undefined,
        order: translate(`ra.sort.${sort.order}`),
        _: label,
    });
    return (React.createElement(Root, { sx: sx, className: (0, clsx_1.default)(className, classNames.root) },
        isXSmall ? (React.createElement(material_1.Tooltip, { title: buttonLabel },
            React.createElement(material_1.IconButton, { "aria-label": buttonLabel, color: "primary", onClick: handleClick, size: "large" }, icon))) : (React.createElement(material_1.Button, { "aria-controls": "simple-menu", "aria-haspopup": "true", color: "primary", onClick: handleClick, startIcon: icon, endIcon: React.createElement(ArrowDropDown_1.default, null), size: "small" }, buttonLabel)),
        React.createElement(material_1.Menu, { id: "simple-menu", anchorEl: anchorEl, keepMounted: true, open: Boolean(anchorEl), onClose: handleClose }, fields.map(field => (React.createElement(material_1.MenuItem, { onClick: handleChangeSort, "data-sort": field, key: field },
            translateLabel({
                resource,
                source: field,
            }),
            ' ',
            translate(`ra.sort.${sort.field === field
                ? inverseOrder(sort.order)
                : 'ASC'}`)))))));
};
const defaultIcon = React.createElement(Sort_1.default, null);
const inverseOrder = (sort) => (sort === 'ASC' ? 'DESC' : 'ASC');
const arePropsEqual = (prevProps, nextProps) => (0, ra_core_1.shallowEqual)(prevProps.fields, nextProps.fields);
const PREFIX = 'RaSortButton';
const classNames = {
    root: `${PREFIX}-root`,
};
const Root = (0, styles_1.styled)('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`.${classNames.root}`]: {},
    '& .MuiButton-sizeSmall': {
        // fix for icon misalignment on small buttons, see https://github.com/mui/material-ui/pull/30240
        lineHeight: 1.5,
    },
    '& .MuiButton-endIcon': { ml: 0 },
});
exports.default = (0, react_1.memo)(SortButton, arePropsEqual);
module.exports = exports.default;
