// ra-ui-materialui/dist/button/SortButton.js
import * as React from 'react';
import { memo } from 'react';
import clsx from 'clsx';
import { Button, Menu, MenuItem, Tooltip, IconButton, useMediaQuery, } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import SortIcon from '@mui/icons-material/Sort';
import ArrowDropDownIcon from '@mui/icons-material/ArrowDropDown';
import { useListSortContext, useTranslate, useTranslateLabel, shallowEqual, } from 'ra-core';
/**
 * A button allowing to change the sort field and order.
 *
 * To be used inside a ListContext (e.g. inside a <List> or <ReferenceManyField>)
 *
 * Expects one 'fields' prop, containing an array of field strings that shall
 * be used and displayed for sorting.
 *
 * When users clicks on the <SortButton>, they see a dropdown list with the
 * proposed sorting fields. Once they click on one of these fields, the related
 * list refreshes, re-sorted.
 *
 * @example
 *
 * import * as React from 'react';
 * import { TopToolbar, SortButton, CreateButton, ExportButton } from 'react-admin';
 *
 * const ListActions = () => (
 *     <TopToolbar>
 *         <SortButton fields={['reference', 'sales', 'stock']} />
 *         <CreateButton />
 *         <ExportButton />
 *     </TopToolbar>
 * );
 */
const SortButton = (inProps) => {
    const props = useThemeProps({
        name: PREFIX,
        props: inProps,
    });
    const { fields, label = 'ra.sort.sort_by', icon = defaultIcon, sx, className, resource: resourceProp, } = props;
    const { resource: resourceFromContext, sort, setSort, } = useListSortContext();
    const resource = resourceProp || resourceFromContext;
    const translate = useTranslate();
    const translateLabel = useTranslateLabel();
    const isXSmall = useMediaQuery((theme) => theme.breakpoints.down('sm'));
    const [anchorEl, setAnchorEl] = React.useState(null);
    const handleClick = (event) => {
        setAnchorEl(event.currentTarget);
    };
    const handleClose = () => {
        setAnchorEl(null);
    };
    const handleChangeSort = (event) => {
        const field = event.currentTarget.dataset.sort;
        if (!field) {
            throw new Error('<SortButton> MenuItems should have a data-sort attribute');
        }
        setSort({
            field,
            order: field === sort.field ? inverseOrder(sort.order) : 'ASC',
        });
        setAnchorEl(null);
    };
    const fieldLabel = translateLabel({
        resource,
        source: sort.field,
    });
    const buttonLabel = translate(label, {
        field: fieldLabel,
        field_lower_first: typeof fieldLabel === 'string'
            ? fieldLabel.charAt(0).toLowerCase() + fieldLabel.slice(1)
            : undefined,
        order: translate(`ra.sort.${sort.order}`),
        _: label,
    });
    return (React.createElement(Root, { sx: sx, className: clsx(className, classNames.root) },
        isXSmall ? (React.createElement(Tooltip, { title: buttonLabel },
            React.createElement(IconButton, { "aria-label": buttonLabel, color: "primary", onClick: handleClick, size: "large" }, icon))) : (React.createElement(Button, { "aria-controls": "simple-menu", "aria-haspopup": "true", color: "primary", onClick: handleClick, startIcon: icon, endIcon: React.createElement(ArrowDropDownIcon, null), size: "small" }, buttonLabel)),
        React.createElement(Menu, { id: "simple-menu", anchorEl: anchorEl, keepMounted: true, open: Boolean(anchorEl), onClose: handleClose }, fields.map(field => (React.createElement(MenuItem, { onClick: handleChangeSort, "data-sort": field, key: field },
            translateLabel({
                resource,
                source: field,
            }),
            ' ',
            translate(`ra.sort.${sort.field === field
                ? inverseOrder(sort.order)
                : 'ASC'}`)))))));
};
const defaultIcon = React.createElement(SortIcon, null);
const inverseOrder = (sort) => (sort === 'ASC' ? 'DESC' : 'ASC');
const arePropsEqual = (prevProps, nextProps) => shallowEqual(prevProps.fields, nextProps.fields);
const PREFIX = 'RaSortButton';
const classNames = {
    root: `${PREFIX}-root`,
};
const Root = styled('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`.${classNames.root}`]: {},
    '& .MuiButton-sizeSmall': {
        // fix for icon misalignment on small buttons, see https://github.com/mui/material-ui/pull/30240
        lineHeight: 1.5,
    },
    '& .MuiButton-endIcon': { ml: 0 },
});
export default memo(SortButton, arePropsEqual);
