// ra-ui-materialui/dist/button/UpdateWithUndoButton.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import ActionUpdate from '@mui/icons-material/Update';
import { useResourceContext, useRecordContext, useTranslate, useGetRecordRepresentation, useResourceTranslation, useUpdateController, } from 'ra-core';
import { humanize, singularize } from 'inflection';
import { Button } from "./Button.js";
export const UpdateWithUndoButton = React.forwardRef(function UpdateWithUndoButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const record = useRecordContext(props);
    const resource = useResourceContext(props);
    const { data, label: labelProp, icon = defaultIcon, onClick, ...rest } = props;
    const { handleUpdate, isPending } = useUpdateController(rest);
    const translate = useTranslate();
    const getRecordRepresentation = useGetRecordRepresentation(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: humanize(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? singularize(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if (React.isValidElement(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.action.update`,
        baseI18nKey: 'ra.action.update',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    const handleClick = e => {
        if (!record) {
            throw new Error('The UpdateWithUndoButton must be used inside a RecordContext.Provider or must be passed a record prop.');
        }
        handleUpdate(data);
        if (typeof onClick === 'function') {
            onClick(e);
        }
        e.stopPropagation();
    };
    return (React.createElement(StyledButton, { ref: ref, onClick: handleClick, 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, disabled: isPending, ...sanitizeRestProps(rest) }, icon));
});
const defaultIcon = React.createElement(ActionUpdate, null);
const sanitizeRestProps = ({ label, mutationOptions, ...rest }) => rest;
const PREFIX = 'RaUpdateWithUndoButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    color: (theme.vars || theme).palette.primary.main,
    '&:hover': {
        backgroundColor: `color-mix(in srgb, ${(theme.vars || theme).palette.primary.main}, transparent 12%)`,
        // Reset on mouse devices
        '@media (hover: none)': {
            backgroundColor: 'transparent',
        },
    },
}));
