// ra-ui-materialui/dist/button/UpdateWithConfirmButton.js
import * as React from 'react';
import { Fragment, useState } from 'react';
import ActionUpdate from '@mui/icons-material/Update';
import { styled, useThemeProps, } from '@mui/material/styles';
import { useTranslate, useResourceContext, useRecordContext, useGetRecordRepresentation, useResourceTranslation, useUpdateController, } from 'ra-core';
import { Confirm } from "../layout/index.js";
import { Button } from "./Button.js";
import { humanize, singularize } from 'inflection';
export const UpdateWithConfirmButton = React.forwardRef(function UpdateWithConfirmButton(inProps, ref) {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const translate = useTranslate();
    const resource = useResourceContext(props);
    const [isOpen, setOpen] = useState(false);
    const record = useRecordContext(props);
    const { confirmTitle: confirmTitleProp, confirmContent: confirmContentProp, data, icon = defaultIcon, label: labelProp, mutationMode = 'pessimistic', onClick, titleTranslateOptions = emptyObject, contentTranslateOptions = emptyObject, ...rest } = props;
    const { handleUpdate, isPending } = useUpdateController({
        ...rest,
        mutationMode,
        mutationOptions: {
            ...rest.mutationOptions,
            onSettled(...args) {
                // In pessimistic mode, we wait for the mutation to be completed (either successfully or with an error) before closing
                if (mutationMode === 'pessimistic') {
                    setOpen(false);
                }
                rest.mutationOptions?.onSettled?.(...args);
            },
        },
    });
    const handleClick = (e) => {
        e.stopPropagation();
        setOpen(true);
    };
    const handleDialogClose = (e) => {
        e?.stopPropagation();
        setOpen(false);
    };
    const handleConfirm = (e) => {
        e.stopPropagation();
        // We close the dialog immediately here for optimistic/undoable modes instead of in onSuccess/onError
        // to avoid reimplementing the default side effects
        if (mutationMode !== 'pessimistic') {
            setOpen(false);
        }
        handleUpdate(data);
        if (typeof onClick === 'function') {
            onClick(e);
        }
    };
    const getRecordRepresentation = useGetRecordRepresentation(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: humanize(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? singularize(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if (React.isValidElement(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.action.update`,
        baseI18nKey: 'ra.action.update',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    const confirmTitle = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.message.bulk_update_title`,
        baseI18nKey: 'ra.message.bulk_update_title',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            smart_count: 1,
            ...titleTranslateOptions,
        },
        userText: confirmTitleProp,
    });
    const confirmContent = useResourceTranslation({
        resourceI18nKey: `resources.${resource}.message.bulk_update_content`,
        baseI18nKey: 'ra.message.bulk_update_content',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            smart_count: 1,
            ...contentTranslateOptions,
        },
        userText: confirmContentProp,
    });
    return (React.createElement(Fragment, null,
        React.createElement(StyledButton, { ref: ref, onClick: handleClick, 
            // avoid double translation
            label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, ...sanitizeRestProps(rest) }, icon),
        React.createElement(Confirm, { isOpen: isOpen, loading: isPending, title: React.createElement(React.Fragment, null, confirmTitle), content: React.createElement(React.Fragment, null, confirmContent), onConfirm: handleConfirm, onClose: handleDialogClose })));
});
const sanitizeRestProps = ({ label, mutationOptions, ...rest }) => rest;
const PREFIX = 'RaUpdateWithConfirmButton';
const StyledButton = styled(Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    color: (theme.vars || theme).palette.primary.main,
    '&:hover': {
        backgroundColor: `color-mix(in srgb, ${(theme.vars || theme).palette.primary.main}, transparent 12%)`,
        // Reset on mouse devices
        '@media (hover: none)': {
            backgroundColor: 'transparent',
        },
    },
}));
const defaultIcon = React.createElement(ActionUpdate, null);
const emptyObject = {};
