// ra-ui-materialui/dist/button/UpdateWithUndoButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateWithUndoButton = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const Update_1 = __importDefault(require("@mui/icons-material/Update"));
const ra_core_1 = require("ra-core");
const inflection_1 = require("inflection");
const Button_1 = require("./Button.cjs");
exports.UpdateWithUndoButton = React.forwardRef(function UpdateWithUndoButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const record = (0, ra_core_1.useRecordContext)(props);
    const resource = (0, ra_core_1.useResourceContext)(props);
    const { data, label: labelProp, icon = defaultIcon, onClick, ...rest } = props;
    const { handleUpdate, isPending } = (0, ra_core_1.useUpdateController)(rest);
    const translate = (0, ra_core_1.useTranslate)();
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: (0, inflection_1.humanize)(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? (0, inflection_1.singularize)(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if (React.isValidElement(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.action.update`,
        baseI18nKey: 'ra.action.update',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    const handleClick = e => {
        if (!record) {
            throw new Error('The UpdateWithUndoButton must be used inside a RecordContext.Provider or must be passed a record prop.');
        }
        handleUpdate(data);
        if (typeof onClick === 'function') {
            onClick(e);
        }
        e.stopPropagation();
    };
    return (React.createElement(StyledButton, { ref: ref, onClick: handleClick, 
        // avoid double translation
        label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, disabled: isPending, ...sanitizeRestProps(rest) }, icon));
});
const defaultIcon = React.createElement(Update_1.default, null);
const sanitizeRestProps = ({ label, mutationOptions, ...rest }) => rest;
const PREFIX = 'RaUpdateWithUndoButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    color: (theme.vars || theme).palette.primary.main,
    '&:hover': {
        backgroundColor: `color-mix(in srgb, ${(theme.vars || theme).palette.primary.main}, transparent 12%)`,
        // Reset on mouse devices
        '@media (hover: none)': {
            backgroundColor: 'transparent',
        },
    },
}));
