// ra-ui-materialui/dist/button/UpdateWithConfirmButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateWithConfirmButton = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const Update_1 = __importDefault(require("@mui/icons-material/Update"));
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const layout_1 = require("../layout/index.cjs");
const Button_1 = require("./Button.cjs");
const inflection_1 = require("inflection");
exports.UpdateWithConfirmButton = React.forwardRef(function UpdateWithConfirmButton(inProps, ref) {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const translate = (0, ra_core_1.useTranslate)();
    const resource = (0, ra_core_1.useResourceContext)(props);
    const [isOpen, setOpen] = (0, react_1.useState)(false);
    const record = (0, ra_core_1.useRecordContext)(props);
    const { confirmTitle: confirmTitleProp, confirmContent: confirmContentProp, data, icon = defaultIcon, label: labelProp, mutationMode = 'pessimistic', onClick, titleTranslateOptions = emptyObject, contentTranslateOptions = emptyObject, ...rest } = props;
    const { handleUpdate, isPending } = (0, ra_core_1.useUpdateController)({
        ...rest,
        mutationMode,
        mutationOptions: {
            ...rest.mutationOptions,
            onSettled(...args) {
                // In pessimistic mode, we wait for the mutation to be completed (either successfully or with an error) before closing
                if (mutationMode === 'pessimistic') {
                    setOpen(false);
                }
                rest.mutationOptions?.onSettled?.(...args);
            },
        },
    });
    const handleClick = (e) => {
        e.stopPropagation();
        setOpen(true);
    };
    const handleDialogClose = (e) => {
        e?.stopPropagation();
        setOpen(false);
    };
    const handleConfirm = (e) => {
        e.stopPropagation();
        // We close the dialog immediately here for optimistic/undoable modes instead of in onSuccess/onError
        // to avoid reimplementing the default side effects
        if (mutationMode !== 'pessimistic') {
            setOpen(false);
        }
        handleUpdate(data);
        if (typeof onClick === 'function') {
            onClick(e);
        }
    };
    const getRecordRepresentation = (0, ra_core_1.useGetRecordRepresentation)(resource);
    let recordRepresentation = getRecordRepresentation(record);
    const resourceName = translate(`resources.${resource}.forcedCaseName`, {
        smart_count: 1,
        _: (0, inflection_1.humanize)(translate(`resources.${resource}.name`, {
            smart_count: 1,
            _: resource ? (0, inflection_1.singularize)(resource) : undefined,
        }), true),
    });
    // We don't support React elements for this
    if (React.isValidElement(recordRepresentation)) {
        recordRepresentation = `#${record?.id}`;
    }
    const label = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.action.update`,
        baseI18nKey: 'ra.action.update',
        options: {
            name: resourceName,
            recordRepresentation,
        },
        userText: labelProp,
    });
    const confirmTitle = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.message.bulk_update_title`,
        baseI18nKey: 'ra.message.bulk_update_title',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            smart_count: 1,
            ...titleTranslateOptions,
        },
        userText: confirmTitleProp,
    });
    const confirmContent = (0, ra_core_1.useResourceTranslation)({
        resourceI18nKey: `resources.${resource}.message.bulk_update_content`,
        baseI18nKey: 'ra.message.bulk_update_content',
        options: {
            recordRepresentation,
            name: resourceName,
            id: record?.id,
            smart_count: 1,
            ...contentTranslateOptions,
        },
        userText: confirmContentProp,
    });
    return (React.createElement(react_1.Fragment, null,
        React.createElement(StyledButton, { ref: ref, onClick: handleClick, 
            // avoid double translation
            label: React.createElement(React.Fragment, null, label), "aria-label": typeof label === 'string' ? label : undefined, ...sanitizeRestProps(rest) }, icon),
        React.createElement(layout_1.Confirm, { isOpen: isOpen, loading: isPending, title: React.createElement(React.Fragment, null, confirmTitle), content: React.createElement(React.Fragment, null, confirmContent), onConfirm: handleConfirm, onClose: handleDialogClose })));
});
const sanitizeRestProps = ({ label, mutationOptions, ...rest }) => rest;
const PREFIX = 'RaUpdateWithConfirmButton';
const StyledButton = (0, styles_1.styled)(Button_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    color: (theme.vars || theme).palette.primary.main,
    '&:hover': {
        backgroundColor: `color-mix(in srgb, ${(theme.vars || theme).palette.primary.main}, transparent 12%)`,
        // Reset on mouse devices
        '@media (hover: none)': {
            backgroundColor: 'transparent',
        },
    },
}));
const defaultIcon = React.createElement(Update_1.default, null);
const emptyObject = {};
