// ra-ui-materialui/dist/preferences/Configurable.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurableClasses = exports.Configurable = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const Settings_1 = __importDefault(require("@mui/icons-material/Settings"));
const clsx_1 = __importDefault(require("clsx"));
/**
 * Wrap any component with this component to make it configurable
 *
 * When the edit mode is enabled, users will see a button to edit the component;
 * when clicked, the inspector will show the editor element.
 *
 * Creates a context for the preference key, so that both the child component
 * and the editor can access it using usePreferenceKey();
 *
 * @example
 * const ConfigurableTextBlock = ({ preferenceKey = "TextBlock", ...props }) => (
 *     <Configurable editor={<TextBlockInspector />} preferenceKey={preferenceKey}>
 *         <TextBlock {...props} />
 *     </Configurable>
 * );
 */
const Configurable = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, editor, preferenceKey, openButtonLabel = 'ra.configurable.customize', sx, } = props;
    const prefixedPreferenceKey = `preferences.${preferenceKey}`;
    const preferencesEditorContext = (0, ra_core_1.usePreferencesEditor)();
    const hasPreferencesEditorContext = !!preferencesEditorContext;
    const translate = (0, ra_core_1.useTranslate)();
    const { isEnabled, setEditor, preferenceKey: currentPreferenceKey, setPreferenceKey, } = preferencesEditorContext || {};
    const isEditorOpen = prefixedPreferenceKey === currentPreferenceKey;
    const editorOpenRef = (0, react_1.useRef)(isEditorOpen);
    const wrapperRef = (0, react_1.useRef)(null);
    const [isCustomizeButtonVisible, setIsCustomizeButtonVisible] = (0, react_1.useState)(false);
    (0, react_1.useEffect)(() => {
        editorOpenRef.current = isEditorOpen;
    }, [isEditorOpen]);
    // on unmount, if selected, remove the editor
    (0, react_1.useEffect)(() => {
        return () => {
            if (!editorOpenRef.current)
                return;
            setPreferenceKey && setPreferenceKey(null);
            setEditor && setEditor(null);
        };
    }, [setEditor, setPreferenceKey]);
    if (!hasPreferencesEditorContext) {
        return children;
    }
    const handleOpenEditor = () => {
        if (!setEditor) {
            throw new Error('Configurable must be used inside a PreferencesEditorContextProvider');
        }
        // include the editorKey as key to force destroy and mount
        // when switching between two identical editors with different editor keys
        // otherwise the editor will see an update and its useStore will return one tick later
        // which would forbid the usage of uncontrolled inputs in the editor
        setEditor((0, react_1.cloneElement)(editor, {
            preferenceKey: prefixedPreferenceKey,
            key: prefixedPreferenceKey,
        }));
        if (!setPreferenceKey) {
            throw new Error('Configurable must be used inside a PreferencesEditorContextProvider');
        }
        // as we modify the editor, isEditorOpen cannot compare the editor element
        // we'll compare the editor key instead
        setPreferenceKey(prefixedPreferenceKey);
    };
    const handleShowButton = () => {
        setIsCustomizeButtonVisible(true);
    };
    const handleHideButton = () => {
        setIsCustomizeButtonVisible(false);
    };
    return (React.createElement(ra_core_1.PreferenceKeyContextProvider, { value: prefixedPreferenceKey },
        React.createElement(Root, { className: (0, clsx_1.default)(exports.ConfigurableClasses.root, isEnabled && exports.ConfigurableClasses.editMode, isEditorOpen && exports.ConfigurableClasses.editorActive), sx: sx, ref: wrapperRef, onMouseEnter: isEnabled ? handleShowButton : undefined, onMouseLeave: isEnabled ? handleHideButton : undefined }, children),
        React.createElement(material_1.Popover, { open: !!isEnabled && (isCustomizeButtonVisible || isEditorOpen), sx: {
                pointerEvents: 'none',
                '& .MuiPaper-root': {
                    pointerEvents: 'auto',
                    borderRadius: 10,
                    padding: '2px',
                    lineHeight: 0,
                    backgroundColor: 'warning.light',
                    color: 'warning.contrastText',
                    '&:hover': {
                        cursor: 'pointer',
                    },
                },
            }, anchorEl: wrapperRef.current, anchorOrigin: {
                vertical: 'top',
                horizontal: 'right',
            }, transformOrigin: {
                vertical: 'center',
                horizontal: 'center',
            }, onClose: handleHideButton, PaperProps: {
                elevation: 1,
                onMouseEnter: handleShowButton,
                onMouseLeave: handleHideButton,
                title: translate(openButtonLabel),
                onClick: handleOpenEditor,
            }, disableAutoFocus: true, disableRestoreFocus: true, disableEnforceFocus: true, disableScrollLock: true, marginThreshold: 8 },
            React.createElement(Settings_1.default
            // @ts-ignore
            , { 
                // @ts-ignore
                fontSize: "12px" }))));
};
exports.Configurable = Configurable;
const PREFIX = 'RaConfigurable';
exports.ConfigurableClasses = {
    root: `${PREFIX}-root`,
    editMode: `${PREFIX}-editMode`,
    editorActive: `${PREFIX}-editorActive`,
};
const Root = (0, styles_1.styled)('span', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    position: 'relative',
    display: 'inline-block',
    [`&.${exports.ConfigurableClasses.editMode}`]: {
        transition: theme.transitions.create('outline'),
        outline: `color-mix(in srgb, ${(theme.vars || theme).palette.warning.main}, transparent 30%) solid 2px`,
    },
    [`&.${exports.ConfigurableClasses.editMode}:hover `]: {
        outline: `color-mix(in srgb, ${(theme.vars || theme).palette.warning.main}, transparent 50%) solid 2px`,
    },
    [`&.${exports.ConfigurableClasses.editMode}.${exports.ConfigurableClasses.editorActive} , &.${exports.ConfigurableClasses.editMode}.${exports.ConfigurableClasses.editorActive}:hover `]: {
        outline: `${(theme.vars || theme).palette.warning.main} solid 2px`,
    },
}));
