// ra-ui-materialui/dist/preferences/FieldsSelector.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldsSelector = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const FieldToggle_1 = require("./FieldToggle.cjs");
/**
 * UI to select / deselect fields, and store the selection in preferences
 */
const FieldsSelector = ({ name = 'columns', availableName = 'availableColumns', }) => {
    const translate = (0, ra_core_1.useTranslate)();
    const [availableFields, setAvailableFields] = (0, ra_core_1.usePreference)(availableName, []);
    const [omit] = (0, ra_core_1.usePreference)('omit', []);
    const [fields, setFields] = (0, ra_core_1.usePreference)(name, availableFields
        .filter(field => !omit?.includes(field.source))
        .map(field => field.index));
    const handleToggle = event => {
        if (event.target.checked) {
            // add the column at the right position
            setFields(availableFields
                .filter(field => field.index === event.target.name ||
                fields.includes(field.index))
                .map(field => field.index));
        }
        else {
            setFields(fields.filter(index => index !== event.target.name));
        }
    };
    const handleMove = (index1, index2) => {
        const index1Pos = availableFields.findIndex(
        // eslint-disable-next-line eqeqeq
        field => field.index == index1);
        const index2Pos = availableFields.findIndex(
        // eslint-disable-next-line eqeqeq
        field => field.index == index2);
        if (index1Pos === -1 || index2Pos === -1) {
            return;
        }
        let newAvailableFields;
        if (index1Pos > index2Pos) {
            newAvailableFields = [
                ...availableFields.slice(0, index2Pos),
                availableFields[index1Pos],
                ...availableFields.slice(index2Pos, index1Pos),
                ...availableFields.slice(index1Pos + 1),
            ];
        }
        else {
            newAvailableFields = [
                ...availableFields.slice(0, index1Pos),
                ...availableFields.slice(index1Pos + 1, index2Pos + 1),
                availableFields[index1Pos],
                ...availableFields.slice(index2Pos + 1),
            ];
        }
        setAvailableFields(newAvailableFields);
        setFields(fields => newAvailableFields
            .filter(field => fields.includes(field.index))
            .map(field => field.index));
    };
    const handleHideAll = () => {
        setFields([]);
    };
    const handleShowAll = () => {
        setFields(availableFields.map(field => field.index));
    };
    return (React.createElement(material_1.Box, { sx: {
            pt: 0.5,
        } },
        React.createElement(material_1.Box, { component: "ul", sx: { paddingInlineStart: 0, m: 0 } }, availableFields.map(field => (React.createElement(FieldToggle_1.FieldToggle, { key: field.index, source: field.source, label: field.label, index: field.index, selected: fields.includes(field.index), onToggle: handleToggle, onMove: handleMove })))),
        React.createElement(material_1.Box, { onDrop: e => e.preventDefault(), sx: {
                display: 'flex',
                justifyContent: 'space-between',
                mx: -0.5,
                mt: 1,
            } },
            React.createElement(material_1.Button, { size: "small", onClick: handleHideAll }, translate('ra.inspector.hideAll', {
                _: 'Hide All',
            })),
            React.createElement(material_1.Button, { size: "small", onClick: handleShowAll }, translate('ra.inspector.showAll', {
                _: 'Show All',
            })))));
};
exports.FieldsSelector = FieldsSelector;
