// ra-ui-materialui/dist/preferences/Inspector.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InspectorClasses = exports.Inspector = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const CancelOutlined_1 = __importDefault(require("@mui/icons-material/CancelOutlined"));
const DeleteOutline_1 = __importDefault(require("@mui/icons-material/DeleteOutline"));
const styles_1 = require("@mui/material/styles");
const InspectorRoot_1 = require("./InspectorRoot.cjs");
const Inspector = () => {
    const { isEnabled, disable, title, titleOptions, editor, preferenceKey } = (0, ra_core_1.usePreferencesEditor)();
    const isDragging = (0, react_1.useRef)(false);
    const removeItems = (0, ra_core_1.useRemoveItemsFromStore)(preferenceKey);
    const theme = (0, styles_1.useTheme)();
    const translate = (0, ra_core_1.useTranslate)();
    const [version, setVersion] = (0, react_1.useState)(0);
    const [dialogPosition, setDialogPosition] = (0, ra_core_1.useStore)('ra.inspector.position', {
        x: 
        // We want it positioned to the far right of the screen
        document?.body.clientWidth -
            // So we remove its size (see the root css class)
            theme.breakpoints.values.sm / 2 -
            // And add a margin
            8,
        y: 8,
    });
    // poor man's drag and drop
    // store click position relative to the dialog position
    const [clickPosition, setClickPosition] = (0, react_1.useState)();
    const handleDragStart = e => {
        // exit if the user drags on anything but the title
        const draggedElement = document?.elementFromPoint(e.clientX, e.clientY);
        if (draggedElement?.id !== 'inspector-dialog-title') {
            return;
        }
        isDragging.current = true;
        e.dataTransfer.effectAllowed = 'move';
        e.dataTransfer.setData('inspector', '');
        setTimeout(() => {
            e.target.classList.add('hide');
        }, 0);
        setClickPosition({
            x: e.clientX - dialogPosition.x,
            y: e.clientY - dialogPosition.y,
        });
    };
    const handleDragEnd = e => {
        if (isDragging.current && clickPosition) {
            setDialogPosition({
                x: e.clientX - clickPosition.x,
                y: e.clientY - clickPosition.y,
            });
            e.target.classList.remove('hide');
            isDragging.current = false;
        }
    };
    // prevent "back to base" animation when the inspector is dropped
    (0, react_1.useEffect)(() => {
        if (!isEnabled)
            return;
        const handleDragover = e => {
            if (e.dataTransfer.types.includes('inspector')) {
                e.preventDefault();
            }
        };
        document?.addEventListener('dragover', handleDragover);
        return () => {
            document?.removeEventListener('dragover', handleDragover);
        };
    }, [isEnabled]);
    // make sure that the dialog is always visible, as the stored position may be outside the screen
    (0, react_1.useEffect)(() => {
        if (!isEnabled)
            return;
        const moveInspectorIfOutsideScreen = () => {
            window?.requestAnimationFrame(() => {
                setDialogPosition(position => ({
                    x: Math.min(position.x, document?.body.clientWidth -
                        theme.breakpoints.values.sm / 2 -
                        8),
                    y: Math.min(position.y, window?.innerHeight - 50),
                }));
            });
        };
        moveInspectorIfOutsideScreen();
        window?.addEventListener('resize', moveInspectorIfOutsideScreen);
        return () => {
            window?.removeEventListener('resize', moveInspectorIfOutsideScreen);
        };
    }, [isEnabled, setDialogPosition, theme.breakpoints.values.sm]);
    const handleReset = () => {
        removeItems();
        // force redraw of the form to use the default values
        setVersion(version => version + 1);
    };
    if (!isEnabled)
        return null;
    return (React.createElement(StyledPaper, { className: exports.InspectorClasses.modal, elevation: 3, draggable: true, onDragStart: handleDragStart, onDragEnd: handleDragEnd, sx: { left: dialogPosition.x, top: dialogPosition.y } },
        React.createElement("div", { className: exports.InspectorClasses.title },
            React.createElement(material_1.Typography, { id: "inspector-dialog-title", variant: "overline", component: "div", sx: {
                    py: 1,
                    px: 2,
                    flex: '1',
                } }, title && translate(title, titleOptions)),
            React.createElement("span", { id: "inspector-toolbar" }),
            preferenceKey && (React.createElement(material_1.IconButton, { "aria-label": translate('ra.action.remove'), onClick: handleReset, size: "small" },
                React.createElement(DeleteOutline_1.default, { fontSize: "inherit" }))),
            React.createElement(material_1.IconButton, { "aria-label": translate('ra.action.close'), onClick: disable, size: "small", sx: { mr: 1 } },
                React.createElement(CancelOutlined_1.default, { fontSize: "inherit" }))),
        React.createElement("div", { className: exports.InspectorClasses.content, key: version },
            React.createElement(ra_core_1.PreferenceKeyContextProvider, { value: preferenceKey }, editor || React.createElement(InspectorRoot_1.InspectorRoot, null)))));
};
exports.Inspector = Inspector;
const PREFIX = 'RaInspector';
exports.InspectorClasses = {
    modal: `${PREFIX}-modal`,
    title: `${PREFIX}-title`,
    content: `${PREFIX}-content`,
};
const StyledPaper = (0, styles_1.styled)(material_1.Paper, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    position: 'fixed',
    zIndex: theme.zIndex.modal + 1,
    width: theme.breakpoints.values.sm / 2,
    transition: theme.transitions.create(['height', 'width']),
    '&.hide': {
        display: 'none',
    },
    [`& .${exports.InspectorClasses.title}`]: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        cursor: 'move',
    },
    [`& .${exports.InspectorClasses.content}`]: {
        overflowY: 'auto',
        maxHeight: '75vh',
        padding: theme.spacing(2),
        paddingTop: 0,
    },
}));
exports.Inspector.displayName = 'Inspector';
