// ra-i18n-polyglot/dist/index.js
import Polyglot from 'node-polyglot';
/**
 * Build a polyglot-based i18nProvider based on a function returning the messages for a locale
 *
 * @example
 *
 * import { Admin, Resource } from 'react-admin';
 * import polyglotI18nProvider from 'ra-i18n-polyglot';
 * import englishMessages from 'ra-language-english';
 * import frenchMessages from 'ra-language-french';
 *
 * const messages = {
 *     fr: frenchMessages,
 *     en: englishMessages,
 * };
 * const i18nProvider = polyglotI18nProvider(
 *     locale => messages[locale],
 *     'en',
 *     [{ locale: 'en', name: 'English' }, { locale: 'fr', name: 'Français' }]
 * )
 */
export default (getMessages, initialLocale = 'en', availableLocales = [{ locale: 'en', name: 'English' }], polyglotOptions = {}) => {
    let locale = initialLocale;
    const messages = getMessages(initialLocale);
    if (messages instanceof Promise) {
        throw new Error(`The i18nProvider returned a Promise for the messages of the default locale (${initialLocale}). Please update your i18nProvider to return the messages of the default locale in a synchronous way.`);
    }
    let availableLocalesFinal, polyglotOptionsFinal;
    if (Array.isArray(availableLocales)) {
        // third argument is an array of locales
        availableLocalesFinal = availableLocales;
        polyglotOptionsFinal = polyglotOptions;
    }
    else {
        // third argument is the polyglotOptions
        availableLocalesFinal = [{ locale: 'en', name: 'English' }];
        polyglotOptionsFinal = availableLocales;
    }
    const polyglot = new Polyglot({
        locale,
        phrases: { '': '', ...messages },
        ...polyglotOptionsFinal,
    });
    let translate = polyglot.t.bind(polyglot);
    return {
        translate: (key, options = {}) => translate(key, options),
        changeLocale: (newLocale) => 
        // We systematically return a Promise for the messages because
        // getMessages may return a Promise
        Promise.resolve(getMessages(newLocale)).then((messages) => {
            locale = newLocale;
            const newPolyglot = new Polyglot({
                locale: newLocale,
                phrases: { '': '', ...messages },
                ...polyglotOptions,
            });
            translate = newPolyglot.t.bind(newPolyglot);
        }),
        getLocale: () => locale,
        getLocales: () => availableLocalesFinal,
    };
};
