// ra-ui-materialui/dist/list/datagrid/Datagrid.js
import * as React from 'react';
import { cloneElement, createElement, isValidElement, useCallback, useRef, useEffect, useMemo, } from 'react';
import { sanitizeListRestProps, useListContextWithProps, OptionalResourceContextProvider, useCanAccess, useResourceContext, } from 'ra-core';
import { Table, useThemeProps } from '@mui/material';
import clsx from 'clsx';
import union from 'lodash/union.js';
import difference from 'lodash/difference.js';
import { DatagridHeader } from "./DatagridHeader.js";
import DatagridLoading from "./DatagridLoading.js";
import DatagridBody, { PureDatagridBody } from "./DatagridBody.js";
import DatagridContextProvider from "./DatagridContextProvider.js";
import { DatagridClasses, DatagridPrefix, DatagridRoot, } from "./useDatagridStyles.js";
import { BulkActionsToolbar } from "../BulkActionsToolbar.js";
import { BulkDeleteButton } from "../../button/index.js";
import { ListNoResults } from "../ListNoResults.js";
const defaultBulkActionButtons = React.createElement(BulkDeleteButton, null);
/**
 * The Datagrid component renders a list of records as a table.
 * It is usually used as a child of the <List> and <ReferenceManyField> components.
 *
 * Props:
 *  - body
 *  - bulkActionToolbar
 *  - bulkActionButtons
 *  - children
 *  - empty
 *  - expand
 *  - header
 *  - hover
 *  - isRowExpandable
 *  - isRowSelectable
 *  - optimized
 *  - rowClick
 *  - rowSx
 *  - size
 *  - sx
 *
 * @example // Display all posts as a datagrid
 * const postRowSx = (record, index) => ({
 *     backgroundColor: record.nb_views >= 500 ? '#efe' : 'white',
 * });
 * export const PostList = () => (
 *     <List>
 *         <Datagrid rowSx={postRowSx}>
 *             <TextField source="id" />
 *             <TextField source="title" />
 *             <TextField source="body" />
 *             <EditButton />
 *         </Datagrid>
 *     </List>
 * );
 *
 * @example // Display all the comments of the current post as a datagrid
 * <ReferenceManyField reference="comments" target="post_id">
 *     <Datagrid>
 *         <TextField source="id" />
 *         <TextField source="body" />
 *         <DateField source="created_at" />
 *         <EditButton />
 *     </Datagrid>
 * </ReferenceManyField>
 *
 * @example // Usage outside of a <List> or a <ReferenceManyField>.
 *
 * const sort = { field: 'published_at', order: 'DESC' };
 *
 * export const MyCustomList = (props) => {
 *     const { data, total, isPending } = useGetList(
 *         'posts',
 *         { pagination: { page: 1, perPage: 10 }, sort: sort }
 *     );
 *
 *     return (
 *         <Datagrid
 *             data={data}
 *             total={total}
 *             isPending={isPending}
 *             sort={sort}
 *             selectedIds={[]}
 *             setSort={() => {
 *                 console.log('set sort');
 *             }}
 *             onSelect={() => {
 *                 console.log('on select');
 *             }}
 *             onToggleItem={() => {
 *                 console.log('on toggle item');
 *             }}
 *         >
 *             <TextField source="id" />
 *             <TextField source="title" />
 *         </Datagrid>
 *     );
 * }
 */
export const Datagrid = React.forwardRef((inProps, ref) => {
    const props = useThemeProps({
        name: DatagridPrefix,
        props: inProps,
    });
    const resourceFromContext = useResourceContext(props);
    const { canAccess: canDelete } = useCanAccess({
        resource: resourceFromContext,
        action: 'delete',
    });
    const { optimized = false, body = optimized ? PureDatagridBody : DatagridBody, header = DatagridHeader, children, className, empty, expand, bulkActionsToolbar, bulkActionButtons = canDelete ? defaultBulkActionButtons : false, hover, isRowSelectable, isRowExpandable, resource, rowClick, rowSx, rowStyle, size = 'small', sx, expandSingle = false, ...rest } = props;
    const { sort, data, isPending, onSelect, onToggleItem, selectedIds, setSort, total, } = useListContextWithProps(props);
    const hasBulkActions = !!bulkActionButtons !== false;
    const contextValue = useMemo(() => ({ isRowExpandable, expandSingle }), [isRowExpandable, expandSingle]);
    const lastSelected = useRef(null);
    useEffect(() => {
        if (!selectedIds || selectedIds.length === 0) {
            lastSelected.current = null;
        }
    }, [JSON.stringify(selectedIds)]); // eslint-disable-line react-hooks/exhaustive-deps
    // we manage row selection at the datagrid level to allow shift+click to select an array of rows
    const handleToggleItem = useCallback((id, event) => {
        if (!data)
            return;
        const ids = data.map(record => record.id);
        const lastSelectedIndex = ids.indexOf(lastSelected.current);
        lastSelected.current = event.target.checked ? id : null;
        if (event.shiftKey && lastSelectedIndex !== -1) {
            const index = ids.indexOf(id);
            const idsBetweenSelections = ids.slice(Math.min(lastSelectedIndex, index), Math.max(lastSelectedIndex, index) + 1);
            const newSelectedIds = event.target.checked
                ? union(selectedIds, idsBetweenSelections)
                : difference(selectedIds, idsBetweenSelections);
            onSelect?.(isRowSelectable
                ? newSelectedIds.filter((id) => isRowSelectable(data.find(record => record.id === id)))
                : newSelectedIds);
        }
        else {
            onToggleItem?.(id);
        }
    }, [data, isRowSelectable, onSelect, onToggleItem, selectedIds]);
    if (isPending === true) {
        return (React.createElement(DatagridLoading, { className: className, expand: expand, hasBulkActions: hasBulkActions, nbChildren: React.Children.count(children), size: size }));
    }
    /**
     * Once loaded, the data for the list may be empty. Instead of
     * displaying the table header with zero data rows,
     * the Datagrid displays the empty component.
     */
    if (data == null || data.length === 0 || total === 0) {
        return empty === undefined ? (React.createElement(ListNoResults, { resource: resource })) : (empty);
    }
    /**
     * After the initial load, if the data for the list isn't empty,
     * and even if the data is refreshing (e.g. after a filter change),
     * the datagrid displays the current data.
     */
    return (React.createElement(DatagridContextProvider, { value: contextValue },
        React.createElement(OptionalResourceContextProvider, { value: resource },
            React.createElement(DatagridRoot, { sx: sx, className: clsx(DatagridClasses.root, className) },
                bulkActionsToolbar ??
                    (bulkActionButtons !== false ? (React.createElement(BulkActionsToolbar, null, isValidElement(bulkActionButtons)
                        ? bulkActionButtons
                        : defaultBulkActionButtons)) : null),
                React.createElement("div", { className: DatagridClasses.tableWrapper },
                    React.createElement(Table, { ref: ref, className: DatagridClasses.table, size: size, ...sanitizeRestProps(rest) },
                        createOrCloneElement(header, {
                            children,
                            sort,
                            data,
                            hasExpand: !!expand,
                            hasBulkActions,
                            isRowSelectable,
                            onSelect,
                            selectedIds,
                            setSort,
                        }, children),
                        createOrCloneElement(body, {
                            expand,
                            rowClick,
                            data,
                            hasBulkActions,
                            hover,
                            onToggleItem: handleToggleItem,
                            resource,
                            rowSx,
                            rowStyle,
                            selectedIds,
                            isRowSelectable,
                        }, children)))))));
});
const createOrCloneElement = (element, props, children) => isValidElement(element)
    ? cloneElement(element, props, children)
    : createElement(element, props, children);
const injectedProps = [
    'isRequired',
    'setFilter',
    'setPagination',
    'limitChoicesToValue',
    'translateChoice',
    // Datagrid may be used as an alternative to SelectInput
    'field',
    'fieldState',
    'formState',
];
const sanitizeRestProps = props => Object.keys(sanitizeListRestProps(props))
    .filter(propName => !injectedProps.includes(propName) || propName === 'ref')
    .reduce((acc, key) => ({ ...acc, [key]: props[key] }), {});
Datagrid.displayName = 'Datagrid';
