// ra-ui-materialui/dist/list/datagrid/Datagrid.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Datagrid = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const union_js_1 = __importDefault(require("lodash/union.js"));
const difference_js_1 = __importDefault(require("lodash/difference.js"));
const DatagridHeader_1 = require("./DatagridHeader.cjs");
const DatagridLoading_1 = __importDefault(require("./DatagridLoading.cjs"));
const DatagridBody_1 = __importStar(require("./DatagridBody.cjs"));
const DatagridContextProvider_1 = __importDefault(require("./DatagridContextProvider.cjs"));
const useDatagridStyles_1 = require("./useDatagridStyles.cjs");
const BulkActionsToolbar_1 = require("../BulkActionsToolbar.cjs");
const button_1 = require("../../button/index.cjs");
const ListNoResults_1 = require("../ListNoResults.cjs");
const defaultBulkActionButtons = React.createElement(button_1.BulkDeleteButton, null);
/**
 * The Datagrid component renders a list of records as a table.
 * It is usually used as a child of the <List> and <ReferenceManyField> components.
 *
 * Props:
 *  - body
 *  - bulkActionToolbar
 *  - bulkActionButtons
 *  - children
 *  - empty
 *  - expand
 *  - header
 *  - hover
 *  - isRowExpandable
 *  - isRowSelectable
 *  - optimized
 *  - rowClick
 *  - rowSx
 *  - size
 *  - sx
 *
 * @example // Display all posts as a datagrid
 * const postRowSx = (record, index) => ({
 *     backgroundColor: record.nb_views >= 500 ? '#efe' : 'white',
 * });
 * export const PostList = () => (
 *     <List>
 *         <Datagrid rowSx={postRowSx}>
 *             <TextField source="id" />
 *             <TextField source="title" />
 *             <TextField source="body" />
 *             <EditButton />
 *         </Datagrid>
 *     </List>
 * );
 *
 * @example // Display all the comments of the current post as a datagrid
 * <ReferenceManyField reference="comments" target="post_id">
 *     <Datagrid>
 *         <TextField source="id" />
 *         <TextField source="body" />
 *         <DateField source="created_at" />
 *         <EditButton />
 *     </Datagrid>
 * </ReferenceManyField>
 *
 * @example // Usage outside of a <List> or a <ReferenceManyField>.
 *
 * const sort = { field: 'published_at', order: 'DESC' };
 *
 * export const MyCustomList = (props) => {
 *     const { data, total, isPending } = useGetList(
 *         'posts',
 *         { pagination: { page: 1, perPage: 10 }, sort: sort }
 *     );
 *
 *     return (
 *         <Datagrid
 *             data={data}
 *             total={total}
 *             isPending={isPending}
 *             sort={sort}
 *             selectedIds={[]}
 *             setSort={() => {
 *                 console.log('set sort');
 *             }}
 *             onSelect={() => {
 *                 console.log('on select');
 *             }}
 *             onToggleItem={() => {
 *                 console.log('on toggle item');
 *             }}
 *         >
 *             <TextField source="id" />
 *             <TextField source="title" />
 *         </Datagrid>
 *     );
 * }
 */
exports.Datagrid = React.forwardRef((inProps, ref) => {
    const props = (0, material_1.useThemeProps)({
        name: useDatagridStyles_1.DatagridPrefix,
        props: inProps,
    });
    const resourceFromContext = (0, ra_core_1.useResourceContext)(props);
    const { canAccess: canDelete } = (0, ra_core_1.useCanAccess)({
        resource: resourceFromContext,
        action: 'delete',
    });
    const { optimized = false, body = optimized ? DatagridBody_1.PureDatagridBody : DatagridBody_1.default, header = DatagridHeader_1.DatagridHeader, children, className, empty, expand, bulkActionsToolbar, bulkActionButtons = canDelete ? defaultBulkActionButtons : false, hover, isRowSelectable, isRowExpandable, resource, rowClick, rowSx, rowStyle, size = 'small', sx, expandSingle = false, ...rest } = props;
    const { sort, data, isPending, onSelect, onToggleItem, selectedIds, setSort, total, } = (0, ra_core_1.useListContextWithProps)(props);
    const hasBulkActions = !!bulkActionButtons !== false;
    const contextValue = (0, react_1.useMemo)(() => ({ isRowExpandable, expandSingle }), [isRowExpandable, expandSingle]);
    const lastSelected = (0, react_1.useRef)(null);
    (0, react_1.useEffect)(() => {
        if (!selectedIds || selectedIds.length === 0) {
            lastSelected.current = null;
        }
    }, [JSON.stringify(selectedIds)]); // eslint-disable-line react-hooks/exhaustive-deps
    // we manage row selection at the datagrid level to allow shift+click to select an array of rows
    const handleToggleItem = (0, react_1.useCallback)((id, event) => {
        if (!data)
            return;
        const ids = data.map(record => record.id);
        const lastSelectedIndex = ids.indexOf(lastSelected.current);
        lastSelected.current = event.target.checked ? id : null;
        if (event.shiftKey && lastSelectedIndex !== -1) {
            const index = ids.indexOf(id);
            const idsBetweenSelections = ids.slice(Math.min(lastSelectedIndex, index), Math.max(lastSelectedIndex, index) + 1);
            const newSelectedIds = event.target.checked
                ? (0, union_js_1.default)(selectedIds, idsBetweenSelections)
                : (0, difference_js_1.default)(selectedIds, idsBetweenSelections);
            onSelect?.(isRowSelectable
                ? newSelectedIds.filter((id) => isRowSelectable(data.find(record => record.id === id)))
                : newSelectedIds);
        }
        else {
            onToggleItem?.(id);
        }
    }, [data, isRowSelectable, onSelect, onToggleItem, selectedIds]);
    if (isPending === true) {
        return (React.createElement(DatagridLoading_1.default, { className: className, expand: expand, hasBulkActions: hasBulkActions, nbChildren: React.Children.count(children), size: size }));
    }
    /**
     * Once loaded, the data for the list may be empty. Instead of
     * displaying the table header with zero data rows,
     * the Datagrid displays the empty component.
     */
    if (data == null || data.length === 0 || total === 0) {
        return empty === undefined ? (React.createElement(ListNoResults_1.ListNoResults, { resource: resource })) : (empty);
    }
    /**
     * After the initial load, if the data for the list isn't empty,
     * and even if the data is refreshing (e.g. after a filter change),
     * the datagrid displays the current data.
     */
    return (React.createElement(DatagridContextProvider_1.default, { value: contextValue },
        React.createElement(ra_core_1.OptionalResourceContextProvider, { value: resource },
            React.createElement(useDatagridStyles_1.DatagridRoot, { sx: sx, className: (0, clsx_1.default)(useDatagridStyles_1.DatagridClasses.root, className) },
                bulkActionsToolbar ??
                    (bulkActionButtons !== false ? (React.createElement(BulkActionsToolbar_1.BulkActionsToolbar, null, (0, react_1.isValidElement)(bulkActionButtons)
                        ? bulkActionButtons
                        : defaultBulkActionButtons)) : null),
                React.createElement("div", { className: useDatagridStyles_1.DatagridClasses.tableWrapper },
                    React.createElement(material_1.Table, { ref: ref, className: useDatagridStyles_1.DatagridClasses.table, size: size, ...sanitizeRestProps(rest) },
                        createOrCloneElement(header, {
                            children,
                            sort,
                            data,
                            hasExpand: !!expand,
                            hasBulkActions,
                            isRowSelectable,
                            onSelect,
                            selectedIds,
                            setSort,
                        }, children),
                        createOrCloneElement(body, {
                            expand,
                            rowClick,
                            data,
                            hasBulkActions,
                            hover,
                            onToggleItem: handleToggleItem,
                            resource,
                            rowSx,
                            rowStyle,
                            selectedIds,
                            isRowSelectable,
                        }, children)))))));
});
const createOrCloneElement = (element, props, children) => (0, react_1.isValidElement)(element)
    ? (0, react_1.cloneElement)(element, props, children)
    : (0, react_1.createElement)(element, props, children);
const injectedProps = [
    'isRequired',
    'setFilter',
    'setPagination',
    'limitChoicesToValue',
    'translateChoice',
    // Datagrid may be used as an alternative to SelectInput
    'field',
    'fieldState',
    'formState',
];
const sanitizeRestProps = props => Object.keys((0, ra_core_1.sanitizeListRestProps)(props))
    .filter(propName => !injectedProps.includes(propName) || propName === 'ref')
    .reduce((acc, key) => ({ ...acc, [key]: props[key] }), {});
exports.Datagrid.displayName = 'Datagrid';
