// ra-ui-materialui/dist/list/datagrid/DatagridBody.js
import * as React from 'react';
import { cloneElement, memo } from 'react';
import { TableBody } from '@mui/material';
import clsx from 'clsx';
import { RecordContextProvider } from 'ra-core';
import { DatagridClasses } from "./useDatagridStyles.js";
import DatagridRow, { PureDatagridRow } from "./DatagridRow.js";
const DatagridBody = React.forwardRef(({ children, className, data = defaultData, expand, hasBulkActions = false, hover, onToggleItem, resource, row = defaultChildren, rowClick, rowSx, rowStyle, selectedIds, isRowSelectable, ...rest }, ref) => (React.createElement(TableBody, { ref: ref, className: clsx('datagrid-body', className, DatagridClasses.tbody), ...rest }, data.map((record, rowIndex) => (React.createElement(RecordContextProvider, { value: record, key: record.id ?? `row${rowIndex}` }, cloneElement(row, {
    className: clsx(DatagridClasses.row, {
        [DatagridClasses.rowEven]: rowIndex % 2 === 0,
        [DatagridClasses.rowOdd]: rowIndex % 2 !== 0,
    }),
    expand,
    hasBulkActions: hasBulkActions && !!selectedIds,
    hover,
    id: record.id ?? `row${rowIndex}`,
    onToggleItem,
    resource,
    rowClick,
    selectable: !isRowSelectable || isRowSelectable(record),
    selected: selectedIds?.includes(record.id),
    sx: rowSx?.(record, rowIndex),
    style: rowStyle?.(record, rowIndex),
}, children)))))));
const defaultChildren = React.createElement(DatagridRow, null);
const defaultData = [];
// trick Material UI Table into thinking this is one of the child type it supports
// @ts-ignore
DatagridBody.muiName = 'TableBody';
export const PureDatagridBody = memo(props => (React.createElement(DatagridBody, { row: React.createElement(PureDatagridRow, null), ...props })));
// trick Material UI Table into thinking this is one of the child type it supports
// @ts-ignore
PureDatagridBody.muiName = 'TableBody';
export default DatagridBody;
