// ra-ui-materialui/dist/list/datagrid/DatagridBody.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PureDatagridBody = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const useDatagridStyles_1 = require("./useDatagridStyles.cjs");
const DatagridRow_1 = __importStar(require("./DatagridRow.cjs"));
const DatagridBody = React.forwardRef(({ children, className, data = defaultData, expand, hasBulkActions = false, hover, onToggleItem, resource, row = defaultChildren, rowClick, rowSx, rowStyle, selectedIds, isRowSelectable, ...rest }, ref) => (React.createElement(material_1.TableBody, { ref: ref, className: (0, clsx_1.default)('datagrid-body', className, useDatagridStyles_1.DatagridClasses.tbody), ...rest }, data.map((record, rowIndex) => (React.createElement(ra_core_1.RecordContextProvider, { value: record, key: record.id ?? `row${rowIndex}` }, (0, react_1.cloneElement)(row, {
    className: (0, clsx_1.default)(useDatagridStyles_1.DatagridClasses.row, {
        [useDatagridStyles_1.DatagridClasses.rowEven]: rowIndex % 2 === 0,
        [useDatagridStyles_1.DatagridClasses.rowOdd]: rowIndex % 2 !== 0,
    }),
    expand,
    hasBulkActions: hasBulkActions && !!selectedIds,
    hover,
    id: record.id ?? `row${rowIndex}`,
    onToggleItem,
    resource,
    rowClick,
    selectable: !isRowSelectable || isRowSelectable(record),
    selected: selectedIds?.includes(record.id),
    sx: rowSx?.(record, rowIndex),
    style: rowStyle?.(record, rowIndex),
}, children)))))));
const defaultChildren = React.createElement(DatagridRow_1.default, null);
const defaultData = [];
// trick Material UI Table into thinking this is one of the child type it supports
// @ts-ignore
DatagridBody.muiName = 'TableBody';
exports.PureDatagridBody = (0, react_1.memo)(props => (React.createElement(DatagridBody, { row: React.createElement(DatagridRow_1.PureDatagridRow, null), ...props })));
// trick Material UI Table into thinking this is one of the child type it supports
// @ts-ignore
exports.PureDatagridBody.muiName = 'TableBody';
exports.default = DatagridBody;
