// ra-ui-materialui/dist/list/datagrid/DatagridHeaderCell.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatagridHeaderCellClasses = exports.DatagridHeaderCell = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const react_1 = require("react");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const oppositeOrder = {
    ASC: 'DESC',
    DESC: 'ASC',
};
const DatagridHeaderCell = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, field, sort, updateSort, isSorting, ...rest } = props;
    const resource = (0, ra_core_1.useResourceContext)();
    const translate = (0, ra_core_1.useTranslate)();
    const translateLabel = (0, ra_core_1.useTranslateLabel)();
    const nextSortOrder = field
        ? sort && sort.field === (field.props.sortBy || field.props.source)
            ? // active sort field, use opposite order
                oppositeOrder[sort.order]
            : // non active sort field, use default order
                field?.props.sortByOrder ?? 'ASC'
        : undefined;
    const fieldLabel = field
        ? translateLabel({
            label: typeof field.props.label === 'string'
                ? field.props.label
                : undefined,
            resource,
            source: field.props.source,
        })
        : undefined;
    const sortLabel = translate('ra.sort.sort_by', {
        field: fieldLabel,
        field_lower_first: typeof fieldLabel === 'string'
            ? fieldLabel.charAt(0).toLowerCase() + fieldLabel.slice(1)
            : undefined,
        order: translate(`ra.sort.${nextSortOrder}`),
        _: translate('ra.action.sort'),
    });
    return (React.createElement(StyledTableCell, { className: (0, clsx_1.default)(className, field?.props.headerClassName), align: field?.props.textAlign || field?.type.textAlign, variant: "head", ...rest }, updateSort &&
        sort &&
        field &&
        field.props.sortable !== false &&
        field.type.sortable !== false &&
        (field.props.sortBy || field.props.source) ? (React.createElement(material_1.Tooltip, { title: sortLabel, placement: field.props.textAlign === 'right' ||
            field.type.textAlign === 'right'
            ? 'bottom-end'
            : 'bottom-start', enterDelay: 300 },
        React.createElement(material_1.TableSortLabel, { active: sort.field ===
                (field.props.sortBy || field.props.source), direction: sort.order === 'ASC' ? 'asc' : 'desc', "data-field": field.props.sortBy || field.props.source, "data-order": field.props.sortByOrder || 'ASC', onClick: updateSort, classes: exports.DatagridHeaderCellClasses },
            React.createElement(ra_core_1.FieldTitle, { label: field.props.label, source: field.props.source, resource: resource })))) : (React.createElement(ra_core_1.FieldTitle, { label: field?.props.label, source: field?.props.source, resource: resource }))));
};
exports.DatagridHeaderCell = DatagridHeaderCell;
exports.default = (0, react_1.memo)(exports.DatagridHeaderCell, (props, nextProps) => props.updateSort === nextProps.updateSort &&
    props.sort?.field === nextProps.sort?.field &&
    props.sort?.order === nextProps.sort?.order &&
    props.isSorting === nextProps.isSorting);
const PREFIX = 'RaDatagridHeaderCell';
exports.DatagridHeaderCellClasses = {
    icon: `${PREFIX}-icon`,
};
// Remove the sort icons when not active
const StyledTableCell = (0, styles_1.styled)(material_1.TableCell, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    [`& .MuiTableSortLabel-icon`]: {
        display: 'none',
    },
    [`& .Mui-active .MuiTableSortLabel-icon`]: {
        display: 'inline',
    },
});
