// ra-ui-materialui/dist/list/datagrid/SelectColumnsButton.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectColumnsButton = void 0;
const React = __importStar(require("react"));
const ra_core_1 = require("ra-core");
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const ViewWeek_1 = __importDefault(require("@mui/icons-material/ViewWeek"));
const preferences_1 = require("../../preferences/index.cjs");
/**
 * Renders a button that lets users show / hide columns in a configurable datagrid
 *
 * @example
 * import { SelectColumnsButton, DatagridConfigurable } from 'react-admin';
 *
 * const PostListActions = () => (
 *   <TopToolbar>
        <SelectColumnsButton />
        <FilterButton />
 *   </TopToolbar>
 * );
 *
 * const PostList = () => (
 *   <List actions={<PostListActions />}>
 *     <DatagridConfigurable>
 *       <TextField source="title" />
 *       <TextField source="author" />
         ...
 *     </DatagridConfigurable>
 *   </List>
 * );
 */
const SelectColumnsButton = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { preferenceKey } = props;
    const resource = (0, ra_core_1.useResourceContext)(props);
    const finalPreferenceKey = preferenceKey || `${resource}.datagrid`;
    const [anchorEl, setAnchorEl] = React.useState(null);
    const [availableColumns, setAvailableColumns] = (0, ra_core_1.useStore)(`preferences.${finalPreferenceKey}.availableColumns`, []);
    const [omit] = (0, ra_core_1.useStore)(`preferences.${finalPreferenceKey}.omit`, []);
    const [columns, setColumns] = (0, ra_core_1.useStore)(`preferences.${finalPreferenceKey}.columns`, availableColumns
        .filter(column => column.source ? !omit?.includes(column.source) : true)
        .map(column => column.index));
    const translate = (0, ra_core_1.useTranslate)();
    const isXSmall = (0, material_1.useMediaQuery)((theme) => theme.breakpoints.down('sm'));
    const title = translate('ra.action.select_columns', { _: 'Columns' });
    const handleClick = (event) => {
        setAnchorEl(event.currentTarget);
    };
    const handleClose = () => {
        setAnchorEl(null);
    };
    const handleToggle = event => {
        if (event.target.checked) {
            // add the column at the right position
            setColumns(availableColumns
                .filter(column => column.index === event.target.name ||
                columns.includes(column.index))
                .map(column => column.index));
        }
        else {
            setColumns(columns.filter(index => index !== event.target.name));
        }
    };
    const handleMove = (index1, index2) => {
        const index1Pos = availableColumns.findIndex(
        // eslint-disable-next-line eqeqeq
        field => field.index == index1);
        const index2Pos = availableColumns.findIndex(
        // eslint-disable-next-line eqeqeq
        field => field.index == index2);
        if (index1Pos === -1 || index2Pos === -1) {
            return;
        }
        let newAvailableColumns;
        if (index1Pos > index2Pos) {
            newAvailableColumns = [
                ...availableColumns.slice(0, index2Pos),
                availableColumns[index1Pos],
                ...availableColumns.slice(index2Pos, index1Pos),
                ...availableColumns.slice(index1Pos + 1),
            ];
        }
        else {
            newAvailableColumns = [
                ...availableColumns.slice(0, index1Pos),
                ...availableColumns.slice(index1Pos + 1, index2Pos + 1),
                availableColumns[index1Pos],
                ...availableColumns.slice(index2Pos + 1),
            ];
        }
        setAvailableColumns(newAvailableColumns);
        setColumns(columns => newAvailableColumns
            .filter(column => columns.includes(column.index))
            .map(column => column.index));
    };
    return (React.createElement(React.Fragment, null,
        isXSmall ? (React.createElement(material_1.Tooltip, { title: title },
            React.createElement(material_1.IconButton, { "aria-label": title, color: "primary", onClick: handleClick, size: "large", ...sanitizeRestProps(props) },
                React.createElement(ViewWeek_1.default, null)))) : (React.createElement(StyledButton, { size: "small", onClick: handleClick, startIcon: React.createElement(ViewWeek_1.default, null), ...sanitizeRestProps(props) }, title)),
        React.createElement(material_1.Popover, { open: Boolean(anchorEl), anchorEl: anchorEl, onClose: handleClose, anchorOrigin: {
                vertical: 'bottom',
                horizontal: 'center',
            }, transformOrigin: {
                vertical: 'top',
                horizontal: 'center',
            } },
            React.createElement(material_1.Box, { component: "ul", sx: {
                    p: 1,
                    my: 0,
                } }, availableColumns.map(column => (React.createElement(preferences_1.FieldToggle, { key: column.index, source: column.source, label: column.label, index: column.index, selected: columns.includes(column.index), onToggle: handleToggle, onMove: handleMove })))))));
};
exports.SelectColumnsButton = SelectColumnsButton;
const PREFIX = 'RaSelectColumnsButton';
const StyledButton = (0, styles_1.styled)(material_1.Button, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    '&.MuiButton-sizeSmall': {
        // fix for icon misalignment on small buttons, see https://github.com/mui/material-ui/pull/30240
        lineHeight: 1.5,
    },
});
const sanitizeRestProps = ({ resource, preferenceKey, ...rest }) => rest;
