// ra-ui-materialui/dist/list/datagrid/useDatagridStyles.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatagridRoot = exports.DatagridClasses = exports.DatagridPrefix = void 0;
const material_1 = require("@mui/material");
exports.DatagridPrefix = 'RaDatagrid';
exports.DatagridClasses = {
    root: `${exports.DatagridPrefix}-root`,
    table: `${exports.DatagridPrefix}-table`,
    tableWrapper: `${exports.DatagridPrefix}-tableWrapper`,
    thead: `${exports.DatagridPrefix}-thead`,
    tbody: `${exports.DatagridPrefix}-tbody`,
    headerRow: `${exports.DatagridPrefix}-headerRow`,
    headerCell: `${exports.DatagridPrefix}-headerCell`,
    checkbox: `${exports.DatagridPrefix}-checkbox`,
    row: `${exports.DatagridPrefix}-row`,
    clickableRow: `${exports.DatagridPrefix}-clickableRow`,
    rowEven: `${exports.DatagridPrefix}-rowEven`,
    rowOdd: `${exports.DatagridPrefix}-rowOdd`,
    rowCell: `${exports.DatagridPrefix}-rowCell`,
    selectable: `${exports.DatagridPrefix}-selectable`,
    expandHeader: `${exports.DatagridPrefix}-expandHeader`,
    expandIconCell: `${exports.DatagridPrefix}-expandIconCell`,
    expandIcon: `${exports.DatagridPrefix}-expandIcon`,
    expandable: `${exports.DatagridPrefix}-expandable`,
    expanded: `${exports.DatagridPrefix}-expanded`,
    expandedPanel: `${exports.DatagridPrefix}-expandedPanel`,
};
exports.DatagridRoot = (0, material_1.styled)('div', {
    name: exports.DatagridPrefix,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.DatagridClasses.table}`]: {
        tableLayout: 'auto',
    },
    [`& .${exports.DatagridClasses.tableWrapper}`]: {},
    [`& .${exports.DatagridClasses.thead}`]: {},
    [`& .${exports.DatagridClasses.tbody}`]: {},
    [`& .${exports.DatagridClasses.headerRow}`]: {},
    [`& .${exports.DatagridClasses.headerCell}`]: {
        position: 'sticky',
        top: 0,
        zIndex: 2,
        backgroundColor: (theme.vars || theme).palette.background.paper,
        '&:first-of-type': {
            borderTopLeftRadius: theme.shape.borderRadius,
        },
        '&:last-child': {
            borderTopRightRadius: theme.shape.borderRadius,
        },
    },
    [`& .${exports.DatagridClasses.checkbox}`]: {},
    [`& .${exports.DatagridClasses.row}`]: {},
    [`& .${exports.DatagridClasses.clickableRow}`]: {
        cursor: 'pointer',
    },
    [`& .${exports.DatagridClasses.rowEven}`]: {},
    [`& .${exports.DatagridClasses.rowOdd}`]: {},
    [`& .${exports.DatagridClasses.rowCell}`]: {},
    [`& .${exports.DatagridClasses.expandHeader}`]: {
        padding: 0,
        width: theme.spacing(6),
    },
    [`& .${exports.DatagridClasses.expandIconCell}`]: {
        width: theme.spacing(6),
    },
    [`& .${exports.DatagridClasses.expandIcon}`]: {
        padding: theme.spacing(1),
        transform: 'rotate(-90deg)',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    [`& .${exports.DatagridClasses.expandIcon}.${exports.DatagridClasses.expanded}`]: {
        transform: 'rotate(0deg)',
    },
    [`& .${exports.DatagridClasses.expandedPanel}`]: {},
}));
