// ra-ui-materialui/dist/list/datagrid/useDatagridStyles.js
import { styled } from '@mui/material';
export const DatagridPrefix = 'RaDatagrid';
export const DatagridClasses = {
    root: `${DatagridPrefix}-root`,
    table: `${DatagridPrefix}-table`,
    tableWrapper: `${DatagridPrefix}-tableWrapper`,
    thead: `${DatagridPrefix}-thead`,
    tbody: `${DatagridPrefix}-tbody`,
    headerRow: `${DatagridPrefix}-headerRow`,
    headerCell: `${DatagridPrefix}-headerCell`,
    checkbox: `${DatagridPrefix}-checkbox`,
    row: `${DatagridPrefix}-row`,
    clickableRow: `${DatagridPrefix}-clickableRow`,
    rowEven: `${DatagridPrefix}-rowEven`,
    rowOdd: `${DatagridPrefix}-rowOdd`,
    rowCell: `${DatagridPrefix}-rowCell`,
    selectable: `${DatagridPrefix}-selectable`,
    expandHeader: `${DatagridPrefix}-expandHeader`,
    expandIconCell: `${DatagridPrefix}-expandIconCell`,
    expandIcon: `${DatagridPrefix}-expandIcon`,
    expandable: `${DatagridPrefix}-expandable`,
    expanded: `${DatagridPrefix}-expanded`,
    expandedPanel: `${DatagridPrefix}-expandedPanel`,
};
export const DatagridRoot = styled('div', {
    name: DatagridPrefix,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${DatagridClasses.table}`]: {
        tableLayout: 'auto',
    },
    [`& .${DatagridClasses.tableWrapper}`]: {},
    [`& .${DatagridClasses.thead}`]: {},
    [`& .${DatagridClasses.tbody}`]: {},
    [`& .${DatagridClasses.headerRow}`]: {},
    [`& .${DatagridClasses.headerCell}`]: {
        position: 'sticky',
        top: 0,
        zIndex: 2,
        backgroundColor: (theme.vars || theme).palette.background.paper,
        '&:first-of-type': {
            borderTopLeftRadius: theme.shape.borderRadius,
        },
        '&:last-child': {
            borderTopRightRadius: theme.shape.borderRadius,
        },
    },
    [`& .${DatagridClasses.checkbox}`]: {},
    [`& .${DatagridClasses.row}`]: {},
    [`& .${DatagridClasses.clickableRow}`]: {
        cursor: 'pointer',
    },
    [`& .${DatagridClasses.rowEven}`]: {},
    [`& .${DatagridClasses.rowOdd}`]: {},
    [`& .${DatagridClasses.rowCell}`]: {},
    [`& .${DatagridClasses.expandHeader}`]: {
        padding: 0,
        width: theme.spacing(6),
    },
    [`& .${DatagridClasses.expandIconCell}`]: {
        width: theme.spacing(6),
    },
    [`& .${DatagridClasses.expandIcon}`]: {
        padding: theme.spacing(1),
        transform: 'rotate(-90deg)',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    [`& .${DatagridClasses.expandIcon}.${DatagridClasses.expanded}`]: {
        transform: 'rotate(0deg)',
    },
    [`& .${DatagridClasses.expandedPanel}`]: {},
}));
