// ra-ui-materialui/dist/layout/AccessDenied.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessDeniedClasses = exports.AccessDenied = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const Lock_1 = __importDefault(require("@mui/icons-material/Lock"));
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const AccessDenied = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, icon = DEFAULT_ICON, textPrimary = 'ra.page.access_denied', textSecondary = 'ra.message.access_denied', ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    return (React.createElement(Root, { className: (0, clsx_1.default)(exports.AccessDeniedClasses.root, className), ...rest },
        React.createElement("div", { className: exports.AccessDeniedClasses.message },
            icon,
            React.createElement(material_1.Typography, { variant: "h5", sx: {
                    mt: 3,
                    color: 'text.secondary',
                } }, translate(textPrimary, { _: textPrimary })),
            React.createElement(material_1.Typography, { variant: "body2" }, translate(textSecondary, {
                _: textSecondary,
            })))));
};
exports.AccessDenied = AccessDenied;
const PREFIX = 'RaAccessDenied';
exports.AccessDeniedClasses = {
    root: `${PREFIX}-root`,
    icon: `${PREFIX}-icon`,
    message: `${PREFIX}-message`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
    [`& .${exports.AccessDeniedClasses.message}`]: {
        textAlign: 'center',
        paddingTop: '1em',
        paddingBottom: '1em',
        opacity: 0.5,
    },
    [`& .${exports.AccessDeniedClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
});
const DEFAULT_ICON = React.createElement(Lock_1.default, { className: exports.AccessDeniedClasses.icon });
