// ra-ui-materialui/dist/layout/AccessDenied.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import LockIcon from '@mui/icons-material/Lock';
import { Typography } from '@mui/material';
import clsx from 'clsx';
import { useTranslate } from 'ra-core';
export const AccessDenied = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, icon = DEFAULT_ICON, textPrimary = 'ra.page.access_denied', textSecondary = 'ra.message.access_denied', ...rest } = props;
    const translate = useTranslate();
    return (React.createElement(Root, { className: clsx(AccessDeniedClasses.root, className), ...rest },
        React.createElement("div", { className: AccessDeniedClasses.message },
            icon,
            React.createElement(Typography, { variant: "h5", sx: {
                    mt: 3,
                    color: 'text.secondary',
                } }, translate(textPrimary, { _: textPrimary })),
            React.createElement(Typography, { variant: "body2" }, translate(textSecondary, {
                _: textSecondary,
            })))));
};
const PREFIX = 'RaAccessDenied';
export const AccessDeniedClasses = {
    root: `${PREFIX}-root`,
    icon: `${PREFIX}-icon`,
    message: `${PREFIX}-message`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
    [`& .${AccessDeniedClasses.message}`]: {
        textAlign: 'center',
        paddingTop: '1em',
        paddingBottom: '1em',
        opacity: 0.5,
    },
    [`& .${AccessDeniedClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
});
const DEFAULT_ICON = React.createElement(LockIcon, { className: AccessDeniedClasses.icon });
