// ra-ui-materialui/dist/layout/AppBar.js
import * as React from 'react';
import { Children, memo } from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { AppBar as MuiAppBar, Toolbar, useMediaQuery, } from '@mui/material';
import { useLocales } from 'ra-core';
import { SidebarToggleButton } from "./SidebarToggleButton.js";
import { LoadingIndicator } from "./LoadingIndicator.js";
import { UserMenu } from "./UserMenu.js";
import { HideOnScroll } from "./HideOnScroll.js";
import { TitlePortal } from "./TitlePortal.js";
import { LocalesMenuButton } from "../button/index.js";
import { useThemesContext } from "../theme/useThemesContext.js";
import { ToggleThemeButton } from "../button/ToggleThemeButton.js";
/**
 * The AppBar component renders a custom MuiAppBar.
 *
 * @param {Object} props
 * @param {ReactNode} props.children React node/s to be rendered as children of the AppBar
 * @param {string} props.className CSS class applied to the MuiAppBar component
 * @param {string} props.color The color of the AppBar
 * @param {Element | boolean} props.userMenu A custom user menu component for the AppBar. <UserMenu/> component by default. Pass false to disable.
 *
 * @example // add a custom button to the AppBar
 *
 * const MyAppBar = () => (
 *   <AppBar>
 *     <TitlePortal />
 *     <MyCustomButton />
 *   </AppBar>
 * );
 *
 * @example // without a user menu
 *
 * const MyAppBar = () => <AppBar userMenu={false} />;
 */
export const AppBar = memo(inProps => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { alwaysOn, children, className, color = 'secondary', toolbar = defaultToolbarElement, userMenu = DefaultUserMenu, container: Container = alwaysOn ? 'div' : HideOnScroll, ...rest } = props;
    const isXSmall = useMediaQuery(theme => theme.breakpoints.down('sm'));
    return (React.createElement(Container, { className: className },
        React.createElement(StyledAppBar, { className: AppBarClasses.appBar, color: color, ...rest },
            React.createElement(Toolbar, { disableGutters: true, variant: isXSmall ? 'regular' : 'dense', className: AppBarClasses.toolbar },
                React.createElement(SidebarToggleButton, { className: AppBarClasses.menuButton }),
                Children.count(children) === 0 ? (React.createElement(TitlePortal, { className: AppBarClasses.title })) : (children),
                toolbar,
                typeof userMenu === 'boolean' ? (userMenu === true ? (React.createElement(UserMenu, null)) : null) : (userMenu)))));
});
const DefaultToolbar = () => {
    const locales = useLocales();
    const { darkTheme } = useThemesContext();
    return (React.createElement(React.Fragment, null,
        locales && locales.length > 1 ? React.createElement(LocalesMenuButton, null) : null,
        darkTheme && React.createElement(ToggleThemeButton, null),
        React.createElement(LoadingIndicator, null)));
};
const defaultToolbarElement = React.createElement(DefaultToolbar, null);
const DefaultUserMenu = React.createElement(UserMenu, null);
const PREFIX = 'RaAppBar';
export const AppBarClasses = {
    appBar: `${PREFIX}-appBar`,
    toolbar: `${PREFIX}-toolbar`,
    menuButton: `${PREFIX}-menuButton`,
    menuButtonIconClosed: `${PREFIX}-menuButtonIconClosed`,
    menuButtonIconOpen: `${PREFIX}-menuButtonIconOpen`,
    title: `${PREFIX}-title`,
};
const StyledAppBar = styled(MuiAppBar, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${AppBarClasses.toolbar}`]: {
        padding: `0 ${theme.spacing(1)}`,
        [theme.breakpoints.down('md')]: {
            minHeight: theme.spacing(6),
        },
    },
    [`& .${AppBarClasses.menuButton}`]: {
        marginRight: '0.2em',
    },
    [`& .${AppBarClasses.title}`]: {},
}));
