// ra-ui-materialui/dist/layout/AppBar.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppBarClasses = exports.AppBar = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const SidebarToggleButton_1 = require("./SidebarToggleButton.cjs");
const LoadingIndicator_1 = require("./LoadingIndicator.cjs");
const UserMenu_1 = require("./UserMenu.cjs");
const HideOnScroll_1 = require("./HideOnScroll.cjs");
const TitlePortal_1 = require("./TitlePortal.cjs");
const button_1 = require("../button/index.cjs");
const useThemesContext_1 = require("../theme/useThemesContext.cjs");
const ToggleThemeButton_1 = require("../button/ToggleThemeButton.cjs");
/**
 * The AppBar component renders a custom MuiAppBar.
 *
 * @param {Object} props
 * @param {ReactNode} props.children React node/s to be rendered as children of the AppBar
 * @param {string} props.className CSS class applied to the MuiAppBar component
 * @param {string} props.color The color of the AppBar
 * @param {Element | boolean} props.userMenu A custom user menu component for the AppBar. <UserMenu/> component by default. Pass false to disable.
 *
 * @example // add a custom button to the AppBar
 *
 * const MyAppBar = () => (
 *   <AppBar>
 *     <TitlePortal />
 *     <MyCustomButton />
 *   </AppBar>
 * );
 *
 * @example // without a user menu
 *
 * const MyAppBar = () => <AppBar userMenu={false} />;
 */
exports.AppBar = (0, react_1.memo)(inProps => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { alwaysOn, children, className, color = 'secondary', toolbar = defaultToolbarElement, userMenu = DefaultUserMenu, container: Container = alwaysOn ? 'div' : HideOnScroll_1.HideOnScroll, ...rest } = props;
    const isXSmall = (0, material_1.useMediaQuery)(theme => theme.breakpoints.down('sm'));
    return (React.createElement(Container, { className: className },
        React.createElement(StyledAppBar, { className: exports.AppBarClasses.appBar, color: color, ...rest },
            React.createElement(material_1.Toolbar, { disableGutters: true, variant: isXSmall ? 'regular' : 'dense', className: exports.AppBarClasses.toolbar },
                React.createElement(SidebarToggleButton_1.SidebarToggleButton, { className: exports.AppBarClasses.menuButton }),
                react_1.Children.count(children) === 0 ? (React.createElement(TitlePortal_1.TitlePortal, { className: exports.AppBarClasses.title })) : (children),
                toolbar,
                typeof userMenu === 'boolean' ? (userMenu === true ? (React.createElement(UserMenu_1.UserMenu, null)) : null) : (userMenu)))));
});
const DefaultToolbar = () => {
    const locales = (0, ra_core_1.useLocales)();
    const { darkTheme } = (0, useThemesContext_1.useThemesContext)();
    return (React.createElement(React.Fragment, null,
        locales && locales.length > 1 ? React.createElement(button_1.LocalesMenuButton, null) : null,
        darkTheme && React.createElement(ToggleThemeButton_1.ToggleThemeButton, null),
        React.createElement(LoadingIndicator_1.LoadingIndicator, null)));
};
const defaultToolbarElement = React.createElement(DefaultToolbar, null);
const DefaultUserMenu = React.createElement(UserMenu_1.UserMenu, null);
const PREFIX = 'RaAppBar';
exports.AppBarClasses = {
    appBar: `${PREFIX}-appBar`,
    toolbar: `${PREFIX}-toolbar`,
    menuButton: `${PREFIX}-menuButton`,
    menuButtonIconClosed: `${PREFIX}-menuButtonIconClosed`,
    menuButtonIconOpen: `${PREFIX}-menuButtonIconOpen`,
    title: `${PREFIX}-title`,
};
const StyledAppBar = (0, styles_1.styled)(material_1.AppBar, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.AppBarClasses.toolbar}`]: {
        padding: `0 ${theme.spacing(1)}`,
        [theme.breakpoints.down('md')]: {
            minHeight: theme.spacing(6),
        },
    },
    [`& .${exports.AppBarClasses.menuButton}`]: {
        marginRight: '0.2em',
    },
    [`& .${exports.AppBarClasses.title}`]: {},
}));
