// ra-ui-materialui/dist/layout/AuthenticationError.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Typography } from '@mui/material';
import WarningAmber from '@mui/icons-material/WarningAmber';
import clsx from 'clsx';
import { useDefaultTitle, useTranslate } from 'ra-core';
import { Title } from "./Title.js";
export const AuthenticationError = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, icon = DEFAULT_ICON, textPrimary = 'ra.page.authentication_error', textSecondary = 'ra.message.authentication_error', ...rest } = props;
    const translate = useTranslate();
    const title = useDefaultTitle();
    return (React.createElement(Root, { className: clsx(AuthenticationErrorClasses.root, className), ...rest },
        React.createElement(Title, { defaultTitle: title }),
        React.createElement("div", { className: AuthenticationErrorClasses.message },
            icon,
            React.createElement(Typography, { variant: "h5", sx: {
                    mt: 3,
                    color: 'text.secondary',
                } }, translate(textPrimary, { _: textPrimary })),
            React.createElement(Typography, { variant: "body2" }, translate(textSecondary, { _: textSecondary })))));
};
const PREFIX = 'RaAuthenticationError';
export const AuthenticationErrorClasses = {
    root: `${PREFIX}-root`,
    icon: `${PREFIX}-icon`,
    message: `${PREFIX}-message`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
    [`& .${AuthenticationErrorClasses.message}`]: {
        textAlign: 'center',
        paddingTop: '1em',
        paddingBottom: '1em',
        opacity: 0.5,
    },
    [`& .${AuthenticationErrorClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
});
const DEFAULT_ICON = (React.createElement(WarningAmber, { className: AuthenticationErrorClasses.icon }));
