// ra-ui-materialui/dist/layout/CardContentInner.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import CardContent from '@mui/material/CardContent';
/**
 * Overrides Material UI CardContent to allow inner content
 *
 * When using several CardContent inside the same Card, the top and bottom
 * padding double the spacing between each CardContent, leading to too much
 * wasted space. Use this component as a CardContent alternative.
 */
export const CardContentInner = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, children } = props;
    return React.createElement(Root, { className: className }, children);
};
const PREFIX = 'RaCardContentInner';
export const CardContentInnerClasses = {
    root: `${PREFIX}-root`,
};
const Root = styled(CardContent, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    paddingTop: 0,
    paddingBottom: 0,
    '&:first-of-type': {
        paddingTop: 16,
    },
    '&:last-child': {
        paddingBottom: 16,
        [theme.breakpoints.only('xs')]: {
            paddingBottom: 70,
        },
    },
}));
