// ra-ui-materialui/dist/layout/Confirm.js
import * as React from 'react';
import { useCallback } from 'react';
import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle, } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
import ActionCheck from '@mui/icons-material/CheckCircle';
import AlertError from '@mui/icons-material/ErrorOutline';
import clsx from 'clsx';
import { useTranslate } from 'ra-core';
/**
 * Confirmation dialog
 *
 * @example
 * <Confirm
 *     isOpen={true}
 *     title="Delete Item"
 *     content="Are you sure you want to delete this item?"
 *     confirm="Yes"
 *     confirmColor="primary"
 *     ConfirmIcon=ActionCheck
 *     CancelIcon=AlertError
 *     cancel="Cancel"
 *     onConfirm={() => { // do something }}
 *     onClose={() => { // do something }}
 * />
 */
export const Confirm = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, isOpen = false, loading, title, content, cancel = 'ra.action.cancel', confirm = 'ra.action.confirm', confirmColor = 'primary', ConfirmIcon = ActionCheck, CancelIcon = AlertError, onClose, onConfirm, translateOptions = {}, titleTranslateOptions = translateOptions, contentTranslateOptions = translateOptions, ...rest } = props;
    const translate = useTranslate();
    const handleConfirm = useCallback(e => {
        e.stopPropagation();
        onConfirm(e);
    }, [onConfirm]);
    const handleClick = useCallback(e => {
        e.stopPropagation();
    }, []);
    return (React.createElement(StyledDialog, { className: className, open: isOpen, onClose: onClose, onClick: handleClick, "aria-labelledby": "alert-dialog-title", ...rest },
        React.createElement(DialogTitle, { id: "alert-dialog-title" }, typeof title === 'string'
            ? translate(title, { _: title, ...titleTranslateOptions })
            : title),
        React.createElement(DialogContent, null, typeof content === 'string' ? (React.createElement(DialogContentText, null, translate(content, {
            _: content,
            ...contentTranslateOptions,
        }))) : (content)),
        React.createElement(DialogActions, null,
            React.createElement(Button, { disabled: loading, onClick: onClose, startIcon: React.createElement(CancelIcon, null) }, translate(cancel, { _: cancel })),
            React.createElement(Button, { disabled: loading, onClick: handleConfirm, className: clsx('ra-confirm', {
                    [ConfirmClasses.confirmWarning]: confirmColor === 'warning',
                    [ConfirmClasses.confirmPrimary]: confirmColor === 'primary',
                }), autoFocus: true, startIcon: React.createElement(ConfirmIcon, null) }, translate(confirm, { _: confirm })))));
};
const PREFIX = 'RaConfirm';
export const ConfirmClasses = {
    confirmPrimary: `${PREFIX}-confirmPrimary`,
    confirmWarning: `${PREFIX}-confirmWarning`,
};
const StyledDialog = styled(Dialog, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${ConfirmClasses.confirmPrimary}`]: {
        color: (theme.vars || theme).palette.primary.main,
    },
    [`& .${ConfirmClasses.confirmWarning}`]: {
        color: (theme.vars || theme).palette.error.main,
        '&:hover': {
            backgroundColor: `color-mix(in srgb, ${(theme.vars || theme).palette.error.main}, transparent 88%)`,
            // Reset on mouse devices
            '@media (hover: none)': {
                backgroundColor: 'transparent',
            },
        },
    },
}));
