// ra-ui-materialui/dist/layout/Confirm.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfirmClasses = exports.Confirm = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const CheckCircle_1 = __importDefault(require("@mui/icons-material/CheckCircle"));
const ErrorOutline_1 = __importDefault(require("@mui/icons-material/ErrorOutline"));
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
/**
 * Confirmation dialog
 *
 * @example
 * <Confirm
 *     isOpen={true}
 *     title="Delete Item"
 *     content="Are you sure you want to delete this item?"
 *     confirm="Yes"
 *     confirmColor="primary"
 *     ConfirmIcon=ActionCheck
 *     CancelIcon=AlertError
 *     cancel="Cancel"
 *     onConfirm={() => { // do something }}
 *     onClose={() => { // do something }}
 * />
 */
const Confirm = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, isOpen = false, loading, title, content, cancel = 'ra.action.cancel', confirm = 'ra.action.confirm', confirmColor = 'primary', ConfirmIcon = CheckCircle_1.default, CancelIcon = ErrorOutline_1.default, onClose, onConfirm, translateOptions = {}, titleTranslateOptions = translateOptions, contentTranslateOptions = translateOptions, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const handleConfirm = (0, react_1.useCallback)(e => {
        e.stopPropagation();
        onConfirm(e);
    }, [onConfirm]);
    const handleClick = (0, react_1.useCallback)(e => {
        e.stopPropagation();
    }, []);
    return (React.createElement(StyledDialog, { className: className, open: isOpen, onClose: onClose, onClick: handleClick, "aria-labelledby": "alert-dialog-title", ...rest },
        React.createElement(material_1.DialogTitle, { id: "alert-dialog-title" }, typeof title === 'string'
            ? translate(title, { _: title, ...titleTranslateOptions })
            : title),
        React.createElement(material_1.DialogContent, null, typeof content === 'string' ? (React.createElement(material_1.DialogContentText, null, translate(content, {
            _: content,
            ...contentTranslateOptions,
        }))) : (content)),
        React.createElement(material_1.DialogActions, null,
            React.createElement(material_1.Button, { disabled: loading, onClick: onClose, startIcon: React.createElement(CancelIcon, null) }, translate(cancel, { _: cancel })),
            React.createElement(material_1.Button, { disabled: loading, onClick: handleConfirm, className: (0, clsx_1.default)('ra-confirm', {
                    [exports.ConfirmClasses.confirmWarning]: confirmColor === 'warning',
                    [exports.ConfirmClasses.confirmPrimary]: confirmColor === 'primary',
                }), autoFocus: true, startIcon: React.createElement(ConfirmIcon, null) }, translate(confirm, { _: confirm })))));
};
exports.Confirm = Confirm;
const PREFIX = 'RaConfirm';
exports.ConfirmClasses = {
    confirmPrimary: `${PREFIX}-confirmPrimary`,
    confirmWarning: `${PREFIX}-confirmWarning`,
};
const StyledDialog = (0, styles_1.styled)(material_1.Dialog, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.ConfirmClasses.confirmPrimary}`]: {
        color: (theme.vars || theme).palette.primary.main,
    },
    [`& .${exports.ConfirmClasses.confirmWarning}`]: {
        color: (theme.vars || theme).palette.error.main,
        '&:hover': {
            backgroundColor: `color-mix(in srgb, ${(theme.vars || theme).palette.error.main}, transparent 88%)`,
            // Reset on mouse devices
            '@media (hover: none)': {
                backgroundColor: 'transparent',
            },
        },
    },
}));
