// ra-ui-materialui/dist/layout/Error.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorClasses = exports.Error = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const Report_1 = __importDefault(require("@mui/icons-material/Report"));
const ExpandMore_1 = __importDefault(require("@mui/icons-material/ExpandMore"));
const History_1 = __importDefault(require("@mui/icons-material/History"));
const ra_core_1 = require("ra-core");
const Title_1 = require("./Title.cjs");
const Error = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { error, errorComponent: ErrorComponent, errorInfo, resetErrorBoundary, className, ...rest } = props;
    const translate = (0, ra_core_1.useTranslate)();
    const title = (0, ra_core_1.useDefaultTitle)();
    (0, ra_core_1.useResetErrorBoundaryOnLocationChange)(resetErrorBoundary);
    if (ErrorComponent) {
        return (React.createElement(ErrorComponent, { error: error, errorInfo: errorInfo, title: title }));
    }
    return (React.createElement(react_1.Fragment, null,
        title && React.createElement(Title_1.Title, { title: title }),
        React.createElement(Root, { className: className, ...rest },
            React.createElement("h1", { className: exports.ErrorClasses.title, role: "alert" },
                React.createElement(Report_1.default, { className: exports.ErrorClasses.icon }),
                translate('ra.page.error')),
            React.createElement("div", null, translate('ra.message.error')),
            process.env.NODE_ENV !== 'production' && (React.createElement(React.Fragment, null,
                React.createElement(material_1.Accordion, { className: exports.ErrorClasses.panel },
                    React.createElement(material_1.AccordionSummary, { expandIcon: React.createElement(ExpandMore_1.default, null), className: exports.ErrorClasses.panelSummary }, translate(error.message, {
                        _: error.message,
                    })),
                    React.createElement(material_1.AccordionDetails, { className: exports.ErrorClasses.panelDetails },
                        React.createElement("p", null, translate(error.message, {
                            _: error.message,
                        })),
                        React.createElement("p", null, errorInfo?.componentStack))),
                React.createElement("div", { className: exports.ErrorClasses.advice },
                    React.createElement(material_1.Typography, { align: "center" }, "Need help with this error? Try the following:"),
                    React.createElement(material_1.Typography, { component: "div" },
                        React.createElement("ul", null,
                            React.createElement("li", null,
                                "Check the",
                                ' ',
                                React.createElement("a", { href: "https://marmelab.com/react-admin/documentation.html" }, "react-admin documentation")),
                            React.createElement("li", null,
                                "Search on",
                                ' ',
                                React.createElement("a", { href: "https://stackoverflow.com/questions/tagged/react-admin" }, "StackOverflow"),
                                ' ',
                                "for community answers"),
                            React.createElement("li", null,
                                "Get help from the core team via",
                                ' ',
                                React.createElement("a", { href: "https://react-admin-ee.marmelab.com/#fromsww" }, "react-admin Enterprise Edition"))))))),
            React.createElement("div", { className: exports.ErrorClasses.toolbar },
                React.createElement(material_1.Button, { variant: "contained", startIcon: React.createElement(History_1.default, null), onClick: goBack }, translate('ra.action.back'))))));
};
exports.Error = Error;
const PREFIX = 'RaError';
exports.ErrorClasses = {
    container: `${PREFIX}-container`,
    title: `${PREFIX}-title`,
    icon: `${PREFIX}-icon`,
    panel: `${PREFIX}-panel`,
    panelSummary: `${PREFIX}-panelSummary`,
    panelDetails: `${PREFIX}-panelDetails`,
    toolbar: `${PREFIX}-toolbar`,
    advice: `${PREFIX}-advice`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    [theme.breakpoints.down('md')]: {
        padding: '1em',
    },
    fontFamily: 'Roboto, sans-serif',
    opacity: 0.5,
    [`& .${exports.ErrorClasses.title}`]: {
        display: 'flex',
        alignItems: 'center',
    },
    [`& .${exports.ErrorClasses.icon}`]: {
        width: '2em',
        height: '2em',
        marginRight: '0.5em',
    },
    [`& .${exports.ErrorClasses.panel}`]: {
        marginTop: '1em',
        maxWidth: '60em',
    },
    [`& .${exports.ErrorClasses.panelSummary}`]: {
        userSelect: 'all',
    },
    [`& .${exports.ErrorClasses.panelDetails}`]: {
        whiteSpace: 'pre-wrap',
    },
    [`& .${exports.ErrorClasses.toolbar}`]: {
        marginTop: '2em',
    },
    [`& .${exports.ErrorClasses.advice}`]: {
        marginTop: '2em',
    },
}));
function goBack() {
    window.history.go(-1);
}
