// ra-ui-materialui/dist/layout/Layout.js
import React, { Suspense, useState, } from 'react';
import { ErrorBoundary } from 'react-error-boundary';
import clsx from 'clsx';
import { styled, useThemeProps, } from '@mui/material/styles';
import { AppBar as DefaultAppBar } from "./AppBar.js";
import { Sidebar as DefaultSidebar } from "./Sidebar.js";
import { Menu as DefaultMenu } from "./Menu.js";
import { Error } from "./Error.js";
import { SkipNavigationButton } from "../button/index.js";
import { Inspector } from "../preferences/index.js";
import { Loading } from "./Loading.js";
export const Layout = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { appBar: AppBar = DefaultAppBar, appBarAlwaysOn, children, className, error: errorComponent, menu: Menu = DefaultMenu, sidebar: Sidebar = DefaultSidebar, ...rest } = props;
    const [errorInfo, setErrorInfo] = useState(undefined);
    const handleError = (error, info) => {
        setErrorInfo(info);
    };
    return (React.createElement(Core, { className: clsx('layout', className), ...rest },
        React.createElement(SkipNavigationButton, null),
        React.createElement("div", { className: LayoutClasses.appFrame },
            React.createElement(AppBar, { alwaysOn: appBarAlwaysOn }),
            React.createElement("main", { className: LayoutClasses.contentWithSidebar },
                React.createElement(Sidebar, { appBarAlwaysOn: appBarAlwaysOn },
                    React.createElement(Menu, null)),
                React.createElement("div", { id: "main-content", className: LayoutClasses.content },
                    React.createElement(ErrorBoundary, { onError: handleError, fallbackRender: ({ error, resetErrorBoundary }) => (React.createElement(Error, { error: error, errorComponent: errorComponent, errorInfo: errorInfo, resetErrorBoundary: resetErrorBoundary })) },
                        React.createElement(Suspense, { fallback: React.createElement(Loading, null) }, children)))),
            React.createElement(Inspector, null))));
};
const PREFIX = 'RaLayout';
export const LayoutClasses = {
    appFrame: `${PREFIX}-appFrame`,
    contentWithSidebar: `${PREFIX}-contentWithSidebar`,
    content: `${PREFIX}-content`,
};
const Core = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    zIndex: 1,
    minHeight: '100vh',
    backgroundColor: (theme.vars || theme).palette.background.default,
    position: 'relative',
    minWidth: 'fit-content',
    width: '100%',
    [`& .${LayoutClasses.appFrame}`]: {
        display: 'flex',
        flexDirection: 'column',
        flexGrow: 1,
        marginTop: theme.spacing(6),
        [theme.breakpoints.down('sm')]: {
            marginTop: theme.spacing(7),
        },
    },
    [`& .${LayoutClasses.contentWithSidebar}`]: {
        display: 'flex',
        flexGrow: 1,
        transition: theme.transitions.create('margin', {
            easing: theme.transitions.easing.easeOut,
            duration: theme.transitions.duration.enteringScreen,
        }),
    },
    [`& .${LayoutClasses.content}`]: {
        backgroundColor: (theme.vars || theme).palette.background.default,
        zIndex: 2,
        display: 'flex',
        flexDirection: 'column',
        flexGrow: 1,
        flexBasis: 0,
        padding: 0,
        [theme.breakpoints.up('xs')]: {
            paddingRight: theme.spacing(1),
            paddingLeft: theme.spacing(1),
        },
    },
}));
