// ra-ui-materialui/dist/layout/Layout.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LayoutClasses = exports.Layout = void 0;
const react_1 = __importStar(require("react"));
const react_error_boundary_1 = require("react-error-boundary");
const clsx_1 = __importDefault(require("clsx"));
const styles_1 = require("@mui/material/styles");
const AppBar_1 = require("./AppBar.cjs");
const Sidebar_1 = require("./Sidebar.cjs");
const Menu_1 = require("./Menu.cjs");
const Error_1 = require("./Error.cjs");
const button_1 = require("../button/index.cjs");
const preferences_1 = require("../preferences/index.cjs");
const Loading_1 = require("./Loading.cjs");
const Layout = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { appBar: AppBar = AppBar_1.AppBar, appBarAlwaysOn, children, className, error: errorComponent, menu: Menu = Menu_1.Menu, sidebar: Sidebar = Sidebar_1.Sidebar, ...rest } = props;
    const [errorInfo, setErrorInfo] = (0, react_1.useState)(undefined);
    const handleError = (error, info) => {
        setErrorInfo(info);
    };
    return (react_1.default.createElement(Core, { className: (0, clsx_1.default)('layout', className), ...rest },
        react_1.default.createElement(button_1.SkipNavigationButton, null),
        react_1.default.createElement("div", { className: exports.LayoutClasses.appFrame },
            react_1.default.createElement(AppBar, { alwaysOn: appBarAlwaysOn }),
            react_1.default.createElement("main", { className: exports.LayoutClasses.contentWithSidebar },
                react_1.default.createElement(Sidebar, { appBarAlwaysOn: appBarAlwaysOn },
                    react_1.default.createElement(Menu, null)),
                react_1.default.createElement("div", { id: "main-content", className: exports.LayoutClasses.content },
                    react_1.default.createElement(react_error_boundary_1.ErrorBoundary, { onError: handleError, fallbackRender: ({ error, resetErrorBoundary }) => (react_1.default.createElement(Error_1.Error, { error: error, errorComponent: errorComponent, errorInfo: errorInfo, resetErrorBoundary: resetErrorBoundary })) },
                        react_1.default.createElement(react_1.Suspense, { fallback: react_1.default.createElement(Loading_1.Loading, null) }, children)))),
            react_1.default.createElement(preferences_1.Inspector, null))));
};
exports.Layout = Layout;
const PREFIX = 'RaLayout';
exports.LayoutClasses = {
    appFrame: `${PREFIX}-appFrame`,
    contentWithSidebar: `${PREFIX}-contentWithSidebar`,
    content: `${PREFIX}-content`,
};
const Core = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    zIndex: 1,
    minHeight: '100vh',
    backgroundColor: (theme.vars || theme).palette.background.default,
    position: 'relative',
    minWidth: 'fit-content',
    width: '100%',
    [`& .${exports.LayoutClasses.appFrame}`]: {
        display: 'flex',
        flexDirection: 'column',
        flexGrow: 1,
        marginTop: theme.spacing(6),
        [theme.breakpoints.down('sm')]: {
            marginTop: theme.spacing(7),
        },
    },
    [`& .${exports.LayoutClasses.contentWithSidebar}`]: {
        display: 'flex',
        flexGrow: 1,
        transition: theme.transitions.create('margin', {
            easing: theme.transitions.easing.easeOut,
            duration: theme.transitions.duration.enteringScreen,
        }),
    },
    [`& .${exports.LayoutClasses.content}`]: {
        backgroundColor: (theme.vars || theme).palette.background.default,
        zIndex: 2,
        display: 'flex',
        flexDirection: 'column',
        flexGrow: 1,
        flexBasis: 0,
        padding: 0,
        [theme.breakpoints.up('xs')]: {
            paddingRight: theme.spacing(1),
            paddingLeft: theme.spacing(1),
        },
    },
}));
