// ra-ui-materialui/dist/layout/Loading.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import { Typography } from '@mui/material';
import CircularProgress from '@mui/material/CircularProgress';
import { useTimeout, useTranslate } from 'ra-core';
export const Loading = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, loadingPrimary = 'ra.page.loading', loadingSecondary = 'ra.message.loading', timeout = 1000, ...rest } = props;
    const oneSecondHasPassed = useTimeout(timeout);
    const translate = useTranslate();
    return oneSecondHasPassed ? (React.createElement(Root, { className: className, ...rest },
        React.createElement("div", { className: LoadingClasses.message },
            React.createElement(CircularProgress, { className: LoadingClasses.icon }),
            React.createElement(Typography, { variant: "h5", sx: {
                    mt: 3,
                    color: 'text.secondary',
                } }, translate(loadingPrimary, { _: loadingPrimary })),
            React.createElement(Typography, { variant: "body2" }, translate(loadingSecondary, { _: loadingSecondary }))))) : null;
};
const PREFIX = 'RaLoading';
export const LoadingClasses = {
    root: `${PREFIX}-root`,
    icon: `${PREFIX}-icon`,
    message: `${PREFIX}-message`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
    [`& .${LoadingClasses.message}`]: {
        textAlign: 'center',
        fontFamily: 'Roboto, sans-serif',
        color: (theme.vars || theme).palette.text.disabled,
        paddingTop: '1em',
        paddingBottom: '1em',
    },
    [`& .${LoadingClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
}));
