// ra-ui-materialui/dist/layout/LoadingIndicator.js
import * as React from 'react';
import { styled, useTheme, useThemeProps, } from '@mui/material/styles';
import clsx from 'clsx';
import CircularProgress from '@mui/material/CircularProgress';
import { useLoading } from 'ra-core';
import { RefreshIconButton } from "../button/index.js";
export const LoadingIndicator = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const { className, onClick, sx, ...rest } = props;
    const loading = useLoading();
    const theme = useTheme();
    return (React.createElement(Root, { className: className, sx: sx },
        React.createElement(RefreshIconButton, { className: `${LoadingIndicatorClasses.loadedIcon} ${loading && LoadingIndicatorClasses.loadedLoading}`, onClick: onClick }),
        loading && (React.createElement(CircularProgress, { className: clsx('app-loader', LoadingIndicatorClasses.loader), color: "inherit", size: theme.spacing(2), thickness: 6, ...rest }))));
};
const PREFIX = 'RaLoadingIndicator';
export const LoadingIndicatorClasses = {
    loader: `${PREFIX}-loader`,
    loadedLoading: `${PREFIX}-loadedLoading`,
    loadedIcon: `${PREFIX}-loadedIcon`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (_, styles) => styles.root,
})({
    position: 'relative',
    [`& .${LoadingIndicatorClasses.loadedIcon}`]: {
        [`&.${LoadingIndicatorClasses.loadedLoading}`]: {
            opacity: 0,
        },
    },
    [`& .${LoadingIndicatorClasses.loader}`]: {
        position: 'absolute',
        top: '30%',
        left: '30%',
    },
});
