// ra-ui-materialui/dist/layout/Loading.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadingClasses = exports.Loading = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const CircularProgress_1 = __importDefault(require("@mui/material/CircularProgress"));
const ra_core_1 = require("ra-core");
const Loading = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, loadingPrimary = 'ra.page.loading', loadingSecondary = 'ra.message.loading', timeout = 1000, ...rest } = props;
    const oneSecondHasPassed = (0, ra_core_1.useTimeout)(timeout);
    const translate = (0, ra_core_1.useTranslate)();
    return oneSecondHasPassed ? (React.createElement(Root, { className: className, ...rest },
        React.createElement("div", { className: exports.LoadingClasses.message },
            React.createElement(CircularProgress_1.default, { className: exports.LoadingClasses.icon }),
            React.createElement(material_1.Typography, { variant: "h5", sx: {
                    mt: 3,
                    color: 'text.secondary',
                } }, translate(loadingPrimary, { _: loadingPrimary })),
            React.createElement(material_1.Typography, { variant: "body2" }, translate(loadingSecondary, { _: loadingSecondary }))))) : null;
};
exports.Loading = Loading;
const PREFIX = 'RaLoading';
exports.LoadingClasses = {
    root: `${PREFIX}-root`,
    icon: `${PREFIX}-icon`,
    message: `${PREFIX}-message`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
    [`& .${exports.LoadingClasses.message}`]: {
        textAlign: 'center',
        fontFamily: 'Roboto, sans-serif',
        color: (theme.vars || theme).palette.text.disabled,
        paddingTop: '1em',
        paddingBottom: '1em',
    },
    [`& .${exports.LoadingClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
}));
