// ra-ui-materialui/dist/layout/LoadingIndicator.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadingIndicatorClasses = exports.LoadingIndicator = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const CircularProgress_1 = __importDefault(require("@mui/material/CircularProgress"));
const ra_core_1 = require("ra-core");
const button_1 = require("../button/index.cjs");
const LoadingIndicator = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, onClick, sx, ...rest } = props;
    const loading = (0, ra_core_1.useLoading)();
    const theme = (0, styles_1.useTheme)();
    return (React.createElement(Root, { className: className, sx: sx },
        React.createElement(button_1.RefreshIconButton, { className: `${exports.LoadingIndicatorClasses.loadedIcon} ${loading && exports.LoadingIndicatorClasses.loadedLoading}`, onClick: onClick }),
        loading && (React.createElement(CircularProgress_1.default, { className: (0, clsx_1.default)('app-loader', exports.LoadingIndicatorClasses.loader), color: "inherit", size: theme.spacing(2), thickness: 6, ...rest }))));
};
exports.LoadingIndicator = LoadingIndicator;
const PREFIX = 'RaLoadingIndicator';
exports.LoadingIndicatorClasses = {
    loader: `${PREFIX}-loader`,
    loadedLoading: `${PREFIX}-loadedLoading`,
    loadedIcon: `${PREFIX}-loadedIcon`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (_, styles) => styles.root,
})({
    position: 'relative',
    [`& .${exports.LoadingIndicatorClasses.loadedIcon}`]: {
        [`&.${exports.LoadingIndicatorClasses.loadedLoading}`]: {
            opacity: 0,
        },
    },
    [`& .${exports.LoadingIndicatorClasses.loader}`]: {
        position: 'absolute',
        top: '30%',
        left: '30%',
    },
});
