// ra-ui-materialui/dist/layout/MenuItemLink.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuItemLinkClasses = exports.MenuItemLink = void 0;
const react_1 = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const ra_core_1 = require("ra-core");
const react_hotkeys_hook_1 = require("react-hotkeys-hook");
const useSidebarState_1 = require("./useSidebarState.cjs");
const KeyboardShortcut_1 = require("../KeyboardShortcut.cjs");
/**
 * Displays a menu item with a label and an icon - or only the icon with a tooltip when the sidebar is minimized.
 * It also handles the automatic closing of the menu on tap on mobile.
 *
 * @typedef {Object} Props the props you can use
 * @prop {string|Location} to The menu item's target. It is passed to a React Router NavLink component.
 * @prop {string|ReactNode} primaryText The menu content, displayed when the menu isn't minimized. |
 * @prop {ReactNode} leftIcon The menu icon
 *
 * Additional props are passed down to the underling Material UI <MenuItem> component
 * @see https://material-ui.com/api/menu-item/#menuitem-api
 *
 * @example // You can create a custom menu component using the <DashboardMenuItem> and <MenuItemLink> components:
 *
 * // in src/Menu.js
 * import * as React from 'react';
 * import { DashboardMenuItem, MenuItemLink } from 'react-admin';
 * import BookIcon from '@mui/icons-material/Book';
 * import ChatBubbleIcon from '@mui/icons-material/ChatBubble';
 * import PeopleIcon from '@mui/icons-material/People';
 * import LabelIcon from '@mui/icons-material/Label';
 *
 * export const Menu = () => (
 *     <div>
 *         <DashboardMenuItem />
 *         <MenuItemLink to="/posts" primaryText="Posts" leftIcon={<BookIcon />}/>
 *         <MenuItemLink to="/comments" primaryText="Comments" leftIcon={<ChatBubbleIcon />}/>
 *         <MenuItemLink to="/users" primaryText="Users" leftIcon={<PeopleIcon />}/>
 *         <MenuItemLink to="/custom-route" primaryText="Miscellaneous" leftIcon={<LabelIcon />}/>
 *     </div>
 * );
 *
 * // to use this custom menu component, pass it to a custom Layout:
 * // in src/Layout.js
 * import { Layout } from 'react-admin';
 * import { Menu } from './Menu';
 *
 * export const Layout = ({ children }) => (
 *     <Layout menu={Menu}>
 *         {children}
 *     </Layout>
 * );
 *
 * // then, use this layout in the <Admin layout> prop:
 * // in src/App.js
 * import { Layout }  from './Layout';
 *
 * const App = () => (
 *     <Admin layout={Layout} dataProvider={simpleRestProvider('http://path.to.my.api')}>
 *         // ...
 *     </Admin>
 * );
 */
exports.MenuItemLink = (0, react_1.forwardRef)((inProps, ref) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, primaryText, leftIcon, onClick, sidebarIsOpen, tooltipProps, children, keyboardShortcut, keyboardShortcutRepresentation, ...rest } = props;
    const isSmall = (0, material_1.useMediaQuery)(theme => theme.breakpoints.down('md'));
    const translate = (0, ra_core_1.useTranslate)();
    const basename = (0, ra_core_1.useBasename)();
    const [open, setOpen] = (0, useSidebarState_1.useSidebarState)();
    const handleMenuTap = (0, react_1.useCallback)(e => {
        if (isSmall) {
            setOpen(false);
        }
        onClick && onClick(e);
    }, [setOpen, isSmall, onClick]);
    const to = (typeof props.to === 'string' ? props.to : props.to.pathname) || '';
    const match = (0, ra_core_1.useMatch)({ path: to, end: to === `${basename}/` });
    const itemRef = (0, react_1.useRef)(null);
    // Use a forked ref allows us to have a ref locally without losing the one passed by users
    const forkedRef = (0, material_1.useForkRef)(itemRef, ref);
    const handleShortcut = (0, ra_core_1.useEvent)(() => itemRef.current?.click());
    (0, react_hotkeys_hook_1.useHotkeys)(keyboardShortcut ?? [], handleShortcut, {
        enabled: keyboardShortcut != null,
    });
    const renderMenuItem = () => {
        return (react_1.default.createElement(StyledMenuItem, { className: (0, clsx_1.default)(className, exports.MenuItemLinkClasses.root, {
                [exports.MenuItemLinkClasses.active]: !!match,
            }), 
            // @ts-ignore
            component: LinkRef, ref: forkedRef, tabIndex: 0, ...rest, onClick: handleMenuTap },
            leftIcon && (react_1.default.createElement(material_1.ListItemIcon, { className: exports.MenuItemLinkClasses.icon }, leftIcon)),
            react_1.default.createElement(material_1.Typography, { variant: "inherit", noWrap: true, sx: { flexGrow: 1 } }, children
                ? children
                : typeof primaryText === 'string'
                    ? translate(primaryText, {
                        _: primaryText,
                    })
                    : primaryText),
            keyboardShortcut
                ? keyboardShortcutRepresentation ?? (react_1.default.createElement(KeyboardShortcut_1.KeyboardShortcut, { className: exports.MenuItemLinkClasses.shortcut, keyboardShortcut: keyboardShortcut }))
                : null));
    };
    if (open) {
        return renderMenuItem();
    }
    return (react_1.default.createElement(material_1.Tooltip, { title: typeof primaryText === 'string'
            ? translate(primaryText, { _: primaryText })
            : primaryText, placement: "right", ...tooltipProps }, renderMenuItem()));
});
const PREFIX = 'RaMenuItemLink';
exports.MenuItemLinkClasses = {
    root: `${PREFIX}-root`,
    active: `${PREFIX}-active`,
    icon: `${PREFIX}-icon`,
    shortcut: `${PREFIX}-shortcut`,
};
const StyledMenuItem = (0, styles_1.styled)(material_1.MenuItem, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    color: (theme.vars || theme).palette.text.secondary,
    [`& .${exports.MenuItemLinkClasses.icon}`]: {
        color: (theme.vars || theme).palette.text.secondary,
    },
    [`& .${exports.MenuItemLinkClasses.shortcut}`]: {
        color: (theme.vars || theme).palette.text.secondary,
        fontSize: theme.typography.body2.fontSize,
        opacity: 0,
        display: 'none',
        transition: 'opacity 0.3s',
    },
    [`&:hover .${exports.MenuItemLinkClasses.shortcut}`]: {
        opacity: 0.7,
        display: 'inline-flex',
    },
    [`&.${exports.MenuItemLinkClasses.active}`]: {
        color: (theme.vars || theme).palette.text.primary,
    },
    [`& .${exports.MenuItemLinkClasses.icon}`]: { minWidth: theme.spacing(5) },
    [`.MuiList-dense > & > .${exports.MenuItemLinkClasses.icon}`]: {
        minWidth: theme.spacing(3.5),
    },
}));
const LinkRef = (0, react_1.forwardRef)((props, ref) => (react_1.default.createElement(ra_core_1.LinkBase, { ref: ref, ...props })));
