// ra-ui-materialui/dist/layout/Menu.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuClasses = exports.Menu = void 0;
const React = __importStar(require("react"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const get_js_1 = __importDefault(require("lodash/get.js"));
const clsx_1 = __importDefault(require("clsx"));
const Sidebar_1 = require("./Sidebar.cjs");
const useSidebarState_1 = require("./useSidebarState.cjs");
const DashboardMenuItem_1 = require("./DashboardMenuItem.cjs");
const MenuItemLink_1 = require("./MenuItemLink.cjs");
const ResourceMenuItem_1 = require("./ResourceMenuItem.cjs");
const ResourceMenuItems_1 = require("./ResourceMenuItems.cjs");
const ra_core_1 = require("ra-core");
/**
 * Renders a menu with one menu item per resource by default. You can also set menu items by hand.
 *
 * @example
 * import * as React from 'react';
 * import { Menu } from 'react-admin';
 *
 * import BookIcon from '@mui/icons-material/Book';
 * import ChatBubbleIcon from '@mui/icons-material/ChatBubble';
 * import PeopleIcon from '@mui/icons-material/People';
 * import LabelIcon from '@mui/icons-material/Label';
 *
 * export const MyMenu = () => (
 *     <Menu>
 *         <Menu.DashboardItem />
 *         <Menu.Item to="/posts" primaryText="Posts" leftIcon={<BookIcon />}/>
 *         <Menu.Item to="/comments" primaryText="Comments" leftIcon={<ChatBubbleIcon />}/>
 *         <Menu.Item to="/users" primaryText="Users" leftIcon={<PeopleIcon />}/>
 *         <Menu.Item to="/custom-route" primaryText="Miscellaneous" leftIcon={<LabelIcon />}/>
 *     </Menu>
 * );
 */
const Menu = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { children, className, hasDashboard: hasDashboardProp, ...rest } = props;
    const hasDashboard = (0, ra_core_1.useHasDashboard)();
    const [open] = (0, useSidebarState_1.useSidebarState)();
    return (React.createElement(Root, { className: (0, clsx_1.default)({
            [exports.MenuClasses.open]: open,
            [exports.MenuClasses.closed]: !open,
        }, className), ...rest },
        hasDashboard && !children && React.createElement(DashboardMenuItem_1.DashboardMenuItem, null),
        children ?? React.createElement(ResourceMenuItems_1.ResourceMenuItems, null)));
};
exports.Menu = Menu;
// re-export MenuItem components for convenience
exports.Menu.Item = MenuItemLink_1.MenuItemLink;
exports.Menu.DashboardItem = DashboardMenuItem_1.DashboardMenuItem;
exports.Menu.ResourceItem = ResourceMenuItem_1.ResourceMenuItem;
exports.Menu.ResourceItems = ResourceMenuItems_1.ResourceMenuItems;
const PREFIX = 'RaMenu';
exports.MenuClasses = {
    open: `${PREFIX}-open`,
    closed: `${PREFIX}-closed`,
};
const Root = (0, styles_1.styled)(material_1.MenuList, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    [theme.breakpoints.only('xs')]: {
        marginTop: 0,
    },
    transition: theme.transitions.create('width', {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.leavingScreen,
    }),
    [`&.${exports.MenuClasses.open}`]: {
        width: (0, get_js_1.default)(theme, 'sidebar.width', Sidebar_1.DRAWER_WIDTH),
    },
    [`&.${exports.MenuClasses.closed}`]: {
        width: (0, get_js_1.default)(theme, 'sidebar.closedWidth', Sidebar_1.CLOSED_DRAWER_WIDTH),
    },
}));
