// ra-ui-materialui/dist/layout/NotFound.js
import * as React from 'react';
import { styled, useThemeProps, } from '@mui/material/styles';
import Button from '@mui/material/Button';
import HotTub from '@mui/icons-material/HotTub';
import History from '@mui/icons-material/History';
import { useAuthenticated, useDefaultTitle, useTranslate } from 'ra-core';
import { Title } from "./Title.js";
import { Loading } from "./Loading.js";
export const NotFound = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const translate = useTranslate();
    const { isPending } = useAuthenticated();
    const title = useDefaultTitle();
    if (isPending)
        return React.createElement(Loading, null);
    return (React.createElement(Root, { ...sanitizeRestProps(props) },
        React.createElement(Title, { defaultTitle: title }),
        React.createElement("div", { className: NotFoundClasses.message },
            React.createElement(HotTub, { className: NotFoundClasses.icon }),
            React.createElement("h1", null, translate('ra.page.not_found')),
            React.createElement("div", null,
                translate('ra.message.not_found'),
                ".")),
        React.createElement("div", { className: NotFoundClasses.toolbar },
            React.createElement(Button, { variant: "contained", startIcon: React.createElement(History, null), onClick: goBack }, translate('ra.action.back')))));
};
const sanitizeRestProps = ({ staticContext, history, location, match, ...rest }) => rest;
const PREFIX = 'RaNotFound';
export const NotFoundClasses = {
    icon: `${PREFIX}-icon`,
    message: `${PREFIX}-message`,
    toolbar: `${PREFIX}-toolbar`,
};
const Root = styled('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    [theme.breakpoints.up('md')]: {
        height: '100%',
    },
    [theme.breakpoints.down('md')]: {
        height: '100vh',
        marginTop: '-3em',
    },
    [`& .${NotFoundClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
    [`& .${NotFoundClasses.message}`]: {
        textAlign: 'center',
        fontFamily: 'Roboto, sans-serif',
        opacity: 0.5,
        margin: '0 1em',
    },
    [`& .${NotFoundClasses.toolbar}`]: {
        textAlign: 'center',
        marginTop: '2em',
    },
}));
function goBack() {
    window.history.go(-1);
}
