// ra-ui-materialui/dist/layout/NotFound.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotFoundClasses = exports.NotFound = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const Button_1 = __importDefault(require("@mui/material/Button"));
const HotTub_1 = __importDefault(require("@mui/icons-material/HotTub"));
const History_1 = __importDefault(require("@mui/icons-material/History"));
const ra_core_1 = require("ra-core");
const Title_1 = require("./Title.cjs");
const Loading_1 = require("./Loading.cjs");
const NotFound = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const translate = (0, ra_core_1.useTranslate)();
    const { isPending } = (0, ra_core_1.useAuthenticated)();
    const title = (0, ra_core_1.useDefaultTitle)();
    if (isPending)
        return React.createElement(Loading_1.Loading, null);
    return (React.createElement(Root, { ...sanitizeRestProps(props) },
        React.createElement(Title_1.Title, { defaultTitle: title }),
        React.createElement("div", { className: exports.NotFoundClasses.message },
            React.createElement(HotTub_1.default, { className: exports.NotFoundClasses.icon }),
            React.createElement("h1", null, translate('ra.page.not_found')),
            React.createElement("div", null,
                translate('ra.message.not_found'),
                ".")),
        React.createElement("div", { className: exports.NotFoundClasses.toolbar },
            React.createElement(Button_1.default, { variant: "contained", startIcon: React.createElement(History_1.default, null), onClick: goBack }, translate('ra.action.back')))));
};
exports.NotFound = NotFound;
const sanitizeRestProps = ({ staticContext, history, location, match, ...rest }) => rest;
const PREFIX = 'RaNotFound';
exports.NotFoundClasses = {
    icon: `${PREFIX}-icon`,
    message: `${PREFIX}-message`,
    toolbar: `${PREFIX}-toolbar`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    [theme.breakpoints.up('md')]: {
        height: '100%',
    },
    [theme.breakpoints.down('md')]: {
        height: '100vh',
        marginTop: '-3em',
    },
    [`& .${exports.NotFoundClasses.icon}`]: {
        width: '9em',
        height: '9em',
    },
    [`& .${exports.NotFoundClasses.message}`]: {
        textAlign: 'center',
        fontFamily: 'Roboto, sans-serif',
        opacity: 0.5,
        margin: '0 1em',
    },
    [`& .${exports.NotFoundClasses.toolbar}`]: {
        textAlign: 'center',
        marginTop: '2em',
    },
}));
function goBack() {
    window.history.go(-1);
}
