// ra-ui-materialui/dist/layout/Notification.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationClasses = exports.Notification = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const react_1 = require("react");
const material_1 = require("@mui/material");
const clsx_1 = __importDefault(require("clsx"));
const ra_core_1 = require("ra-core");
const defaultAnchorOrigin = {
    vertical: 'bottom',
    horizontal: 'center',
};
/**
 * Provides a way to show a notification.
 * @see useNotify
 *
 * @example <caption>Basic usage</caption>
 * <Notification />
 *
 * @param props The component props
 * @param {string} props.type The notification type. Defaults to 'info'.
 * @param {number} props.autoHideDuration Duration in milliseconds to wait until hiding a given notification. Defaults to 4000.
 * @param {boolean} props.multiLine Set it to `true` if the notification message should be shown in more than one line.
 */
const Notification = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { className, type = 'info', autoHideDuration = 4000, multiLine = false, anchorOrigin = defaultAnchorOrigin, ...rest } = props;
    const { notifications, takeNotification } = (0, ra_core_1.useNotificationContext)();
    const takeMutation = (0, ra_core_1.useTakeUndoableMutation)();
    const [open, setOpen] = (0, react_1.useState)(false);
    const [currentNotification, setCurrentNotification] = React.useState(undefined);
    const translate = (0, ra_core_1.useTranslate)();
    (0, react_1.useEffect)(() => {
        if (notifications.length && !currentNotification) {
            // Set a new snack when we don't have an active one
            const notification = takeNotification();
            if (notification) {
                setCurrentNotification(notification);
                setOpen(true);
            }
        }
        if (currentNotification) {
            const beforeunload = (e) => {
                e.preventDefault();
                const confirmationMessage = '';
                e.returnValue = confirmationMessage;
                return confirmationMessage;
            };
            if (currentNotification?.notificationOptions?.undoable) {
                window.addEventListener('beforeunload', beforeunload);
                return () => {
                    window.removeEventListener('beforeunload', beforeunload);
                };
            }
        }
    }, [notifications, currentNotification, open, takeNotification]);
    const handleRequestClose = (0, react_1.useCallback)(() => {
        setOpen(false);
    }, [setOpen]);
    const handleExited = (0, react_1.useCallback)(() => {
        if (currentNotification &&
            currentNotification.notificationOptions?.undoable) {
            const mutation = takeMutation();
            if (mutation) {
                mutation({ isUndo: false });
            }
            else {
                // FIXME kept for BC: remove in v6
                ra_core_1.undoableEventEmitter.emit('end', { isUndo: false });
            }
        }
        setCurrentNotification(undefined);
    }, [currentNotification, takeMutation]);
    const handleUndo = (0, react_1.useCallback)(() => {
        const mutation = takeMutation();
        if (mutation) {
            mutation({ isUndo: true });
        }
        else {
            // FIXME kept for BC: remove in v6
            ra_core_1.undoableEventEmitter.emit('end', { isUndo: true });
        }
        setOpen(false);
    }, [takeMutation]);
    if (!currentNotification)
        return null;
    const { message, type: typeFromMessage, notificationOptions, } = currentNotification;
    const { autoHideDuration: autoHideDurationFromMessage, messageArgs, multiLine: multilineFromMessage, undoable, ...options } = notificationOptions || {};
    return (React.createElement(ra_core_1.CloseNotificationContext.Provider, { value: handleRequestClose },
        React.createElement(StyledSnackbar, { className: className, open: open, message: message &&
                typeof message === 'string' &&
                translate(message, messageArgs), autoHideDuration: 
            // Only apply the default autoHideDuration when autoHideDurationFromMessage is undefined
            // as 0 and null are valid values
            autoHideDurationFromMessage === undefined
                ? autoHideDuration
                : autoHideDurationFromMessage ?? undefined, disableWindowBlurListener: undoable, TransitionProps: { onExited: handleExited }, onClose: handleRequestClose, ContentProps: {
                className: (0, clsx_1.default)(exports.NotificationClasses[typeFromMessage || type], {
                    [exports.NotificationClasses.multiLine]: multilineFromMessage || multiLine,
                }),
            }, action: undoable ? (React.createElement(material_1.Button, { color: "primary", className: exports.NotificationClasses.undo, size: "small", onClick: handleUndo },
                React.createElement(React.Fragment, null, translate('ra.action.undo')))) : null, anchorOrigin: anchorOrigin, ...rest, ...options }, message &&
            typeof message !== 'string' &&
            React.isValidElement(message) ? (
        // Wrap the element in a div to avoid forcing our users to wrap their component with forwardRef
        // which is required for animating the Snackbar
        React.createElement("div", null, message)) : undefined)));
};
exports.Notification = Notification;
const PREFIX = 'RaNotification';
exports.NotificationClasses = {
    success: `${PREFIX}-success`,
    error: `${PREFIX}-error`,
    warning: `${PREFIX}-warning`,
    undo: `${PREFIX}-undo`,
    multiLine: `${PREFIX}-multiLine`,
};
const StyledSnackbar = (0, styles_1.styled)(material_1.Snackbar, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme, type }) => ({
    [`& .${exports.NotificationClasses.success}`]: {
        backgroundColor: theme?.palette.success.main,
        color: theme?.palette.success.contrastText,
    },
    [`& .${exports.NotificationClasses.error}`]: {
        backgroundColor: theme?.palette.error.main,
        color: theme?.palette.error.contrastText,
    },
    [`& .${exports.NotificationClasses.warning}`]: {
        backgroundColor: theme?.palette.warning.main,
        color: theme?.palette.warning.contrastText,
    },
    [`& .${exports.NotificationClasses.undo}`]: {
        color: type === 'success'
            ? theme?.palette.success.contrastText
            : theme?.palette.primary.light,
    },
    [`& .${exports.NotificationClasses.multiLine}`]: {
        whiteSpace: 'pre-wrap',
    },
}));
