// ra-ui-materialui/dist/layout/Sidebar.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLOSED_DRAWER_WIDTH = exports.DRAWER_WIDTH = exports.SidebarClasses = exports.Sidebar = void 0;
const React = __importStar(require("react"));
const styles_1 = require("@mui/material/styles");
const clsx_1 = __importDefault(require("clsx"));
const material_1 = require("@mui/material");
const get_js_1 = __importDefault(require("lodash/get.js"));
const ra_core_1 = require("ra-core");
const useSidebarState_1 = require("./useSidebarState.cjs");
const Sidebar = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const { appBarAlwaysOn, children, closedSize, size, ...rest } = props;
    const isXSmall = (0, material_1.useMediaQuery)(theme => theme.breakpoints.down('sm'));
    const [open, setOpen] = (0, useSidebarState_1.useSidebarState)();
    (0, ra_core_1.useLocale)(); // force redraw on locale change
    const trigger = (0, material_1.useScrollTrigger)();
    const toggleSidebar = () => setOpen(!open);
    return isXSmall ? (React.createElement(StyledDrawer, { variant: "temporary", open: open, onClose: toggleSidebar, classes: exports.SidebarClasses, ...rest }, children)) : (React.createElement(StyledDrawer, { variant: "permanent", open: open, onClose: toggleSidebar, classes: exports.SidebarClasses, className: (0, clsx_1.default)(trigger && !appBarAlwaysOn
            ? exports.SidebarClasses.appBarCollapsed
            : '', open ? OPEN_CLASS : CLOSED_CLASS), ...rest },
        React.createElement("div", { className: exports.SidebarClasses.fixed }, children)));
};
exports.Sidebar = Sidebar;
const PREFIX = 'RaSidebar';
exports.SidebarClasses = {
    docked: `${PREFIX}-docked`,
    paper: `${PREFIX}-paper`,
    paperAnchorLeft: `${PREFIX}-paperAnchorLeft`,
    paperAnchorRight: `${PREFIX}-paperAnchorRight`,
    paperAnchorTop: `${PREFIX}-paperAnchorTop`,
    paperAnchorBottom: `${PREFIX}-paperAnchorBottom`,
    paperAnchorDockedLeft: `${PREFIX}-paperAnchorDockedLeft`,
    paperAnchorDockedTop: `${PREFIX}-paperAnchorDockedTop`,
    paperAnchorDockedRight: `${PREFIX}-paperAnchorDockedRight`,
    paperAnchorDockedBottom: `${PREFIX}-paperAnchorDockedBottom`,
    modal: `${PREFIX}-modal`,
    fixed: `${PREFIX}-fixed`,
    appBarCollapsed: `${PREFIX}-appBarCollapsed`,
};
const OPEN_CLASS = `${PREFIX}-open`;
const CLOSED_CLASS = `${PREFIX}-closed`;
const StyledDrawer = (0, styles_1.styled)(material_1.Drawer, {
    name: PREFIX,
    slot: 'Root',
    overridesResolver: (props, styles) => styles.root,
    shouldForwardProp: () => true,
})(({ open, theme }) => ({
    height: 'calc(100vh - 3em)',
    marginTop: 0,
    transition: theme.transitions.create('margin', {
        easing: theme.transitions.easing.easeOut,
        duration: theme.transitions.duration.enteringScreen,
    }),
    [`&.${exports.SidebarClasses.appBarCollapsed}`]: {
        // compensate the margin of the Layout appFrame instead of removing it in the Layout
        // because otherwise, the appFrame content without margin may revert the scrollTrigger,
        // leading to a visual jiggle
        marginTop: theme.spacing(-6),
        [theme.breakpoints.down('sm')]: {
            marginTop: theme.spacing(-7),
        },
        transition: theme.transitions.create('margin', {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.leavingScreen,
        }),
    },
    [`& .${exports.SidebarClasses.docked}`]: {},
    [`& .${exports.SidebarClasses.paper}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorLeft}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorRight}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorTop}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorBottom}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorDockedLeft}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorDockedTop}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorDockedRight}`]: {},
    [`& .${exports.SidebarClasses.paperAnchorDockedBottom}`]: {},
    [`& .${exports.SidebarClasses.modal}`]: {},
    [`& .${exports.SidebarClasses.fixed}`]: {
        position: 'fixed',
        height: 'calc(100vh - 3em)',
        overflowX: 'hidden',
        // hide scrollbar
        scrollbarWidth: 'none',
        msOverflowStyle: 'none',
        '&::-webkit-scrollbar': {
            display: 'none',
        },
    },
    [`& .MuiPaper-root`]: {
        position: 'relative',
        width: open
            ? (0, get_js_1.default)(theme, 'sidebar.width', exports.DRAWER_WIDTH)
            : (0, get_js_1.default)(theme, 'sidebar.closedWidth', exports.CLOSED_DRAWER_WIDTH),
        transition: theme.transitions.create('width', {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.leavingScreen,
        }),
        backgroundColor: 'transparent',
        borderRight: 'none',
        [theme.breakpoints.only('xs')]: {
            marginTop: 0,
            height: '100vh',
            position: 'inherit',
            backgroundColor: (theme.vars || theme).palette.background.default,
        },
        [theme.breakpoints.up('md')]: {
            border: 'none',
        },
        zIndex: 'inherit',
    },
}));
exports.DRAWER_WIDTH = 240;
exports.CLOSED_DRAWER_WIDTH = 55;
