// ra-ui-materialui/dist/layout/TopToolbar.js
import * as React from 'react';
import { useMediaQuery, Toolbar, } from '@mui/material';
import { styled, useThemeProps, } from '@mui/material/styles';
export const TopToolbar = (inProps) => {
    const props = useThemeProps({
        props: inProps,
        name: PREFIX,
    });
    const isXSmall = useMediaQuery(theme => theme.breakpoints.down('sm'));
    return (React.createElement(StyledToolbar, { disableGutters: true, variant: isXSmall ? 'regular' : 'dense', ...sanitizeToolbarRestProps(props) }));
};
export default TopToolbar;
const PREFIX = 'RaTopToolbar';
const StyledToolbar = styled(Toolbar, {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    display: 'flex',
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
    gap: theme.spacing(1),
    whiteSpace: 'nowrap',
    flex: '0 1 auto',
    padding: theme.spacing(0.5),
    paddingBottom: theme.spacing(1),
    [theme.breakpoints.down('md')]: {
        flex: '0 1 100%',
    },
    [theme.breakpoints.down('sm')]: {
        backgroundColor: (theme.vars || theme).palette.background.paper,
        padding: 0,
        paddingBottom: 0,
    },
}));
const sanitizeToolbarRestProps = ({ hasCreate, ...props }) => props;
