// ra-ui-materialui/dist/layout/UserMenu.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserMenuClasses = exports.UserMenu = void 0;
const AccountCircle_1 = __importDefault(require("@mui/icons-material/AccountCircle"));
const material_1 = require("@mui/material");
const styles_1 = require("@mui/material/styles");
const ra_core_1 = require("ra-core");
const React = __importStar(require("react"));
const react_1 = require("react");
const Logout_1 = require("../auth/Logout.cjs");
/**
 * The UserMenu component renders a Mui Button that shows a Menu.
 * It accepts children that must be Mui MenuItem components.
 *
 * @example
 * import { Logout, UserMenu, useUserMenu } from 'react-admin';
 * import MenuItem from '@mui/material/MenuItem';
 * import ListItemIcon from '@mui/material/ListItemIcon';
 * import ListItemText from '@mui/material/ListItemText';
 * import SettingsIcon from '@mui/icons-material/Settings';

 * const ConfigurationMenu = React.forwardRef((props, ref) => {
 *     const { onClose } = useUserMenu();
 *     return (
 *         <MenuItem
 *             ref={ref}
 *             {...props}
 *             to="/configuration"
 *             onClick={onClose}
 *         >
 *             <ListItemIcon>
 *                 <SettingsIcon />
 *             </ListItemIcon>
 *             <ListItemText>Configuration</ListItemText>
 *         </MenuItem>
 *     );
 * });
 *
 * export const MyUserMenu = () => (
 *     <UserMenu>
 *         <ConfigurationMenu />
 *         <Logout />
 *     </UserMenu>
 * );
 * @param props
 * @param {ReactNode} props.children React node/s to be rendered as children of the UserMenu. Must be Mui MenuItem components
 * @param {string} props.className CSS class applied to the MuiAppBar component
 * @param {string} props.label The label of the UserMenu button. Accepts translation keys
 * @param {Element} props.icon The icon of the UserMenu button.
 *
 */
const UserMenu = (inProps) => {
    const props = (0, styles_1.useThemeProps)({
        props: inProps,
        name: PREFIX,
    });
    const [anchorEl, setAnchorEl] = (0, react_1.useState)(null);
    const translate = (0, ra_core_1.useTranslate)();
    const { isPending, identity } = (0, ra_core_1.useGetIdentity)();
    const authProvider = (0, ra_core_1.useAuthProvider)();
    const isLargeEnough = (0, material_1.useMediaQuery)(theme => theme.breakpoints.up('sm'));
    const { children = authProvider ? React.createElement(Logout_1.Logout, null) : null, className, label = 'ra.auth.user_menu', icon = defaultIcon, } = props;
    const handleMenu = event => setAnchorEl(event.currentTarget);
    const handleClose = (0, react_1.useCallback)(() => setAnchorEl(null), []);
    const context = (0, react_1.useMemo)(() => ({ onClose: handleClose }), [handleClose]);
    if (!children)
        return null;
    const open = Boolean(anchorEl);
    return (React.createElement(Root, { className: className },
        isLargeEnough && !isPending && identity?.fullName ? (React.createElement(material_1.Button, { "aria-label": label && translate(label, { _: label }), className: exports.UserMenuClasses.userButton, color: "inherit", startIcon: identity.avatar ? (React.createElement(material_1.Avatar, { className: exports.UserMenuClasses.avatar, src: identity.avatar, alt: identity.fullName })) : (icon), onClick: handleMenu, variant: "text" }, identity.fullName)) : (React.createElement(material_1.Tooltip, { title: label && translate(label, { _: 'Profile' }) },
            React.createElement(material_1.IconButton, { "aria-label": label && translate(label, { _: 'Profile' }), "aria-owns": open ? 'menu-appbar' : undefined, "aria-haspopup": true, color: "inherit", onClick: handleMenu }, !isPending && identity?.avatar ? (React.createElement(material_1.Avatar, { className: exports.UserMenuClasses.avatar, src: identity.avatar, alt: identity.fullName })) : (icon)))),
        React.createElement(ra_core_1.UserMenuContextProvider, { value: context },
            React.createElement(material_1.Menu, { id: "menu-appbar", disableScrollLock: true, anchorEl: anchorEl, anchorOrigin: AnchorOrigin, transformOrigin: TransformOrigin, open: open, onClose: handleClose }, children))));
};
exports.UserMenu = UserMenu;
const PREFIX = 'RaUserMenu';
exports.UserMenuClasses = {
    userButton: `${PREFIX}-userButton`,
    avatar: `${PREFIX}-avatar`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${exports.UserMenuClasses.userButton}`]: {
        textTransform: 'none',
        marginInlineStart: theme.spacing(0.5),
    },
    [`& .${exports.UserMenuClasses.avatar}`]: {
        width: theme.spacing(3),
        height: theme.spacing(3),
    },
}));
const defaultIcon = React.createElement(AccountCircle_1.default, null);
const AnchorOrigin = {
    vertical: 'bottom',
    horizontal: 'right',
};
const TransformOrigin = {
    vertical: 'top',
    horizontal: 'right',
};
