// ra-ui-materialui/dist/input/InPlaceEditor/InPlaceEditor.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InPlaceEditor = void 0;
const React = __importStar(require("react"));
const react_1 = require("react");
const ra_core_1 = require("ra-core");
const isEqual_js_1 = __importDefault(require("lodash/isEqual.js"));
const styles_1 = require("@mui/material/styles");
const material_1 = require("@mui/material");
const Save_1 = __importDefault(require("@mui/icons-material/Save"));
const Close_1 = __importDefault(require("@mui/icons-material/Close"));
const TextInput_1 = require("../TextInput.cjs");
const field_1 = require("../../field/index.cjs");
/**
 * Renders a value, and on click it turns into an editable field.
 *
 * The editable field is rendered inside a Form component, so InPlaceEditor
 * cannot be used inside another Form component.
 */
const InPlaceEditor = (props) => {
    const { source, mutationMode, mutationOptions = {}, sx, cancelOnBlur, children = source ? (React.createElement(field_1.TextField, { source: source, variant: "body1", component: "div", sx: { marginBottom: '5px' } })) : null, editor = source ? (React.createElement(TextInput_1.TextInput, { source: source, size: "small", margin: "none", label: false, variant: "standard", autoFocus: true, helperText: false })) : null, showButtons, notifyOnSuccess, } = props;
    if (!source && !children && !editor) {
        throw new Error('InPlaceEditor requires either a source prop or children or editor prop');
    }
    if (mutationMode === 'undoable' && !notifyOnSuccess) {
        throw new Error('InPlaceEditor requires notifyOnSuccess to be true when mutationMode is undoable');
    }
    const submitButtonRef = (0, react_1.useRef)(null);
    const [state, dispatch] = (0, react_1.useReducer)((_, action) => {
        switch (action.type) {
            case 'edit':
                return { state: 'editing' };
            case 'save':
                return { state: 'saving', values: action.values };
            case 'error':
            case 'success':
            case 'cancel':
                return { state: 'reading' };
            default:
                throw new Error('Unhandled action');
        }
    }, { state: 'reading' });
    const record = (0, ra_core_1.useRecordContext)();
    const resource = (0, ra_core_1.useResourceContext)(props);
    const notify = (0, ra_core_1.useNotify)();
    const translate = (0, ra_core_1.useTranslate)();
    const [update] = (0, ra_core_1.useUpdate)();
    const { meta: mutationMeta, onSuccess = () => {
        dispatch({ type: 'success' });
        if (mutationMode !== 'undoable' && !notifyOnSuccess)
            return;
        notify(`resources.${resource}.notifications.updated`, {
            type: 'info',
            messageArgs: {
                smart_count: 1,
                _: translate('ra.notification.updated', {
                    smart_count: 1,
                }),
            },
            undoable: mutationMode === 'undoable',
        });
    }, onError = error => {
        notify('ra.notification.http_error', {
            type: 'error',
            messageArgs: { _: error.message },
        });
        dispatch({ type: 'error', error });
    }, ...otherMutationOptions } = mutationOptions;
    const handleSave = async (values) => {
        if (!record) {
            throw new Error('No record found');
        }
        if ((0, isEqual_js_1.default)(values, record)) {
            dispatch({ type: 'cancel' });
            return;
        }
        dispatch({ type: 'save', values });
        update(resource, {
            id: record.id,
            data: values,
            previousData: record,
            meta: mutationMeta,
        }, {
            onSuccess,
            onError,
            mutationMode,
            ...otherMutationOptions,
        });
    };
    const handleEdit = () => {
        dispatch({ type: 'edit' });
    };
    const handleCancel = () => {
        dispatch({ type: 'cancel' });
    };
    const handleKeyDown = (event) => {
        if (event.key === 'Escape') {
            dispatch({ type: 'cancel' });
        }
    };
    const handleBlur = (event) => {
        if (event.relatedTarget) {
            return;
        }
        if (cancelOnBlur) {
            dispatch({ type: 'cancel' });
            return;
        }
        if (state.state === 'editing') {
            // trigger the parent form submit
            // to save the changes
            submitButtonRef.current.click();
        }
    };
    const renderContent = () => {
        switch (state.state) {
            case 'reading':
                return (React.createElement(material_1.Box, { onClick: handleEdit, className: InPlaceEditorClasses.reading }, children));
            case 'editing':
                return (React.createElement(ra_core_1.Form, { onSubmit: handleSave },
                    React.createElement(material_1.Box, { onKeyDown: handleKeyDown, onBlur: handleBlur, className: InPlaceEditorClasses.editing },
                        editor,
                        showButtons ? (React.createElement(React.Fragment, null,
                            React.createElement(material_1.IconButton, { size: "small", type: "submit", ref: submitButtonRef, "aria-label": translate('ra.action.save') },
                                React.createElement(Save_1.default, { fontSize: "small" })),
                            React.createElement(material_1.IconButton, { size: "small", onClick: handleCancel, "aria-label": translate('ra.action.cancel') },
                                React.createElement(Close_1.default, { fontSize: "small" })))) : (React.createElement("button", { type: "submit", style: { display: 'none' }, ref: submitButtonRef })))));
            case 'saving':
                // set a custom record context with the new values
                // to avoid flickering
                return (React.createElement(ra_core_1.RecordContextProvider, { value: state.values },
                    React.createElement(material_1.Box, { className: InPlaceEditorClasses.saving }, children)));
            default:
                throw new Error('Unhandled state');
        }
    };
    return React.createElement(Root, { sx: sx }, renderContent());
};
exports.InPlaceEditor = InPlaceEditor;
const PREFIX = 'RaInPlaceEditor';
const InPlaceEditorClasses = {
    reading: `${PREFIX}-reading`,
    editing: `${PREFIX}-editing`,
    saving: `${PREFIX}-saving`,
};
const Root = (0, styles_1.styled)('div', {
    name: PREFIX,
    overridesResolver: (props, styles) => styles.root,
})(({ theme }) => ({
    [`& .${InPlaceEditorClasses.reading}`]: {
        cursor: 'pointer',
        '&:hover': {
            backgroundColor: theme.palette.action.hover,
        },
    },
    [`& .${InPlaceEditorClasses.editing}`]: {
        display: 'flex',
        alignItems: 'center',
        gap: 1,
    },
    [`& .${InPlaceEditorClasses.saving}`]: { opacity: 0.5 },
}));
